{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "American_palm_data.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/mftorres/APP/blob/main/American_palm_data.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5fCEDCU_qrC0"
      },
      "source": [
        "# Leaving this here in case you need info about Colaboratory in Google\n",
        "\n",
        "[link text](https://)<p><img alt=\"Colaboratory logo\" height=\"45px\" src=\"/img/colab_favicon.ico\" align=\"left\" hspace=\"10px\" vspace=\"0px\"></p>\n",
        "\n",
        "<h1>What is Colaboratory?</h1>\n",
        "\n",
        "Colaboratory, or \"Colab\" for short, allows you to write and execute Python in your browser, with \n",
        "- Zero configuration required\n",
        "- Free access to GPUs\n",
        "- Easy sharing\n",
        "\n",
        "Whether you're a **student**, a **data scientist** or an **AI researcher**, Colab can make your work easier. Watch [Introduction to Colab](https://www.youtube.com/watch?v=inN8seMm7UI) to learn more, or just get started below!"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GJBs_flRovLc"
      },
      "source": [
        "----\n",
        "\n",
        "----\n",
        "\n",
        "# **American Palm Phylogeny data exploration/management**\n",
        "\n",
        "Document authored by **Maria Fernanda Torres Jimenez**  \n",
        "Date: 2021 May 05\n",
        "\n",
        "____\n",
        "\n",
        "\n",
        "The aim of this notebook is to show you how to retrieve data from the American Palm Phylogeny metadata, how to query, and how to update records.\n",
        "\n",
        "Maintaining metadata associated to files requires you to do version control, properly dated, and to keep backups at all times. You should be prepared to accidentally change/delete metadata and you should always be able to go back to a stable version (but that depends on you and your data practices).\n",
        "\n",
        "\n",
        "## 0. Importing packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eZXbXzVW8Bj6"
      },
      "source": [
        "import pandas as pd\n",
        "import re\n",
        "import matplotlib as mpl\n",
        "from matplotlib import pyplot as plt\n",
        "import requests as rs\n",
        "import datetime\n",
        "import numpy as np\n",
        "\n",
        "# !pip install fuzzywuzzy # install package to do partial text match\n",
        "from fuzzywuzzy import fuzz\n",
        "from fuzzywuzzy import process"
      ],
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rKut_-9OVXnU"
      },
      "source": [
        "The database is stored in a google spreadsheet. It is good to have it in the cloud, google has version control and everyone can have access to the document without it staying with a single person.\n",
        "\n",
        "Creating a function to download the data:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 626
        },
        "id": "gJr_9dXGpJ05",
        "outputId": "a3889624-58a8-4076-8ce3-9791b27dae95"
      },
      "source": [
        "def get_data():\n",
        "  return pd.read_csv('./palm_metadata_%s.txt'%(str(datetime.datetime.now()).split(' ')[0]), sep = '\\t')\n",
        "\n",
        "# try reading the file if exists, keep in mind the date\n",
        "try:\n",
        "  df = get_data()\n",
        "except:\n",
        "  url = 'https://raw.githubusercontent.com/mftorres/APP/main/data/Metadata_palm_gap_2021-05-05.txt' # shared link careful who you share this with\n",
        "  print('Downloading')\n",
        "  res = rs.get(url = url)\n",
        "  with open('./palm_metadata_%s.txt'%(str(datetime.datetime.now()).split(' ')[0]), 'wb') as file: # Writing document in binary mode so python doesn't make changes\n",
        "    file.write(res.content)\n",
        "  df = get_data()\n",
        "\n",
        "df # check the database"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Downloading\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>filename</th>\n",
              "      <th>sense</th>\n",
              "      <th>filecode</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "      <th>botanic_garden</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>ifmorphotype</th>\n",
              "      <th>ifpopulation</th>\n",
              "      <th>sent_to_Cano</th>\n",
              "      <th>newfilename</th>\n",
              "      <th>raw_reads</th>\n",
              "      <th>reads_trimed_paired_flag</th>\n",
              "      <th>reads_trimed_paired</th>\n",
              "      <th>reads_trimed_single</th>\n",
              "      <th>percentage_lost_tosingles</th>\n",
              "      <th>pcr_filtered</th>\n",
              "      <th>collection_year</th>\n",
              "      <th>Continent</th>\n",
              "      <th>country</th>\n",
              "      <th>long</th>\n",
              "      <th>lat</th>\n",
              "      <th>flag</th>\n",
              "      <th>notes</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AKG91_P3252_1182_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AKG91_P3252_1182</td>\n",
              "      <td>AKG91_1182</td>\n",
              "      <td>CTAGCT</td>\n",
              "      <td>JBP</td>\n",
              "      <td>Lorenzi_&amp;_Soares_6762</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_emensis</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_eme_AKG911182_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3044</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_vulgata_Standley67344_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>Standley67344</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_vulgata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3045</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>Azofeifa484</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3046</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>Azofeifa484</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3047</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...</td>\n",
              "      <td>R1</td>\n",
              "      <td>Azofeifa484_S36_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3048</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...</td>\n",
              "      <td>R2</td>\n",
              "      <td>Azofeifa484_S36_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3049 rows Ã— 30 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "            dataset        provider pi_code  ... lat flag notes\n",
              "0              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "1              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "2              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "3              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "4              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "...             ...             ...     ...  ...  ..  ...   ...\n",
              "3044  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3045  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3046  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3047  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3048  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "\n",
              "[3049 rows x 30 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MjHzGbeechIv"
      },
      "source": [
        "The first thing to check is the columns on the data and what they mean"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HBaHmjmGckXD",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bb5509c0-7892-4ce2-e9ec-014ada398086"
      },
      "source": [
        "list(df.columns)"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['dataset',\n",
              " 'provider',\n",
              " 'pi_code',\n",
              " 'originalmetadata',\n",
              " 'filename',\n",
              " 'sense',\n",
              " 'filecode',\n",
              " 'samplecode',\n",
              " 'library_index',\n",
              " 'botanic_garden',\n",
              " 'voucher',\n",
              " 'taxgenus',\n",
              " 'taxspecies',\n",
              " 'ifmorphotype',\n",
              " 'ifpopulation',\n",
              " 'sent_to_Cano',\n",
              " 'newfilename',\n",
              " 'raw_reads',\n",
              " 'reads_trimed_paired_flag',\n",
              " 'reads_trimed_paired',\n",
              " 'reads_trimed_single',\n",
              " 'percentage_lost_tosingles',\n",
              " 'pcr_filtered',\n",
              " 'collection_year',\n",
              " 'Continent',\n",
              " 'country',\n",
              " 'long',\n",
              " 'lat',\n",
              " 'flag',\n",
              " 'notes']"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tD7Ifx2zx0Jv"
      },
      "source": [
        "**dataset** bundle in which files are provided. e.g. CANO: sequences extracted by Angela Cano  \n",
        "**provider** who provided the files or is responsible for sending the files  \n",
        "**pi_code** two-letter code to shorten the *provider*  \n",
        "**originalmetadata** name of original metadata provided by provider. That file should never be modified and only annotated  \n",
        "**filename** original file name, usually as delivered by the sequencing facility  \n",
        "**sense** whether file contains forward, reverse, interleaved, single reads    \n",
        "**filecode** code provided by sequencing facility, often links to sequencing experiment, lane and plate information  \n",
        "**samplecode** sample ID in the plate or as delivered to the sequencing facility\n",
        "**library_index** sequencing index if provided or if in header of fastq. When double, each index is separated with a '+'  \n",
        "**botanic_garden** if sample was obtained from a botanic garden rather than collected  \n",
        "**voucher** voucher of the tissue or collecting code  \n",
        "**taxgenus** taxonomic genus of the sample  \n",
        "**taxspecies** taxonomic species with genus, separated by an underscore '_'  \n",
        "**ifmorphotype** if the sample has a morphotype, variety or subspecies assigned  \n",
        "**ifpopulation** if the sample belongs to a population-wide sampling effort, which population it was collected from  \n",
        "**sent_to_Cano** files sent to Angela Cano who's analysing the data and leading the project  \n",
        "**newfilename** new file name assigned to the sample. This is to annotate provider, dataset, species and voucher in the name. See Appendix for the code used to name files  \n",
        "**raw_reads** number of raw reads before processing the data  -- filled as analysis progresses  \n",
        "**reads_trimed_paired_flag** if trimmed -- filled as analysis progresses  \n",
        "**reads_trimed_paired** number of reads after trimming -- filled as analysis progresses  \n",
        "**reads_trimed_single** number of unpaired reads after trimming -- filled as analysis progresses  \n",
        "**percentage_lost_tosingles** percentage of unpaired reads from the total of reads -- filled as analysis progresses  \n",
        "**pcr_filtered** if PCR duplicate reads were filtered out using samtools or picard-- filled as analysis progresses  \n",
        "**collection_year** year that sample was collected  \n",
        "**Continent** continent. The project involves species from America but the metadata keeps track of other data from other continents  \n",
        "**country** which country was the sample collected at if known from collection or voucher  \n",
        "**long** longitude coordinates if available, should be standardised to decimals  \n",
        "**lat**  latitude coordinates if available, should be standardised to decimals  \n",
        "**flag** single word descriptors separated by spaces (never colons, semicolons or tabs) that describe the flaws of the file. e.g. misingtaxspecies  \n",
        "**notes** elaborated notes if needed but avoid using commas, semicolons or tabs.  "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2fhs6GZ4qFMx"
      },
      "source": [
        "The second interesting thing to do is filtering unique species for which we have data. The files here represent species that might or might not be good, but at this point of the analysis (which will be on a sequence by sequence basis) we can only know what we have sequenced.\n",
        "\n",
        "Another thing is to filter out species that are not in America."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-gE-Ez1qtyIA",
        "outputId": "a9361bd7-26ea-4f4b-91cf-d0c1df3f25c3"
      },
      "source": [
        "df['Continent'].unique()"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['americas', 'out_americas', nan], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1GQlEP_TiZ4u"
      },
      "source": [
        "Some species don't have a continent specified. Let's see which ones:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Y9MDwf5NioT4",
        "outputId": "66131d75-3ccd-48dd-eef5-7bd57e82fc71"
      },
      "source": [
        "# filtering pandas dataframe\n",
        "df[df['Continent'].isna()]['taxspecies'].unique()"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Ceroxylon_alpinum', 'Ceroxylon_ceriferum',\n",
              "       'Ceroxylon_echinulatum', 'Ceroxylon_parvifrons',\n",
              "       'Ceroxylon_parvum', 'Ceroxylon_quindiuense',\n",
              "       'Ceroxylon_ventricosum', 'Ceroxylon_vogelianum', 'Chamaedorea',\n",
              "       'Cocos', nan, 'Phytelephas_aecuatorialis',\n",
              "       'Phytelephas_macrocarpa', 'Phytelephas', 'Phytelephas_tumacana',\n",
              "       'Ravenea_sambiranensis', 'Chamaedorea_nationsiana',\n",
              "       'Chamaedorea_arenbergiana', 'Chamaedorea_binderi',\n",
              "       'Chamaedorea_brachyclada', 'Chamaedorea_seifrizii',\n",
              "       'Chamaedorea_fractiflexa', 'Chamaedorea_graminifolia',\n",
              "       'Chamaedorea_hodelii', 'Chamaedorea_ibarrae',\n",
              "       'Chamaedorea_incrustata', 'Chamaedorea_keelerorum',\n",
              "       'Chamaedorea_lehmannii', 'Chamaedorea_liebmannii',\n",
              "       'Chamaedorea_pachenoana', 'Chamaedorea_parvisecta',\n",
              "       'Chamaedorea_pauciflora', 'Chamaedorea_piscifolia',\n",
              "       'Chamaedorea_pumila', 'Chamaedorea_costaricana',\n",
              "       'Chamaedorea_rigida', 'Chamaedorea_schideana',\n",
              "       'Chamaedorea_simplex', 'Chamaedorea_skutchii',\n",
              "       'Chamaedorea_stenocarpa', 'Chamaedorea_subjectifolia',\n",
              "       'Chamaedorea_tenerrima', 'Chamaedorea_verapazensis',\n",
              "       'Chamaedorea_volcanensis', 'Chamaedorea_vulgata',\n",
              "       'Chamaedorea_zamorae'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3d2HoUqji2WB",
        "outputId": "9e8510a3-6086-470b-b104-2284a65d99a9"
      },
      "source": [
        "# using a diferent syntax that can help with readability\n",
        "df[df.Continent.isna()]['taxspecies'].unique()"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Ceroxylon_alpinum', 'Ceroxylon_ceriferum',\n",
              "       'Ceroxylon_echinulatum', 'Ceroxylon_parvifrons',\n",
              "       'Ceroxylon_parvum', 'Ceroxylon_quindiuense',\n",
              "       'Ceroxylon_ventricosum', 'Ceroxylon_vogelianum', 'Chamaedorea',\n",
              "       'Cocos', nan, 'Phytelephas_aecuatorialis',\n",
              "       'Phytelephas_macrocarpa', 'Phytelephas', 'Phytelephas_tumacana',\n",
              "       'Ravenea_sambiranensis', 'Chamaedorea_nationsiana',\n",
              "       'Chamaedorea_arenbergiana', 'Chamaedorea_binderi',\n",
              "       'Chamaedorea_brachyclada', 'Chamaedorea_seifrizii',\n",
              "       'Chamaedorea_fractiflexa', 'Chamaedorea_graminifolia',\n",
              "       'Chamaedorea_hodelii', 'Chamaedorea_ibarrae',\n",
              "       'Chamaedorea_incrustata', 'Chamaedorea_keelerorum',\n",
              "       'Chamaedorea_lehmannii', 'Chamaedorea_liebmannii',\n",
              "       'Chamaedorea_pachenoana', 'Chamaedorea_parvisecta',\n",
              "       'Chamaedorea_pauciflora', 'Chamaedorea_piscifolia',\n",
              "       'Chamaedorea_pumila', 'Chamaedorea_costaricana',\n",
              "       'Chamaedorea_rigida', 'Chamaedorea_schideana',\n",
              "       'Chamaedorea_simplex', 'Chamaedorea_skutchii',\n",
              "       'Chamaedorea_stenocarpa', 'Chamaedorea_subjectifolia',\n",
              "       'Chamaedorea_tenerrima', 'Chamaedorea_verapazensis',\n",
              "       'Chamaedorea_volcanensis', 'Chamaedorea_vulgata',\n",
              "       'Chamaedorea_zamorae'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ag5uTCqAjIIi",
        "outputId": "51a65fcf-8cd0-418a-cb3d-ad0c8d21271b"
      },
      "source": [
        "df[df.Continent.isna()]['taxgenus'].unique()"
      ],
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Ceroxylon', 'Chamaedorea', 'Cocos', nan, 'Phytelephas', 'Ravenea'],\n",
              "      dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IQmksWQZjDGx"
      },
      "source": [
        "Most species are from the genera *Ceroxylon*, *Chamaedorea*, *Cocos*, *Phytelephas*, and *Ravenea*. I can already remember the dataset they come from and I can tell they don't have a continent assigned because I haven't updated the dataframe (as is not my main priority)\n",
        "\n",
        "we can totally fix that today though. Let's iterate through rows in the dataset to assign the continent information. We also know all but *Cocos* (which is cosmopolita) are from America."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WIpoqq1ojBWq",
        "outputId": "94afeb95-43a6-4cb4-a167-d107a3cad275"
      },
      "source": [
        "# itertuples iterates through tuples (pairs) of index,row and different columns can be acceced using the column label. Similar to the iterrows() method but faster.\n",
        "# itertuples() only allows for the \"dot\" syntax for filtering pandas columns.\n",
        "\n",
        "for row in df[df['Continent'].isna()].itertuples():\n",
        "  print(row.Continent,'empty') # test\n",
        "  df.loc[row.Index,'Continent']\n",
        "  # else:\n",
        "    # print(row.Continent)"
      ],
      "execution_count": 31,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n",
            "nan empty\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4_kCnsPUqS6o"
      },
      "source": [
        "You can import your own data into Colab notebooks from your Google Drive account, including from spreadsheets, as well as from Github and many other sources. To learn more about importing data, and how Colab can be used for data science, see the links below under [Working with Data](#working-with-data)."
      ]
    }
  ]
}