{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "American_palm_data.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/mftorres/APP/blob/main/American_palm_data.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GJBs_flRovLc"
      },
      "source": [
        "----\n",
        "\n",
        "----\n",
        "\n",
        "# **American Palm Phylogeny data exploration/management**\n",
        "\n",
        "Document authored by **Maria Fernanda Torres Jimenez**  \n",
        "Date: 2021 May 05\n",
        "\n",
        "____\n",
        "\n",
        "\n",
        "The aim of this notebook is to show you how to retrieve data from the American Palm Phylogeny metadata, how to query, and how to update records.\n",
        "\n",
        "Maintaining metadata associated to files requires you to do version control, properly dated, and to keep backups at all times. You should be prepared to accidentally change/delete metadata and you should always be able to go back to a stable version (but that depends on you and your data practices).\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qr2DGggLNlAp"
      },
      "source": [
        "## 0. Importing packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eZXbXzVW8Bj6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "46231754-d0e0-4c48-bd23-7032eb9ad241"
      },
      "source": [
        "import pandas as pd\n",
        "import re\n",
        "import matplotlib as mpl\n",
        "from matplotlib import pyplot as plt\n",
        "import requests as rs\n",
        "import datetime\n",
        "import numpy as np\n",
        "\n",
        "!pip install fuzzywuzzy # install package to do partial text match\n",
        "from fuzzywuzzy import fuzz\n",
        "from fuzzywuzzy import process"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting fuzzywuzzy\n",
            "  Downloading https://files.pythonhosted.org/packages/43/ff/74f23998ad2f93b945c0309f825be92e04e0348e062026998b5eefef4c33/fuzzywuzzy-0.18.0-py2.py3-none-any.whl\n",
            "Installing collected packages: fuzzywuzzy\n",
            "Successfully installed fuzzywuzzy-0.18.0\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/fuzzywuzzy/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning\n",
            "  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rKut_-9OVXnU"
      },
      "source": [
        "\n",
        "## 1. Importing main metadata file\n",
        "\n",
        "Creating a function to download the data:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 626
        },
        "id": "gJr_9dXGpJ05",
        "outputId": "462f431a-cafd-4861-bc30-947154fd512a"
      },
      "source": [
        "def get_data():\n",
        "  return pd.read_csv('./palm_metadata_%s.txt'%(str(datetime.datetime.now()).split(' ')[0]), sep = '\\t')\n",
        "\n",
        "# try reading the file if exists, keep in mind the date\n",
        "try:\n",
        "  df = get_data()\n",
        "except:\n",
        "  url = 'https://raw.githubusercontent.com/mftorres/APP/main/data/Metadata_palm_gap_2021-05-05.txt' # shared link careful who you share this with\n",
        "  print('Downloading')\n",
        "  res = rs.get(url = url)\n",
        "  with open('./palm_metadata_%s.txt'%(str(datetime.datetime.now()).split(' ')[0]), 'wb') as file: # Writing document in binary mode so python doesn't make changes\n",
        "    file.write(res.content)\n",
        "  df = get_data()\n",
        "\n",
        "df # check the database"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Downloading\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>filename</th>\n",
              "      <th>sense</th>\n",
              "      <th>filecode</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "      <th>botanic_garden</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>ifmorphotype</th>\n",
              "      <th>ifpopulation</th>\n",
              "      <th>sent_to_Cano</th>\n",
              "      <th>newfilename</th>\n",
              "      <th>raw_reads</th>\n",
              "      <th>reads_trimed_paired_flag</th>\n",
              "      <th>reads_trimed_paired</th>\n",
              "      <th>reads_trimed_single</th>\n",
              "      <th>percentage_lost_tosingles</th>\n",
              "      <th>pcr_filtered</th>\n",
              "      <th>collection_year</th>\n",
              "      <th>Continent</th>\n",
              "      <th>country</th>\n",
              "      <th>long</th>\n",
              "      <th>lat</th>\n",
              "      <th>flag</th>\n",
              "      <th>notes</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AKG91_P3252_1182_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AKG91_P3252_1182</td>\n",
              "      <td>AKG91_1182</td>\n",
              "      <td>CTAGCT</td>\n",
              "      <td>JBP</td>\n",
              "      <td>Lorenzi_&amp;_Soares_6762</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_emensis</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_eme_AKG911182_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3044</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_vulgata_Standley67344_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>Standley67344</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_vulgata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3045</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>Azofeifa484</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3046</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>Azofeifa484</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3047</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...</td>\n",
              "      <td>R1</td>\n",
              "      <td>Azofeifa484_S36_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3048</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...</td>\n",
              "      <td>R2</td>\n",
              "      <td>Azofeifa484_S36_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3049 rows Ã— 30 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "            dataset        provider pi_code  ... lat flag notes\n",
              "0              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "1              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "2              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "3              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "4              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "...             ...             ...     ...  ...  ..  ...   ...\n",
              "3044  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3045  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3046  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3047  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3048  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "\n",
              "[3049 rows x 30 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 2
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MjHzGbeechIv"
      },
      "source": [
        "The first thing to check is the columns on the data and what they mean"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HBaHmjmGckXD",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8b0026f2-32ee-4915-97eb-648c1076d8a3"
      },
      "source": [
        "list(df.columns)"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['dataset',\n",
              " 'provider',\n",
              " 'pi_code',\n",
              " 'originalmetadata',\n",
              " 'filename',\n",
              " 'sense',\n",
              " 'filecode',\n",
              " 'samplecode',\n",
              " 'library_index',\n",
              " 'botanic_garden',\n",
              " 'voucher',\n",
              " 'taxgenus',\n",
              " 'taxspecies',\n",
              " 'ifmorphotype',\n",
              " 'ifpopulation',\n",
              " 'sent_to_Cano',\n",
              " 'newfilename',\n",
              " 'raw_reads',\n",
              " 'reads_trimed_paired_flag',\n",
              " 'reads_trimed_paired',\n",
              " 'reads_trimed_single',\n",
              " 'percentage_lost_tosingles',\n",
              " 'pcr_filtered',\n",
              " 'collection_year',\n",
              " 'Continent',\n",
              " 'country',\n",
              " 'long',\n",
              " 'lat',\n",
              " 'flag',\n",
              " 'notes']"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tD7Ifx2zx0Jv"
      },
      "source": [
        "**dataset** bundle in which files are provided. e.g. CANO: sequences extracted by Angela Cano  \n",
        "**provider** who provided the files or is responsible for sending the files  \n",
        "**pi_code** two-letter code to shorten the *provider*  \n",
        "**originalmetadata** name of original metadata provided by provider. That file should never be modified and only annotated  \n",
        "**filename** original file name, usually as delivered by the sequencing facility  \n",
        "**sense** whether file contains forward, reverse, interleaved, single reads    \n",
        "**filecode** code provided by sequencing facility, often links to sequencing experiment, lane and plate information  \n",
        "**samplecode** sample ID in the plate or as delivered to the sequencing facility\n",
        "**library_index** sequencing index if provided or if in header of fastq. When double, each index is separated with a '+'  \n",
        "**botanic_garden** if sample was obtained from a botanic garden rather than collected  \n",
        "**voucher** voucher of the tissue or collecting code  \n",
        "**taxgenus** taxonomic genus of the sample  \n",
        "**taxspecies** taxonomic species with genus, separated by an underscore '_'  \n",
        "**ifmorphotype** if the sample has a morphotype, variety or subspecies assigned  \n",
        "**ifpopulation** if the sample belongs to a population-wide sampling effort, which population it was collected from  \n",
        "**sent_to_Cano** files sent to Angela Cano who's analysing the data and leading the project  \n",
        "**newfilename** new file name assigned to the sample. This is to annotate provider, dataset, species and voucher in the name. See Appendix for the code used to name files  \n",
        "**raw_reads** number of raw reads before processing the data  -- filled as analysis progresses  \n",
        "**reads_trimed_paired_flag** if trimmed -- filled as analysis progresses  \n",
        "**reads_trimed_paired** number of reads after trimming -- filled as analysis progresses  \n",
        "**reads_trimed_single** number of unpaired reads after trimming -- filled as analysis progresses  \n",
        "**percentage_lost_tosingles** percentage of unpaired reads from the total of reads -- filled as analysis progresses  \n",
        "**pcr_filtered** if PCR duplicate reads were filtered out using samtools or picard-- filled as analysis progresses  \n",
        "**collection_year** year that sample was collected  \n",
        "**Continent** continent. The project involves species from America but the metadata keeps track of other data from other continents  \n",
        "**country** which country was the sample collected at if known from collection or voucher  \n",
        "**long** longitude coordinates if available, should be standardised to decimals  \n",
        "**lat**  latitude coordinates if available, should be standardised to decimals  \n",
        "**flag** single word descriptors separated by spaces (never colons, semicolons or tabs) that describe the flaws of the file. e.g. misingtaxspecies  \n",
        "**notes** elaborated notes if needed but avoid using commas, semicolons or tabs.  "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2fhs6GZ4qFMx"
      },
      "source": [
        "## 2. Filtering\n",
        "\n",
        "The second interesting thing to do is filtering unique species for which we have data. The files here represent species that might or might not be good, but at this point of the analysis (which will be on a sequence by sequence basis) we can only know what we have sequenced.\n",
        "\n",
        "Another thing is to filter out species that are not in America."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-gE-Ez1qtyIA",
        "outputId": "cb76b188-a2d9-4a30-f63e-24ae6ec202e5"
      },
      "source": [
        "df['Continent'].unique()"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['americas', 'out_americas', nan], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1GQlEP_TiZ4u"
      },
      "source": [
        "Some species don't have a continent specified. Let's see which ones:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Y9MDwf5NioT4",
        "outputId": "1a759199-0a7a-4bf9-8daf-f82caa86d110"
      },
      "source": [
        "# filtering pandas dataframe\n",
        "df[df['Continent'].isna()]['taxspecies'].unique()"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Ceroxylon_alpinum', 'Ceroxylon_ceriferum',\n",
              "       'Ceroxylon_echinulatum', 'Ceroxylon_parvifrons',\n",
              "       'Ceroxylon_parvum', 'Ceroxylon_quindiuense',\n",
              "       'Ceroxylon_ventricosum', 'Ceroxylon_vogelianum', 'Chamaedorea',\n",
              "       'Cocos', nan, 'Phytelephas_aecuatorialis',\n",
              "       'Phytelephas_macrocarpa', 'Phytelephas', 'Phytelephas_tumacana',\n",
              "       'Ravenea_sambiranensis', 'Chamaedorea_nationsiana',\n",
              "       'Chamaedorea_arenbergiana', 'Chamaedorea_binderi',\n",
              "       'Chamaedorea_brachyclada', 'Chamaedorea_seifrizii',\n",
              "       'Chamaedorea_fractiflexa', 'Chamaedorea_graminifolia',\n",
              "       'Chamaedorea_hodelii', 'Chamaedorea_ibarrae',\n",
              "       'Chamaedorea_incrustata', 'Chamaedorea_keelerorum',\n",
              "       'Chamaedorea_lehmannii', 'Chamaedorea_liebmannii',\n",
              "       'Chamaedorea_pachenoana', 'Chamaedorea_parvisecta',\n",
              "       'Chamaedorea_pauciflora', 'Chamaedorea_piscifolia',\n",
              "       'Chamaedorea_pumila', 'Chamaedorea_costaricana',\n",
              "       'Chamaedorea_rigida', 'Chamaedorea_schideana',\n",
              "       'Chamaedorea_simplex', 'Chamaedorea_skutchii',\n",
              "       'Chamaedorea_stenocarpa', 'Chamaedorea_subjectifolia',\n",
              "       'Chamaedorea_tenerrima', 'Chamaedorea_verapazensis',\n",
              "       'Chamaedorea_volcanensis', 'Chamaedorea_vulgata',\n",
              "       'Chamaedorea_zamorae'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3d2HoUqji2WB",
        "outputId": "075cd2c7-15fe-4eee-abf4-5059083e1a6e"
      },
      "source": [
        "# using a diferent syntax that can help with readability\n",
        "df[df.Continent.isna()]['taxspecies'].unique()"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Ceroxylon_alpinum', 'Ceroxylon_ceriferum',\n",
              "       'Ceroxylon_echinulatum', 'Ceroxylon_parvifrons',\n",
              "       'Ceroxylon_parvum', 'Ceroxylon_quindiuense',\n",
              "       'Ceroxylon_ventricosum', 'Ceroxylon_vogelianum', 'Chamaedorea',\n",
              "       'Cocos', nan, 'Phytelephas_aecuatorialis',\n",
              "       'Phytelephas_macrocarpa', 'Phytelephas', 'Phytelephas_tumacana',\n",
              "       'Ravenea_sambiranensis', 'Chamaedorea_nationsiana',\n",
              "       'Chamaedorea_arenbergiana', 'Chamaedorea_binderi',\n",
              "       'Chamaedorea_brachyclada', 'Chamaedorea_seifrizii',\n",
              "       'Chamaedorea_fractiflexa', 'Chamaedorea_graminifolia',\n",
              "       'Chamaedorea_hodelii', 'Chamaedorea_ibarrae',\n",
              "       'Chamaedorea_incrustata', 'Chamaedorea_keelerorum',\n",
              "       'Chamaedorea_lehmannii', 'Chamaedorea_liebmannii',\n",
              "       'Chamaedorea_pachenoana', 'Chamaedorea_parvisecta',\n",
              "       'Chamaedorea_pauciflora', 'Chamaedorea_piscifolia',\n",
              "       'Chamaedorea_pumila', 'Chamaedorea_costaricana',\n",
              "       'Chamaedorea_rigida', 'Chamaedorea_schideana',\n",
              "       'Chamaedorea_simplex', 'Chamaedorea_skutchii',\n",
              "       'Chamaedorea_stenocarpa', 'Chamaedorea_subjectifolia',\n",
              "       'Chamaedorea_tenerrima', 'Chamaedorea_verapazensis',\n",
              "       'Chamaedorea_volcanensis', 'Chamaedorea_vulgata',\n",
              "       'Chamaedorea_zamorae'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ag5uTCqAjIIi",
        "outputId": "d0ceeb43-299d-4c6b-b0ae-e2772c527d69"
      },
      "source": [
        "df[df.Continent.isna()]['taxgenus'].unique()"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Ceroxylon', 'Chamaedorea', 'Cocos', nan, 'Phytelephas', 'Ravenea'],\n",
              "      dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IQmksWQZjDGx"
      },
      "source": [
        "Most species are from the genera *Ceroxylon*, *Chamaedorea*, *Cocos*, *Phytelephas*, and *Ravenea*. I can already remember the dataset they come from and I can tell they don't have a continent assigned because I haven't updated the dataframe (as is not my main priority)\n",
        "\n",
        "we can totally fix that today though. Let's iterate through rows in the dataset to assign the continent information. We also know all but *Cocos* (which is cosmopolita) are from America."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WIpoqq1ojBWq",
        "outputId": "8ba2aeea-86cd-41f8-8830-94537e94e3eb"
      },
      "source": [
        "# itertuples iterates through tuples (pairs) of index,row and different columns can be acceced using the column label. Similar to the iterrows() method but faster.\n",
        "# itertuples() only allows for the \"dot\" syntax for filtering pandas columns.\n",
        "\n",
        "for row in df[df['Continent'].isna()].itertuples():\n",
        "  # print(row.Continent,'empty') # test\n",
        "  df.loc[row.Index,'Continent'] = 'americas' # loc assigns a value in a cell with coordinates [row, column] labels\n",
        "df['Continent'].unique() # no nan should appear"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['americas', 'out_americas'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZDgD093GuFbP"
      },
      "source": [
        "Nice. Now we can filter all species in America and count/list them"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sljrTqz4uLuO",
        "outputId": "a4ba61ab-117b-40c6-f50d-75cf4ec0c11b"
      },
      "source": [
        "df[df['Continent'] == 'americas']['taxspecies'].unique()"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Acrocomia_aculeata', 'Acrocomia_crispa', 'Acrocomia_emensis',\n",
              "       'Acrocomia_glaucescens', 'Acrocomia_intumescens',\n",
              "       'Acrocomia_totai', 'Aiphanes_acaulis', 'Aiphanes_buenaventurae',\n",
              "       'Aiphanes_concinna', 'Aiphanes_erinacea', 'Aiphanes_gelatinosa',\n",
              "       'Aiphanes_hirsuta', 'Aiphanes_horrida', 'Aiphanes_killipii',\n",
              "       'Aiphanes_leiostachys', 'Aiphanes_lindeniana', 'Aiphanes_linearis',\n",
              "       'Aiphanes_macroloba', 'Aiphanes_minima', 'Aiphanes_parvifolia',\n",
              "       'Aiphanes_pilaris', 'Aiphanes_simplex', 'Aiphanes_tricuspidata',\n",
              "       'Aiphanes_ulei', 'Allagoptera_arenaria', 'Allagoptera_brevicalyx',\n",
              "       'Allagoptera_caudescens', 'Allagoptera_leucocalyx',\n",
              "       'Allagoptera_pectinata', 'Ammandra_decasperma', 'Aphandra_natalia',\n",
              "       'Astrocaryum_acaule', 'Astrocaryum_aculeatum',\n",
              "       'Astrocaryum_alatum', 'Astrocaryum_campestre',\n",
              "       'Astrocaryum_carnosum', 'Astrocaryum_chambira',\n",
              "       'Astrocaryum_chonta', 'Astrocaryum_ciliatum',\n",
              "       'Astrocaryum_echinatum', 'Astrocaryum_faranae',\n",
              "       'Astrocaryum_ferrugineum', 'Astrocaryum_gratum',\n",
              "       'Astrocaryum_gynacanthum', 'Astrocaryum_huicungo',\n",
              "       'Astrocaryum_javarense', 'Astrocaryum_macrocalyx',\n",
              "       'Astrocaryum_malybo', 'Asterogyne_martiana',\n",
              "       'Astrocaryum_mexicanum', 'Astrocaryum_murumuru',\n",
              "       'Astrocaryum_perangustatum', 'Astrocaryum_sciophilum',\n",
              "       'Astrocaryum_scopatum', 'Astrocaryum_standleyanum',\n",
              "       'Astrocaryum_triandrum', 'Astrocaryum_ulei',\n",
              "       'Astrocaryum_urostachys', 'Attalea_allenii', 'Attalea_amygdalina',\n",
              "       'Attalea_anisitsiana', 'Attalea_apoda', 'Attalea_attaleoides',\n",
              "       'Attalea_barreirensis', 'Attalea_brasiliensis',\n",
              "       'Attalea_brejinhoensis', 'Attalea_burretiana', 'Attalea_butyracea',\n",
              "       'Attalea_cohune', 'Attalea_colenda', 'Attalea_crassispatha',\n",
              "       'Attalea_cuatrecasana', 'Attalea_cuatrecasa', 'Attalea_dubia',\n",
              "       'Attalea_eichleri', 'Attalea_funifera', 'Attalea_gearensis',\n",
              "       'Attalea_geraensis', 'Attalea_guacuyule', 'Attalea_humilis',\n",
              "       'Attalea_iguadummat', 'Attalea_insignis', 'Attalea_leandroana',\n",
              "       'Attalea_luetzelburgii', 'Attalea_maripa', 'Attalea_microcarpa',\n",
              "       'Attalea_moorei', 'Attalea_nucifera', 'Attalea_oleifera',\n",
              "       'Attalea_phalerata', 'Attalea_pindobassu', 'Attalea_plowmanii',\n",
              "       'Attalea_princeps', 'Attalea_sagotii', 'Attalea_seabrensis',\n",
              "       'Attalea_spectabilis', 'Attalea_speciosa', 'Attalea_tessmannii',\n",
              "       'Attalea_vitrivir', 'Bactris_barronis', 'Bactris_bidentula',\n",
              "       'Bactris_brongniartii', 'Bactris_caryotifolia',\n",
              "       'Bactris_charnleyae', 'Bactris_coloniata', 'Bactris_coloradonis',\n",
              "       'Bactris_concinna', 'Bactris_elegans', 'Bactris_fissifrons',\n",
              "       'Bactris_gasipaes', 'Bactris_glandulosa', 'Bactris_glaucescens',\n",
              "       'Bactris_gracilior', 'Bactris_hirta', 'Bactris_killipii',\n",
              "       'Bactris_macroacantha', 'Bactris_major', 'Bactris_manriquei',\n",
              "       'Bactris_maraja', 'Bactris_mexicana', 'Bactris_militaris',\n",
              "       'Bactris_neomilitaris', 'Bactris_obovata', 'Bactris_panamensis',\n",
              "       'Bactris_pickellii', 'Bactris_pilosa', 'Bactris_plumeriana',\n",
              "       'Bactris_setulosa', 'Bactris_setosa', 'Bactris_simplicifrons',\n",
              "       'Barcella_odora', 'Brahea_aculeata', 'Brahea_armata',\n",
              "       'Brahea_bella', 'Brahea_dulcis', 'Brahea_nitida',\n",
              "       'Brahea_sarukhanii', 'Butia_archeri', 'Butia_campicola',\n",
              "       'Butia_capitata', 'Butia_catharinensis', 'Butia_eriospatha',\n",
              "       'Butia_exospadix', 'Butia_lallemantii', 'Butia_leiospatha',\n",
              "       'Butia_lepidotispatha', 'Butia_leptospatha', 'Butia_marmori',\n",
              "       'Butia_matogrossensis', 'Butia_microspadix', 'Butia_noblickii',\n",
              "       'Butia_odorata', 'Butia_paraguayensis', 'Butia_pubispatha',\n",
              "       'Butia_purpurascens', 'Butia_witecki', 'Butia_yatay',\n",
              "       'Calyptrogyne_allenii', 'Calyptrogyne_anomala',\n",
              "       'Calyptrogyne_costatifrons', 'Calyptrogyne_ghiesbreghtiana',\n",
              "       'Calyptronoma_occidentalis', 'Calyptrogyne_panamensis',\n",
              "       'Calyptronoma_rivalis', 'Calyptrogyne_tutensis',\n",
              "       'Ceroxylon_alpinum', 'Ceroxylon_amazonicum', 'Ceroxylon_ceriferum',\n",
              "       'Ceroxylon_echinulatum', 'Ceroxylon_parvifrons',\n",
              "       'Ceroxylon_parvum', 'Ceroxylon_peruvianum',\n",
              "       'Ceroxylon_pityrophyllum', 'Ceroxylon_quindiuense',\n",
              "       'Ceroxylon_sasaimae', 'Ceroxylon_ventricosum',\n",
              "       'Ceroxylon_vogelianum', 'Chamaedorea_adscendens',\n",
              "       'Chamaedorea_allenii', 'Chamaedorea_alternans',\n",
              "       'Chamaedorea_anemophila', 'Chamaedorea_arenbergiana',\n",
              "       'Chamaedorea_benziei', 'Chamaedorea_brachypoda',\n",
              "       'Chamaedorea_carchensis', 'Chamaedorea_cataractarum',\n",
              "       'Chamaedorea_correae', 'Chamaedorea_costaricana',\n",
              "       'Chamaedorea_crucensis', 'Chamaedorea_dammeriana',\n",
              "       'Chamaedorea_deckeriana', 'Chamaedorea_deneversiana',\n",
              "       'Chamaedorea_elatior', 'Chamaedorea_elegans',\n",
              "       'Chamaedorea_ernesti-augustii', 'Chamaedorea_falcifera',\n",
              "       'Chamaedorea_fragrans', 'Chamaedorea_frondosa',\n",
              "       'Chamaedorea_geonomiformis', 'Chamaedorea_glaucifolia',\n",
              "       'Chamaedorea_graminifolia', 'Chamaedorea_guntheriana',\n",
              "       'Chamaedorea_hooperiana', 'Chamaedorea_klotzschiana',\n",
              "       'Chamaedorea_linearis', 'Chamaedorea_lucidifrons',\n",
              "       'Chamaedorea_macrospadix', 'Chamaedorea_matae',\n",
              "       'Chamaedorea_metallica', 'Chamaedorea_microspadix',\n",
              "       'Chamaedorea_microphylla', 'Chamaedorea_murriensis',\n",
              "       'Chamaedorea_nationsiana', 'Chamaedorea_neurochlamys',\n",
              "       'Chamaedorea_nubium', 'Chamaedorea_oblongata',\n",
              "       'Chamaedorea_oreophila', 'Chamaedorea_palmeriana',\n",
              "       'Chamaedorea_parvifolia', 'Chamaedorea_pedunculata',\n",
              "       'Chamaedorea_pinnatifrons', 'Chamaedorea_pittieri',\n",
              "       'Chamaedorea_plumosa', 'Chamaedorea_pochutlensis',\n",
              "       'Chamaedorea_ponderosa', 'Chamaedorea_pygmaea',\n",
              "       'Chamaedorea_quezalteca', 'Chamaedorea_radicalis',\n",
              "       'Chamaedorea_recurvata', 'Chamaedorea_rhizomatosa',\n",
              "       'Chamaedorea_ricardoi', 'Chamaedorea_robertii',\n",
              "       'Chamaedorea_rojasiana', 'Chamaedorea_rossteniorum',\n",
              "       'Chamaedorea_sartorii', 'Chamaedorea_scheryi',\n",
              "       'Chamaedorea_schippii', 'Chamaedorea_schiediana',\n",
              "       'Chamaedorea_seifrizii', 'Chamaedorea_spp', 'Chamaedorea',\n",
              "       'Chamaedorea_stolonifera', 'Chamaedorea_stricta',\n",
              "       'Chamaedorea_sullivaniorum', 'Chamaedorea_tenella',\n",
              "       'Chamaedorea_tepejilote', 'Chamaedorea_tuerckheimii',\n",
              "       'Chamaedorea_warscewiczii', 'Chamaedorea_verecunda',\n",
              "       'Chamaedorea_woodsoniana', 'Chamaedorea_zamorae',\n",
              "       'Chelyocarpus_chuco', 'Chelyocarpus_dianeurus',\n",
              "       'Chelyocarpus_repens', 'Chelyocarpus_ulei',\n",
              "       'Coccothrinax_alexandria', 'Coccothrinax_alta',\n",
              "       'Coccothrinax_argentata', 'Coccothrinax_argentea',\n",
              "       'Coccothrinax_barbadensis', 'Coccothrinax_borhidiana',\n",
              "       'Coccothrinax_clarensis', 'Coccothrinax_crinita',\n",
              "       'Coccothrinax_cupularis', 'Coccothrinax_ekmanii',\n",
              "       'Coccothrinax_elegans', 'Coccothrinax_fragrans',\n",
              "       'Coccothrinax_gracilis', 'Coccothrinax_guantanamensis',\n",
              "       'Coccothrinax_gundlachii', 'Coccothrinax_hioramii',\n",
              "       'Coccothrinax_inaguensis', 'Coccothrinax_jamaicensis',\n",
              "       'Coccothrinax_litoralis', 'Coccothrinax_macroglossa',\n",
              "       'Coccothrinax_miraguama', 'Coccothrinax_montana',\n",
              "       'Coccothrinax_munizii', 'Coccothrinax_orientalis',\n",
              "       'Coccothrinax_proctorii', 'Coccothrinax_readii',\n",
              "       'Coccothrinax_salvatoris', 'Coccothrinax_saxicola',\n",
              "       'Coccothrinax_spissa', 'Cocos', 'Coccothrinax_torrida',\n",
              "       'Colpothrinax_aphanopetala', 'Colpothrinax_wrightii',\n",
              "       'Copernicia_alba', 'Copernicia_baileyana', 'Copernicia_berteroana',\n",
              "       'Copernicia_ekmanii', 'Copernicia_fallaensis', 'Copernicia_gigas',\n",
              "       'Copernicia_glabrescens', 'Copernicia_hospita',\n",
              "       'Copernicia_macroglossa', 'Copernicia_prunifera',\n",
              "       'Copernicia_rigida', 'Copernicia_tectorum', 'Copernicia_yarey',\n",
              "       'Cryosophila_bartlettii', 'Cryosophila_guagara',\n",
              "       'Cryosophila_kalbreyeri', 'Cryosophila_nana',\n",
              "       'Cryosophila_stauracantha', 'Cryosophila_warscewiczii',\n",
              "       'Cryosophila_williamsii', 'Desmoncus_mitis',\n",
              "       'Desmoncus_orthacanthos', 'Desmoncus_polyacanthos',\n",
              "       'Dictyocaryum_lamarckianum', 'Elaeis_oleifera',\n",
              "       'Euterpe_broadwayi', 'Euterpe_catinga', 'Euterpe_edulis',\n",
              "       'Euterpe_espiritosantensis', 'Euterpe_longivaginata',\n",
              "       'Euterpe_oleracea', 'Euterpe_precatoria', 'Gaussia_attenuata',\n",
              "       'Gaussia_gomez-pompae', 'Gaussia_maya', 'Gaussia_princeps',\n",
              "       'Gaussia_spirituana', 'Geonoma_baculifera', 'Geonoma_brongniartii',\n",
              "       'Geonoma_calyptrogynoidea', 'Geonoma_camana',\n",
              "       'Geonoma_chlamydostachys', 'Geonoma_concinna', 'Geonoma_congesta',\n",
              "       'Geonoma_cuneata', 'Geonoma_deversa', 'Geonoma_elegans',\n",
              "       'Geonoma_fosteri', 'Geonoma_frontinensis', 'Geonoma_galeanoae',\n",
              "       'Geonoma_hugonis', 'Geonoma_interrupta', 'Geonoma_lehmannii',\n",
              "       'Geonoma_longivaginata', 'Geonoma_macrostachys', 'Geonoma_maxima',\n",
              "       'Geonoma_monospatha', 'Geonoma_mooreana', 'Geonoma_multisecta',\n",
              "       'Geonoma_occidentalis', 'Geonoma_orbignyana', 'Geonoma_paradoxa',\n",
              "       'Geonoma_pauciflora', 'Geonoma_pohliana', 'Geonoma_procumbens',\n",
              "       'Geonoma_rubescens', 'Geonoma_santanderensis',\n",
              "       'Geonoma_schottiana', 'Geonoma_stricta', 'Geonoma_triandra',\n",
              "       'Geonoma_undata', 'Hemithrinax_ekmaniana', 'Hyospathe_elegans',\n",
              "       'Hyospathe_pittieri', 'Iriartea_deltoidea', 'Itaya_amicorum',\n",
              "       'Jubaea_chilensis', 'Leopoldinia_pulchra', 'Lepidocaryum_tenue',\n",
              "       'Leucothrinax_morrisii', 'Manicaria_saccifera',\n",
              "       'Mauritiella_aculeata', 'Mauritiella_armata', 'Mauritia_Carana',\n",
              "       'Mauritiella_disticha', 'Mauritia_flexuosa',\n",
              "       'Mauritiella_macroclada', 'Mauritiella_pumila', nan,\n",
              "       'Neonicholsonia_watsonii', 'Oenocarpus_balickii',\n",
              "       'Oenocarpus_bataua', 'Oenocarpus_distichus', 'Oenocarpus_mapora',\n",
              "       'Oenocarpus_minor', 'Pholidostachys_dactyloides',\n",
              "       'Pholidostachys_kalbreyeri', 'Pholidostachys_panamensis',\n",
              "       'Pholidostachys_pulchra', 'Pholidostachys_sanluisensis',\n",
              "       'Phytelephas_aecuatorialis', 'Phytelephas_aequatorialis',\n",
              "       'Phytelephas_macrocarpa', 'Phytelephas', 'Phytelephas_tenuicaulis',\n",
              "       'Phytelephas_tumacana', 'Prestoea_acuminata', 'Prestoea_decurrens',\n",
              "       'Prestoea_ensiformis', 'Prestoea_longepetiolata',\n",
              "       'Prestoea_pubens', 'Prestoea_schultzeana', 'Pseudophoenix_ekmanii',\n",
              "       'Pseudophoenix_lediniana', 'Pseudophoenix_sargentii',\n",
              "       'Pseudophoenix_vinifera', 'Raphia_taedigera',\n",
              "       'Ravenea_sambiranensis', 'Reinhardtia_gracilis',\n",
              "       'Reinhardtia_koschnyana', 'Reinhardtia_simplex',\n",
              "       'Rhapidophyllum_hystrix', 'Roystonea_altissima',\n",
              "       'Roystonea_borinquena', 'Roystonea_lenis', 'Roystonea_maisiana',\n",
              "       'Roystonea_oleracea', 'Roystonea_regia', 'Sabal_bermudana',\n",
              "       'Sabal_causiarum', 'Sabal_dominguensis', 'Sabal_etonia',\n",
              "       'Sabal_guatemalensis', 'Sabinaria_magnifica', 'Sabal_maritima',\n",
              "       'Sabal_mauritiiformis', 'Sabal_mexicana', 'Sabal_minor',\n",
              "       'Sabal_palmetto', 'Sabal_pumos', 'Sabal_rosei', 'Sabal_uresana',\n",
              "       'Sabal_yapa', 'Schippia_concolor', 'Serenoa_repens',\n",
              "       'Socratea_exhorriza', 'Socratea_exorrhiza',\n",
              "       'Syagrus_allagopteroides', 'Syagrus_amara', 'Syagrus_angustifolia',\n",
              "       'Syagrus_botryophora', 'Syagrus_caerulescens',\n",
              "       'Syagrus_campylospatha', 'Syagrus_campestris',\n",
              "       'Syagrus_cardenasii', 'Syagrus_cerqueirana', 'Syagrus_cocoides',\n",
              "       'Syagrus_comosa', 'Syagrus_coronata', 'Syagrus_deflexa',\n",
              "       'Syagrus_duartei', 'Syagrus_evansiana', 'Syagrus_flexuosa',\n",
              "       'Syagrus_glaucescens', 'Syagrus_glazioviana', 'Syagrus_gouveiana',\n",
              "       'Syagrus_graminifolia', 'Syagrus_guimaraensis', 'Syagrus_harleyi',\n",
              "       'Syagrus_hoehnei', 'Syagrus_inajai', 'Syagrus_insignis',\n",
              "       'Syagrus_itapebiensis', 'Syagrus_itacambirana', 'Syagrus_kellyana',\n",
              "       'Syagrus_liliputiana', 'Syagrus_loefgrenii',\n",
              "       'Syagrus_longipedunculata', 'Syagrus_lorenzoniorum',\n",
              "       'Syagrus_macrocarpa', 'Syagrus_menzeliana', 'Syagrus_mendanhensis',\n",
              "       'Syagrus_microphylla', 'Syagrus_minor', 'Syagrus_oleracea',\n",
              "       'Syagrus_orinocensis', 'Syagrus_petraea', 'Syagrus_picrophylla',\n",
              "       'Syagrus_pleiocladoides', 'Syagrus_procumbens',\n",
              "       'Syagrus_pseudococos', 'Syagrus_romanzoffiana', 'Syagrus_rupicola',\n",
              "       'Syagrus_ruschiana', 'Syagrus_santosii', 'Syagrus_sancona',\n",
              "       'Syagrus_schizophylla', 'Syagrus_smithii', 'Syagrus_smith',\n",
              "       'Syagrus_stenopetala', 'Syagrus_stratincola', 'Syagrus_vagans',\n",
              "       'Syagrus_weddelianum', 'Syagrus_werdermanii',\n",
              "       'Syagrus_vermicularis', 'Syagrus_yungascensis',\n",
              "       'Synechanthus_fibrosus', 'Synechanthus_warscewiczianus',\n",
              "       'Thrinax_parviflora', 'Thrinax_radiata', 'Trithrinax_brasiliensis',\n",
              "       'Trithrinax_campestris', 'Trithrinax_schizophylla',\n",
              "       'Washingtonia_filifera', 'Welfia_regia', 'Wendlandiella_gracilis',\n",
              "       'Wettinia_augusta', 'Wettinia_aequalis', 'Zombia_antillarum',\n",
              "       'Chamaedorea_binderi', 'Chamaedorea_brachyclada',\n",
              "       'Chamaedorea_fractiflexa', 'Chamaedorea_hodelii',\n",
              "       'Chamaedorea_ibarrae', 'Chamaedorea_incrustata',\n",
              "       'Chamaedorea_keelerorum', 'Chamaedorea_lehmannii',\n",
              "       'Chamaedorea_liebmannii', 'Chamaedorea_pachenoana',\n",
              "       'Chamaedorea_parvisecta', 'Chamaedorea_pauciflora',\n",
              "       'Chamaedorea_piscifolia', 'Chamaedorea_pumila',\n",
              "       'Chamaedorea_rigida', 'Chamaedorea_schideana',\n",
              "       'Chamaedorea_simplex', 'Chamaedorea_skutchii',\n",
              "       'Chamaedorea_stenocarpa', 'Chamaedorea_subjectifolia',\n",
              "       'Chamaedorea_tenerrima', 'Chamaedorea_verapazensis',\n",
              "       'Chamaedorea_volcanensis', 'Chamaedorea_vulgata'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Z9NZthOoN0is"
      },
      "source": [
        "## 3. Counting"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rG4SjP7_-dzt",
        "outputId": "ee0fdfed-0a07-4727-8025-eec2b2ca2bb6"
      },
      "source": [
        "print('Number of american palm species sequenced:')\n",
        "len(df[df['Continent'] == 'americas']['taxspecies'].unique())"
      ],
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Number of american palm species sequenced:\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "531"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nJYca_hk_WmU"
      },
      "source": [
        "We need to know how many species are American species. I initially used this list passed on to me (but it is something to check out more formaly). The Kew checklist, for example, doesn't include some Sabal species for which we have sequences. There has to be a consensus about which list to compare against."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KxW-Tm5Q_kEW",
        "outputId": "9be9b4b6-fcde-4316-eaf3-bbf3ee874bf9"
      },
      "source": [
        "palms_americas=['Schippia_concolor','Pseudophoenix_ekmanii','Pseudophoenix_lediniana','Pseudophoenix_sargentii','Pseudophoenix_vinifera','Acoelorraphe_wrightii','Pholidostachys_amazonensis','Pholidostachys_dactyloides','Pholidostachys_kalbreyeri','Pholidostachys_occidentalis','Pholidostachys_panamensis','Pholidostachys_pulchra','Pholidostachys_sanluisensis','Pholidostachys_synanthera','Lepidocaryum_tenue','Hyospathe_elegans','Hyospathe_frontinensis','Hyospathe_macrorhachis','Hyospathe_peruviana','Hyospathe_pittieri','Hyospathe_wendlandiana','Chelyocarpus_chuco','Chelyocarpus_dianeurus','Chelyocarpus_repens','Chelyocarpus_ulei','Acrocomia_aculeata','Acrocomia_crispa','Acrocomia_emensis','Acrocomia_glaucescens','Acrocomia_hassleri','Acrocomia_intumescens','Acrocomia_media','Acrocomia_totai','Aiphanes_acanthophylla','Aiphanes_acaulis','Aiphanes_bicornis','Aiphanes_buenaventurae','Aiphanes_chiribogensis','Aiphanes_concinna','Aiphanes_deltoidea','Aiphanes_duquei','Aiphanes_eggersii','Aiphanes_erinacea','Aiphanes_gelatinosa','Aiphanes_graminifolia','Aiphanes_grandis','Aiphanes_hirsuta','Aiphanes_horrida','Aiphanes_killipii','Aiphanes_leiostachys','Aiphanes_lindeniana','Aiphanes_linearis','Aiphanes_macroloba','Aiphanes_minima','Aiphanes_multiplex','Aiphanes_parvifolia','Aiphanes_pilaris','Aiphanes_simplex','Aiphanes_spicata','Aiphanes_stergiosii','Aiphanes_tricuspidata','Aiphanes_ulei','Aiphanes_verrucosa','Aiphanes_weberbaueri','Allagoptera_arenaria','Allagoptera_brevicalyx','Allagoptera_campestris','Allagoptera_caudescens','Allagoptera_leucocalyx','Ammandra_decasperma','Aphandra_natalia','Asterogyne_guianensis','Asterogyne_martiana','Asterogyne_ramosa','Asterogyne_spicata','Asterogyne_yaracuyense','Astrocaryum_acaule','Astrocaryum_aculeatissimum','Astrocaryum_aculeatum','Astrocaryum_alatum','Astrocaryum_campestre','Astrocaryum_carnosum','Astrocaryum_chambira','Astrocaryum_chonta','Astrocaryum_ciliatum','Astrocaryum_confertum','Astrocaryum_cuatrecasanum','Astrocaryum_echinatum','Astrocaryum_faranae','Astrocaryum_farinosum','Astrocaryum_ferrugineum','Astrocaryum_giganteum','Astrocaryum_gratum','Astrocaryum_gynacanthum','Astrocaryum_huaimi','Astrocaryum_huicungo','Astrocaryum_jauari','Astrocaryum_javarense','Astrocaryum_macrocalyx','Astrocaryum_malybo','Astrocaryum_mexicanum','Astrocaryum_minus','Astrocaryum_murumuru','Astrocaryum_paramaca','Astrocaryum_perangustatum','Astrocaryum_rodriguesii','Astrocaryum_sciophilum','Astrocaryum_scopatum','Astrocaryum_sociale','Astrocaryum_standleyanum','Astrocaryum_triandrum','Astrocaryum_tucuma','Astrocaryum_ulei','Astrocaryum_urostachys','Astrocaryum_vulgare','Attalea_allenii','Attalea_amygdalina','Attalea_amylacea','Attalea_anisitsiana','Attalea_apoda','Attalea_attaleoides','Attalea_barreirensis','Attalea_bassleriana','Attalea_blepharopus','Attalea_brasiliensis','Attalea_brejinhoensis','Attalea_burretiana','Attalea_butyracea','Attalea_camopiensis','Attalea_cephalotus','Attalea_cohune','Attalea_colenda','Attalea_compta','Attalea_crassispatha','Attalea_cuatrecasana','Attalea_dahlgreniana','Attalea_degranvillei','Attalea_dubia','Attalea_eichleri','Attalea_exigua','Attalea_fairchildensis','Attalea_funifera','Attalea_geraensis','Attalea_guacuyule','Attalea_guianensis','Attalea_hoehnei','Attalea_huebneri','Attalea_humilis','Attalea_iguadummat','Attalea_insignis','Attalea_kewensis','Attalea_lauromuelleriana','Attalea_leandroana','Attalea_luetzelburgii','Attalea_macrolepis','Attalea_magdalenica','Attalea_maracaibensis','Attalea_maripa','Attalea_maripensis','Attalea_microcarpa','Attalea_moorei','Attalea_nucifera','Attalea_oleifera','Attalea_osmantha','Attalea_peruviana','Attalea_phalerata','Attalea_pindobassu','Attalea_plowmanii','Attalea_princeps','Attalea_racemosa','Attalea_rhynchocarpa','Attalea_rostrata','Attalea_salazarii','Attalea_salvadorensis','Attalea_seabrensis','Attalea_septuagenata','Attalea_speciosa','Attalea_spectabilis','Attalea_tessmannii','Attalea_vitrivir','Attalea_weberbaueri','Attalea_wesselsboeri','Bactris_acanthocarpa','Bactris_acanthocarpoides','Bactris_ana-juliae','Bactris_aubletiana','Bactris_bahiensis','Bactris_balanophora','Bactris_barronis','Bactris_bidentula','Bactris_bifida','Bactris_brongniartii','Bactris_campestris','Bactris_caryotifolia','Bactris_caudata','Bactris_charnleyae','Bactris_chaveziae','Bactris_coloniata','Bactris_coloradonis','Bactris_concinna','Bactris_constanciae','Bactris_corossilla','Bactris_cubensis','Bactris_cuspidata','Bactris_dianeura','Bactris_elegans','Bactris_faucium','Bactris_ferruginea','Bactris_fissifrons','Bactris_gasipaes','Bactris_gastoniana','Bactris_glandulosa','Bactris_glassmanii','Bactris_glaucescens','Bactris_gracilior','Bactris_grayumii','Bactris_guineensis','Bactris_halmoorei','Bactris_hatschbachii','Bactris_herrerana','Bactris_hirta','Bactris_hondurensis','Bactris_horridispatha','Bactris_jamaicana','Bactris_killipii','Bactris_kunorum','Bactris_longiseta','Bactris_macroacantha','Bactris_major','Bactris_maraja','Bactris_martiana','Bactris_mexicana','Bactris_militaris','Bactris_nancibaensis','Bactris_obovata','Bactris_oligocarpa','Bactris_oligoclada','Bactris_panamensis','Bactris_pickelii','Bactris_pilosa','Bactris_pliniana','Bactris_plumeriana','Bactris_polystachya','Bactris_ptariana','Bactris_rhaphidacantha','Bactris_riparia','Bactris_rostrata','Bactris_schultesii','Bactris_setiflora','Bactris_setosa','Bactris_setulosa','Bactris_simplicifrons','Bactris_soeiroana','Bactris_sphaerocarpa','Bactris_syagroides','Bactris_tefensis','Bactris_timbuiensis','Bactris_tomentosa','Bactris_turbinocarpa','Bactris_vulgaris','Brahea_aculeata','Brahea_armata','Brahea_brandegeei','Brahea_calcarea','Brahea_decumbens','Brahea_dulcis','Brahea_edulis','Brahea_moorei','Brahea_pimo','Brahea_salvadorensis','Brahea_sarukhanii','Butia_archeri','Butia_campicola','Butia_capitata','Butia_catarinensis','Butia_eriospatha','Butia_exilata','Butia_exospadix','Butia_lallemantii','Butia_leptospatha','Butia_marmorii','Butia_matogrossensis','Butia_microspadix','Butia_missionera','Butia_noblickii','Butia_odorata','Butia_paraguayensis','Butia_purpurascens','Butia_quaraimana','Butia_stolonifera','Butia_yatay','Calyptrogyne_allenii','Calyptrogyne_anomala','Calyptrogyne_baudensis','Calyptrogyne_coloradensis','Calyptrogyne_condensata','Calyptrogyne_costatifrons','Calyptrogyne_deneversii','Calyptrogyne_fortunensis','Calyptrogyne_ghiesbreghtiana','Calyptrogyne_herrerae','Calyptrogyne_kunorum','Calyptrogyne_osensis','Calyptrogyne_panamensis','Calyptrogyne_pubescens','Calyptrogyne_sanblasensis','Calyptrogyne_trichostachys','Calyptrogyne_tutensis','Calyptronoma_occidentalis','Calyptronoma_plumeriana','Calyptronoma_rivalis','Ceroxylon_alpinum','Ceroxylon_amazonicum','Ceroxylon_ceriferum','Ceroxylon_echinulatum','Ceroxylon_parvifrons','Ceroxylon_parvum','Ceroxylon_peruvianum','Ceroxylon_pityrophyllum','Ceroxylon_quindiuense','Ceroxylon_sasaimae','Ceroxylon_ventricosum','Ceroxylon_vogelianum','Chamaedorea_adscendens','Chamaedorea_allenii','Chamaedorea_alternans','Chamaedorea_amabilis','Chamaedorea_anemophila','Chamaedorea_angustisecta','Chamaedorea_arenbergiana','Chamaedorea_atrovirens','Chamaedorea_benziei','Chamaedorea_binderi','Chamaedorea_brachyclada','Chamaedorea_brachypoda','Chamaedorea_carchensis','Chamaedorea_castillo-montii','Chamaedorea_cataractarum','Chamaedorea_christinae','Chamaedorea_correae','Chamaedorea_costaricana','Chamaedorea_crucensis','Chamaedorea_dammeriana','Chamaedorea_deckeriana','Chamaedorea_deneversiana','Chamaedorea_elatior','Chamaedorea_elegans','Chamaedorea_ernesti-augusti','Chamaedorea_falcifera','Chamaedorea_foveata','Chamaedorea_fractiflexa','Chamaedorea_fragrans','Chamaedorea_frondosa','Chamaedorea_geonomiformis','Chamaedorea_glaucifolia','Chamaedorea_graminifolia','Chamaedorea_guntheriana','Chamaedorea_hodelii','Chamaedorea_hooperiana','Chamaedorea_ibarrae','Chamaedorea_incrustata','Chamaedorea_keelerorum','Chamaedorea_klotzschiana','Chamaedorea_latisecta','Chamaedorea_lehmannii','Chamaedorea_liebmannii','Chamaedorea_linearis','Chamaedorea_lucidifrons','Chamaedorea_macrospadix','Chamaedorea_matae','Chamaedorea_metallica','Chamaedorea_microphylla','Chamaedorea_microspadix','Chamaedorea_moliniana','Chamaedorea_murriensis','Chamaedorea_nationsiana','Chamaedorea_neurochlamys','Chamaedorea_nubium','Chamaedorea_oblongata','Chamaedorea_oreophila','Chamaedorea_pachecoana','Chamaedorea_palmeriana','Chamaedorea_parvifolia','Chamaedorea_parvisecta','Chamaedorea_pauciflora','Chamaedorea_pedunculata','Chamaedorea_pinnatifrons','Chamaedorea_piscifolia','Chamaedorea_pittieri','Chamaedorea_plumosa','Chamaedorea_pochutlensis','Chamaedorea_ponderosa','Chamaedorea_pumila','Chamaedorea_pygmaea','Chamaedorea_queroana','Chamaedorea_radicalis','Chamaedorea_recurvata','Chamaedorea_rhizomatosa','Chamaedorea_ricardoi','Chamaedorea_rigida','Chamaedorea_robertii','Chamaedorea_rojasiana','Chamaedorea_rosibeliae','Chamaedorea_rossteniorum','Chamaedorea_sartorii','Chamaedorea_scheryi','Chamaedorea_schiedeana','Chamaedorea_schippii','Chamaedorea_seifrizii','Chamaedorea_serpens','Chamaedorea_simplex','Chamaedorea_skutchii','Chamaedorea_smithii','Chamaedorea_stenocarpa','Chamaedorea_stolonifera','Chamaedorea_stricta','Chamaedorea_subjectifolia','Chamaedorea_tenerrima','Chamaedorea_tepejilote','Chamaedorea_tuerckheimii','Chamaedorea_undulatifolia','Chamaedorea_verapazensis','Chamaedorea_verecunda','Chamaedorea_volcanensis','Chamaedorea_vulgata','Chamaedorea_warscewiczii','Chamaedorea_whitelockiana','Chamaedorea_woodsoniana','Chamaedorea_zamorae','Coccothrinax_acunana','Coccothrinax_alexandri','Coccothrinax_argentata','Coccothrinax_argentea','Coccothrinax_baracoensis','Coccothrinax_barbadensis','Coccothrinax_bermudezii','Coccothrinax_borhidiana','Coccothrinax_boschiana','Coccothrinax_camagueyana','Coccothrinax_clarensis','Coccothrinax_concolor','Coccothrinax_crinita','Coccothrinax_cupularis','Coccothrinax_ekmanii','Coccothrinax_elegans','Coccothrinax_fagildei','Coccothrinax_fragrans','Coccothrinax_garciana','Coccothrinax_gracilis','Coccothrinax_guantanamensis','Coccothrinax_gundlachii','Coccothrinax_hioramii','Coccothrinax_inaguensis','Coccothrinax_jamaicensis','Coccothrinax_leonis','Coccothrinax_litoralis','Coccothrinax_macroglossa','Coccothrinax_microphylla','Coccothrinax_miraguama','Coccothrinax_moaensis','Coccothrinax_montana','Coccothrinax_munizii','Coccothrinax_muricata','Coccothrinax_nipensis','Coccothrinax_orientalis','Coccothrinax_pauciramosa','Coccothrinax_proctorii','Coccothrinax_pseudorigida','Coccothrinax_pumila','Coccothrinax_readii','Coccothrinax_rigida','Coccothrinax_salvatoris','Coccothrinax_saxicola','Coccothrinax_scoparia','Coccothrinax_spissa','Coccothrinax_torrida','Coccothrinax_trinitensis','Coccothrinax_victorini','Coccothrinax_yunquensis','Coccothrinax_yuraguana','Colpothrinax_aphanopetala','Colpothrinax_cookii','Colpothrinax_wrightii','Copernicia_alba','Copernicia_baileyana','Copernicia_berteroana','Copernicia_brittonorum','Copernicia_cowellii','Copernicia_curbeloi','Copernicia_curtissii','Copernicia_ekmanii','Copernicia_fallaensis','Copernicia_gigas','Copernicia_glabrescens','Copernicia_hospita','Copernicia_humicola','Copernicia_longiglossa','Copernicia_macroglossa','Copernicia_molineti','Copernicia_oxycalyx','Copernicia_prunifera','Copernicia_rigida','Copernicia_roigii','Copernicia_tectorum','Copernicia_yarey','Cryosophila_bartlettii','Cryosophila_cookii','Cryosophila_grayumii','Cryosophila_guagara','Cryosophila_kalbreyeri','Cryosophila_macrocarpa','Cryosophila_nana','Cryosophila_stauracantha','Cryosophila_warscewiczii','Cryosophila_williamsii','Desmoncus_chinantlensis','Desmoncus_cirrhifer','Desmoncus_costaricensis','Desmoncus_giganteus','Desmoncus_horridus','Desmoncus_interjectus','Desmoncus_kunarius','Desmoncus_latisectus','Desmoncus_leptoclonos','Desmoncus_loretanus','Desmoncus_madrensis','Desmoncus_mitis','Desmoncus_moorei','Desmoncus_myriacanthos','Desmoncus_obovoideus','Desmoncus_orthacanthos','Desmoncus_osensis','Desmoncus_parvulus','Desmoncus_polyacanthos','Desmoncus_prunifer','Desmoncus_pumilus','Desmoncus_setosus','Desmoncus_stans','Desmoncus_vacivus','Dictyocaryum_fuscum','Dictyocaryum_lamarckianum','Dictyocaryum_ptarianum','Elaeis_oleifera','Euterpe_broadwayi','Euterpe_catinga','Euterpe_edulis','Euterpe_longibracteata','Euterpe_luminosa','Euterpe_oleracea','Euterpe_precatoria','Gaussia_attenuata','Gaussia_gomez-pompae','Gaussia_maya','Gaussia_princeps','Gaussia_spirituana','Geonoma_aspidiifolia','Geonoma_baculifera','Geonoma_bernalii','Geonoma_braunii','Geonoma_brenesii','Geonoma_brongniartii','Geonoma_calyptrogynoidea','Geonoma_camana','Geonoma_chlamydostachys','Geonoma_chococola','Geonoma_concinna','Geonoma_concinnoidea','Geonoma_congesta','Geonoma_cuneata','Geonoma_deneversii','Geonoma_deversa','Geonoma_dindoensis','Geonoma_divisa','Geonoma_elegans','Geonoma_epetiolata','Geonoma_euspatha','Geonoma_ferruginea','Geonoma_fosteri','Geonoma_frontinensis','Geonoma_galeanoae','Geonoma_gentryi','Geonoma_hollinensis','Geonoma_hugonis','Geonoma_interrupta','Geonoma_lanata','Geonoma_laxiflora','Geonoma_lehmannii','Geonoma_leptospadix','Geonoma_longipedunculata','Geonoma_longivaginata','Geonoma_macrostachys','Geonoma_maxima','Geonoma_monospatha','Geonoma_mooreana','Geonoma_multisecta','Geonoma_occidentalis','Geonoma_oldemanii','Geonoma_oligoclona','Geonoma_operculata','Geonoma_orbignyana','Geonoma_paradoxa','Geonoma_pauciflora','Geonoma_peruviana','Geonoma_pinnatifrons','Geonoma_poeppigiana','Geonoma_pohliana','Geonoma_poiteauana','Geonoma_sanmartinensis','Geonoma_santanderensis','Geonoma_schizocarpa','Geonoma_schottiana','Geonoma_scoparia','Geonoma_simplicifrons','Geonoma_spinescens','Geonoma_stricta','Geonoma_talamancana','Geonoma_tenuissima','Geonoma_triandra','Geonoma_triglochin','Geonoma_trigona','Geonoma_umbraculiformis','Geonoma_undata','Geonoma_venosa','Guihaia_argyrata','Guihaia_grossifibrosa','Guihaia_lancifolia','Hemithrinax_compacta','Hemithrinax_ekmaniana','Hemithrinax_rivularis','Iriartea_deltoidea','Iriartella_setigera','Iriartella_stenocarpa','Itaya_amicorum','Juania_australis','Jubaea_chilensis','Leopoldinia_piassaba','Leopoldinia_pulchra','Leucothrinax_morrisii','Lytocaryum_hoehnei','Lytocaryum_insigne','Lytocaryum_itapebiense','Lytocaryum_weddellianum','Manicaria_martiana','Manicaria_saccifera','Mauritia_carana','Mauritia_flexuosa','Mauritiella_aculeata','Mauritiella_armata','Mauritiella_macroclada','Mauritiella_pumila','Neonicholsonia_watsonii','Oenocarpus_bacaba','Oenocarpus_balickii','Oenocarpus_bataua','Oenocarpus_circumtextus','Oenocarpus_distichus','Oenocarpus_makeru','Oenocarpus_mapora','Oenocarpus_minor','Oenocarpus_simplex','Phytelephas_macrocarpa','Phytelephas_schottii','Phytelephas_seemannii','Phytelephas_tenuicaulis','Phytelephas_tumacana','Prestoea_acuminata','Prestoea_carderi','Prestoea_decurrens','Prestoea_ensiformis','Prestoea_longepetiolata','Prestoea_pubens','Prestoea_pubigera','Prestoea_schultzeana','Prestoea_simplicifolia','Prestoea_tenuiramosa','Raphia_taedigera','Reinhardtia_elegans','Reinhardtia_gracilis','Reinhardtia_koschnyana','Reinhardtia_latisecta','Reinhardtia_paiewonskiana','Reinhardtia_simplex','Rhapidophyllum_hystrix','Roystonea_altissima','Roystonea_borinquena','Roystonea_dunlapiana','Roystonea_lenis','Roystonea_maisiana','Roystonea_oleracea','Roystonea_princeps','Roystonea_regia','Roystonea_stellata','Roystonea_violacea','Sabal_bermudana','Sabal_causiarum','Sabal_domingensis','Sabal_etonia','Sabal_gretherae','Sabal_maritima','Sabal_mauritiiformis','Sabal_mexicana','Sabal_minor','Sabal_palmetto','Sabal_pumos','Sabal_rosei','Sabal_uresana','Sabal_yapa','Serenoa_repens','Socratea_exorrhiza','Socratea_hecatonandra','Socratea_montana','Socratea_rostrata','Socratea_salazarii','Syagrus_allagopteroides','Syagrus_amara','Syagrus_angustifolia','Syagrus_botryophora','Syagrus_caerulescens','Syagrus_campestris','Syagrus_campylospatha','Syagrus_cardenasii','Syagrus_cataphracta','Syagrus_cearensis','Syagrus_cerqueirana','Syagrus_cocoides','Syagrus_comosa','Syagrus_coronata','Syagrus_deflexa','Syagrus_duartei','Syagrus_emasensis','Syagrus_evansiana','Syagrus_flexuosa','Syagrus_glaucescens','Syagrus_glazioviana','Syagrus_gouveiana','Syagrus_graminifolia','Syagrus_harleyi','Syagrus_hoehnei','Syagrus_inajai','Syagrus_insignis','Syagrus_itacambirana','Syagrus_itapebiensis','Syagrus_kellyana','Syagrus_lilliputiana','Syagrus_loefgrenii','Syagrus_longipedunculata','Syagrus_lorenzoniorum','Syagrus_macrocarpa','Syagrus_mendanhensis','Syagrus_menzeliana','Syagrus_microphylla','Syagrus_minor','Syagrus_oleracea','Syagrus_orinocensis','Syagrus_petraea','Syagrus_picrophylla','Syagrus_pimentae','Syagrus_pleioclada','Syagrus_pleiocladoides','Syagrus_pompeoi','Syagrus_procumbens','Syagrus_pseudococos','Syagrus_romanzoffiana','Syagrus_rupicola','Syagrus_ruschiana','Syagrus_sancona','Syagrus_santosii','Syagrus_schizophylla','Syagrus_smithii','Syagrus_stenopetala','Syagrus_stratincola','Syagrus_vagans','Syagrus_vermicularis','Syagrus_werdermannii','Syagrus_yungasensis','Synechanthus_fibrosus','Synechanthus_warscewiczianus','Thrinax_excelsa','Thrinax_parviflora','Thrinax_radiata','Trithrinax_acanthocoma','Trithrinax_brasiliensis','Trithrinax_campestris','Trithrinax_schizophylla','Washingtonia_filifera','Washingtonia_robusta','Welfia_regia','Wendlandiella_gracilis','Wettinia_aequalis','Wettinia_aequatorialis','Wettinia_anomala','Wettinia_augusta','Wettinia_castanea','Wettinia_disticha','Wettinia_drudei','Wettinia_fascicularis','Wettinia_hirsuta','Wettinia_kalbreyeri','Wettinia_lanata','Wettinia_longipetala','Wettinia_maynensis','Wettinia_microcarpa','Wettinia_minima','Wettinia_oxycarpa','Wettinia_panamensis','Wettinia_praemorsa','Wettinia_quinaria','Wettinia_radiata','Wettinia_verruculosa','Zombia_antillarum']\n",
        "\n",
        "print('Number of american palms:')\n",
        "len(palms_americas)"
      ],
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Number of american palms:\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "810"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yTdA2VFAA6PK"
      },
      "source": [
        "So, what have we sequenced?"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "o-3Ff_prCUOo",
        "outputId": "b3508722-3f73-4b2d-b1a7-d3bc4d98c73f"
      },
      "source": [
        "# first, we create sets of items to compare, using list comprehention. Sets should be defined within curly brackets\n",
        "sppseqed = {spp for spp in df[df['Continent'] == 'americas']['taxspecies'].unique()} #list comprehension are loops of a single line\n",
        "apspp = {spp for spp in palms_americas} # american palm species\n",
        "# we can apply set theory to find the differences (not sequenced or sequenced but not in the list of american palms)\n",
        "# items in setA that are not shared with setB and viceversa\n",
        "# symmetric difference \n",
        "print(apspp - sppseqed) # niceee, nothing is missing\n",
        "sppmissing = len(apspp - sppseqed)\n",
        "print('\\nAlleged missing species: %s'%(sppmissing))\n",
        "print('Percentage of species missing: %s'%((sppmissing * 100)/len(palms_americas)))\n",
        "print('Percentage of species sequenced: %s'%(100-(sppmissing * 100)/len(palms_americas)))"
      ],
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "{'Attalea_exigua', 'Attalea_compta', 'Bactris_ferruginea', 'Astrocaryum_rodriguesii', 'Geonoma_triglochin', 'Attalea_osmantha', 'Attalea_camopiensis', 'Bactris_faucium', 'Pholidostachys_amazonensis', 'Wettinia_microcarpa', 'Astrocaryum_confertum', 'Copernicia_roigii', 'Bactris_constanciae', 'Socratea_salazarii', 'Bactris_sphaerocarpa', 'Bactris_hondurensis', 'Leopoldinia_piassaba', 'Geonoma_poiteauana', 'Aiphanes_stergiosii', 'Desmoncus_leptoclonos', 'Bactris_hatschbachii', 'Coccothrinax_acunana', 'Coccothrinax_fagildei', 'Astrocaryum_farinosum', 'Brahea_pimo', 'Cryosophila_cookii', 'Butia_quaraimana', 'Wettinia_minima', 'Bactris_oligoclada', 'Attalea_racemosa', 'Geonoma_laxiflora', 'Bactris_herrerana', 'Brahea_edulis', 'Copernicia_curbeloi', 'Bactris_corossilla', 'Bactris_syagroides', 'Acrocomia_hassleri', 'Bactris_setiflora', 'Aiphanes_acanthophylla', 'Syagrus_werdermannii', 'Chamaedorea_smithii', 'Coccothrinax_bermudezii', 'Cryosophila_macrocarpa', 'Bactris_schultesii', 'Bactris_acanthocarpoides', 'Socratea_montana', 'Cryosophila_grayumii', 'Reinhardtia_paiewonskiana', 'Wettinia_fascicularis', 'Chamaedorea_pachecoana', 'Geonoma_hollinensis', 'Desmoncus_parvulus', 'Wettinia_anomala', 'Attalea_septuagenata', 'Coccothrinax_moaensis', 'Copernicia_cowellii', 'Asterogyne_spicata', 'Oenocarpus_bacaba', 'Aiphanes_duquei', 'Phytelephas_schottii', 'Calyptrogyne_osensis', 'Desmoncus_moorei', 'Hemithrinax_rivularis', 'Bactris_gastoniana', 'Lytocaryum_insigne', 'Desmoncus_costaricensis', 'Aiphanes_bicornis', 'Geonoma_aspidiifolia', 'Prestoea_simplicifolia', 'Acrocomia_media', 'Desmoncus_osensis', 'Coccothrinax_yunquensis', 'Geonoma_brenesii', 'Asterogyne_yaracuyense', 'Asterogyne_ramosa', 'Wettinia_praemorsa', 'Geonoma_umbraculiformis', 'Copernicia_curtissii', 'Astrocaryum_tucuma', 'Oenocarpus_makeru', 'Astrocaryum_vulgare', 'Wettinia_oxycarpa', 'Astrocaryum_giganteum', 'Copernicia_longiglossa', 'Hemithrinax_compacta', 'Syagrus_pleioclada', 'Coccothrinax_concolor', 'Geonoma_braunii', 'Wettinia_panamensis', 'Bactris_bahiensis', 'Desmoncus_setosus', 'Bactris_polystachya', 'Desmoncus_stans', 'Attalea_salvadorensis', 'Desmoncus_pumilus', 'Hyospathe_frontinensis', 'Sabal_domingensis', 'Acoelorraphe_wrightii', 'Syagrus_emasensis', 'Desmoncus_obovoideus', 'Chamaedorea_amabilis', 'Attalea_fairchildensis', 'Calyptronoma_plumeriana', 'Bactris_pickelii', 'Bactris_glassmanii', 'Copernicia_brittonorum', 'Bactris_kunorum', 'Butia_stolonifera', 'Roystonea_violacea', 'Phytelephas_seemannii', 'Butia_missionera', 'Geonoma_spinescens', 'Desmoncus_madrensis', 'Geonoma_ferruginea', 'Attalea_huebneri', 'Lytocaryum_hoehnei', 'Bactris_riparia', 'Desmoncus_prunifer', 'Wettinia_longipetala', 'Attalea_magdalenica', 'Calyptrogyne_fortunensis', 'Coccothrinax_garciana', 'Calyptrogyne_coloradensis', 'Bactris_balanophora', 'Chamaedorea_ernesti-augusti', 'Coccothrinax_boschiana', 'Geonoma_tenuissima', 'Astrocaryum_aculeatissimum', 'Chamaedorea_serpens', 'Geonoma_bernalii', 'Chamaedorea_rosibeliae', 'Attalea_dahlgreniana', 'Wettinia_kalbreyeri', 'Syagrus_pimentae', 'Attalea_rostrata', 'Attalea_weberbaueri', 'Mauritia_carana', 'Bactris_tefensis', 'Reinhardtia_latisecta', 'Bactris_rostrata', 'Wettinia_castanea', 'Calyptrogyne_sanblasensis', 'Chamaedorea_atrovirens', 'Attalea_macrolepis', 'Chamaedorea_schiedeana', 'Coccothrinax_nipensis', 'Hyospathe_wendlandiana', 'Coccothrinax_muricata', 'Attalea_rhynchocarpa', 'Bactris_halmoorei', 'Geonoma_talamancana', 'Geonoma_oldemanii', 'Desmoncus_loretanus', 'Geonoma_concinnoidea', 'Oenocarpus_circumtextus', 'Lytocaryum_weddellianum', 'Geonoma_scoparia', 'Bactris_soeiroana', 'Coccothrinax_yuraguana', 'Astrocaryum_cuatrecasanum', 'Iriartella_stenocarpa', 'Attalea_degranvillei', 'Aiphanes_chiribogensis', 'Sabal_gretherae', 'Oenocarpus_simplex', 'Pholidostachys_synanthera', 'Astrocaryum_paramaca', 'Chamaedorea_angustisecta', 'Geonoma_dindoensis', 'Attalea_guianensis', 'Bactris_jamaicana', 'Asterogyne_guianensis', 'Desmoncus_latisectus', 'Aiphanes_deltoidea', 'Geonoma_lanata', 'Euterpe_luminosa', 'Prestoea_carderi', 'Coccothrinax_camagueyana', 'Bactris_acanthocarpa', 'Desmoncus_interjectus', 'Bactris_grayumii', 'Wettinia_drudei', 'Aiphanes_weberbaueri', 'Bactris_chaveziae', 'Guihaia_lancifolia', 'Coccothrinax_leonis', 'Geonoma_epetiolata', 'Chamaedorea_whitelockiana', 'Syagrus_pompeoi', 'Coccothrinax_victorini', 'Chamaedorea_undulatifolia', 'Hyospathe_macrorhachis', 'Geonoma_oligoclona', 'Dictyocaryum_fuscum', 'Wettinia_hirsuta', 'Lytocaryum_itapebiense', 'Wettinia_radiata', 'Copernicia_oxycalyx', 'Coccothrinax_rigida', 'Geonoma_schizocarpa', 'Colpothrinax_cookii', 'Syagrus_lilliputiana', 'Geonoma_chococola', 'Geonoma_operculata', 'Geonoma_divisa', 'Attalea_salazarii', 'Trithrinax_acanthocoma', 'Brahea_brandegeei', 'Geonoma_euspatha', 'Chamaedorea_queroana', 'Attalea_blepharopus', 'Bactris_timbuiensis', 'Chamaedorea_latisecta', 'Attalea_kewensis', 'Desmoncus_horridus', 'Wettinia_aequatorialis', 'Bactris_pliniana', 'Desmoncus_chinantlensis', 'Guihaia_argyrata', 'Syagrus_cataphracta', 'Wettinia_verruculosa', 'Bactris_aubletiana', 'Bactris_tomentosa', 'Syagrus_yungasensis', 'Attalea_wesselsboeri', 'Geonoma_longipedunculata', 'Bactris_dianeura', 'Calyptrogyne_baudensis', 'Bactris_guineensis', 'Geonoma_venosa', 'Reinhardtia_elegans', 'Calyptrogyne_kunorum', 'Bactris_cuspidata', 'Chamaedorea_castillo-montii', 'Pholidostachys_occidentalis', 'Geonoma_simplicifrons', 'Bactris_nancibaensis', 'Bactris_longiseta', 'Hyospathe_peruviana', 'Geonoma_poeppigiana', 'Chamaedorea_foveata', 'Roystonea_princeps', 'Geonoma_gentryi', 'Desmoncus_cirrhifer', 'Coccothrinax_baracoensis', 'Geonoma_trigona', 'Aiphanes_graminifolia', 'Bactris_martiana', 'Bactris_ptariana', 'Socratea_hecatonandra', 'Coccothrinax_pseudorigida', 'Coccothrinax_pauciramosa', 'Coccothrinax_alexandri', 'Calyptrogyne_trichostachys', 'Allagoptera_campestris', 'Prestoea_pubigera', 'Geonoma_pinnatifrons', 'Bactris_vulgaris', 'Roystonea_stellata', 'Bactris_oligocarpa', 'Copernicia_humicola', 'Desmoncus_giganteus', 'Calyptrogyne_pubescens', 'Socratea_rostrata', 'Attalea_maracaibensis', 'Coccothrinax_pumila', 'Chamaedorea_christinae', 'Attalea_cephalotus', 'Attalea_maripensis', 'Desmoncus_vacivus', 'Brahea_decumbens', 'Guihaia_grossifibrosa', 'Wettinia_maynensis', 'Brahea_calcarea', 'Calyptrogyne_deneversii', 'Juania_australis', 'Dictyocaryum_ptarianum', 'Syagrus_cearensis', 'Geonoma_sanmartinensis', 'Thrinax_excelsa', 'Aiphanes_spicata', 'Brahea_moorei', 'Bactris_ana-juliae', 'Astrocaryum_jauari', 'Geonoma_deneversii', 'Butia_marmorii', 'Bactris_caudata', 'Bactris_turbinocarpa', 'Attalea_peruviana', 'Wettinia_quinaria', 'Aiphanes_multiplex', 'Coccothrinax_scoparia', 'Wettinia_disticha', 'Wettinia_lanata', 'Bactris_horridispatha', 'Euterpe_longibracteata', 'Calyptrogyne_herrerae', 'Manicaria_martiana', 'Desmoncus_myriacanthos', 'Bactris_bifida', 'Geonoma_peruviana', 'Attalea_bassleriana', 'Attalea_amylacea', 'Aiphanes_verrucosa', 'Butia_catarinensis', 'Bactris_cubensis', 'Coccothrinax_microphylla', 'Aiphanes_grandis', 'Bactris_rhaphidacantha', 'Calyptrogyne_condensata', 'Roystonea_dunlapiana', 'Washingtonia_robusta', 'Astrocaryum_huaimi', 'Copernicia_molineti', 'Attalea_lauromuelleriana', 'Bactris_campestris', 'Geonoma_leptospadix', 'Astrocaryum_sociale', 'Aiphanes_eggersii', 'Coccothrinax_trinitensis', 'Attalea_hoehnei', 'Butia_exilata', 'Brahea_salvadorensis', 'Iriartella_setigera', 'Chamaedorea_moliniana', 'Desmoncus_kunarius', 'Astrocaryum_minus', 'Prestoea_tenuiramosa'}\n",
            "\n",
            "Alleged missing species: 328\n",
            "Percentage of species missing: 40.49382716049383\n",
            "Percentage of species sequenced: 59.50617283950617\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "c5I5tHfWlTyh"
      },
      "source": [
        "## 4. Adding more data\n",
        "\n",
        "That looks like a very low percentage. But there are a bunch of sequences we have not logged in the metadata. We can have a look at lists of species other people sequenced and count them in.\n",
        "\n",
        "The datasets that are missing:  \n",
        "- Chamaedoreas from Elliot Gardner\n",
        "- Species extracted by Paola Lima\n",
        "- Acrocomia from Brazil researchers\n",
        "- Geonoma from Loiseau et al. 2019\n",
        "\n",
        "We can check lists of files and metadata for those datasets\n",
        "\n",
        "First, Chamaedoreas. They are conveniently named already. All I'll do is to create a list of files in a text file and work from there.\n",
        "\n",
        "If running this on a local disk, you can use jupyter notebook's magic commands for bash\n",
        "\n",
        "\n",
        "```\n",
        "%%bash\n",
        "\n",
        "ls *gz > files_chamaedorea.txt\n",
        "```\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 436
        },
        "id": "6IeqWbaqCU7o",
        "outputId": "87b114c3-86f0-490f-b087-7af464393cb2"
      },
      "source": [
        "def get_data():\n",
        "  return pd.read_csv('./files_chamaedorea.txt', sep = '\\t', names = ['file']) # the file doesn't have a header\n",
        "\n",
        "# try reading the file if exists, keep in mind the date\n",
        "try:\n",
        "  eg = get_data()\n",
        "except:\n",
        "  url = 'https://raw.githubusercontent.com/mftorres/APP/main/data/files_chamaedorea.txt' # shared link careful who you share this with\n",
        "  print('Downloading')\n",
        "  res = rs.get(url = url)\n",
        "  with open('./files_chamaedorea.txt', 'wb') as file: # Writing document in binary mode so python doesn't make changes\n",
        "    file.write(res.content)\n",
        "  eg = get_data()\n",
        "\n",
        "eg # check the database"
      ],
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Downloading\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>file</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Chamaedorea_aff_nationsiana_Steyermark41863_S2...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Chamaedorea_aff_nationsiana_Steyermark41863_S2...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Chamaedorea_arenbergiana_Hawkins228_R1_001.fas...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Chamaedorea_arenbergiana_Hawkins228_R2_001.fas...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Chamaedorea_binderi_Grayum3351_R1_001.fastq.gz*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>75</th>\n",
              "      <td>Chamaedorea_vulgata_Standley67344_R2_001.fastq...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>76</th>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>77</th>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>79</th>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>80 rows Ã— 1 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                 file\n",
              "0   Chamaedorea_aff_nationsiana_Steyermark41863_S2...\n",
              "1   Chamaedorea_aff_nationsiana_Steyermark41863_S2...\n",
              "2   Chamaedorea_arenbergiana_Hawkins228_R1_001.fas...\n",
              "3   Chamaedorea_arenbergiana_Hawkins228_R2_001.fas...\n",
              "4     Chamaedorea_binderi_Grayum3351_R1_001.fastq.gz*\n",
              "..                                                ...\n",
              "75  Chamaedorea_vulgata_Standley67344_R2_001.fastq...\n",
              "76   Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz*\n",
              "77   Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz*\n",
              "78  Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...\n",
              "79  Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...\n",
              "\n",
              "[80 rows x 1 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TIpiqQXW0Rqk"
      },
      "source": [
        "That file has 80 files of compressed paired-end reads, forward and reverse.\n",
        "We need to add the file names into the database. We keep every file in the database to be able to mix single and paired-end reads.\n",
        "\n",
        "The column for files looks like"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yXVqI43XCU-q",
        "outputId": "7d968d60-49c3-4da4-bcd6-303b8ff02d5b"
      },
      "source": [
        "df.columns\n",
        "df['filename']"
      ],
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0          1_151124_000000000-AJE6N_P3252_1001_1.fastq.gz\n",
              "1          1_151124_000000000-AJE6N_P3252_1001_2.fastq.gz\n",
              "2          1_151130_000000000-AK5EU_P3252_1126_1.fastq.gz\n",
              "3          1_151130_000000000-AK5EU_P3252_1126_2.fastq.gz\n",
              "4          1_151130_000000000-AKG91_P3252_1182_1.fastq.gz\n",
              "                              ...                        \n",
              "3044    Chamaedorea_vulgata_Standley67344_R2_001.fastq.gz\n",
              "3045      Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz\n",
              "3046      Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz\n",
              "3047    Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...\n",
              "3048    Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...\n",
              "Name: filename, Length: 3049, dtype: object"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "S-Hd1aem0-cc"
      },
      "source": [
        "well, looks like I did update that data on the metadata. Let's see if the other fields are filled. Looks pretty much complete. We can go ahead with other datasets then."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 609
        },
        "id": "q20OOxN8CVDx",
        "outputId": "76f712df-58f8-4c7a-f37a-327aecbfb988"
      },
      "source": [
        "df[df['filename'].str.contains('Chamaedorea')] # filters the database by looking a string match"
      ],
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>filename</th>\n",
              "      <th>sense</th>\n",
              "      <th>filecode</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "      <th>botanic_garden</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>ifmorphotype</th>\n",
              "      <th>ifpopulation</th>\n",
              "      <th>sent_to_Cano</th>\n",
              "      <th>newfilename</th>\n",
              "      <th>raw_reads</th>\n",
              "      <th>reads_trimed_paired_flag</th>\n",
              "      <th>reads_trimed_paired</th>\n",
              "      <th>reads_trimed_single</th>\n",
              "      <th>percentage_lost_tosingles</th>\n",
              "      <th>pcr_filtered</th>\n",
              "      <th>collection_year</th>\n",
              "      <th>Continent</th>\n",
              "      <th>country</th>\n",
              "      <th>long</th>\n",
              "      <th>lat</th>\n",
              "      <th>flag</th>\n",
              "      <th>notes</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>2969</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_aff_nationsiana_Steyermark41863_S2...</td>\n",
              "      <td>R1</td>\n",
              "      <td>Steyermark41863_S26_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_nationsiana</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2970</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_aff_nationsiana_Steyermark41863_S2...</td>\n",
              "      <td>R2</td>\n",
              "      <td>Steyermark41863_S26_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_nationsiana</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2971</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_arenbergiana_Hawkins228_R1_001.fas...</td>\n",
              "      <td>R1</td>\n",
              "      <td>Hawkins228</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_arenbergiana</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2972</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_arenbergiana_Hawkins228_R2_001.fas...</td>\n",
              "      <td>R2</td>\n",
              "      <td>Hawkins228</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_arenbergiana</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2973</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_binderi_Grayum3351_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>Grayum3351</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_binderi</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3044</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_vulgata_Standley67344_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>Standley67344</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_vulgata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3045</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>Azofeifa484</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3046</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>Azofeifa484</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3047</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...</td>\n",
              "      <td>R1</td>\n",
              "      <td>Azofeifa484_S36_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3048</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...</td>\n",
              "      <td>R2</td>\n",
              "      <td>Azofeifa484_S36_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>80 rows Ã— 30 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "            dataset        provider pi_code  ... lat flag notes\n",
              "2969  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "2970  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "2971  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "2972  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "2973  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "...             ...             ...     ...  ...  ..  ...   ...\n",
              "3044  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3045  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3046  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3047  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3048  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "\n",
              "[80 rows x 30 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qx1KxDSY23k9"
      },
      "source": [
        "Let's check the data sequenced by Paola. There's a list of species "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 623
        },
        "id": "ri4iuevFCVHF",
        "outputId": "8f2882d5-d9cf-4fbf-9b5d-06b96fa44a66"
      },
      "source": [
        "def get_data():\n",
        "  return pd.read_csv('./metadata_machinereadable_Paola_2020.txt', sep = '\\t') # the file doesn't have a header\n",
        "\n",
        "# try reading the file if exists, keep in mind the date\n",
        "try:\n",
        "  seqp = get_data()\n",
        "except:\n",
        "  url = 'https://raw.githubusercontent.com/mftorres/APP/main/data/metadata_machinereadable_Paola_2020.txt' # shared link careful who you share this with\n",
        "  print('Downloading')\n",
        "  res = rs.get(url = url)\n",
        "  with open('./metadata_machinereadable_Paola_2020.txt', 'wb') as file: # Writing document in binary mode so python doesn't make changes\n",
        "    file.write(res.content)\n",
        "  seqp = get_data()\n",
        "\n",
        "seqp # check the database"
      ],
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Downloading\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plate</th>\n",
              "      <th>Rapid_Genomics_Identification</th>\n",
              "      <th>Plate</th>\n",
              "      <th>Column Sort</th>\n",
              "      <th>Silica number</th>\n",
              "      <th>Species</th>\n",
              "      <th>Collector and Number</th>\n",
              "      <th>Final Concentration (ng/ul)</th>\n",
              "      <th>Nanodrop (260/280)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA01</td>\n",
              "      <td>A1</td>\n",
              "      <td>1</td>\n",
              "      <td>3174</td>\n",
              "      <td>Acoelorrhaphe wrightii</td>\n",
              "      <td>H. Balslev 8156</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.76</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA02</td>\n",
              "      <td>A2</td>\n",
              "      <td>9</td>\n",
              "      <td>1156</td>\n",
              "      <td>Aiphanes bicornis</td>\n",
              "      <td>J. West s.n.</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.82</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA03</td>\n",
              "      <td>A3</td>\n",
              "      <td>17</td>\n",
              "      <td>261</td>\n",
              "      <td>Aiphanes chiribogensis</td>\n",
              "      <td>F. Borchsenius 652</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.80</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA04</td>\n",
              "      <td>A4</td>\n",
              "      <td>25</td>\n",
              "      <td>3563</td>\n",
              "      <td>Aiphanes deltoidea</td>\n",
              "      <td>L.A. NuÃ±ez s.n.</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.84</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA05</td>\n",
              "      <td>A5</td>\n",
              "      <td>33</td>\n",
              "      <td>397</td>\n",
              "      <td>Aiphanes duquei</td>\n",
              "      <td>R. Bernal 2852</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.78</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>118</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P002_WB11</td>\n",
              "      <td>B11</td>\n",
              "      <td>82</td>\n",
              "      <td>431</td>\n",
              "      <td>Wettinia praemorsa</td>\n",
              "      <td>R. Bernal 2883</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.82</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>119</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P002_WB12</td>\n",
              "      <td>B12</td>\n",
              "      <td>90</td>\n",
              "      <td>434</td>\n",
              "      <td>Wettinia quinaria</td>\n",
              "      <td>R. Bernal 2491</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.73</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>120</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P002_WC01</td>\n",
              "      <td>C1</td>\n",
              "      <td>3</td>\n",
              "      <td>436</td>\n",
              "      <td>Wettinia radiata</td>\n",
              "      <td>R. Bernal 2190</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.75</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P002_WC02</td>\n",
              "      <td>C2</td>\n",
              "      <td>11</td>\n",
              "      <td>3514</td>\n",
              "      <td>Wettinia verruculosa</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.90</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>122</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P002_WC03</td>\n",
              "      <td>C3</td>\n",
              "      <td>19</td>\n",
              "      <td>2299</td>\n",
              "      <td>Mauritiella armata</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>50.0</td>\n",
              "      <td>2.11</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>123 rows Ã— 9 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     plate  ... Nanodrop (260/280)\n",
              "0        1  ...               1.76\n",
              "1        1  ...               1.82\n",
              "2        1  ...               1.80\n",
              "3        1  ...               1.84\n",
              "4        1  ...               1.78\n",
              "..     ...  ...                ...\n",
              "118      2  ...               1.82\n",
              "119      2  ...               1.73\n",
              "120      2  ...               1.75\n",
              "121      2  ...               1.90\n",
              "122      2  ...               2.11\n",
              "\n",
              "[123 rows x 9 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 796
        },
        "id": "lRmy13YlB4PQ",
        "outputId": "406284d5-0d82-4baf-f561-692383db382f"
      },
      "source": [
        "# now the rapid genomics data\n",
        "def get_data():\n",
        "  return pd.read_csv('./GOT_130404_SampleSheet.csv', sep = '\\t') # the file doesn't have a header\n",
        "\n",
        "# try reading the file if exists, keep in mind the date\n",
        "try:\n",
        "  seqprapid = get_data()\n",
        "except:\n",
        "  url = 'https://raw.githubusercontent.com/mftorres/APP/main/data/GOT_130404_SampleSheet.csv' # shared link careful who you share this with\n",
        "  print('Downloading')\n",
        "  res = rs.get(url = url)\n",
        "  with open('./GOT_130404_SampleSheet.csv', 'wb') as file: # Writing document in binary mode so python doesn't make changes\n",
        "    file.write(res.content)\n",
        "  seqprapid = get_data()\n",
        "\n",
        "seqprapid # check the database"
      ],
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Downloading\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>RG_Sample_Code</th>\n",
              "      <th>Customer_Code</th>\n",
              "      <th>i5_Barcode_Seq</th>\n",
              "      <th>i7_Barcode_Seq</th>\n",
              "      <th>Sequence_Name</th>\n",
              "      <th>Sequencing_Cycle</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>GOT_130404_P001_WA01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>GTGTTCTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>GOT_130404_P001_WA02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>AGTCACTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>GOT_130404_P001_WA03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>ATTGGCTC</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>GOT_130404_P001_WA04</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>CAGATCTG</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA04_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>GOT_130404_P001_WA05</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>CCGTGAGA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA05_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>118</th>\n",
              "      <td>GOT_130404_P002_WB11</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>ACCACTGT</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WB11_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>119</th>\n",
              "      <td>GOT_130404_P002_WB12</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>ACAGATTC</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WB12_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>120</th>\n",
              "      <td>GOT_130404_P002_WC01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGAACAA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121</th>\n",
              "      <td>GOT_130404_P002_WC02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGTGGTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>122</th>\n",
              "      <td>GOT_130404_P002_WC03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>CGAACTTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>123 rows Ã— 6 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "           RG_Sample_Code  ...  Sequencing_Cycle\n",
              "0    GOT_130404_P001_WA01  ...             2x150\n",
              "1    GOT_130404_P001_WA02  ...             2x150\n",
              "2    GOT_130404_P001_WA03  ...             2x150\n",
              "3    GOT_130404_P001_WA04  ...             2x150\n",
              "4    GOT_130404_P001_WA05  ...             2x150\n",
              "..                    ...  ...               ...\n",
              "118  GOT_130404_P002_WB11  ...             2x150\n",
              "119  GOT_130404_P002_WB12  ...             2x150\n",
              "120  GOT_130404_P002_WC01  ...             2x150\n",
              "121  GOT_130404_P002_WC02  ...             2x150\n",
              "122  GOT_130404_P002_WC03  ...             2x150\n",
              "\n",
              "[123 rows x 6 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wfH1Jc63CVJ6",
        "outputId": "7b7b57ba-13ac-427c-ab7d-e658e40d506d"
      },
      "source": [
        "df.columns\n",
        "df['dataset'].unique() # looks like these sequences aren't there. We can fill up the dataset then."
      ],
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Cano', 'Lausanne_Cerox', 'American Palms', 'Astrocaryum',\n",
              "       'Mauritia', 'Zizka', 'Raphia', 'Geonoma', 'EliottGardner'],\n",
              "      dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 18
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_ux58BM0-qya"
      },
      "source": [
        "We can create new columns in seqp to contain the information needed in df and then concatenate both.\n",
        "\n",
        "File names can be extractede from the 'Sequence_name' column in seqpraid. We can add the species and voucher from seqp to seqpraid, then duplicate rows to create the R2 read entries, rename columns and concatenate with df"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8f8kXdXD-oQl",
        "outputId": "12f4ca6d-70a3-4173-acc1-d3054a1a8095"
      },
      "source": [
        "# create a column with file information, forward and reverse\n",
        "# will create empty lists that I will join in a dataframe later\n",
        "\n",
        "# first, a dictionary onject linking sample codes with species and vouchers\n",
        "# a dictionary has keys (unique) and values, the latest can be strings, lists, other dictionaries, etc.\n",
        "\n",
        "# in this case keys are the rapid genomics identifications\n",
        "# values are tuples of species (replacing space with underscore) and voucher\n",
        "id_spp_dict = dict(zip(seqp['Rapid_Genomics_Identification'],zip(seqp['Species'].str.replace(' ','_'),seqp['Collector and Number']))) # think of zip as a jacket zip: makes tuples with values in the same indexing\n",
        "\n",
        "# looks like this\n",
        "id_spp_dict"
      ],
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'GOT_130404_P001_WA01': ('Acoelorrhaphe_wrightii', 'H. Balslev 8156'),\n",
              " 'GOT_130404_P001_WA02': ('Aiphanes_bicornis', 'J. West s.n.'),\n",
              " 'GOT_130404_P001_WA03': ('Aiphanes_chiribogensis', 'F. Borchsenius 652'),\n",
              " 'GOT_130404_P001_WA04': ('Aiphanes_deltoidea', 'L.A. NuÃ±ez s.n.'),\n",
              " 'GOT_130404_P001_WA05': ('Aiphanes_duquei', 'R. Bernal 2852'),\n",
              " 'GOT_130404_P001_WA06': ('Aiphanes_eggersii', 'J.C.Pintaud 1476'),\n",
              " 'GOT_130404_P001_WA07': ('Aiphanes_graminifolia', 'R.Bernal 4825'),\n",
              " 'GOT_130404_P001_WA08': ('Aiphanes_grandis', 'F. Borchsenius 648'),\n",
              " 'GOT_130404_P001_WA09': ('Aiphanes_verrucosa', 'F. Borchsenius 641'),\n",
              " 'GOT_130404_P001_WA10': ('Aiphanes_weberbaueri', 'H.Balslev 7769'),\n",
              " 'GOT_130404_P001_WA11': ('Asterogyne_guianensis', 'J. Roncal 406'),\n",
              " 'GOT_130404_P001_WA12': ('Asterogyne_ramosa', 'Private collection'),\n",
              " 'GOT_130404_P001_WB01': ('Astrocaryum_chambira', 'T. Emilio 864'),\n",
              " 'GOT_130404_P001_WB02': ('Astrocaryum_faranae', 'H. Balslev 7897'),\n",
              " 'GOT_130404_P001_WB03': ('Astrocaryum_huaimi', 'T.P.L. Couvreur 289'),\n",
              " 'GOT_130404_P001_WB04': ('Astrocaryum_jauari', 'H. Balslev 7347'),\n",
              " 'GOT_130404_P001_WB05': ('Mauritiella_pumila', 'R. Bernal 4373'),\n",
              " 'GOT_130404_P001_WB06': ('Astrocaryum_minus', 'De Granville, J.-J. 14511'),\n",
              " 'GOT_130404_P001_WB07': ('Astrocaryum_paramaca', 'J.C.Pintaud 901'),\n",
              " 'GOT_130404_P001_WB08': ('Astrocaryum_sciophilum', 'B. Maguire 54126'),\n",
              " 'GOT_130404_P001_WB09': ('Astrocaryum_urostachys',\n",
              "  'F. Borchsenius, no voucher'),\n",
              " 'GOT_130404_P001_WB10': ('Astrocaryum_vulgare', 'M.J. Balic 1626'),\n",
              " 'GOT_130404_P001_WB11': ('Attalea_funifera', 'L.A. Matos Silva 747'),\n",
              " 'GOT_130404_P001_WB12': ('Attalea_racemosa', 'H. Balslev 7272'),\n",
              " 'GOT_130404_P001_WC01': ('Bactris_acanthocarpoides', 'T. Emilio 831'),\n",
              " 'GOT_130404_P001_WC02': ('Bactris_aubletiana', 'ENV 1593'),\n",
              " 'GOT_130404_P001_WC03': ('Bactris_balanophora', 'R. Bernal 4417'),\n",
              " 'GOT_130404_P001_WC04': ('Bactris_bifida', 'H. Balslev 6594'),\n",
              " 'GOT_130404_P001_WC05': ('Bactris_campestris', 'J.C. Pintaud 1448'),\n",
              " 'GOT_130404_P001_WC06': ('Bactris_chaveziae', 'T.L.P. Couvreur 279'),\n",
              " 'GOT_130404_P001_WC07': ('Bactris_constanciae', 'De Granville 4850'),\n",
              " 'GOT_130404_P001_WC08': ('Bactris_corossilla', 'H. Balslev 7294'),\n",
              " 'GOT_130404_P001_WC09': ('Bactris_gastoniana', 'J.C.Pintaud 1446'),\n",
              " 'GOT_130404_P001_WC10': ('Bactris_guineensis', 'JBP s.n.'),\n",
              " 'GOT_130404_P001_WC11': ('Bactris_halmoorei', 'H. Balslev 7267'),\n",
              " 'GOT_130404_P001_WC12': ('Bactris_killipii', 'J.C.Pintaud 1428'),\n",
              " 'GOT_130404_P001_WD01': ('Bactris_longiseta', 'J.C.Pintaud 330'),\n",
              " 'GOT_130404_P001_WD02': ('Bactris_maraja', 'H. Balslev 6565'),\n",
              " 'GOT_130404_P001_WD03': ('Bactris_martiana', 'H. Balslev 6549'),\n",
              " 'GOT_130404_P001_WD04': ('Bactris_oligocarpa', 'T. Emilio 528'),\n",
              " 'GOT_130404_P001_WD05': ('Bactris_pliniana', 'J.C.Pintaud 1452'),\n",
              " 'GOT_130404_P001_WD06': ('Bactris_riparia', 'H. Balslev 7953'),\n",
              " 'GOT_130404_P001_WD07': ('Bactris_schultesii', 'H. Balslev 7284'),\n",
              " 'GOT_130404_P001_WD08': ('Bactris_setiflora', 'F. Borchsenius 629'),\n",
              " 'GOT_130404_P001_WD09': ('Bactris_simplicifrons', 'H. Balslev 6780'),\n",
              " 'GOT_130404_P001_WD10': ('Bactris_syagroides', 'T. Emilio 801'),\n",
              " 'GOT_130404_P001_WD11': ('Calyptrogyne_baudensis', 'SamÃ­n et al. 05'),\n",
              " 'GOT_130404_P001_WD12': ('Calyptrogyne_fortunensis', 'J. Roncal 370'),\n",
              " 'GOT_130404_P001_WE01': ('Calyptrogyne_trichostachys', 'J. Roncal 414'),\n",
              " 'GOT_130404_P001_WE02': ('Calyptronoma_occidentalis', 'FTG 71375D'),\n",
              " 'GOT_130404_P001_WE03': ('Calyptronoma_plumeriana', 'FTG 82492D'),\n",
              " 'GOT_130404_P001_WE04': ('Ceroxylon_peruvianum', 'SamÃ­n et al. 05'),\n",
              " 'GOT_130404_P001_WE05': ('Chamaedorea_crucensis', 'Hod No ID'),\n",
              " 'GOT_130404_P001_WE06': ('Chamaedorea_fragrans', 'J.C. Pintaud 1490'),\n",
              " 'GOT_130404_P001_WE07': ('Desmoncus_cirrhifer', 'R. Bernal 4989'),\n",
              " 'GOT_130404_P001_WE08': ('Desmoncus_giganteus', 'L.A. NuÃ±ez s.n.'),\n",
              " 'GOT_130404_P001_WE09': ('Desmoncus_myriacanthos', 'R. Bernal 4945'),\n",
              " 'GOT_130404_P001_WE10': ('Desmoncus_vacivus', 'MFT 703'),\n",
              " 'GOT_130404_P001_WE11': ('Dictyocaryon_ptarianum', 'Pintaud ENV 1185'),\n",
              " 'GOT_130404_P001_WE12': ('Geonoma_aspdiifolia', 'C. Farney 1752'),\n",
              " 'GOT_130404_P001_WF01': ('Geonoma_atrovirens_(Chamaedorea)', 'T. Emilio 761'),\n",
              " 'GOT_130404_P001_WF02': ('Geonoma_baculifera', 'M.J. Jansen-Jacobs 2386'),\n",
              " 'GOT_130404_P001_WF03': ('Geonoma_epetiolata', 'J. Roncal 316'),\n",
              " 'GOT_130404_P001_WF04': ('Geonoma_euspatha', 'J.J.Granville 11989'),\n",
              " 'GOT_130404_P001_WF05': ('Geonoma_hollinensis', 'H. Balslev 6418'),\n",
              " 'GOT_130404_P001_WF06': ('Geonoma_hugonis', 'J. Roncal 358'),\n",
              " 'GOT_130404_P001_WF07': ('Geonoma_lanata', 'F. Borchsenius 334'),\n",
              " 'GOT_130404_P001_WF08': ('Geonoma_laxiflora', 'H. Balslev 7273'),\n",
              " 'GOT_130404_P001_WF09': ('Geonoma_leptospadix', 'T. Emilio 714'),\n",
              " 'GOT_130404_P001_WF10': ('Geonoma_longepedunculata', 'Gloria Galeano 9186'),\n",
              " 'GOT_130404_P001_WF11': ('Geonoma_oldemanii', 'J. Roncal 403'),\n",
              " 'GOT_130404_P001_WF12': ('Geonoma_oligoclona', 'T. Emilio 731'),\n",
              " 'GOT_130404_P001_WG01': ('Geonoma_pinnatifrons', 'BG 534'),\n",
              " 'GOT_130404_P001_WG02': ('Geonoma_poeppigiana', 'R. Bernal 4621'),\n",
              " 'GOT_130404_P001_WG03': ('Mauritiella_armata', 'R. Bernal 4469'),\n",
              " 'GOT_130404_P001_WG04': ('Geonoma_triglochin', 'T. Emilio 820'),\n",
              " 'GOT_130404_P001_WG05': ('Geonoma_umbraculiformis', 'T. Emilio 780'),\n",
              " 'GOT_130404_P001_WG06': ('Guiahia_argyrata', '3757'),\n",
              " 'GOT_130404_P001_WG07': ('Hyospathe_frontinensis', 'R. Bernal 4955'),\n",
              " 'GOT_130404_P001_WG08': ('Iriartella_setigera', 'R. Bernal 4462'),\n",
              " 'GOT_130404_P001_WG09': ('Iriartella_stenocarpa', 'MFT 704'),\n",
              " 'GOT_130404_P001_WG10': ('Juania_australis', 'JCP 1454'),\n",
              " 'GOT_130404_P001_WG11': ('Manicaria_saccifera', 'H. Balslev 8114'),\n",
              " 'GOT_130404_P001_WG12': ('Oenocarpus_bacaba', 'M.J. Jansen-Jacobs 2387'),\n",
              " 'GOT_130404_P001_WH01': ('Oenocarpus_simplex', 'Rodrigo CÃ¡mara-Leret 1711'),\n",
              " 'GOT_130404_P001_WH02': ('Parajubaea_cocoides', 'Austria 80 **'),\n",
              " 'GOT_130404_P001_WH03': ('Pholidostachys_synanthera', 'T. Couvreur 119'),\n",
              " 'GOT_130404_P001_WH04': ('Phytelephas_tumacana', 'F.Borchsenius 659'),\n",
              " 'GOT_130404_P001_WH05': ('Prestoea_pubigera', 'A. Henderson 1623'),\n",
              " 'GOT_130404_P001_WH06': ('Pritchardia_minor', 'Kwood 7367'),\n",
              " 'GOT_130404_P001_WH07': ('Pritchardia_pericularum', 'JVM 1 (Niau?)'),\n",
              " 'GOT_130404_P001_WH08': ('Pritchardia_viscosa', 'SP Perlman 18712'),\n",
              " 'GOT_130404_P001_WH09': ('Pritchardia_vuylstekeana', 'JVM 2 (Makatea?)'),\n",
              " 'GOT_130404_P001_WH10': ('Sabal_antillensis', 'M.P. Griffith 388'),\n",
              " 'GOT_130404_P001_WH11': ('Sabal_gretherae', 'H. Balslev 8170'),\n",
              " 'GOT_130404_P001_WH12': ('Sabal_lougheediana', 'M.P. Griffith 392'),\n",
              " 'GOT_130404_P002_WA01': ('Sabal_minor', 'M. Eason 1142'),\n",
              " 'GOT_130404_P002_WA02': ('Sabal_yapa', 'H. Balslev 8166'),\n",
              " 'GOT_130404_P002_WA03': ('Schippia_concolor', '59-1200 D'),\n",
              " 'GOT_130404_P002_WA04': ('Socratea_hecatonandra', 'GG 8119'),\n",
              " 'GOT_130404_P002_WA05': ('Socratea_rostrata', 'R. Bernal 2493'),\n",
              " 'GOT_130404_P002_WA06': ('Socratea_salazarii', 'H. Balslev 7298'),\n",
              " 'GOT_130404_P002_WA07': ('Syagrus_amara', 'Noblick 5560'),\n",
              " 'GOT_130404_P002_WA08': ('Washingtonia_robusta', 'Austria 81 **'),\n",
              " 'GOT_130404_P002_WA09': ('Wettinia_aequatorialis', 'H. Balslev 6462'),\n",
              " 'GOT_130404_P002_WA10': ('Wettinia_anomala', 'R. Bernal 2487'),\n",
              " 'GOT_130404_P002_WA11': ('Wettinia_disticha', 'R. Bernal 2838'),\n",
              " 'GOT_130404_P002_WA12': ('Wettinia_castanea', 'R. Bernal 2825'),\n",
              " 'GOT_130404_P002_WB01': ('Wettinia_drudei', 'T. Emilio 810'),\n",
              " 'GOT_130404_P002_WB02': ('Wettinia_fascicularis', 'R. Bernal 2830'),\n",
              " 'GOT_130404_P002_WB03': ('Wettinia_hirsuta', 'R. Bernal 4737'),\n",
              " 'GOT_130404_P002_WB04': ('Wettinia_kalbreyeri', 'R. Bernal 4726'),\n",
              " 'GOT_130404_P002_WB05': ('Wettinia_lanata', 'R. Bernal 4775'),\n",
              " 'GOT_130404_P002_WB06': ('Wettinia_longipetala', 'H. Balslev 6545'),\n",
              " 'GOT_130404_P002_WB07': ('Wettinia_maynensis', 'A. Barford 60538'),\n",
              " 'GOT_130404_P002_WB08': ('Wettinia_microcarpa', 'R. Bernal 3481'),\n",
              " 'GOT_130404_P002_WB09': ('Wettinia_oxycarpa', 'R. Bernal 2498'),\n",
              " 'GOT_130404_P002_WB10': ('Wettinia_panamensis', 'A.J. Henderson 2045'),\n",
              " 'GOT_130404_P002_WB11': ('Wettinia_praemorsa', 'R. Bernal 2883'),\n",
              " 'GOT_130404_P002_WB12': ('Wettinia_quinaria', 'R. Bernal 2491'),\n",
              " 'GOT_130404_P002_WC01': ('Wettinia_radiata', 'R. Bernal 2190'),\n",
              " 'GOT_130404_P002_WC02': ('Wettinia_verruculosa', 'R. Bernal 2500'),\n",
              " 'GOT_130404_P002_WC03': ('Mauritiella_armata', 'J.C.Pintaud 1488')}"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 19
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "cRo7MFt9l5EH",
        "outputId": "70280fa9-0c83-4063-9704-a1413d74e1ab"
      },
      "source": [
        "# now, a dictionary for sample code\n",
        "id_code_dict = dict(zip(seqp['Rapid_Genomics_Identification'],seqp['Plate']+'_'+seqp['Collector and Number'].str.replace('[^a-zA-Z \\d]','').str.replace(' ','')))\n",
        "id_code_dict"
      ],
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'GOT_130404_P001_WA01': 'A1_HBalslev8156',\n",
              " 'GOT_130404_P001_WA02': 'A2_JWestsn',\n",
              " 'GOT_130404_P001_WA03': 'A3_FBorchsenius652',\n",
              " 'GOT_130404_P001_WA04': 'A4_LANuezsn',\n",
              " 'GOT_130404_P001_WA05': 'A5_RBernal2852',\n",
              " 'GOT_130404_P001_WA06': 'A6_JCPintaud1476',\n",
              " 'GOT_130404_P001_WA07': 'A7_RBernal4825',\n",
              " 'GOT_130404_P001_WA08': 'A8_FBorchsenius648',\n",
              " 'GOT_130404_P001_WA09': 'A9_FBorchsenius641',\n",
              " 'GOT_130404_P001_WA10': 'A10_HBalslev7769',\n",
              " 'GOT_130404_P001_WA11': 'A11_JRoncal406',\n",
              " 'GOT_130404_P001_WA12': 'A12_Privatecollection',\n",
              " 'GOT_130404_P001_WB01': 'B1_TEmilio864',\n",
              " 'GOT_130404_P001_WB02': 'B2_HBalslev7897',\n",
              " 'GOT_130404_P001_WB03': 'B3_TPLCouvreur289',\n",
              " 'GOT_130404_P001_WB04': 'B4_HBalslev7347',\n",
              " 'GOT_130404_P001_WB05': 'B5_RBernal4373',\n",
              " 'GOT_130404_P001_WB06': 'B6_DeGranvilleJJ14511',\n",
              " 'GOT_130404_P001_WB07': 'B7_JCPintaud901',\n",
              " 'GOT_130404_P001_WB08': 'B8_BMaguire54126',\n",
              " 'GOT_130404_P001_WB09': 'B9_FBorchseniusnovoucher',\n",
              " 'GOT_130404_P001_WB10': 'B10_MJBalic1626',\n",
              " 'GOT_130404_P001_WB11': 'B11_LAMatosSilva747',\n",
              " 'GOT_130404_P001_WB12': 'B12_HBalslev7272',\n",
              " 'GOT_130404_P001_WC01': 'C1_TEmilio831',\n",
              " 'GOT_130404_P001_WC02': 'C2_ENV1593',\n",
              " 'GOT_130404_P001_WC03': 'C3_RBernal4417',\n",
              " 'GOT_130404_P001_WC04': 'C4_HBalslev6594',\n",
              " 'GOT_130404_P001_WC05': 'C5_JCPintaud1448',\n",
              " 'GOT_130404_P001_WC06': 'C6_TLPCouvreur279',\n",
              " 'GOT_130404_P001_WC07': 'C7_DeGranville4850',\n",
              " 'GOT_130404_P001_WC08': 'C8_HBalslev7294',\n",
              " 'GOT_130404_P001_WC09': 'C9_JCPintaud1446',\n",
              " 'GOT_130404_P001_WC10': 'C10_JBPsn',\n",
              " 'GOT_130404_P001_WC11': 'C11_HBalslev7267',\n",
              " 'GOT_130404_P001_WC12': 'C12_JCPintaud1428',\n",
              " 'GOT_130404_P001_WD01': 'D1_JCPintaud330',\n",
              " 'GOT_130404_P001_WD02': 'D2_HBalslev6565',\n",
              " 'GOT_130404_P001_WD03': 'D3_HBalslev6549',\n",
              " 'GOT_130404_P001_WD04': 'D4_TEmilio528',\n",
              " 'GOT_130404_P001_WD05': 'D5_JCPintaud1452',\n",
              " 'GOT_130404_P001_WD06': 'D6_HBalslev7953',\n",
              " 'GOT_130404_P001_WD07': 'D7_HBalslev7284',\n",
              " 'GOT_130404_P001_WD08': 'D8_FBorchsenius629',\n",
              " 'GOT_130404_P001_WD09': 'D9_HBalslev6780',\n",
              " 'GOT_130404_P001_WD10': 'D10_TEmilio801',\n",
              " 'GOT_130404_P001_WD11': 'D11_Samnetal05',\n",
              " 'GOT_130404_P001_WD12': 'D12_JRoncal370',\n",
              " 'GOT_130404_P001_WE01': 'E1_JRoncal414',\n",
              " 'GOT_130404_P001_WE02': 'E2_FTG71375D',\n",
              " 'GOT_130404_P001_WE03': 'E3_FTG82492D',\n",
              " 'GOT_130404_P001_WE04': 'E4_Samnetal05',\n",
              " 'GOT_130404_P001_WE05': 'E5_HodNoID',\n",
              " 'GOT_130404_P001_WE06': 'E6_JCPintaud1490',\n",
              " 'GOT_130404_P001_WE07': 'E7_RBernal4989',\n",
              " 'GOT_130404_P001_WE08': 'E8_LANuezsn',\n",
              " 'GOT_130404_P001_WE09': 'E9_RBernal4945',\n",
              " 'GOT_130404_P001_WE10': 'E10_MFT703',\n",
              " 'GOT_130404_P001_WE11': 'E11_PintaudENV1185',\n",
              " 'GOT_130404_P001_WE12': 'E12_CFarney1752',\n",
              " 'GOT_130404_P001_WF01': 'F1_TEmilio761',\n",
              " 'GOT_130404_P001_WF02': 'F2_MJJansenJacobs2386',\n",
              " 'GOT_130404_P001_WF03': 'F3_JRoncal316',\n",
              " 'GOT_130404_P001_WF04': 'F4_JJGranville11989',\n",
              " 'GOT_130404_P001_WF05': 'F5_HBalslev6418',\n",
              " 'GOT_130404_P001_WF06': 'F6_JRoncal358',\n",
              " 'GOT_130404_P001_WF07': 'F7_FBorchsenius334',\n",
              " 'GOT_130404_P001_WF08': 'F8_HBalslev7273',\n",
              " 'GOT_130404_P001_WF09': 'F9_TEmilio714',\n",
              " 'GOT_130404_P001_WF10': 'F10_GloriaGaleano9186',\n",
              " 'GOT_130404_P001_WF11': 'F11_JRoncal403',\n",
              " 'GOT_130404_P001_WF12': 'F12_TEmilio731',\n",
              " 'GOT_130404_P001_WG01': 'G1_BG534',\n",
              " 'GOT_130404_P001_WG02': 'G2_RBernal4621',\n",
              " 'GOT_130404_P001_WG03': 'G3_RBernal4469',\n",
              " 'GOT_130404_P001_WG04': 'G4_TEmilio820',\n",
              " 'GOT_130404_P001_WG05': 'G5_TEmilio780',\n",
              " 'GOT_130404_P001_WG06': 'G6_3757',\n",
              " 'GOT_130404_P001_WG07': 'G7_RBernal4955',\n",
              " 'GOT_130404_P001_WG08': 'G8_RBernal4462',\n",
              " 'GOT_130404_P001_WG09': 'G9_MFT704',\n",
              " 'GOT_130404_P001_WG10': 'G10_JCP1454',\n",
              " 'GOT_130404_P001_WG11': 'G11_HBalslev8114',\n",
              " 'GOT_130404_P001_WG12': 'G12_MJJansenJacobs2387',\n",
              " 'GOT_130404_P001_WH01': 'H1_RodrigoCmaraLeret1711',\n",
              " 'GOT_130404_P001_WH02': 'H2_Austria80',\n",
              " 'GOT_130404_P001_WH03': 'H3_TCouvreur119',\n",
              " 'GOT_130404_P001_WH04': 'H4_FBorchsenius659',\n",
              " 'GOT_130404_P001_WH05': 'H5_AHenderson1623',\n",
              " 'GOT_130404_P001_WH06': 'H6_Kwood7367',\n",
              " 'GOT_130404_P001_WH07': 'H7_JVM1Niau',\n",
              " 'GOT_130404_P001_WH08': 'H8_SPPerlman18712',\n",
              " 'GOT_130404_P001_WH09': 'H9_JVM2Makatea',\n",
              " 'GOT_130404_P001_WH10': 'H10_MPGriffith388',\n",
              " 'GOT_130404_P001_WH11': 'H11_HBalslev8170',\n",
              " 'GOT_130404_P001_WH12': 'H12_MPGriffith392',\n",
              " 'GOT_130404_P002_WA01': 'A1_MEason1142',\n",
              " 'GOT_130404_P002_WA02': 'A2_HBalslev8166',\n",
              " 'GOT_130404_P002_WA03': 'A3_591200D',\n",
              " 'GOT_130404_P002_WA04': 'A4_GG8119',\n",
              " 'GOT_130404_P002_WA05': 'A5_RBernal2493',\n",
              " 'GOT_130404_P002_WA06': 'A6_HBalslev7298',\n",
              " 'GOT_130404_P002_WA07': 'A7_Noblick5560',\n",
              " 'GOT_130404_P002_WA08': 'A8_Austria81',\n",
              " 'GOT_130404_P002_WA09': 'A9_HBalslev6462',\n",
              " 'GOT_130404_P002_WA10': 'A10_RBernal2487',\n",
              " 'GOT_130404_P002_WA11': 'A11_RBernal2838',\n",
              " 'GOT_130404_P002_WA12': 'A12_RBernal2825',\n",
              " 'GOT_130404_P002_WB01': 'B1_TEmilio810',\n",
              " 'GOT_130404_P002_WB02': 'B2_RBernal2830',\n",
              " 'GOT_130404_P002_WB03': 'B3_RBernal4737',\n",
              " 'GOT_130404_P002_WB04': 'B4_RBernal4726',\n",
              " 'GOT_130404_P002_WB05': 'B5_RBernal4775',\n",
              " 'GOT_130404_P002_WB06': 'B6_HBalslev6545',\n",
              " 'GOT_130404_P002_WB07': 'B7_ABarford60538',\n",
              " 'GOT_130404_P002_WB08': 'B8_RBernal3481',\n",
              " 'GOT_130404_P002_WB09': 'B9_RBernal2498',\n",
              " 'GOT_130404_P002_WB10': 'B10_AJHenderson2045',\n",
              " 'GOT_130404_P002_WB11': 'B11_RBernal2883',\n",
              " 'GOT_130404_P002_WB12': 'B12_RBernal2491',\n",
              " 'GOT_130404_P002_WC01': 'C1_RBernal2190',\n",
              " 'GOT_130404_P002_WC02': 'C2_RBernal2500',\n",
              " 'GOT_130404_P002_WC03': 'C3_JCPintaud1488'}"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 20
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rH8gRCEHCVMk",
        "outputId": "b2826e49-c8d3-4d20-8c62-0fbebfed5fbc"
      },
      "source": [
        "# you can access the values for a particular key like this\n",
        "# dictionaries are a cool object that R doesn't have..jejeje\n",
        "# GOT_13etc is the key\n",
        "print(id_spp_dict['GOT_130404_P001_WA01'])\n",
        "\n",
        "# to access only the species:\n",
        "print('\\n',id_spp_dict['GOT_130404_P001_WA01'][0]) # python, unlike R, is zero-index. The first item on anything has index 0\n",
        "\n",
        "# to access only the voucher:\n",
        "print('\\n',id_spp_dict['GOT_130404_P001_WA01'][1])"
      ],
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "('Acoelorrhaphe_wrightii', 'H. Balslev 8156')\n",
            "\n",
            " Acoelorrhaphe_wrightii\n",
            "\n",
            " H. Balslev 8156\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 779
        },
        "id": "ekf8vKl_CVPY",
        "outputId": "63972b27-dd70-4684-c100-6155d0f154f0"
      },
      "source": [
        "# creating a column with species and voucher in the rapidgenomics dataframe\n",
        "seqprapid.columns # print them to make sure you aren't replacing any column\n",
        "\n",
        "seqprapid['taxspecies'] = ''\n",
        "seqprapid['voucher'] = ''\n",
        "seqprapid['taxgenus'] = ''\n",
        "for row in seqprapid.itertuples():\n",
        "  seqprapid.loc[row.Index,'taxspecies'] = id_spp_dict[row.RG_Sample_Code][0]\n",
        "  seqprapid.loc[row.Index,'taxgenus'] = id_spp_dict[row.RG_Sample_Code][0].split('_')[0]\n",
        "  seqprapid.loc[row.Index,'voucher'] = id_spp_dict[row.RG_Sample_Code][1]\n",
        "seqprapid['dataset'] = 'APP'\n",
        "seqprapid['provider'] = 'Paola Lima'\n",
        "seqprapid['pi_code'] = 'PL'\n",
        "seqprapid['Continent'] = 'americas'\n",
        "seqprapid['sense'] = 'R1'\n",
        "\n",
        "seqprapid['samplecode'] = seqprapid['RG_Sample_Code'].map(id_code_dict)\n",
        "\n",
        "seqprapid['originalmetadata'] = 'GOT_130404_SampleSheet metadata_machinereadable_Paola_2020'\n",
        "seqprapid['library_index'] = seqprapid['i5_Barcode_Seq']+'+'+seqprapid['i7_Barcode_Seq']\n",
        "\n",
        "seqprapid.rename(columns = {'Sequence_Name':'filename'}, inplace = True) # uses a dictionary with keys (old name) and value (new name)\n",
        "seqprapid"
      ],
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>RG_Sample_Code</th>\n",
              "      <th>Customer_Code</th>\n",
              "      <th>i5_Barcode_Seq</th>\n",
              "      <th>i7_Barcode_Seq</th>\n",
              "      <th>filename</th>\n",
              "      <th>Sequencing_Cycle</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>Continent</th>\n",
              "      <th>sense</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>library_index</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>GOT_130404_P001_WA01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>GTGTTCTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Acoelorrhaphe_wrightii</td>\n",
              "      <td>H. Balslev 8156</td>\n",
              "      <td>Acoelorrhaphe</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>A1_HBalslev8156</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TATGCAGT+GTGTTCTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>GOT_130404_P001_WA02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>AGTCACTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_bicornis</td>\n",
              "      <td>J. West s.n.</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>A2_JWestsn</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TATGCAGT+AGTCACTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>GOT_130404_P001_WA03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>ATTGGCTC</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_chiribogensis</td>\n",
              "      <td>F. Borchsenius 652</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>A3_FBorchsenius652</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TATGCAGT+ATTGGCTC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>GOT_130404_P001_WA04</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>CAGATCTG</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA04_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_deltoidea</td>\n",
              "      <td>L.A. NuÃ±ez s.n.</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>A4_LANuezsn</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TATGCAGT+CAGATCTG</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>GOT_130404_P001_WA05</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>CCGTGAGA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA05_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_duquei</td>\n",
              "      <td>R. Bernal 2852</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>A5_RBernal2852</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TATGCAGT+CCGTGAGA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>118</th>\n",
              "      <td>GOT_130404_P002_WB11</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>ACCACTGT</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WB11_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_praemorsa</td>\n",
              "      <td>R. Bernal 2883</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>B11_RBernal2883</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TACTCCTT+ACCACTGT</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>119</th>\n",
              "      <td>GOT_130404_P002_WB12</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>ACAGATTC</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WB12_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_quinaria</td>\n",
              "      <td>R. Bernal 2491</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>B12_RBernal2491</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TACTCCTT+ACAGATTC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>120</th>\n",
              "      <td>GOT_130404_P002_WC01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGAACAA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_radiata</td>\n",
              "      <td>R. Bernal 2190</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>C1_RBernal2190</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TACTCCTT+TGGAACAA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121</th>\n",
              "      <td>GOT_130404_P002_WC02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGTGGTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_verruculosa</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>C2_RBernal2500</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TACTCCTT+TGGTGGTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>122</th>\n",
              "      <td>GOT_130404_P002_WC03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>CGAACTTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Mauritiella_armata</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>Mauritiella</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>C3_JCPintaud1488</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TACTCCTT+CGAACTTA</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>123 rows Ã— 17 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "           RG_Sample_Code  ...      library_index\n",
              "0    GOT_130404_P001_WA01  ...  TATGCAGT+GTGTTCTA\n",
              "1    GOT_130404_P001_WA02  ...  TATGCAGT+AGTCACTA\n",
              "2    GOT_130404_P001_WA03  ...  TATGCAGT+ATTGGCTC\n",
              "3    GOT_130404_P001_WA04  ...  TATGCAGT+CAGATCTG\n",
              "4    GOT_130404_P001_WA05  ...  TATGCAGT+CCGTGAGA\n",
              "..                    ...  ...                ...\n",
              "118  GOT_130404_P002_WB11  ...  TACTCCTT+ACCACTGT\n",
              "119  GOT_130404_P002_WB12  ...  TACTCCTT+ACAGATTC\n",
              "120  GOT_130404_P002_WC01  ...  TACTCCTT+TGGAACAA\n",
              "121  GOT_130404_P002_WC02  ...  TACTCCTT+TGGTGGTA\n",
              "122  GOT_130404_P002_WC03  ...  TACTCCTT+CGAACTTA\n",
              "\n",
              "[123 rows x 17 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 22
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MlCxAxl3IzcQ",
        "outputId": "8d5084ef-78a7-410c-a8ee-ef96fc6bab4e"
      },
      "source": [
        "# duplicate the dataframe to make up for the file with reverse files\n",
        "seqprapidR2 = seqprapid.copy(deep = True)\n",
        "seqprapidR2['filename'] = seqprapidR2['filename'].str.replace('_R1_','_R2_')\n",
        "seqprapidR2['sense'] = seqprapidR2['sense'].str.replace('R1','R2')\n",
        "\n",
        "list(seqprapidR2['filename'])"
      ],
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['RAPiD-Genomics_F147_GOT_130404_P001_WA01_i5-514_i7-59_S42_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA02_i5-514_i7-27_S43_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA03_i5-514_i7-82_S44_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA04_i5-514_i7-7_S45_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA05_i5-514_i7-38_S46_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA06_i5-514_i7-74_S47_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA07_i5-514_i7-77_S48_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA08_i5-514_i7-16_S49_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA09_i5-514_i7-36_S50_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA10_i5-514_i7-54_S51_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA11_i5-514_i7-25_S52_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA12_i5-514_i7-23_S53_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB01_i5-514_i7-31_S54_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB02_i5-514_i7-39_S55_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB03_i5-514_i7-2_S56_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB04_i5-514_i7-20_S57_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB05_i5-514_i7-90_S58_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB06_i5-514_i7-6_S59_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB07_i5-514_i7-29_S60_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB08_i5-514_i7-57_S61_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB09_i5-514_i7-63_S62_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB10_i5-514_i7-55_S63_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB11_i5-514_i7-5_S64_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB12_i5-514_i7-75_S65_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC01_i5-514_i7-65_S66_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC02_i5-514_i7-67_S67_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC03_i5-514_i7-40_S68_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC04_i5-514_i7-19_S69_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC05_i5-514_i7-81_S70_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC06_i5-514_i7-3_S71_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC07_i5-514_i7-62_S72_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC08_i5-514_i7-17_S73_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC09_i5-514_i7-93_S74_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC10_i5-514_i7-47_S75_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC11_i5-514_i7-70_S76_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC12_i5-514_i7-43_S77_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD01_i5-514_i7-52_S78_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD02_i5-514_i7-72_S79_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD03_i5-514_i7-30_S80_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD04_i5-514_i7-91_S81_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD05_i5-514_i7-71_S82_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD06_i5-514_i7-50_S83_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD07_i5-514_i7-86_S84_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD08_i5-514_i7-80_S85_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD09_i5-514_i7-64_S86_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD10_i5-514_i7-8_S87_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD11_i5-514_i7-58_S88_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD12_i5-514_i7-9_S89_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE01_i5-514_i7-48_S90_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE02_i5-514_i7-89_S91_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE03_i5-514_i7-87_S92_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE04_i5-514_i7-21_S93_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE05_i5-514_i7-94_S94_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE06_i5-514_i7-35_S95_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE07_i5-514_i7-44_S96_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE08_i5-514_i7-4_S97_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE09_i5-514_i7-24_S98_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE10_i5-514_i7-61_S99_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE11_i5-514_i7-13_S100_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE12_i5-514_i7-76_S101_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF01_i5-514_i7-28_S102_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF02_i5-514_i7-78_S103_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF03_i5-514_i7-79_S104_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF04_i5-514_i7-45_S150_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF05_i5-514_i7-92_S105_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF06_i5-514_i7-60_S106_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF07_i5-514_i7-33_S107_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF08_i5-514_i7-53_S108_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF09_i5-514_i7-46_S109_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF10_i5-514_i7-26_S110_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF11_i5-514_i7-51_S111_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF12_i5-514_i7-66_S112_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG01_i5-514_i7-84_S113_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG02_i5-514_i7-12_S114_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG03_i5-514_i7-22_S115_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG04_i5-514_i7-10_S116_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG05_i5-514_i7-69_S117_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG06_i5-514_i7-14_S118_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG07_i5-514_i7-18_S119_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG08_i5-514_i7-68_S151_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG09_i5-514_i7-41_S120_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG10_i5-514_i7-37_S121_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG11_i5-514_i7-96_S122_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG12_i5-514_i7-42_S123_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH01_i5-514_i7-11_S124_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH02_i5-514_i7-32_S125_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH03_i5-514_i7-88_S126_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH04_i5-514_i7-73_S127_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH05_i5-514_i7-15_S128_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH06_i5-514_i7-49_S129_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH07_i5-514_i7-1_S130_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH08_i5-514_i7-85_S131_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH09_i5-514_i7-83_S132_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH10_i5-514_i7-34_S133_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH11_i5-514_i7-95_S134_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH12_i5-514_i7-56_S152_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA01_i5-515_i7-59_S135_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA02_i5-515_i7-27_S136_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA03_i5-515_i7-82_S137_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA04_i5-515_i7-7_S138_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA05_i5-515_i7-38_S139_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA06_i5-515_i7-74_S140_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA07_i5-515_i7-77_S141_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA08_i5-515_i7-16_S142_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA09_i5-515_i7-36_S143_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA10_i5-515_i7-54_S144_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA11_i5-515_i7-25_S145_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA12_i5-515_i7-23_S146_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB01_i5-515_i7-31_S147_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB02_i5-515_i7-39_S148_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB03_i5-515_i7-2_S149_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB04_i5-515_i7-20_S153_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB05_i5-515_i7-90_S154_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB06_i5-515_i7-6_S155_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB07_i5-515_i7-29_S156_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB08_i5-515_i7-57_S157_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB09_i5-515_i7-63_S158_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB10_i5-515_i7-55_S159_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB11_i5-515_i7-5_S160_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB12_i5-515_i7-75_S161_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WC01_i5-515_i7-65_S162_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-515_i7-67_S163_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-515_i7-40_S164_L001_R2_001.fastq.gz']"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 23
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 779
        },
        "id": "DFo3wwcVKZ5H",
        "outputId": "6ad623ba-b57b-4e37-8886-38c1aaf737f2"
      },
      "source": [
        "# concatenate dataframes for forward and reverse files and sort by filename\n",
        "seqprapidall = pd.concat([seqprapid, seqprapidR2]).sort_values(by = 'filename').reset_index()\n",
        "seqprapidall"
      ],
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>index</th>\n",
              "      <th>RG_Sample_Code</th>\n",
              "      <th>Customer_Code</th>\n",
              "      <th>i5_Barcode_Seq</th>\n",
              "      <th>i7_Barcode_Seq</th>\n",
              "      <th>filename</th>\n",
              "      <th>Sequencing_Cycle</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>Continent</th>\n",
              "      <th>sense</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>library_index</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>GOT_130404_P001_WA01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>GTGTTCTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Acoelorrhaphe_wrightii</td>\n",
              "      <td>H. Balslev 8156</td>\n",
              "      <td>Acoelorrhaphe</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>A1_HBalslev8156</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TATGCAGT+GTGTTCTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>0</td>\n",
              "      <td>GOT_130404_P001_WA01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>GTGTTCTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Acoelorrhaphe_wrightii</td>\n",
              "      <td>H. Balslev 8156</td>\n",
              "      <td>Acoelorrhaphe</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R2</td>\n",
              "      <td>A1_HBalslev8156</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TATGCAGT+GTGTTCTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>AGTCACTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_bicornis</td>\n",
              "      <td>J. West s.n.</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>A2_JWestsn</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TATGCAGT+AGTCACTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>AGTCACTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_bicornis</td>\n",
              "      <td>J. West s.n.</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R2</td>\n",
              "      <td>A2_JWestsn</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TATGCAGT+AGTCACTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P001_WA03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>ATTGGCTC</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_chiribogensis</td>\n",
              "      <td>F. Borchsenius 652</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>A3_FBorchsenius652</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TATGCAGT+ATTGGCTC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>241</th>\n",
              "      <td>120</td>\n",
              "      <td>GOT_130404_P002_WC01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGAACAA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_radiata</td>\n",
              "      <td>R. Bernal 2190</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R2</td>\n",
              "      <td>C1_RBernal2190</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TACTCCTT+TGGAACAA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>242</th>\n",
              "      <td>121</td>\n",
              "      <td>GOT_130404_P002_WC02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGTGGTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_verruculosa</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>C2_RBernal2500</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TACTCCTT+TGGTGGTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>243</th>\n",
              "      <td>121</td>\n",
              "      <td>GOT_130404_P002_WC02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGTGGTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_verruculosa</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R2</td>\n",
              "      <td>C2_RBernal2500</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TACTCCTT+TGGTGGTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>244</th>\n",
              "      <td>122</td>\n",
              "      <td>GOT_130404_P002_WC03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>CGAACTTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Mauritiella_armata</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>Mauritiella</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>C3_JCPintaud1488</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TACTCCTT+CGAACTTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>245</th>\n",
              "      <td>122</td>\n",
              "      <td>GOT_130404_P002_WC03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>CGAACTTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Mauritiella_armata</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>Mauritiella</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R2</td>\n",
              "      <td>C3_JCPintaud1488</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>TACTCCTT+CGAACTTA</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>246 rows Ã— 18 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     index  ...      library_index\n",
              "0        0  ...  TATGCAGT+GTGTTCTA\n",
              "1        0  ...  TATGCAGT+GTGTTCTA\n",
              "2        1  ...  TATGCAGT+AGTCACTA\n",
              "3        1  ...  TATGCAGT+AGTCACTA\n",
              "4        2  ...  TATGCAGT+ATTGGCTC\n",
              "..     ...  ...                ...\n",
              "241    120  ...  TACTCCTT+TGGAACAA\n",
              "242    121  ...  TACTCCTT+TGGTGGTA\n",
              "243    121  ...  TACTCCTT+TGGTGGTA\n",
              "244    122  ...  TACTCCTT+CGAACTTA\n",
              "245    122  ...  TACTCCTT+CGAACTTA\n",
              "\n",
              "[246 rows x 18 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 24
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hIz2L-PyIRl0"
      },
      "source": [
        "Now, we can concatenate the dataframes"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 694
        },
        "id": "lq-9XIitH-fB",
        "outputId": "388a3ab8-cd68-477f-b20a-1aeef1cce8e6"
      },
      "source": [
        "df = pd.concat([df, seqprapidall[['filename','taxspecies', 'voucher','taxgenus',\n",
        "       'dataset', 'provider', 'pi_code', 'Continent', 'sense',\n",
        "       'originalmetadata', 'samplecode', 'library_index' ]]])\n",
        "df"
      ],
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>filename</th>\n",
              "      <th>sense</th>\n",
              "      <th>filecode</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "      <th>botanic_garden</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>ifmorphotype</th>\n",
              "      <th>ifpopulation</th>\n",
              "      <th>sent_to_Cano</th>\n",
              "      <th>newfilename</th>\n",
              "      <th>raw_reads</th>\n",
              "      <th>reads_trimed_paired_flag</th>\n",
              "      <th>reads_trimed_paired</th>\n",
              "      <th>reads_trimed_single</th>\n",
              "      <th>percentage_lost_tosingles</th>\n",
              "      <th>pcr_filtered</th>\n",
              "      <th>collection_year</th>\n",
              "      <th>Continent</th>\n",
              "      <th>country</th>\n",
              "      <th>long</th>\n",
              "      <th>lat</th>\n",
              "      <th>flag</th>\n",
              "      <th>notes</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AKG91_P3252_1182_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AKG91_P3252_1182</td>\n",
              "      <td>AKG91_1182</td>\n",
              "      <td>CTAGCT</td>\n",
              "      <td>JBP</td>\n",
              "      <td>Lorenzi_&amp;_Soares_6762</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_emensis</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_eme_AKG911182_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>241</th>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC01_i5-51...</td>\n",
              "      <td>R2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>C1_RBernal2190</td>\n",
              "      <td>TACTCCTT+TGGAACAA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>R. Bernal 2190</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>Wettinia_radiata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>242</th>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>R1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>C2_RBernal2500</td>\n",
              "      <td>TACTCCTT+TGGTGGTA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>Wettinia_verruculosa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>243</th>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>R2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>C2_RBernal2500</td>\n",
              "      <td>TACTCCTT+TGGTGGTA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>Wettinia_verruculosa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>244</th>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>R1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>C3_JCPintaud1488</td>\n",
              "      <td>TACTCCTT+CGAACTTA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>Mauritiella</td>\n",
              "      <td>Mauritiella_armata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>245</th>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>R2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>C3_JCPintaud1488</td>\n",
              "      <td>TACTCCTT+CGAACTTA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>Mauritiella</td>\n",
              "      <td>Mauritiella_armata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3295 rows Ã— 30 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "    dataset     provider pi_code  ... lat flag notes\n",
              "0      Cano  Angela Cano      AC  ... NaN  NaN   NaN\n",
              "1      Cano  Angela Cano      AC  ... NaN  NaN   NaN\n",
              "2      Cano  Angela Cano      AC  ... NaN  NaN   NaN\n",
              "3      Cano  Angela Cano      AC  ... NaN  NaN   NaN\n",
              "4      Cano  Angela Cano      AC  ... NaN  NaN   NaN\n",
              "..      ...          ...     ...  ...  ..  ...   ...\n",
              "241     APP   Paola Lima      PL  ... NaN  NaN   NaN\n",
              "242     APP   Paola Lima      PL  ... NaN  NaN   NaN\n",
              "243     APP   Paola Lima      PL  ... NaN  NaN   NaN\n",
              "244     APP   Paola Lima      PL  ... NaN  NaN   NaN\n",
              "245     APP   Paola Lima      PL  ... NaN  NaN   NaN\n",
              "\n",
              "[3295 rows x 30 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 25
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8tmr99mGLlGV"
      },
      "source": [
        "## 5. Counting again\n",
        "\n",
        "Let's count species again"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GQZ5zBZYCVRt",
        "outputId": "e8353346-0d5a-4afe-bc1e-a1191eb84c45"
      },
      "source": [
        "# first, we create sets of items to compare, using list comprehention. Sets should be defined within curly brackets\n",
        "sppseqed = {spp for spp in df[df['Continent'] == 'americas']['taxspecies'].unique()} #list comprehension are loops of a single line\n",
        "apspp = {spp for spp in palms_americas} # american palm species\n",
        "# we can apply set theory to find the differences (not sequenced or sequenced but not in the list of american palms)\n",
        "# items in setA that are not shared with setB and viceversa\n",
        "# symmetric difference \n",
        "print(apspp - sppseqed) # niceee, nothing is missing\n",
        "sppmissing = len(apspp - sppseqed)\n",
        "print('\\nAlleged missing species: %s'%(sppmissing))\n",
        "print('Percentage of species missing: %s'%((sppmissing * 100)/len(palms_americas)))\n",
        "print('Percentage of species sequenced: %s'%(100-(sppmissing * 100)/len(palms_americas)))"
      ],
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "{'Desmoncus_latisectus', 'Bactris_ptariana', 'Attalea_exigua', 'Attalea_compta', 'Coccothrinax_pseudorigida', 'Bactris_ferruginea', 'Astrocaryum_rodriguesii', 'Coccothrinax_pauciramosa', 'Bactris_bahiensis', 'Desmoncus_setosus', 'Coccothrinax_alexandri', 'Bactris_polystachya', 'Desmoncus_stans', 'Attalea_salvadorensis', 'Attalea_osmantha', 'Euterpe_luminosa', 'Desmoncus_pumilus', 'Prestoea_carderi', 'Coccothrinax_camagueyana', 'Sabal_domingensis', 'Attalea_camopiensis', 'Bactris_faucium', 'Acoelorraphe_wrightii', 'Allagoptera_campestris', 'Syagrus_emasensis', 'Desmoncus_obovoideus', 'Pholidostachys_amazonensis', 'Astrocaryum_confertum', 'Chamaedorea_amabilis', 'Copernicia_roigii', 'Bactris_acanthocarpa', 'Attalea_fairchildensis', 'Desmoncus_interjectus', 'Bactris_vulgaris', 'Bactris_hondurensis', 'Bactris_sphaerocarpa', 'Leopoldinia_piassaba', 'Bactris_pickelii', 'Bactris_glassmanii', 'Copernicia_brittonorum', 'Bactris_kunorum', 'Bactris_grayumii', 'Roystonea_stellata', 'Geonoma_poiteauana', 'Aiphanes_stergiosii', 'Desmoncus_leptoclonos', 'Butia_stolonifera', 'Roystonea_violacea', 'Copernicia_humicola', 'Bactris_hatschbachii', 'Coccothrinax_acunana', 'Phytelephas_seemannii', 'Guihaia_lancifolia', 'Coccothrinax_leonis', 'Coccothrinax_fagildei', 'Butia_missionera', 'Astrocaryum_farinosum', 'Chamaedorea_whitelockiana', 'Calyptrogyne_pubescens', 'Geonoma_spinescens', 'Syagrus_pompeoi', 'Attalea_maracaibensis', 'Coccothrinax_victorini', 'Chamaedorea_undulatifolia', 'Coccothrinax_pumila', 'Hyospathe_macrorhachis', 'Chamaedorea_christinae', 'Brahea_pimo', 'Cryosophila_cookii', 'Desmoncus_madrensis', 'Attalea_cephalotus', 'Geonoma_ferruginea', 'Attalea_huebneri', 'Lytocaryum_hoehnei', 'Dictyocaryum_fuscum', 'Butia_quaraimana', 'Wettinia_minima', 'Desmoncus_prunifer', 'Lytocaryum_itapebiense', 'Attalea_magdalenica', 'Copernicia_oxycalyx', 'Coccothrinax_garciana', 'Coccothrinax_rigida', 'Attalea_maripensis', 'Geonoma_schizocarpa', 'Colpothrinax_cookii', 'Calyptrogyne_coloradensis', 'Syagrus_lilliputiana', 'Brahea_decumbens', 'Guihaia_grossifibrosa', 'Bactris_oligoclada', 'Bactris_herrerana', 'Brahea_edulis', 'Brahea_calcarea', 'Copernicia_curbeloi', 'Calyptrogyne_deneversii', 'Chamaedorea_ernesti-augusti', 'Coccothrinax_boschiana', 'Geonoma_tenuissima', 'Bactris_jamaicana', 'Astrocaryum_aculeatissimum', 'Chamaedorea_serpens', 'Dictyocaryum_ptarianum', 'Geonoma_bernalii', 'Geonoma_chococola', 'Acrocomia_hassleri', 'Chamaedorea_rosibeliae', 'Attalea_dahlgreniana', 'Syagrus_cearensis', 'Geonoma_sanmartinensis', 'Syagrus_pimentae', 'Thrinax_excelsa', 'Attalea_rostrata', 'Aiphanes_spicata', 'Brahea_moorei', 'Bactris_ana-juliae', 'Aiphanes_acanthophylla', 'Attalea_weberbaueri', 'Geonoma_operculata', 'Mauritia_carana', 'Geonoma_divisa', 'Attalea_salazarii', 'Bactris_tefensis', 'Geonoma_deneversii', 'Syagrus_werdermannii', 'Butia_marmorii', 'Chamaedorea_smithii', 'Bactris_caudata', 'Coccothrinax_bermudezii', 'Cryosophila_macrocarpa', 'Socratea_montana', 'Trithrinax_acanthocoma', 'Bactris_turbinocarpa', 'Attalea_peruviana', 'Brahea_brandegeei', 'Reinhardtia_latisecta', 'Cryosophila_grayumii', 'Reinhardtia_paiewonskiana', 'Chamaedorea_queroana', 'Chamaedorea_pachecoana', 'Bactris_rostrata', 'Aiphanes_multiplex', 'Coccothrinax_scoparia', 'Calyptrogyne_sanblasensis', 'Chamaedorea_atrovirens', 'Bactris_horridispatha', 'Attalea_blepharopus', 'Desmoncus_parvulus', 'Attalea_macrolepis', 'Attalea_septuagenata', 'Coccothrinax_moaensis', 'Chamaedorea_schiedeana', 'Euterpe_longibracteata', 'Copernicia_cowellii', 'Bactris_timbuiensis', 'Asterogyne_spicata', 'Chamaedorea_latisecta', 'Coccothrinax_nipensis', 'Calyptrogyne_herrerae', 'Hyospathe_wendlandiana', 'Manicaria_martiana', 'Coccothrinax_muricata', 'Attalea_rhynchocarpa', 'Attalea_kewensis', 'Geonoma_peruviana', 'Attalea_bassleriana', 'Phytelephas_schottii', 'Calyptrogyne_osensis', 'Desmoncus_moorei', 'Geonoma_talamancana', 'Desmoncus_horridus', 'Hemithrinax_rivularis', 'Attalea_amylacea', 'Desmoncus_loretanus', 'Geonoma_concinnoidea', 'Butia_catarinensis', 'Oenocarpus_circumtextus', 'Desmoncus_chinantlensis', 'Guihaia_argyrata', 'Syagrus_cataphracta', 'Bactris_cubensis', 'Lytocaryum_weddellianum', 'Lytocaryum_insigne', 'Desmoncus_costaricensis', 'Bactris_tomentosa', 'Geonoma_scoparia', 'Coccothrinax_microphylla', 'Syagrus_yungasensis', 'Attalea_wesselsboeri', 'Geonoma_longipedunculata', 'Geonoma_aspidiifolia', 'Bactris_rhaphidacantha', 'Bactris_dianeura', 'Calyptrogyne_condensata', 'Roystonea_dunlapiana', 'Prestoea_simplicifolia', 'Acrocomia_media', 'Desmoncus_osensis', 'Bactris_soeiroana', 'Coccothrinax_yunquensis', 'Geonoma_brenesii', 'Coccothrinax_yuraguana', 'Astrocaryum_cuatrecasanum', 'Copernicia_molineti', 'Geonoma_venosa', 'Reinhardtia_elegans', 'Attalea_degranvillei', 'Asterogyne_yaracuyense', 'Calyptrogyne_kunorum', 'Attalea_lauromuelleriana', 'Copernicia_curtissii', 'Bactris_cuspidata', 'Astrocaryum_sociale', 'Chamaedorea_castillo-montii', 'Astrocaryum_tucuma', 'Pholidostachys_occidentalis', 'Coccothrinax_trinitensis', 'Oenocarpus_makeru', 'Attalea_hoehnei', 'Butia_exilata', 'Geonoma_simplicifrons', 'Chamaedorea_angustisecta', 'Bactris_nancibaensis', 'Astrocaryum_giganteum', 'Brahea_salvadorensis', 'Hyospathe_peruviana', 'Chamaedorea_moliniana', 'Copernicia_longiglossa', 'Geonoma_dindoensis', 'Chamaedorea_foveata', 'Hemithrinax_compacta', 'Roystonea_princeps', 'Geonoma_gentryi', 'Desmoncus_kunarius', 'Coccothrinax_baracoensis', 'Syagrus_pleioclada', 'Attalea_guianensis', 'Coccothrinax_concolor', 'Prestoea_tenuiramosa', 'Geonoma_trigona', 'Geonoma_braunii'}\n",
            "\n",
            "Alleged missing species: 241\n",
            "Percentage of species missing: 29.753086419753085\n",
            "Percentage of species sequenced: 70.24691358024691\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "G4IH81SXL42z"
      },
      "source": [
        "Looks better, huh?\n",
        "We can also double check that we are not missing species because of mislabeled continents. We already have the list of american species and that column is redundant... but better to have it"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pDK-iVJZi-KA",
        "outputId": "640e7b4a-7ced-4a3e-b42d-cb49ac286669"
      },
      "source": [
        "# first, we create sets of items to compare, using list comprehention. Sets should be defined within curly brackets\n",
        "sppseqed = {spp for spp in df['taxspecies'].unique()} #list comprehension are loops of a single line\n",
        "apspp = {spp for spp in palms_americas} # american palm species\n",
        "# we can apply set theory to find the differences (not sequenced or sequenced but not in the list of american palms)\n",
        "# items in setA that are not shared with setB and viceversa\n",
        "# symmetric difference \n",
        "print(apspp - sppseqed) # niceee, nothing is missing\n",
        "sppmissing = len(apspp - sppseqed)\n",
        "print('\\nAlleged missing species: %s'%(sppmissing))\n",
        "print('Percentage of species missing: %s'%((sppmissing * 100)/len(palms_americas)))\n",
        "print('Percentage of species sequenced: %s'%(100-(sppmissing * 100)/len(palms_americas)))"
      ],
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "{'Desmoncus_latisectus', 'Bactris_ptariana', 'Attalea_exigua', 'Attalea_compta', 'Coccothrinax_pseudorigida', 'Bactris_ferruginea', 'Astrocaryum_rodriguesii', 'Coccothrinax_pauciramosa', 'Bactris_bahiensis', 'Desmoncus_setosus', 'Coccothrinax_alexandri', 'Bactris_polystachya', 'Desmoncus_stans', 'Attalea_salvadorensis', 'Attalea_osmantha', 'Euterpe_luminosa', 'Desmoncus_pumilus', 'Prestoea_carderi', 'Coccothrinax_camagueyana', 'Sabal_domingensis', 'Attalea_camopiensis', 'Bactris_faucium', 'Acoelorraphe_wrightii', 'Allagoptera_campestris', 'Syagrus_emasensis', 'Desmoncus_obovoideus', 'Pholidostachys_amazonensis', 'Astrocaryum_confertum', 'Chamaedorea_amabilis', 'Copernicia_roigii', 'Bactris_acanthocarpa', 'Attalea_fairchildensis', 'Desmoncus_interjectus', 'Bactris_vulgaris', 'Bactris_hondurensis', 'Bactris_sphaerocarpa', 'Leopoldinia_piassaba', 'Bactris_pickelii', 'Bactris_glassmanii', 'Copernicia_brittonorum', 'Bactris_kunorum', 'Bactris_grayumii', 'Roystonea_stellata', 'Geonoma_poiteauana', 'Aiphanes_stergiosii', 'Desmoncus_leptoclonos', 'Butia_stolonifera', 'Roystonea_violacea', 'Copernicia_humicola', 'Bactris_hatschbachii', 'Coccothrinax_acunana', 'Phytelephas_seemannii', 'Guihaia_lancifolia', 'Coccothrinax_leonis', 'Coccothrinax_fagildei', 'Butia_missionera', 'Astrocaryum_farinosum', 'Chamaedorea_whitelockiana', 'Calyptrogyne_pubescens', 'Geonoma_spinescens', 'Syagrus_pompeoi', 'Attalea_maracaibensis', 'Coccothrinax_victorini', 'Chamaedorea_undulatifolia', 'Coccothrinax_pumila', 'Hyospathe_macrorhachis', 'Chamaedorea_christinae', 'Brahea_pimo', 'Cryosophila_cookii', 'Desmoncus_madrensis', 'Attalea_cephalotus', 'Geonoma_ferruginea', 'Attalea_huebneri', 'Lytocaryum_hoehnei', 'Dictyocaryum_fuscum', 'Butia_quaraimana', 'Wettinia_minima', 'Desmoncus_prunifer', 'Lytocaryum_itapebiense', 'Attalea_magdalenica', 'Copernicia_oxycalyx', 'Coccothrinax_garciana', 'Coccothrinax_rigida', 'Attalea_maripensis', 'Geonoma_schizocarpa', 'Colpothrinax_cookii', 'Calyptrogyne_coloradensis', 'Syagrus_lilliputiana', 'Brahea_decumbens', 'Guihaia_grossifibrosa', 'Bactris_oligoclada', 'Bactris_herrerana', 'Brahea_edulis', 'Brahea_calcarea', 'Copernicia_curbeloi', 'Calyptrogyne_deneversii', 'Chamaedorea_ernesti-augusti', 'Coccothrinax_boschiana', 'Geonoma_tenuissima', 'Bactris_jamaicana', 'Astrocaryum_aculeatissimum', 'Chamaedorea_serpens', 'Dictyocaryum_ptarianum', 'Geonoma_bernalii', 'Geonoma_chococola', 'Acrocomia_hassleri', 'Chamaedorea_rosibeliae', 'Attalea_dahlgreniana', 'Syagrus_cearensis', 'Geonoma_sanmartinensis', 'Syagrus_pimentae', 'Thrinax_excelsa', 'Attalea_rostrata', 'Aiphanes_spicata', 'Brahea_moorei', 'Bactris_ana-juliae', 'Aiphanes_acanthophylla', 'Attalea_weberbaueri', 'Geonoma_operculata', 'Mauritia_carana', 'Geonoma_divisa', 'Attalea_salazarii', 'Bactris_tefensis', 'Geonoma_deneversii', 'Syagrus_werdermannii', 'Butia_marmorii', 'Chamaedorea_smithii', 'Bactris_caudata', 'Coccothrinax_bermudezii', 'Cryosophila_macrocarpa', 'Socratea_montana', 'Trithrinax_acanthocoma', 'Bactris_turbinocarpa', 'Attalea_peruviana', 'Brahea_brandegeei', 'Reinhardtia_latisecta', 'Cryosophila_grayumii', 'Reinhardtia_paiewonskiana', 'Chamaedorea_queroana', 'Chamaedorea_pachecoana', 'Bactris_rostrata', 'Aiphanes_multiplex', 'Coccothrinax_scoparia', 'Calyptrogyne_sanblasensis', 'Chamaedorea_atrovirens', 'Bactris_horridispatha', 'Attalea_blepharopus', 'Desmoncus_parvulus', 'Attalea_macrolepis', 'Attalea_septuagenata', 'Coccothrinax_moaensis', 'Chamaedorea_schiedeana', 'Euterpe_longibracteata', 'Copernicia_cowellii', 'Bactris_timbuiensis', 'Asterogyne_spicata', 'Chamaedorea_latisecta', 'Coccothrinax_nipensis', 'Calyptrogyne_herrerae', 'Hyospathe_wendlandiana', 'Manicaria_martiana', 'Coccothrinax_muricata', 'Attalea_rhynchocarpa', 'Attalea_kewensis', 'Geonoma_peruviana', 'Attalea_bassleriana', 'Phytelephas_schottii', 'Calyptrogyne_osensis', 'Desmoncus_moorei', 'Geonoma_talamancana', 'Desmoncus_horridus', 'Hemithrinax_rivularis', 'Attalea_amylacea', 'Desmoncus_loretanus', 'Geonoma_concinnoidea', 'Butia_catarinensis', 'Oenocarpus_circumtextus', 'Desmoncus_chinantlensis', 'Guihaia_argyrata', 'Syagrus_cataphracta', 'Bactris_cubensis', 'Lytocaryum_weddellianum', 'Lytocaryum_insigne', 'Desmoncus_costaricensis', 'Bactris_tomentosa', 'Geonoma_scoparia', 'Coccothrinax_microphylla', 'Syagrus_yungasensis', 'Attalea_wesselsboeri', 'Geonoma_longipedunculata', 'Geonoma_aspidiifolia', 'Bactris_rhaphidacantha', 'Bactris_dianeura', 'Calyptrogyne_condensata', 'Roystonea_dunlapiana', 'Prestoea_simplicifolia', 'Acrocomia_media', 'Desmoncus_osensis', 'Bactris_soeiroana', 'Coccothrinax_yunquensis', 'Geonoma_brenesii', 'Coccothrinax_yuraguana', 'Astrocaryum_cuatrecasanum', 'Copernicia_molineti', 'Geonoma_venosa', 'Reinhardtia_elegans', 'Attalea_degranvillei', 'Asterogyne_yaracuyense', 'Calyptrogyne_kunorum', 'Attalea_lauromuelleriana', 'Copernicia_curtissii', 'Bactris_cuspidata', 'Astrocaryum_sociale', 'Chamaedorea_castillo-montii', 'Astrocaryum_tucuma', 'Pholidostachys_occidentalis', 'Coccothrinax_trinitensis', 'Oenocarpus_makeru', 'Attalea_hoehnei', 'Butia_exilata', 'Geonoma_simplicifrons', 'Chamaedorea_angustisecta', 'Bactris_nancibaensis', 'Astrocaryum_giganteum', 'Brahea_salvadorensis', 'Hyospathe_peruviana', 'Chamaedorea_moliniana', 'Copernicia_longiglossa', 'Geonoma_dindoensis', 'Chamaedorea_foveata', 'Hemithrinax_compacta', 'Roystonea_princeps', 'Geonoma_gentryi', 'Desmoncus_kunarius', 'Coccothrinax_baracoensis', 'Syagrus_pleioclada', 'Attalea_guianensis', 'Coccothrinax_concolor', 'Prestoea_tenuiramosa', 'Geonoma_trigona', 'Geonoma_braunii'}\n",
            "\n",
            "Alleged missing species: 241\n",
            "Percentage of species missing: 29.753086419753085\n",
            "Percentage of species sequenced: 70.24691358024691\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7i6Y9yx4MOL5"
      },
      "source": [
        "well.. same thing. But it is always GOOD to double check things.\n",
        "\n",
        "The oher thing is that we have one species set as 'nan'. I know already this is a sample for which the species wasn;t determined in the voucher or was a mismated sample that we could recover via phylogenies. Let's check it"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7eTq9wFcOxCQ",
        "outputId": "07c26fb8-2225-4095-9ee6-bbe8b521caf2"
      },
      "source": [
        "df[df['taxspecies'].isna()][['pi_code','filename','samplecode','taxgenus','taxspecies']]\n",
        "df[df['taxspecies'].isna()]['pi_code'].unique() # TC are Raphias, not relevant here. RC are probably mauritias and overrepresented. MJS are Ceroxylon. We need to check on that but later."
      ],
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['RC', 'TC', 'MJS'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 28
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ux-bY0AqOi9I"
      },
      "source": [
        "Next thing to do are the Acrocromias. First... how many species do we have until now?"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "A77ZxZhnMXfq",
        "outputId": "82056eca-3d8c-4810-8306-b4ae540b1218"
      },
      "source": [
        "# let's ignore the nans by filtering them out\n",
        "dfall = df[~df['taxspecies'].isna()]\n",
        "dfall[dfall['taxspecies'].str.contains('Acrocomia')]['taxspecies'].unique()\n",
        "# df.columns"
      ],
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Acrocomia_aculeata', 'Acrocomia_crispa', 'Acrocomia_emensis',\n",
              "       'Acrocomia_glaucescens', 'Acrocomia_intumescens',\n",
              "       'Acrocomia_totai'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 29
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2I97mwClUH9i",
        "outputId": "45910f5a-8ea3-46b8-c05c-d49b9f1a903d"
      },
      "source": [
        "# there are how many Acrocomias in the list of american species?\n",
        "print([x for x in apspp if 'Acrocomia' in str(x)])"
      ],
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "['Acrocomia_emensis', 'Acrocomia_hassleri', 'Acrocomia_media', 'Acrocomia_glaucescens', 'Acrocomia_intumescens', 'Acrocomia_totai', 'Acrocomia_crispa', 'Acrocomia_aculeata']\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "W6l_LKcdWSW2",
        "outputId": "918ceecb-b5f2-425e-e80d-6168d6725729"
      },
      "source": [
        "# the species sent by R. Collevatti\n",
        "rc_acrocomia = ['Acrocomia_aculeata', 'Acrocomia_totai', 'Acrocomia_intumescens']\n",
        "\n",
        "# acrocomias sequenced\n",
        "seqacroc = rc_acrocomia + list(dfall[dfall['taxspecies'].str.contains('Acrocomia')]['taxspecies'].unique())\n",
        "print('Sequenced: ',seqacroc)\n",
        "print('\\nAcrocomias missing: ',[x for x in [x for x in apspp if 'Acrocomia' in str(x)] if x not in seqacroc])"
      ],
      "execution_count": 31,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Sequenced:  ['Acrocomia_aculeata', 'Acrocomia_totai', 'Acrocomia_intumescens', 'Acrocomia_aculeata', 'Acrocomia_crispa', 'Acrocomia_emensis', 'Acrocomia_glaucescens', 'Acrocomia_intumescens', 'Acrocomia_totai']\n",
            "\n",
            "Acrocomias missing:  ['Acrocomia_hassleri', 'Acrocomia_media']\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7wAM2nvcYEUF"
      },
      "source": [
        "For now, and only to keep track of the species sequenced and check what's missing. But you will have to add the files, vouchers, metadata on the dataframe to keep the metadata UPDATED"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DRp4YitLdoQy",
        "outputId": "21e31c33-56f1-420a-af2f-f4a6653495ea"
      },
      "source": [
        "df.columns"
      ],
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['dataset', 'provider', 'pi_code', 'originalmetadata', 'filename',\n",
              "       'sense', 'filecode', 'samplecode', 'library_index', 'botanic_garden',\n",
              "       'voucher', 'taxgenus', 'taxspecies', 'ifmorphotype', 'ifpopulation',\n",
              "       'sent_to_Cano', 'newfilename', 'raw_reads', 'reads_trimed_paired_flag',\n",
              "       'reads_trimed_paired', 'reads_trimed_single',\n",
              "       'percentage_lost_tosingles', 'pcr_filtered', 'collection_year',\n",
              "       'Continent', 'country', 'long', 'lat', 'flag', 'notes'],\n",
              "      dtype='object')"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 34
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 142
        },
        "id": "Cq4Ph-XTYsXj",
        "outputId": "d96e8425-601d-4537-9aee-675b8a6f40e7"
      },
      "source": [
        "# sequenced = list(df['taxspecies'].unique()) + seqacroc\n",
        "# print(len(sequenced),'\\n')\n",
        "# sequenced\n",
        "\n",
        "# the species sent by R. Collevatti\n",
        "rc_acrocomia = ['Acrocomia_aculeata', 'Acrocomia_totai', 'Acrocomia_intumescens']\n",
        "rc_acrocomia_pi = ['RC'] * len(rc_acrocomia)\n",
        "rc_acrocomia_dataset = ['AcrocomiaRC'] * len(rc_acrocomia)\n",
        "rc_acrocomia_flag = ['filesnotadded'] * len(rc_acrocomia)\n",
        "rc_acrocomia_prov= ['Rosanne Collevatti'] * len(rc_acrocomia)\n",
        "rc_acrocomia_taxgenus = [x.split('_')[0] for x in rc_acrocomia]\n",
        "\n",
        "ac = pd.DataFrame(list(zip(rc_acrocomia,rc_acrocomia_pi,rc_acrocomia_dataset,rc_acrocomia_taxgenus,rc_acrocomia_prov,rc_acrocomia_flag)),\n",
        "                  columns = ['taxspecies','pi_code','dataset','taxgenus','provider','flag'])\n",
        "ac"
      ],
      "execution_count": 40,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>dataset</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>provider</th>\n",
              "      <th>flag</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>RC</td>\n",
              "      <td>AcrocomiaRC</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Rosanne Collevatti</td>\n",
              "      <td>filesnotadded</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Acrocomia_totai</td>\n",
              "      <td>RC</td>\n",
              "      <td>AcrocomiaRC</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Rosanne Collevatti</td>\n",
              "      <td>filesnotadded</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Acrocomia_intumescens</td>\n",
              "      <td>RC</td>\n",
              "      <td>AcrocomiaRC</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Rosanne Collevatti</td>\n",
              "      <td>filesnotadded</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "              taxspecies pi_code  ...            provider           flag\n",
              "0     Acrocomia_aculeata      RC  ...  Rosanne Collevatti  filesnotadded\n",
              "1        Acrocomia_totai      RC  ...  Rosanne Collevatti  filesnotadded\n",
              "2  Acrocomia_intumescens      RC  ...  Rosanne Collevatti  filesnotadded\n",
              "\n",
              "[3 rows x 6 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 40
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 575
        },
        "id": "_sy5xv8TeT1x",
        "outputId": "258bfe04-79a2-4f77-e4d5-7179071cec08"
      },
      "source": [
        "# concatenate to main df\n",
        "df = pd.concat([df,ac])\n",
        "df"
      ],
      "execution_count": 41,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>filename</th>\n",
              "      <th>sense</th>\n",
              "      <th>filecode</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "      <th>botanic_garden</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>ifmorphotype</th>\n",
              "      <th>ifpopulation</th>\n",
              "      <th>sent_to_Cano</th>\n",
              "      <th>newfilename</th>\n",
              "      <th>raw_reads</th>\n",
              "      <th>reads_trimed_paired_flag</th>\n",
              "      <th>reads_trimed_paired</th>\n",
              "      <th>reads_trimed_single</th>\n",
              "      <th>percentage_lost_tosingles</th>\n",
              "      <th>pcr_filtered</th>\n",
              "      <th>collection_year</th>\n",
              "      <th>Continent</th>\n",
              "      <th>country</th>\n",
              "      <th>long</th>\n",
              "      <th>lat</th>\n",
              "      <th>flag</th>\n",
              "      <th>notes</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AKG91_P3252_1182_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AKG91_P3252_1182</td>\n",
              "      <td>AKG91_1182</td>\n",
              "      <td>CTAGCT</td>\n",
              "      <td>JBP</td>\n",
              "      <td>Lorenzi_&amp;_Soares_6762</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_emensis</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_eme_AKG911182_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>AcrocomiaRC</td>\n",
              "      <td>NaN</td>\n",
              "      <td>RC</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_totai</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>filesnotadded</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>AcrocomiaRC</td>\n",
              "      <td>NaN</td>\n",
              "      <td>RC</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_intumescens</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>filesnotadded</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>AcrocomiaRC</td>\n",
              "      <td>Rosanne Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>filesnotadded</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>AcrocomiaRC</td>\n",
              "      <td>Rosanne Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_totai</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>filesnotadded</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>AcrocomiaRC</td>\n",
              "      <td>Rosanne Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_intumescens</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>filesnotadded</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3301 rows Ã— 30 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "        dataset            provider pi_code  ... lat           flag notes\n",
              "0          Cano         Angela Cano      AC  ... NaN            NaN   NaN\n",
              "1          Cano         Angela Cano      AC  ... NaN            NaN   NaN\n",
              "2          Cano         Angela Cano      AC  ... NaN            NaN   NaN\n",
              "3          Cano         Angela Cano      AC  ... NaN            NaN   NaN\n",
              "4          Cano         Angela Cano      AC  ... NaN            NaN   NaN\n",
              "..          ...                 ...     ...  ...  ..            ...   ...\n",
              "1   AcrocomiaRC                 NaN      RC  ... NaN  filesnotadded   NaN\n",
              "2   AcrocomiaRC                 NaN      RC  ... NaN  filesnotadded   NaN\n",
              "0   AcrocomiaRC  Rosanne Collevatti      RC  ... NaN  filesnotadded   NaN\n",
              "1   AcrocomiaRC  Rosanne Collevatti      RC  ... NaN  filesnotadded   NaN\n",
              "2   AcrocomiaRC  Rosanne Collevatti      RC  ... NaN  filesnotadded   NaN\n",
              "\n",
              "[3301 rows x 30 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 41
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Lor2ii3zZoI1"
      },
      "source": [
        "Lastly, let's try to solve the Ceroxylon data"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "BYjmFFaSZsI1",
        "outputId": "9aa3f038-6d6f-43d1-d6ed-00da6ab49b4a"
      },
      "source": [
        "# first, R. Collevatti ones\n",
        "df[(df['taxspecies'].isna()) & (df['dataset'] == 'Mauritia')]"
      ],
      "execution_count": 44,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>filename</th>\n",
              "      <th>sense</th>\n",
              "      <th>filecode</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "      <th>botanic_garden</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>ifmorphotype</th>\n",
              "      <th>ifpopulation</th>\n",
              "      <th>sent_to_Cano</th>\n",
              "      <th>newfilename</th>\n",
              "      <th>raw_reads</th>\n",
              "      <th>reads_trimed_paired_flag</th>\n",
              "      <th>reads_trimed_paired</th>\n",
              "      <th>reads_trimed_single</th>\n",
              "      <th>percentage_lost_tosingles</th>\n",
              "      <th>pcr_filtered</th>\n",
              "      <th>collection_year</th>\n",
              "      <th>Continent</th>\n",
              "      <th>country</th>\n",
              "      <th>long</th>\n",
              "      <th>lat</th>\n",
              "      <th>flag</th>\n",
              "      <th>notes</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>2075</th>\n",
              "      <td>Mauritia</td>\n",
              "      <td>Rosane Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>Mauritia_sample_data_MFT_050119.xlsx GOI_42030...</td>\n",
              "      <td>GOI-420302-P01-WA01-i5-509-i7-59_S944_L006_R1_...</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOI-420302-P01-WA01</td>\n",
              "      <td>G302P01WA01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GOI_420302 folder same as above with different...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2076</th>\n",
              "      <td>Mauritia</td>\n",
              "      <td>Rosane Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>Mauritia_sample_data_MFT_050119.xlsx GOI_42030...</td>\n",
              "      <td>GOI-420302-P01-WA01-i5-509-i7-59_S944_L006_R2_...</td>\n",
              "      <td>R2</td>\n",
              "      <td>GOI-420302-P01-WA01</td>\n",
              "      <td>G302P01WA01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GOI_420302 folder same as above with different...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2077</th>\n",
              "      <td>Mauritia</td>\n",
              "      <td>Rosane Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>Mauritia_sample_data_MFT_050119.xlsx GOI_42030...</td>\n",
              "      <td>GOI-420302-P01-WA02-i5-509-i7-27_S945_L006_R1_...</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOI-420302-P01-WA02</td>\n",
              "      <td>G302P01WA02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GOI_420302 folder same as above with different...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2078</th>\n",
              "      <td>Mauritia</td>\n",
              "      <td>Rosane Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>Mauritia_sample_data_MFT_050119.xlsx GOI_42030...</td>\n",
              "      <td>GOI-420302-P01-WA02-i5-509-i7-27_S945_L006_R2_...</td>\n",
              "      <td>R2</td>\n",
              "      <td>GOI-420302-P01-WA02</td>\n",
              "      <td>G302P01WA02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GOI_420302 folder same as above with different...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2079</th>\n",
              "      <td>Mauritia</td>\n",
              "      <td>Rosane Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>Mauritia_sample_data_MFT_050119.xlsx GOI_42030...</td>\n",
              "      <td>GOI-420302-P01-WA03-i5-509-i7-82_S946_L006_R1_...</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOI-420302-P01-WA03</td>\n",
              "      <td>G302P01WA03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GOI_420302 folder same as above with different...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2454</th>\n",
              "      <td>Mauritia</td>\n",
              "      <td>Rosane Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>Mauritia_sample_data_MFT_050119.xlsx GOI_42030...</td>\n",
              "      <td>GOI-420302-P02-WH10-i5-510-i7-34_S1133_L006_R2...</td>\n",
              "      <td>R2</td>\n",
              "      <td>GOI-420302-P02-WH10</td>\n",
              "      <td>G302P02WH10</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GOI_420302 folder same as above with different...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2455</th>\n",
              "      <td>Mauritia</td>\n",
              "      <td>Rosane Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>Mauritia_sample_data_MFT_050119.xlsx GOI_42030...</td>\n",
              "      <td>GOI-420302-P02-WH11-i5-510-i7-95_S1134_L006_R1...</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOI-420302-P02-WH11</td>\n",
              "      <td>G302P02WH11</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GOI_420302 folder same as above with different...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2456</th>\n",
              "      <td>Mauritia</td>\n",
              "      <td>Rosane Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>Mauritia_sample_data_MFT_050119.xlsx GOI_42030...</td>\n",
              "      <td>GOI-420302-P02-WH11-i5-510-i7-95_S1134_L006_R2...</td>\n",
              "      <td>R2</td>\n",
              "      <td>GOI-420302-P02-WH11</td>\n",
              "      <td>G302P02WH11</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GOI_420302 folder same as above with different...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2457</th>\n",
              "      <td>Mauritia</td>\n",
              "      <td>Rosane Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>Mauritia_sample_data_MFT_050119.xlsx GOI_42030...</td>\n",
              "      <td>GOI-420302-P02-WH12-i5-510-i7-56_S1135_L006_R1...</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOI-420302-P02-WH12</td>\n",
              "      <td>G302P02WH12</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GOI_420302 folder same as above with different...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2458</th>\n",
              "      <td>Mauritia</td>\n",
              "      <td>Rosane Collevatti</td>\n",
              "      <td>RC</td>\n",
              "      <td>Mauritia_sample_data_MFT_050119.xlsx GOI_42030...</td>\n",
              "      <td>GOI-420302-P02-WH12-i5-510-i7-56_S1135_L006_R2...</td>\n",
              "      <td>R2</td>\n",
              "      <td>GOI-420302-P02-WH12</td>\n",
              "      <td>G302P02WH12</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GOI_420302 folder same as above with different...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>384 rows Ã— 30 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "       dataset  ...                                              notes\n",
              "2075  Mauritia  ...  GOI_420302 folder same as above with different...\n",
              "2076  Mauritia  ...  GOI_420302 folder same as above with different...\n",
              "2077  Mauritia  ...  GOI_420302 folder same as above with different...\n",
              "2078  Mauritia  ...  GOI_420302 folder same as above with different...\n",
              "2079  Mauritia  ...  GOI_420302 folder same as above with different...\n",
              "...        ...  ...                                                ...\n",
              "2454  Mauritia  ...  GOI_420302 folder same as above with different...\n",
              "2455  Mauritia  ...  GOI_420302 folder same as above with different...\n",
              "2456  Mauritia  ...  GOI_420302 folder same as above with different...\n",
              "2457  Mauritia  ...  GOI_420302 folder same as above with different...\n",
              "2458  Mauritia  ...  GOI_420302 folder same as above with different...\n",
              "\n",
              "[384 rows x 30 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 44
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 589
        },
        "id": "zATprpqTfikM",
        "outputId": "a91b6c83-38e4-4692-f54c-94cf7903c6f5"
      },
      "source": [
        "# Now, MJ Sanin\n",
        "df[(df['taxspecies'].isna()) & (df['pi_code'] == 'MJS')] # didn't find which species does that sample belongs to."
      ],
      "execution_count": 46,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>filename</th>\n",
              "      <th>sense</th>\n",
              "      <th>filecode</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "      <th>botanic_garden</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>ifmorphotype</th>\n",
              "      <th>ifpopulation</th>\n",
              "      <th>sent_to_Cano</th>\n",
              "      <th>newfilename</th>\n",
              "      <th>raw_reads</th>\n",
              "      <th>reads_trimed_paired_flag</th>\n",
              "      <th>reads_trimed_paired</th>\n",
              "      <th>reads_trimed_single</th>\n",
              "      <th>percentage_lost_tosingles</th>\n",
              "      <th>pcr_filtered</th>\n",
              "      <th>collection_year</th>\n",
              "      <th>Continent</th>\n",
              "      <th>country</th>\n",
              "      <th>long</th>\n",
              "      <th>lat</th>\n",
              "      <th>flag</th>\n",
              "      <th>notes</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>2961</th>\n",
              "      <td>Lausanne_Cerox</td>\n",
              "      <td>Maria Jose Sanin</td>\n",
              "      <td>MJS</td>\n",
              "      <td>HybSeq_Palms.xlsx</td>\n",
              "      <td>523B_rep_S161_L004_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>523B_rep_S161_L004</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>#DIV/0!</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2962</th>\n",
              "      <td>Lausanne_Cerox</td>\n",
              "      <td>Maria Jose Sanin</td>\n",
              "      <td>MJS</td>\n",
              "      <td>HybSeq_Palms.xlsx</td>\n",
              "      <td>523B_rep_S161_L004_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>523B_rep_S161_L004</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>#DIV/0!</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2963</th>\n",
              "      <td>Lausanne_Cerox</td>\n",
              "      <td>Maria Jose Sanin</td>\n",
              "      <td>MJS</td>\n",
              "      <td>HybSeq_Palms.xlsx</td>\n",
              "      <td>523B_rep_S161_L005_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>523B_rep_S161_L005</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>#DIV/0!</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2964</th>\n",
              "      <td>Lausanne_Cerox</td>\n",
              "      <td>Maria Jose Sanin</td>\n",
              "      <td>MJS</td>\n",
              "      <td>HybSeq_Palms.xlsx</td>\n",
              "      <td>523B_rep_S161_L005_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>523B_rep_S161_L005</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>#DIV/0!</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2965</th>\n",
              "      <td>Lausanne_Cerox</td>\n",
              "      <td>Maria Jose Sanin</td>\n",
              "      <td>MJS</td>\n",
              "      <td>HybSeq_Palms.xlsx</td>\n",
              "      <td>523B_S68_L004_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>523B_S68_L004</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>#DIV/0!</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2966</th>\n",
              "      <td>Lausanne_Cerox</td>\n",
              "      <td>Maria Jose Sanin</td>\n",
              "      <td>MJS</td>\n",
              "      <td>HybSeq_Palms.xlsx</td>\n",
              "      <td>523B_S68_L004_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>523B_S68_L004</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>#DIV/0!</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2967</th>\n",
              "      <td>Lausanne_Cerox</td>\n",
              "      <td>Maria Jose Sanin</td>\n",
              "      <td>MJS</td>\n",
              "      <td>HybSeq_Palms.xlsx</td>\n",
              "      <td>523B_S68_L005_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>523B_S68_L005</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>#DIV/0!</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2968</th>\n",
              "      <td>Lausanne_Cerox</td>\n",
              "      <td>Maria Jose Sanin</td>\n",
              "      <td>MJS</td>\n",
              "      <td>HybSeq_Palms.xlsx</td>\n",
              "      <td>523B_S68_L005_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>523B_S68_L005</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>#DIV/0!</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "             dataset          provider pi_code  ... lat flag notes\n",
              "2961  Lausanne_Cerox  Maria Jose Sanin     MJS  ... NaN  NaN   NaN\n",
              "2962  Lausanne_Cerox  Maria Jose Sanin     MJS  ... NaN  NaN   NaN\n",
              "2963  Lausanne_Cerox  Maria Jose Sanin     MJS  ... NaN  NaN   NaN\n",
              "2964  Lausanne_Cerox  Maria Jose Sanin     MJS  ... NaN  NaN   NaN\n",
              "2965  Lausanne_Cerox  Maria Jose Sanin     MJS  ... NaN  NaN   NaN\n",
              "2966  Lausanne_Cerox  Maria Jose Sanin     MJS  ... NaN  NaN   NaN\n",
              "2967  Lausanne_Cerox  Maria Jose Sanin     MJS  ... NaN  NaN   NaN\n",
              "2968  Lausanne_Cerox  Maria Jose Sanin     MJS  ... NaN  NaN   NaN\n",
              "\n",
              "[8 rows x 30 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 46
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PNMiy8gelQab"
      },
      "source": [
        "Finally, an example of fuzzy match to detect typos\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "noKjbtmOlJC7",
        "outputId": "4ae51e7c-e50a-4b46-eb2c-f0e7fe77ff71"
      },
      "source": [
        "# we need two lists. one with american species, one with sequenced species\n",
        "\n",
        "apspp # american palms\n",
        "sequenced = list(df['taxspecies'].dropna().unique())\n",
        "\n",
        "print(len(apspp), 'total')\n",
        "print(len(sequenced), 'sequenced')"
      ],
      "execution_count": 56,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "810 total\n",
            "685 sequenced\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "weM8SD7Cmspc"
      },
      "source": [
        "# now, we search for species matching between 95% and 100%\n",
        "#  FuzzyWuzzy has, just like the Levenshtein package\n",
        "# a ratio function that computes the standard Levenshtein distance similarity ratio between two sequences\n",
        "\n",
        "import time\n",
        "from itertools import product\n",
        "\n",
        "# iterate trough all combinations between species\n",
        "for x,y in product(apspp, sequenced):\n",
        "  # print(x,y)\n",
        "  ratio = fuzz.ratio(x.lower(),y.lower()) # we make sure all characters are lower case. it is case sensitive\n",
        "  if 95 >= ratio > 100:\n",
        "    print(x,y)\n",
        "  # time.sleep(0.2)"
      ],
      "execution_count": 59,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 520
        },
        "id": "PGpDaiDboxFB",
        "outputId": "eb11e317-8268-4e55-a847-fe317784c15a"
      },
      "source": [
        "import time\n",
        "from itertools import product\n",
        "\n",
        "# iterate trough all combinations between species\n",
        "for x,y in product(apspp, sequenced):\n",
        "  # print(x,y)\n",
        "  ratio = fuzz.ratio(x.lower(),y.lower()) # we make sure all characters are lower case. it is case sensitive\n",
        "  print(x,y,ratio)\n",
        "  time.sleep(0.2)"
      ],
      "execution_count": 62,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Chamaedorea_oblongata Acrocomia_aculeata 41\n",
            "Chamaedorea_oblongata Acrocomia_crispa 27\n",
            "Chamaedorea_oblongata Acrocomia_emensis 26\n",
            "Chamaedorea_oblongata Acrocomia_glaucescens 29\n",
            "Chamaedorea_oblongata Acrocomia_intumescens 29\n",
            "Chamaedorea_oblongata Acrocomia_totai 39\n",
            "Chamaedorea_oblongata Aiphanes_acaulis 27\n",
            "Chamaedorea_oblongata Aiphanes_buenaventurae 19\n",
            "Chamaedorea_oblongata Aiphanes_concinna 37\n",
            "Chamaedorea_oblongata Aiphanes_erinacea 26\n",
            "Chamaedorea_oblongata Aiphanes_gelatinosa 40\n",
            "Chamaedorea_oblongata Aiphanes_hirsuta 32\n",
            "Chamaedorea_oblongata Aiphanes_horrida 32\n",
            "Chamaedorea_oblongata Aiphanes_killipii 26\n",
            "Chamaedorea_oblongata Aiphanes_leiostachys 10\n",
            "Chamaedorea_oblongata Aiphanes_lindeniana 20\n",
            "Chamaedorea_oblongata Aiphanes_linearis 26\n",
            "Chamaedorea_oblongata Aiphanes_macroloba 41\n",
            "Chamaedorea_oblongata Aiphanes_minima 22\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "error",
          "ename": "KeyboardInterrupt",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-62-10378b3dcb20>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m   \u001b[0mratio\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfuzz\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mratio\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlower\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlower\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# we make sure all characters are lower case. it is case sensitive\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m   \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mratio\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m   \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0.2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5FywrWwXrB3Q",
        "outputId": "06f814a0-3d3f-4093-b95f-d4e10c10cead"
      },
      "source": [
        "import time\n",
        "from itertools import product\n",
        "\n",
        "# iterate trough all combinations between species\n",
        "for x,y in product([spp for spp in apspp if 'Geonoma' in spp], [spp for spp in sequenced if 'Geonoma' in spp]):\n",
        "  # print(x,y)\n",
        "  ratio = fuzz.ratio(x.lower(),y.lower()) # we make sure all characters are lower case. it is case sensitive\n",
        "  print(x,y,ratio)\n",
        "  time.sleep(0.2)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Geonoma_triglochin Geonoma_baculifera 50\n",
            "Geonoma_triglochin Geonoma_brongniartii 58\n",
            "Geonoma_triglochin Geonoma_calyptrogynoidea 52\n",
            "Geonoma_triglochin Geonoma_camana 62\n",
            "Geonoma_triglochin Geonoma_chlamydostachys 49\n",
            "Geonoma_triglochin Geonoma_concinna 71\n",
            "Geonoma_triglochin Geonoma_congesta 53\n",
            "Geonoma_triglochin Geonoma_cuneata 55\n",
            "Geonoma_triglochin Geonoma_deversa 55\n",
            "Geonoma_triglochin Geonoma_elegans 61\n",
            "Geonoma_triglochin Geonoma_fosteri 67\n",
            "Geonoma_triglochin Geonoma_frontinensis 58\n",
            "Geonoma_triglochin Geonoma_galeanoae 63\n",
            "Geonoma_triglochin Geonoma_hugonis 55\n",
            "Geonoma_triglochin Geonoma_interrupta 56\n",
            "Geonoma_triglochin Geonoma_lehmannii 57\n",
            "Geonoma_triglochin Geonoma_longivaginata 62\n",
            "Geonoma_triglochin Geonoma_macrostachys 58\n",
            "Geonoma_triglochin Geonoma_maxima 56\n",
            "Geonoma_triglochin Geonoma_monospatha 56\n",
            "Geonoma_triglochin Geonoma_mooreana 59\n",
            "Geonoma_triglochin Geonoma_multisecta 61\n",
            "Geonoma_triglochin Geonoma_occidentalis 63\n",
            "Geonoma_triglochin Geonoma_orbignyana 67\n",
            "Geonoma_triglochin Geonoma_paradoxa 59\n",
            "Geonoma_triglochin Geonoma_pauciflora 61\n",
            "Geonoma_triglochin Geonoma_pohliana 59\n",
            "Geonoma_triglochin Geonoma_procumbens 67\n",
            "Geonoma_triglochin Geonoma_rubescens 63\n",
            "Geonoma_triglochin Geonoma_santanderensis 55\n",
            "Geonoma_triglochin Geonoma_schottiana 67\n",
            "Geonoma_triglochin Geonoma_stricta 73\n",
            "Geonoma_triglochin Geonoma_triandra 71\n",
            "Geonoma_triglochin Geonoma_undata 56\n",
            "Geonoma_triglochin Geonoma_aspdiifolia 59\n",
            "Geonoma_triglochin Geonoma_atrovirens_(Chamaedorea) 52\n",
            "Geonoma_triglochin Geonoma_epetiolata 61\n",
            "Geonoma_triglochin Geonoma_euspatha 59\n",
            "Geonoma_triglochin Geonoma_hollinensis 59\n",
            "Geonoma_triglochin Geonoma_lanata 56\n",
            "Geonoma_triglochin Geonoma_laxiflora 63\n",
            "Geonoma_triglochin Geonoma_leptospadix 54\n",
            "Geonoma_triglochin Geonoma_longepedunculata 52\n",
            "Geonoma_triglochin Geonoma_oldemanii 57\n",
            "Geonoma_triglochin Geonoma_oligoclona 72\n",
            "Geonoma_triglochin Geonoma_pinnatifrons 53\n",
            "Geonoma_triglochin Geonoma_poeppigiana 65\n",
            "Geonoma_triglochin Geonoma_triglochin 100\n",
            "Geonoma_triglochin Geonoma_umbraculiformis 59\n",
            "Geonoma_monospatha Geonoma_baculifera 56\n",
            "Geonoma_monospatha Geonoma_brongniartii 63\n",
            "Geonoma_monospatha Geonoma_calyptrogynoidea 57\n",
            "Geonoma_monospatha Geonoma_camana 69\n",
            "Geonoma_monospatha Geonoma_chlamydostachys 63\n",
            "Geonoma_monospatha Geonoma_concinna 65\n",
            "Geonoma_monospatha Geonoma_congesta 71\n",
            "Geonoma_monospatha Geonoma_cuneata 73\n",
            "Geonoma_monospatha Geonoma_deversa 61\n",
            "Geonoma_monospatha Geonoma_elegans 61\n",
            "Geonoma_monospatha Geonoma_fosteri 67\n",
            "Geonoma_monospatha Geonoma_frontinensis 58\n",
            "Geonoma_monospatha Geonoma_galeanoae 63\n",
            "Geonoma_monospatha Geonoma_hugonis 67\n",
            "Geonoma_monospatha Geonoma_interrupta 61\n",
            "Geonoma_monospatha Geonoma_lehmannii 57\n",
            "Geonoma_monospatha Geonoma_longivaginata 67\n",
            "Geonoma_monospatha Geonoma_macrostachys 68\n",
            "Geonoma_monospatha Geonoma_maxima 69\n",
            "Geonoma_monospatha Geonoma_monospatha 100\n",
            "Geonoma_monospatha Geonoma_mooreana 71\n",
            "Geonoma_monospatha Geonoma_multisecta 61\n",
            "Geonoma_monospatha Geonoma_occidentalis 58\n",
            "Geonoma_monospatha Geonoma_orbignyana 67\n",
            "Geonoma_monospatha Geonoma_paradoxa 65\n",
            "Geonoma_monospatha Geonoma_pauciflora 61\n",
            "Geonoma_monospatha Geonoma_pohliana 65\n",
            "Geonoma_monospatha Geonoma_procumbens 61\n",
            "Geonoma_monospatha Geonoma_rubescens 57\n",
            "Geonoma_monospatha Geonoma_santanderensis 50\n",
            "Geonoma_monospatha Geonoma_schottiana 61\n",
            "Geonoma_monospatha Geonoma_stricta 61\n",
            "Geonoma_monospatha Geonoma_triandra 59\n",
            "Geonoma_monospatha Geonoma_undata 75\n",
            "Geonoma_monospatha Geonoma_aspdiifolia 59\n",
            "Geonoma_monospatha Geonoma_atrovirens_(Chamaedorea) 48\n",
            "Geonoma_monospatha Geonoma_epetiolata 67\n",
            "Geonoma_monospatha Geonoma_euspatha 82\n",
            "Geonoma_monospatha Geonoma_hollinensis 59\n",
            "Geonoma_monospatha Geonoma_lanata 75\n",
            "Geonoma_monospatha Geonoma_laxiflora 57\n",
            "Geonoma_monospatha Geonoma_leptospadix 65\n",
            "Geonoma_monospatha Geonoma_longepedunculata 67\n",
            "Geonoma_monospatha Geonoma_oldemanii 57\n",
            "Geonoma_monospatha Geonoma_oligoclona 61\n",
            "Geonoma_monospatha Geonoma_pinnatifrons 58\n",
            "Geonoma_monospatha Geonoma_poeppigiana 59\n",
            "Geonoma_monospatha Geonoma_triglochin 56\n",
            "Geonoma_monospatha Geonoma_umbraculiformis 54\n",
            "Geonoma_occidentalis Geonoma_baculifera 58\n",
            "Geonoma_occidentalis Geonoma_brongniartii 60\n",
            "Geonoma_occidentalis Geonoma_calyptrogynoidea 59\n",
            "Geonoma_occidentalis Geonoma_camana 65\n",
            "Geonoma_occidentalis Geonoma_chlamydostachys 56\n",
            "Geonoma_occidentalis Geonoma_concinna 72\n",
            "Geonoma_occidentalis Geonoma_congesta 67\n",
            "Geonoma_occidentalis Geonoma_cuneata 69\n",
            "Geonoma_occidentalis Geonoma_deversa 63\n",
            "Geonoma_occidentalis Geonoma_elegans 63\n",
            "Geonoma_occidentalis Geonoma_fosteri 57\n",
            "Geonoma_occidentalis Geonoma_frontinensis 70\n",
            "Geonoma_occidentalis Geonoma_galeanoae 54\n",
            "Geonoma_occidentalis Geonoma_hugonis 69\n",
            "Geonoma_occidentalis Geonoma_interrupta 63\n",
            "Geonoma_occidentalis Geonoma_lehmannii 54\n",
            "Geonoma_occidentalis Geonoma_longivaginata 63\n",
            "Geonoma_occidentalis Geonoma_macrostachys 60\n",
            "Geonoma_occidentalis Geonoma_maxima 59\n",
            "Geonoma_occidentalis Geonoma_monospatha 63\n",
            "Geonoma_occidentalis Geonoma_mooreana 67\n",
            "Geonoma_occidentalis Geonoma_multisecta 58\n",
            "Geonoma_occidentalis Geonoma_occidentalis 100\n",
            "Geonoma_occidentalis Geonoma_orbignyana 63\n",
            "Geonoma_occidentalis Geonoma_paradoxa 56\n",
            "Geonoma_occidentalis Geonoma_pauciflora 58\n",
            "Geonoma_occidentalis Geonoma_pohliana 61\n",
            "Geonoma_occidentalis Geonoma_procumbens 68\n",
            "Geonoma_occidentalis Geonoma_rubescens 65\n",
            "Geonoma_occidentalis Geonoma_santanderensis 62\n",
            "Geonoma_occidentalis Geonoma_schottiana 63\n",
            "Geonoma_occidentalis Geonoma_stricta 63\n",
            "Geonoma_occidentalis Geonoma_triandra 61\n",
            "Geonoma_occidentalis Geonoma_undata 65\n",
            "Geonoma_occidentalis Geonoma_aspdiifolia 56\n",
            "Geonoma_occidentalis Geonoma_atrovirens_(Chamaedorea) 50\n",
            "Geonoma_occidentalis Geonoma_epetiolata 58\n",
            "Geonoma_occidentalis Geonoma_euspatha 61\n",
            "Geonoma_occidentalis Geonoma_hollinensis 72\n",
            "Geonoma_occidentalis Geonoma_lanata 65\n",
            "Geonoma_occidentalis Geonoma_laxiflora 54\n",
            "Geonoma_occidentalis Geonoma_leptospadix 51\n",
            "Geonoma_occidentalis Geonoma_longepedunculata 55\n",
            "Geonoma_occidentalis Geonoma_oldemanii 70\n",
            "Geonoma_occidentalis Geonoma_oligoclona 58\n",
            "Geonoma_occidentalis Geonoma_pinnatifrons 55\n",
            "Geonoma_occidentalis Geonoma_poeppigiana 62\n",
            "Geonoma_occidentalis Geonoma_triglochin 63\n",
            "Geonoma_occidentalis Geonoma_umbraculiformis 56\n",
            "Geonoma_stricta Geonoma_baculifera 61\n",
            "Geonoma_stricta Geonoma_brongniartii 57\n",
            "Geonoma_stricta Geonoma_calyptrogynoidea 62\n",
            "Geonoma_stricta Geonoma_camana 69\n",
            "Geonoma_stricta Geonoma_chlamydostachys 58\n",
            "Geonoma_stricta Geonoma_concinna 65\n",
            "Geonoma_stricta Geonoma_congesta 71\n",
            "Geonoma_stricta Geonoma_cuneata 73\n",
            "Geonoma_stricta Geonoma_deversa 67\n",
            "Geonoma_stricta Geonoma_elegans 60\n",
            "Geonoma_stricta Geonoma_fosteri 80\n",
            "Geonoma_stricta Geonoma_frontinensis 57\n",
            "Geonoma_stricta Geonoma_galeanoae 56\n",
            "Geonoma_stricta Geonoma_hugonis 60\n",
            "Geonoma_stricta Geonoma_interrupta 73\n",
            "Geonoma_stricta Geonoma_lehmannii 56\n",
            "Geonoma_stricta Geonoma_longivaginata 61\n",
            "Geonoma_stricta Geonoma_macrostachys 63\n",
            "Geonoma_stricta Geonoma_maxima 69\n",
            "Geonoma_stricta Geonoma_monospatha 67\n",
            "Geonoma_stricta Geonoma_mooreana 65\n",
            "Geonoma_stricta Geonoma_multisecta 73\n",
            "Geonoma_stricta Geonoma_occidentalis 63\n",
            "Geonoma_stricta Geonoma_orbignyana 67\n",
            "Geonoma_stricta Geonoma_paradoxa 65\n",
            "Geonoma_stricta Geonoma_pauciflora 61\n",
            "Geonoma_stricta Geonoma_pohliana 65\n",
            "Geonoma_stricta Geonoma_procumbens 55\n",
            "Geonoma_stricta Geonoma_rubescens 62\n",
            "Geonoma_stricta Geonoma_santanderensis 59\n",
            "Geonoma_stricta Geonoma_schottiana 73\n",
            "Geonoma_stricta Geonoma_stricta 100\n",
            "Geonoma_stricta Geonoma_triandra 77\n",
            "Geonoma_stricta Geonoma_undata 69\n",
            "Geonoma_stricta Geonoma_aspdiifolia 65\n",
            "Geonoma_stricta Geonoma_atrovirens_(Chamaedorea) 55\n",
            "Geonoma_stricta Geonoma_epetiolata 73\n",
            "Geonoma_stricta Geonoma_euspatha 71\n",
            "Geonoma_stricta Geonoma_hollinensis 59\n",
            "Geonoma_stricta Geonoma_lanata 69\n",
            "Geonoma_stricta Geonoma_laxiflora 62\n",
            "Geonoma_stricta Geonoma_leptospadix 59\n",
            "Geonoma_stricta Geonoma_longepedunculata 56\n",
            "Geonoma_stricta Geonoma_oldemanii 56\n",
            "Geonoma_stricta Geonoma_oligoclona 67\n",
            "Geonoma_stricta Geonoma_pinnatifrons 51\n",
            "Geonoma_stricta Geonoma_poeppigiana 59\n",
            "Geonoma_stricta Geonoma_triglochin 73\n",
            "Geonoma_stricta Geonoma_umbraculiformis 47\n",
            "Geonoma_poiteauana Geonoma_baculifera 61\n",
            "Geonoma_poiteauana Geonoma_brongniartii 58\n",
            "Geonoma_poiteauana Geonoma_calyptrogynoidea 62\n",
            "Geonoma_poiteauana Geonoma_camana 75\n",
            "Geonoma_poiteauana Geonoma_chlamydostachys 54\n",
            "Geonoma_poiteauana Geonoma_concinna 71\n",
            "Geonoma_poiteauana Geonoma_congesta 65\n",
            "Geonoma_poiteauana Geonoma_cuneata 67\n",
            "Geonoma_poiteauana Geonoma_deversa 61\n",
            "Geonoma_poiteauana Geonoma_elegans 67\n",
            "Geonoma_poiteauana Geonoma_fosteri 67\n",
            "Geonoma_poiteauana Geonoma_frontinensis 63\n",
            "Geonoma_poiteauana Geonoma_galeanoae 63\n",
            "Geonoma_poiteauana Geonoma_hugonis 61\n",
            "Geonoma_poiteauana Geonoma_interrupta 67\n",
            "Geonoma_poiteauana Geonoma_lehmannii 63\n",
            "Geonoma_poiteauana Geonoma_longivaginata 67\n",
            "Geonoma_poiteauana Geonoma_macrostachys 58\n",
            "Geonoma_poiteauana Geonoma_maxima 62\n",
            "Geonoma_poiteauana Geonoma_monospatha 61\n",
            "Geonoma_poiteauana Geonoma_mooreana 76\n",
            "Geonoma_poiteauana Geonoma_multisecta 61\n",
            "Geonoma_poiteauana Geonoma_occidentalis 63\n",
            "Geonoma_poiteauana Geonoma_orbignyana 72\n",
            "Geonoma_poiteauana Geonoma_paradoxa 65\n",
            "Geonoma_poiteauana Geonoma_pauciflora 67\n",
            "Geonoma_poiteauana Geonoma_pohliana 82\n",
            "Geonoma_poiteauana Geonoma_procumbens 67\n",
            "Geonoma_poiteauana Geonoma_rubescens 57\n",
            "Geonoma_poiteauana Geonoma_santanderensis 55\n",
            "Geonoma_poiteauana Geonoma_schottiana 72\n",
            "Geonoma_poiteauana Geonoma_stricta 67\n",
            "Geonoma_poiteauana Geonoma_triandra 71\n",
            "Geonoma_poiteauana Geonoma_undata 62\n",
            "Geonoma_poiteauana Geonoma_aspdiifolia 65\n",
            "Geonoma_poiteauana Geonoma_atrovirens_(Chamaedorea) 48\n",
            "Geonoma_poiteauana Geonoma_epetiolata 67\n",
            "Geonoma_poiteauana Geonoma_euspatha 65\n",
            "Geonoma_poiteauana Geonoma_hollinensis 65\n",
            "Geonoma_poiteauana Geonoma_lanata 69\n",
            "Geonoma_poiteauana Geonoma_laxiflora 57\n",
            "Geonoma_poiteauana Geonoma_leptospadix 59\n",
            "Geonoma_poiteauana Geonoma_longepedunculata 52\n",
            "Geonoma_poiteauana Geonoma_oldemanii 69\n",
            "Geonoma_poiteauana Geonoma_oligoclona 67\n",
            "Geonoma_poiteauana Geonoma_pinnatifrons 63\n",
            "Geonoma_poiteauana Geonoma_poeppigiana 76\n",
            "Geonoma_poiteauana Geonoma_triglochin 61\n",
            "Geonoma_poiteauana Geonoma_umbraculiformis 49\n",
            "Geonoma_laxiflora Geonoma_baculifera 74\n",
            "Geonoma_laxiflora Geonoma_brongniartii 54\n",
            "Geonoma_laxiflora Geonoma_calyptrogynoidea 49\n",
            "Geonoma_laxiflora Geonoma_camana 65\n",
            "Geonoma_laxiflora Geonoma_chlamydostachys 60\n",
            "Geonoma_laxiflora Geonoma_concinna 55\n",
            "Geonoma_laxiflora Geonoma_congesta 55\n",
            "Geonoma_laxiflora Geonoma_cuneata 62\n",
            "Geonoma_laxiflora Geonoma_deversa 56\n",
            "Geonoma_laxiflora Geonoma_elegans 62\n",
            "Geonoma_laxiflora Geonoma_fosteri 56\n",
            "Geonoma_laxiflora Geonoma_frontinensis 49\n",
            "Geonoma_laxiflora Geonoma_galeanoae 71\n",
            "Geonoma_laxiflora Geonoma_hugonis 56\n",
            "Geonoma_laxiflora Geonoma_interrupta 51\n",
            "Geonoma_laxiflora Geonoma_lehmannii 65\n",
            "Geonoma_laxiflora Geonoma_longivaginata 63\n",
            "Geonoma_laxiflora Geonoma_macrostachys 59\n",
            "Geonoma_laxiflora Geonoma_maxima 77\n",
            "Geonoma_laxiflora Geonoma_monospatha 57\n",
            "Geonoma_laxiflora Geonoma_mooreana 67\n",
            "Geonoma_laxiflora Geonoma_multisecta 57\n",
            "Geonoma_laxiflora Geonoma_occidentalis 54\n",
            "Geonoma_laxiflora Geonoma_orbignyana 63\n",
            "Geonoma_laxiflora Geonoma_paradoxa 67\n",
            "Geonoma_laxiflora Geonoma_pauciflora 86\n",
            "Geonoma_laxiflora Geonoma_pohliana 67\n",
            "Geonoma_laxiflora Geonoma_procumbens 51\n",
            "Geonoma_laxiflora Geonoma_rubescens 53\n",
            "Geonoma_laxiflora Geonoma_santanderensis 51\n",
            "Geonoma_laxiflora Geonoma_schottiana 57\n",
            "Geonoma_laxiflora Geonoma_stricta 56\n",
            "Geonoma_laxiflora Geonoma_triandra 67\n",
            "Geonoma_laxiflora Geonoma_undata 65\n",
            "Geonoma_laxiflora Geonoma_aspdiifolia 72\n",
            "Geonoma_laxiflora Geonoma_atrovirens_(Chamaedorea) 53\n",
            "Geonoma_laxiflora Geonoma_epetiolata 63\n",
            "Geonoma_laxiflora Geonoma_euspatha 61\n",
            "Geonoma_laxiflora Geonoma_hollinensis 56\n",
            "Geonoma_laxiflora Geonoma_lanata 71\n",
            "Geonoma_laxiflora Geonoma_laxiflora 100\n",
            "Geonoma_laxiflora Geonoma_leptospadix 61\n",
            "Geonoma_laxiflora Geonoma_longepedunculata 54\n",
            "Geonoma_laxiflora Geonoma_oldemanii 65\n",
            "Geonoma_laxiflora Geonoma_oligoclona 74\n",
            "Geonoma_laxiflora Geonoma_pinnatifrons 65\n",
            "Geonoma_laxiflora Geonoma_poeppigiana 56\n",
            "Geonoma_laxiflora Geonoma_triglochin 63\n",
            "Geonoma_laxiflora Geonoma_umbraculiformis 65\n",
            "Geonoma_mooreana Geonoma_baculifera 59\n",
            "Geonoma_mooreana Geonoma_brongniartii 56\n",
            "Geonoma_mooreana Geonoma_calyptrogynoidea 60\n",
            "Geonoma_mooreana Geonoma_camana 80\n",
            "Geonoma_mooreana Geonoma_chlamydostachys 56\n",
            "Geonoma_mooreana Geonoma_concinna 69\n",
            "Geonoma_mooreana Geonoma_congesta 69\n",
            "Geonoma_mooreana Geonoma_cuneata 71\n",
            "Geonoma_mooreana Geonoma_deversa 65\n",
            "Geonoma_mooreana Geonoma_elegans 71\n",
            "Geonoma_mooreana Geonoma_fosteri 65\n",
            "Geonoma_mooreana Geonoma_frontinensis 61\n",
            "Geonoma_mooreana Geonoma_galeanoae 73\n",
            "Geonoma_mooreana Geonoma_hugonis 65\n",
            "Geonoma_mooreana Geonoma_interrupta 59\n",
            "Geonoma_mooreana Geonoma_lehmannii 67\n",
            "Geonoma_mooreana Geonoma_longivaginata 65\n",
            "Geonoma_mooreana Geonoma_macrostachys 61\n",
            "Geonoma_mooreana Geonoma_maxima 73\n",
            "Geonoma_mooreana Geonoma_monospatha 76\n",
            "Geonoma_mooreana Geonoma_mooreana 100\n",
            "Geonoma_mooreana Geonoma_multisecta 65\n",
            "Geonoma_mooreana Geonoma_occidentalis 61\n",
            "Geonoma_mooreana Geonoma_orbignyana 76\n",
            "Geonoma_mooreana Geonoma_paradoxa 62\n",
            "Geonoma_mooreana Geonoma_pauciflora 65\n",
            "Geonoma_mooreana Geonoma_pohliana 75\n",
            "Geonoma_mooreana Geonoma_procumbens 65\n",
            "Geonoma_mooreana Geonoma_rubescens 67\n",
            "Geonoma_mooreana Geonoma_santanderensis 58\n",
            "Geonoma_mooreana Geonoma_schottiana 71\n",
            "Geonoma_mooreana Geonoma_stricta 65\n",
            "Geonoma_mooreana Geonoma_triandra 75\n",
            "Geonoma_mooreana Geonoma_undata 67\n",
            "Geonoma_mooreana Geonoma_aspdiifolia 57\n",
            "Geonoma_mooreana Geonoma_atrovirens_(Chamaedorea) 54\n",
            "Geonoma_mooreana Geonoma_epetiolata 65\n",
            "Geonoma_mooreana Geonoma_euspatha 69\n",
            "Geonoma_mooreana Geonoma_hollinensis 63\n",
            "Geonoma_mooreana Geonoma_lanata 73\n",
            "Geonoma_mooreana Geonoma_laxiflora 67\n",
            "Geonoma_mooreana Geonoma_leptospadix 57\n",
            "Geonoma_mooreana Geonoma_longepedunculata 60\n",
            "Geonoma_mooreana Geonoma_oldemanii 67\n",
            "Geonoma_mooreana Geonoma_oligoclona 71\n",
            "Geonoma_mooreana Geonoma_pinnatifrons 56\n",
            "Geonoma_mooreana Geonoma_poeppigiana 74\n",
            "Geonoma_mooreana Geonoma_triglochin 59\n",
            "Geonoma_mooreana Geonoma_umbraculiformis 56\n",
            "Geonoma_undata Geonoma_baculifera 62\n",
            "Geonoma_undata Geonoma_brongniartii 65\n",
            "Geonoma_undata Geonoma_calyptrogynoidea 58\n",
            "Geonoma_undata Geonoma_camana 71\n",
            "Geonoma_undata Geonoma_chlamydostachys 59\n",
            "Geonoma_undata Geonoma_concinna 67\n",
            "Geonoma_undata Geonoma_congesta 73\n",
            "Geonoma_undata Geonoma_cuneata 90\n",
            "Geonoma_undata Geonoma_deversa 69\n",
            "Geonoma_undata Geonoma_elegans 62\n",
            "Geonoma_undata Geonoma_fosteri 62\n",
            "Geonoma_undata Geonoma_frontinensis 59\n",
            "Geonoma_undata Geonoma_galeanoae 65\n",
            "Geonoma_undata Geonoma_hugonis 69\n",
            "Geonoma_undata Geonoma_interrupta 69\n",
            "Geonoma_undata Geonoma_lehmannii 58\n",
            "Geonoma_undata Geonoma_longivaginata 69\n",
            "Geonoma_undata Geonoma_macrostachys 65\n",
            "Geonoma_undata Geonoma_maxima 71\n",
            "Geonoma_undata Geonoma_monospatha 75\n",
            "Geonoma_undata Geonoma_mooreana 67\n",
            "Geonoma_undata Geonoma_multisecta 69\n",
            "Geonoma_undata Geonoma_occidentalis 65\n",
            "Geonoma_undata Geonoma_orbignyana 69\n",
            "Geonoma_undata Geonoma_paradoxa 67\n",
            "Geonoma_undata Geonoma_pauciflora 62\n",
            "Geonoma_undata Geonoma_pohliana 67\n",
            "Geonoma_undata Geonoma_procumbens 62\n",
            "Geonoma_undata Geonoma_rubescens 65\n",
            "Geonoma_undata Geonoma_santanderensis 56\n",
            "Geonoma_undata Geonoma_schottiana 62\n",
            "Geonoma_undata Geonoma_stricta 69\n",
            "Geonoma_undata Geonoma_triandra 73\n",
            "Geonoma_undata Geonoma_undata 100\n",
            "Geonoma_undata Geonoma_aspdiifolia 61\n",
            "Geonoma_undata Geonoma_atrovirens_(Chamaedorea) 48\n",
            "Geonoma_undata Geonoma_epetiolata 69\n",
            "Geonoma_undata Geonoma_euspatha 80\n",
            "Geonoma_undata Geonoma_hollinensis 55\n",
            "Geonoma_undata Geonoma_lanata 86\n",
            "Geonoma_undata Geonoma_laxiflora 65\n",
            "Geonoma_undata Geonoma_leptospadix 55\n",
            "Geonoma_undata Geonoma_longepedunculata 68\n",
            "Geonoma_undata Geonoma_oldemanii 58\n",
            "Geonoma_undata Geonoma_oligoclona 62\n",
            "Geonoma_undata Geonoma_pinnatifrons 65\n",
            "Geonoma_undata Geonoma_poeppigiana 61\n",
            "Geonoma_undata Geonoma_triglochin 56\n",
            "Geonoma_undata Geonoma_umbraculiformis 54\n",
            "Geonoma_hollinensis Geonoma_baculifera 59\n",
            "Geonoma_hollinensis Geonoma_brongniartii 56\n",
            "Geonoma_hollinensis Geonoma_calyptrogynoidea 51\n",
            "Geonoma_hollinensis Geonoma_camana 55\n",
            "Geonoma_hollinensis Geonoma_chlamydostachys 57\n",
            "Geonoma_hollinensis Geonoma_concinna 69\n",
            "Geonoma_hollinensis Geonoma_congesta 69\n",
            "Geonoma_hollinensis Geonoma_cuneata 59\n",
            "Geonoma_hollinensis Geonoma_deversa 59\n",
            "Geonoma_hollinensis Geonoma_elegans 71\n",
            "Geonoma_hollinensis Geonoma_fosteri 59\n",
            "Geonoma_hollinensis Geonoma_frontinensis 82\n",
            "Geonoma_hollinensis Geonoma_galeanoae 56\n",
            "Geonoma_hollinensis Geonoma_hugonis 76\n",
            "Geonoma_hollinensis Geonoma_interrupta 59\n",
            "Geonoma_hollinensis Geonoma_lehmannii 61\n",
            "Geonoma_hollinensis Geonoma_longivaginata 55\n",
            "Geonoma_hollinensis Geonoma_macrostachys 51\n",
            "Geonoma_hollinensis Geonoma_maxima 55\n",
            "Geonoma_hollinensis Geonoma_monospatha 49\n",
            "Geonoma_hollinensis Geonoma_mooreana 57\n",
            "Geonoma_hollinensis Geonoma_multisecta 59\n",
            "Geonoma_hollinensis Geonoma_occidentalis 62\n",
            "Geonoma_hollinensis Geonoma_orbignyana 65\n",
            "Geonoma_hollinensis Geonoma_paradoxa 51\n",
            "Geonoma_hollinensis Geonoma_pauciflora 49\n",
            "Geonoma_hollinensis Geonoma_pohliana 69\n",
            "Geonoma_hollinensis Geonoma_procumbens 65\n",
            "Geonoma_hollinensis Geonoma_rubescens 61\n",
            "Geonoma_hollinensis Geonoma_santanderensis 68\n",
            "Geonoma_hollinensis Geonoma_schottiana 65\n",
            "Geonoma_hollinensis Geonoma_stricta 53\n",
            "Geonoma_hollinensis Geonoma_triandra 57\n",
            "Geonoma_hollinensis Geonoma_undata 55\n",
            "Geonoma_hollinensis Geonoma_aspdiifolia 58\n",
            "Geonoma_hollinensis Geonoma_atrovirens_(Chamaedorea) 51\n",
            "Geonoma_hollinensis Geonoma_epetiolata 54\n",
            "Geonoma_hollinensis Geonoma_euspatha 51\n",
            "Geonoma_hollinensis Geonoma_hollinensis 100\n",
            "Geonoma_hollinensis Geonoma_lanata 61\n",
            "Geonoma_hollinensis Geonoma_laxiflora 50\n",
            "Geonoma_hollinensis Geonoma_leptospadix 53\n",
            "Geonoma_hollinensis Geonoma_longepedunculata 47\n",
            "Geonoma_hollinensis Geonoma_oldemanii 67\n",
            "Geonoma_hollinensis Geonoma_oligoclona 65\n",
            "Geonoma_hollinensis Geonoma_pinnatifrons 62\n",
            "Geonoma_hollinensis Geonoma_poeppigiana 58\n",
            "Geonoma_hollinensis Geonoma_triglochin 59\n",
            "Geonoma_hollinensis Geonoma_umbraculiformis 57\n",
            "Geonoma_lehmannii Geonoma_baculifera 63\n",
            "Geonoma_lehmannii Geonoma_brongniartii 65\n",
            "Geonoma_lehmannii Geonoma_calyptrogynoidea 54\n",
            "Geonoma_lehmannii Geonoma_camana 71\n",
            "Geonoma_lehmannii Geonoma_chlamydostachys 50\n",
            "Geonoma_lehmannii Geonoma_concinna 61\n",
            "Geonoma_lehmannii Geonoma_congesta 61\n",
            "Geonoma_lehmannii Geonoma_cuneata 62\n",
            "Geonoma_lehmannii Geonoma_deversa 62\n",
            "Geonoma_lehmannii Geonoma_elegans 75\n",
            "Geonoma_lehmannii Geonoma_fosteri 62\n",
            "Geonoma_lehmannii Geonoma_frontinensis 59\n",
            "Geonoma_lehmannii Geonoma_galeanoae 71\n",
            "Geonoma_lehmannii Geonoma_hugonis 69\n",
            "Geonoma_lehmannii Geonoma_interrupta 57\n",
            "Geonoma_lehmannii Geonoma_lehmannii 100\n",
            "Geonoma_lehmannii Geonoma_longivaginata 58\n",
            "Geonoma_lehmannii Geonoma_macrostachys 54\n",
            "Geonoma_lehmannii Geonoma_maxima 71\n",
            "Geonoma_lehmannii Geonoma_monospatha 57\n",
            "Geonoma_lehmannii Geonoma_mooreana 67\n",
            "Geonoma_lehmannii Geonoma_multisecta 63\n",
            "Geonoma_lehmannii Geonoma_occidentalis 54\n",
            "Geonoma_lehmannii Geonoma_orbignyana 57\n",
            "Geonoma_lehmannii Geonoma_paradoxa 55\n",
            "Geonoma_lehmannii Geonoma_pauciflora 57\n",
            "Geonoma_lehmannii Geonoma_pohliana 67\n",
            "Geonoma_lehmannii Geonoma_procumbens 57\n",
            "Geonoma_lehmannii Geonoma_rubescens 59\n",
            "Geonoma_lehmannii Geonoma_santanderensis 62\n",
            "Geonoma_lehmannii Geonoma_schottiana 63\n",
            "Geonoma_lehmannii Geonoma_stricta 56\n",
            "Geonoma_lehmannii Geonoma_triandra 61\n",
            "Geonoma_lehmannii Geonoma_undata 58\n",
            "Geonoma_lehmannii Geonoma_aspdiifolia 61\n",
            "Geonoma_lehmannii Geonoma_atrovirens_(Chamaedorea) 49\n",
            "Geonoma_lehmannii Geonoma_epetiolata 57\n",
            "Geonoma_lehmannii Geonoma_euspatha 67\n",
            "Geonoma_lehmannii Geonoma_hollinensis 67\n",
            "Geonoma_lehmannii Geonoma_lanata 71\n",
            "Geonoma_lehmannii Geonoma_laxiflora 65\n",
            "Geonoma_lehmannii Geonoma_leptospadix 67\n",
            "Geonoma_lehmannii Geonoma_longepedunculata 54\n",
            "Geonoma_lehmannii Geonoma_oldemanii 88\n",
            "Geonoma_lehmannii Geonoma_oligoclona 57\n",
            "Geonoma_lehmannii Geonoma_pinnatifrons 59\n",
            "Geonoma_lehmannii Geonoma_poeppigiana 61\n",
            "Geonoma_lehmannii Geonoma_triglochin 63\n",
            "Geonoma_lehmannii Geonoma_umbraculiformis 55\n",
            "Geonoma_aspidiifolia Geonoma_baculifera 63\n",
            "Geonoma_aspidiifolia Geonoma_brongniartii 55\n",
            "Geonoma_aspidiifolia Geonoma_calyptrogynoidea 59\n",
            "Geonoma_aspidiifolia Geonoma_camana 59\n",
            "Geonoma_aspidiifolia Geonoma_chlamydostachys 51\n",
            "Geonoma_aspidiifolia Geonoma_concinna 50\n",
            "Geonoma_aspidiifolia Geonoma_congesta 50\n",
            "Geonoma_aspidiifolia Geonoma_cuneata 57\n",
            "Geonoma_aspidiifolia Geonoma_deversa 51\n",
            "Geonoma_aspidiifolia Geonoma_elegans 57\n",
            "Geonoma_aspidiifolia Geonoma_fosteri 63\n",
            "Geonoma_aspidiifolia Geonoma_frontinensis 50\n",
            "Geonoma_aspidiifolia Geonoma_galeanoae 59\n",
            "Geonoma_aspidiifolia Geonoma_hugonis 51\n",
            "Geonoma_aspidiifolia Geonoma_interrupta 47\n",
            "Geonoma_aspidiifolia Geonoma_lehmannii 59\n",
            "Geonoma_aspidiifolia Geonoma_longivaginata 54\n",
            "Geonoma_aspidiifolia Geonoma_macrostachys 55\n",
            "Geonoma_aspidiifolia Geonoma_maxima 65\n",
            "Geonoma_aspidiifolia Geonoma_monospatha 58\n",
            "Geonoma_aspidiifolia Geonoma_mooreana 56\n",
            "Geonoma_aspidiifolia Geonoma_multisecta 47\n",
            "Geonoma_aspidiifolia Geonoma_occidentalis 60\n",
            "Geonoma_aspidiifolia Geonoma_orbignyana 53\n",
            "Geonoma_aspidiifolia Geonoma_paradoxa 67\n",
            "Geonoma_aspidiifolia Geonoma_pauciflora 68\n",
            "Geonoma_aspidiifolia Geonoma_pohliana 72\n",
            "Geonoma_aspidiifolia Geonoma_procumbens 47\n",
            "Geonoma_aspidiifolia Geonoma_rubescens 49\n",
            "Geonoma_aspidiifolia Geonoma_santanderensis 52\n",
            "Geonoma_aspidiifolia Geonoma_schottiana 63\n",
            "Geonoma_aspidiifolia Geonoma_stricta 51\n",
            "Geonoma_aspidiifolia Geonoma_triandra 56\n",
            "Geonoma_aspidiifolia Geonoma_undata 59\n",
            "Geonoma_aspidiifolia Geonoma_aspdiifolia 97\n",
            "Geonoma_aspidiifolia Geonoma_atrovirens_(Chamaedorea) 50\n",
            "Geonoma_aspidiifolia Geonoma_epetiolata 68\n",
            "Geonoma_aspidiifolia Geonoma_euspatha 61\n",
            "Geonoma_aspidiifolia Geonoma_hollinensis 56\n",
            "Geonoma_aspidiifolia Geonoma_lanata 59\n",
            "Geonoma_aspidiifolia Geonoma_laxiflora 70\n",
            "Geonoma_aspidiifolia Geonoma_leptospadix 62\n",
            "Geonoma_aspidiifolia Geonoma_longepedunculata 45\n",
            "Geonoma_aspidiifolia Geonoma_oldemanii 59\n",
            "Geonoma_aspidiifolia Geonoma_oligoclona 63\n",
            "Geonoma_aspidiifolia Geonoma_pinnatifrons 65\n",
            "Geonoma_aspidiifolia Geonoma_poeppigiana 62\n",
            "Geonoma_aspidiifolia Geonoma_triglochin 53\n",
            "Geonoma_aspidiifolia Geonoma_umbraculiformis 60\n",
            "Geonoma_brenesii Geonoma_baculifera 59\n",
            "Geonoma_brenesii Geonoma_brongniartii 72\n",
            "Geonoma_brenesii Geonoma_calyptrogynoidea 50\n",
            "Geonoma_brenesii Geonoma_camana 60\n",
            "Geonoma_brenesii Geonoma_chlamydostachys 46\n",
            "Geonoma_brenesii Geonoma_concinna 62\n",
            "Geonoma_brenesii Geonoma_congesta 69\n",
            "Geonoma_brenesii Geonoma_cuneata 65\n",
            "Geonoma_brenesii Geonoma_deversa 65\n",
            "Geonoma_brenesii Geonoma_elegans 71\n",
            "Geonoma_brenesii Geonoma_fosteri 65\n",
            "Geonoma_brenesii Geonoma_frontinensis 72\n",
            "Geonoma_brenesii Geonoma_galeanoae 67\n",
            "Geonoma_brenesii Geonoma_hugonis 65\n",
            "Geonoma_brenesii Geonoma_interrupta 53\n",
            "Geonoma_brenesii Geonoma_lehmannii 73\n",
            "Geonoma_brenesii Geonoma_longivaginata 59\n",
            "Geonoma_brenesii Geonoma_macrostachys 56\n",
            "Geonoma_brenesii Geonoma_maxima 60\n",
            "Geonoma_brenesii Geonoma_monospatha 59\n",
            "Geonoma_brenesii Geonoma_mooreana 69\n",
            "Geonoma_brenesii Geonoma_multisecta 53\n",
            "Geonoma_brenesii Geonoma_occidentalis 61\n",
            "Geonoma_brenesii Geonoma_orbignyana 59\n",
            "Geonoma_brenesii Geonoma_paradoxa 56\n",
            "Geonoma_brenesii Geonoma_pauciflora 53\n",
            "Geonoma_brenesii Geonoma_pohliana 56\n",
            "Geonoma_brenesii Geonoma_procumbens 71\n",
            "Geonoma_brenesii Geonoma_rubescens 73\n",
            "Geonoma_brenesii Geonoma_santanderensis 68\n",
            "Geonoma_brenesii Geonoma_schottiana 53\n",
            "Geonoma_brenesii Geonoma_stricta 65\n",
            "Geonoma_brenesii Geonoma_triandra 62\n",
            "Geonoma_brenesii Geonoma_undata 60\n",
            "Geonoma_brenesii Geonoma_aspdiifolia 63\n",
            "Geonoma_brenesii Geonoma_atrovirens_(Chamaedorea) 50\n",
            "Geonoma_brenesii Geonoma_epetiolata 65\n",
            "Geonoma_brenesii Geonoma_euspatha 62\n",
            "Geonoma_brenesii Geonoma_hollinensis 69\n",
            "Geonoma_brenesii Geonoma_lanata 60\n",
            "Geonoma_brenesii Geonoma_laxiflora 55\n",
            "Geonoma_brenesii Geonoma_leptospadix 63\n",
            "Geonoma_brenesii Geonoma_longepedunculata 50\n",
            "Geonoma_brenesii Geonoma_oldemanii 73\n",
            "Geonoma_brenesii Geonoma_oligoclona 53\n",
            "Geonoma_brenesii Geonoma_pinnatifrons 61\n",
            "Geonoma_brenesii Geonoma_poeppigiana 57\n",
            "Geonoma_brenesii Geonoma_triglochin 59\n",
            "Geonoma_brenesii Geonoma_umbraculiformis 56\n",
            "Geonoma_umbraculiformis Geonoma_baculifera 78\n",
            "Geonoma_umbraculiformis Geonoma_brongniartii 60\n",
            "Geonoma_umbraculiformis Geonoma_calyptrogynoidea 43\n",
            "Geonoma_umbraculiformis Geonoma_camana 54\n",
            "Geonoma_umbraculiformis Geonoma_chlamydostachys 52\n",
            "Geonoma_umbraculiformis Geonoma_concinna 46\n",
            "Geonoma_umbraculiformis Geonoma_congesta 46\n",
            "Geonoma_umbraculiformis Geonoma_cuneata 53\n",
            "Geonoma_umbraculiformis Geonoma_deversa 53\n",
            "Geonoma_umbraculiformis Geonoma_elegans 53\n",
            "Geonoma_umbraculiformis Geonoma_fosteri 63\n",
            "Geonoma_umbraculiformis Geonoma_frontinensis 56\n",
            "Geonoma_umbraculiformis Geonoma_galeanoae 55\n",
            "Geonoma_umbraculiformis Geonoma_hugonis 63\n",
            "Geonoma_umbraculiformis Geonoma_interrupta 49\n",
            "Geonoma_umbraculiformis Geonoma_lehmannii 60\n",
            "Geonoma_umbraculiformis Geonoma_longivaginata 45\n",
            "Geonoma_umbraculiformis Geonoma_macrostachys 60\n",
            "Geonoma_umbraculiformis Geonoma_maxima 65\n",
            "Geonoma_umbraculiformis Geonoma_monospatha 49\n",
            "Geonoma_umbraculiformis Geonoma_mooreana 56\n",
            "Geonoma_umbraculiformis Geonoma_multisecta 63\n",
            "Geonoma_umbraculiformis Geonoma_occidentalis 56\n",
            "Geonoma_umbraculiformis Geonoma_orbignyana 54\n",
            "Geonoma_umbraculiformis Geonoma_paradoxa 56\n",
            "Geonoma_umbraculiformis Geonoma_pauciflora 54\n",
            "Geonoma_umbraculiformis Geonoma_pohliana 51\n",
            "Geonoma_umbraculiformis Geonoma_procumbens 59\n",
            "Geonoma_umbraculiformis Geonoma_rubescens 60\n",
            "Geonoma_umbraculiformis Geonoma_santanderensis 49\n",
            "Geonoma_umbraculiformis Geonoma_schottiana 44\n",
            "Geonoma_umbraculiformis Geonoma_stricta 53\n",
            "Geonoma_umbraculiformis Geonoma_triandra 51\n",
            "Geonoma_umbraculiformis Geonoma_undata 54\n",
            "Geonoma_umbraculiformis Geonoma_aspdiifolia 62\n",
            "Geonoma_umbraculiformis Geonoma_atrovirens_(Chamaedorea) 44\n",
            "Geonoma_umbraculiformis Geonoma_epetiolata 44\n",
            "Geonoma_umbraculiformis Geonoma_euspatha 51\n",
            "Geonoma_umbraculiformis Geonoma_hollinensis 57\n",
            "Geonoma_umbraculiformis Geonoma_lanata 49\n",
            "Geonoma_umbraculiformis Geonoma_laxiflora 50\n",
            "Geonoma_umbraculiformis Geonoma_leptospadix 48\n",
            "Geonoma_umbraculiformis Geonoma_longepedunculata 51\n",
            "Geonoma_umbraculiformis Geonoma_oldemanii 60\n",
            "Geonoma_umbraculiformis Geonoma_oligoclona 54\n",
            "Geonoma_umbraculiformis Geonoma_pinnatifrons 60\n",
            "Geonoma_umbraculiformis Geonoma_poeppigiana 43\n",
            "Geonoma_umbraculiformis Geonoma_triglochin 54\n",
            "Geonoma_umbraculiformis Geonoma_umbraculiformis 100\n",
            "Geonoma_braunii Geonoma_baculifera 67\n",
            "Geonoma_braunii Geonoma_brongniartii 74\n",
            "Geonoma_braunii Geonoma_calyptrogynoidea 51\n",
            "Geonoma_braunii Geonoma_camana 69\n",
            "Geonoma_braunii Geonoma_chlamydostachys 47\n",
            "Geonoma_braunii Geonoma_concinna 58\n",
            "Geonoma_braunii Geonoma_congesta 58\n",
            "Geonoma_braunii Geonoma_cuneata 67\n",
            "Geonoma_braunii Geonoma_deversa 67\n",
            "Geonoma_braunii Geonoma_elegans 67\n",
            "Geonoma_braunii Geonoma_fosteri 67\n",
            "Geonoma_braunii Geonoma_frontinensis 69\n",
            "Geonoma_braunii Geonoma_galeanoae 62\n",
            "Geonoma_braunii Geonoma_hugonis 73\n",
            "Geonoma_braunii Geonoma_interrupta 61\n",
            "Geonoma_braunii Geonoma_lehmannii 75\n",
            "Geonoma_braunii Geonoma_longivaginata 56\n",
            "Geonoma_braunii Geonoma_macrostachys 57\n",
            "Geonoma_braunii Geonoma_maxima 69\n",
            "Geonoma_braunii Geonoma_monospatha 55\n",
            "Geonoma_braunii Geonoma_mooreana 71\n",
            "Geonoma_braunii Geonoma_multisecta 55\n",
            "Geonoma_braunii Geonoma_occidentalis 57\n",
            "Geonoma_braunii Geonoma_orbignyana 67\n",
            "Geonoma_braunii Geonoma_paradoxa 65\n",
            "Geonoma_braunii Geonoma_pauciflora 61\n",
            "Geonoma_braunii Geonoma_pohliana 65\n",
            "Geonoma_braunii Geonoma_procumbens 61\n",
            "Geonoma_braunii Geonoma_rubescens 62\n",
            "Geonoma_braunii Geonoma_santanderensis 59\n",
            "Geonoma_braunii Geonoma_schottiana 61\n",
            "Geonoma_braunii Geonoma_stricta 67\n",
            "Geonoma_braunii Geonoma_triandra 65\n",
            "Geonoma_braunii Geonoma_undata 69\n",
            "Geonoma_braunii Geonoma_aspdiifolia 65\n",
            "Geonoma_braunii Geonoma_atrovirens_(Chamaedorea) 43\n",
            "Geonoma_braunii Geonoma_epetiolata 55\n",
            "Geonoma_braunii Geonoma_euspatha 58\n",
            "Geonoma_braunii Geonoma_hollinensis 59\n",
            "Geonoma_braunii Geonoma_lanata 69\n",
            "Geonoma_braunii Geonoma_laxiflora 62\n",
            "Geonoma_braunii Geonoma_leptospadix 59\n",
            "Geonoma_braunii Geonoma_longepedunculata 51\n",
            "Geonoma_braunii Geonoma_oldemanii 75\n",
            "Geonoma_braunii Geonoma_oligoclona 55\n",
            "Geonoma_braunii Geonoma_pinnatifrons 57\n",
            "Geonoma_braunii Geonoma_poeppigiana 59\n",
            "Geonoma_braunii Geonoma_triglochin 61\n",
            "Geonoma_braunii Geonoma_umbraculiformis 74\n",
            "Geonoma_macrostachys Geonoma_baculifera 63\n",
            "Geonoma_macrostachys Geonoma_brongniartii 55\n",
            "Geonoma_macrostachys Geonoma_calyptrogynoidea 55\n",
            "Geonoma_macrostachys Geonoma_camana 65\n",
            "Geonoma_macrostachys Geonoma_chlamydostachys 79\n",
            "Geonoma_macrostachys Geonoma_concinna 50\n",
            "Geonoma_macrostachys Geonoma_congesta 72\n",
            "Geonoma_macrostachys Geonoma_cuneata 63\n",
            "Geonoma_macrostachys Geonoma_deversa 51\n",
            "Geonoma_macrostachys Geonoma_elegans 57\n",
            "Geonoma_macrostachys Geonoma_fosteri 63\n",
            "Geonoma_macrostachys Geonoma_frontinensis 60\n",
            "Geonoma_macrostachys Geonoma_galeanoae 59\n",
            "Geonoma_macrostachys Geonoma_hugonis 57\n",
            "Geonoma_macrostachys Geonoma_interrupta 58\n",
            "Geonoma_macrostachys Geonoma_lehmannii 54\n",
            "Geonoma_macrostachys Geonoma_longivaginata 54\n",
            "Geonoma_macrostachys Geonoma_macrostachys 100\n",
            "Geonoma_macrostachys Geonoma_maxima 65\n",
            "Geonoma_macrostachys Geonoma_monospatha 68\n",
            "Geonoma_macrostachys Geonoma_mooreana 61\n",
            "Geonoma_macrostachys Geonoma_multisecta 63\n",
            "Geonoma_macrostachys Geonoma_occidentalis 60\n",
            "Geonoma_macrostachys Geonoma_orbignyana 53\n",
            "Geonoma_macrostachys Geonoma_paradoxa 67\n",
            "Geonoma_macrostachys Geonoma_pauciflora 63\n",
            "Geonoma_macrostachys Geonoma_pohliana 56\n",
            "Geonoma_macrostachys Geonoma_procumbens 58\n",
            "Geonoma_macrostachys Geonoma_rubescens 54\n",
            "Geonoma_macrostachys Geonoma_santanderensis 57\n",
            "Geonoma_macrostachys Geonoma_schottiana 58\n",
            "Geonoma_macrostachys Geonoma_stricta 63\n",
            "Geonoma_macrostachys Geonoma_triandra 61\n",
            "Geonoma_macrostachys Geonoma_undata 65\n",
            "Geonoma_macrostachys Geonoma_aspdiifolia 56\n",
            "Geonoma_macrostachys Geonoma_atrovirens_(Chamaedorea) 46\n",
            "Geonoma_macrostachys Geonoma_epetiolata 58\n",
            "Geonoma_macrostachys Geonoma_euspatha 61\n",
            "Geonoma_macrostachys Geonoma_hollinensis 56\n",
            "Geonoma_macrostachys Geonoma_lanata 65\n",
            "Geonoma_macrostachys Geonoma_laxiflora 59\n",
            "Geonoma_macrostachys Geonoma_leptospadix 56\n",
            "Geonoma_macrostachys Geonoma_longepedunculata 50\n",
            "Geonoma_macrostachys Geonoma_oldemanii 54\n",
            "Geonoma_macrostachys Geonoma_oligoclona 47\n",
            "Geonoma_macrostachys Geonoma_pinnatifrons 60\n",
            "Geonoma_macrostachys Geonoma_poeppigiana 51\n",
            "Geonoma_macrostachys Geonoma_triglochin 63\n",
            "Geonoma_macrostachys Geonoma_umbraculiformis 60\n",
            "Geonoma_spinescens Geonoma_baculifera 56\n",
            "Geonoma_spinescens Geonoma_brongniartii 47\n",
            "Geonoma_spinescens Geonoma_calyptrogynoidea 52\n",
            "Geonoma_spinescens Geonoma_camana 56\n",
            "Geonoma_spinescens Geonoma_chlamydostachys 49\n",
            "Geonoma_spinescens Geonoma_concinna 65\n",
            "Geonoma_spinescens Geonoma_congesta 65\n",
            "Geonoma_spinescens Geonoma_cuneata 61\n",
            "Geonoma_spinescens Geonoma_deversa 55\n",
            "Geonoma_spinescens Geonoma_elegans 73\n",
            "Geonoma_spinescens Geonoma_fosteri 61\n",
            "Geonoma_spinescens Geonoma_frontinensis 68\n",
            "Geonoma_spinescens Geonoma_galeanoae 57\n",
            "Geonoma_spinescens Geonoma_hugonis 55\n",
            "Geonoma_spinescens Geonoma_interrupta 61\n",
            "Geonoma_spinescens Geonoma_lehmannii 51\n",
            "Geonoma_spinescens Geonoma_longivaginata 51\n",
            "Geonoma_spinescens Geonoma_macrostachys 53\n",
            "Geonoma_spinescens Geonoma_maxima 56\n",
            "Geonoma_spinescens Geonoma_monospatha 56\n",
            "Geonoma_spinescens Geonoma_mooreana 53\n",
            "Geonoma_spinescens Geonoma_multisecta 61\n",
            "Geonoma_spinescens Geonoma_occidentalis 63\n",
            "Geonoma_spinescens Geonoma_orbignyana 61\n",
            "Geonoma_spinescens Geonoma_paradoxa 53\n",
            "Geonoma_spinescens Geonoma_pauciflora 56\n",
            "Geonoma_spinescens Geonoma_pohliana 65\n",
            "Geonoma_spinescens Geonoma_procumbens 72\n",
            "Geonoma_spinescens Geonoma_rubescens 80\n",
            "Geonoma_spinescens Geonoma_santanderensis 70\n",
            "Geonoma_spinescens Geonoma_schottiana 61\n",
            "Geonoma_spinescens Geonoma_stricta 67\n",
            "Geonoma_spinescens Geonoma_triandra 59\n",
            "Geonoma_spinescens Geonoma_undata 56\n",
            "Geonoma_spinescens Geonoma_aspdiifolia 59\n",
            "Geonoma_spinescens Geonoma_atrovirens_(Chamaedorea) 48\n",
            "Geonoma_spinescens Geonoma_epetiolata 56\n",
            "Geonoma_spinescens Geonoma_euspatha 59\n",
            "Geonoma_spinescens Geonoma_hollinensis 70\n",
            "Geonoma_spinescens Geonoma_lanata 56\n",
            "Geonoma_spinescens Geonoma_laxiflora 51\n",
            "Geonoma_spinescens Geonoma_leptospadix 59\n",
            "Geonoma_spinescens Geonoma_longepedunculata 52\n",
            "Geonoma_spinescens Geonoma_oldemanii 51\n",
            "Geonoma_spinescens Geonoma_oligoclona 56\n",
            "Geonoma_spinescens Geonoma_pinnatifrons 68\n",
            "Geonoma_spinescens Geonoma_poeppigiana 59\n",
            "Geonoma_spinescens Geonoma_triglochin 56\n",
            "Geonoma_spinescens Geonoma_umbraculiformis 44\n",
            "Geonoma_ferruginea Geonoma_baculifera 67\n",
            "Geonoma_ferruginea Geonoma_brongniartii 58\n",
            "Geonoma_ferruginea Geonoma_calyptrogynoidea 62\n",
            "Geonoma_ferruginea Geonoma_camana 62\n",
            "Geonoma_ferruginea Geonoma_chlamydostachys 44\n",
            "Geonoma_ferruginea Geonoma_concinna 65\n",
            "Geonoma_ferruginea Geonoma_congesta 59\n",
            "Geonoma_ferruginea Geonoma_cuneata 73\n",
            "Geonoma_ferruginea Geonoma_deversa 67\n",
            "Geonoma_ferruginea Geonoma_elegans 67\n",
            "Geonoma_ferruginea Geonoma_fosteri 73\n",
            "Geonoma_ferruginea Geonoma_frontinensis 68\n",
            "Geonoma_ferruginea Geonoma_galeanoae 63\n",
            "Geonoma_ferruginea Geonoma_hugonis 67\n",
            "Geonoma_ferruginea Geonoma_interrupta 72\n",
            "Geonoma_ferruginea Geonoma_lehmannii 57\n",
            "Geonoma_ferruginea Geonoma_longivaginata 62\n",
            "Geonoma_ferruginea Geonoma_macrostachys 53\n",
            "Geonoma_ferruginea Geonoma_maxima 62\n",
            "Geonoma_ferruginea Geonoma_monospatha 56\n",
            "Geonoma_ferruginea Geonoma_mooreana 65\n",
            "Geonoma_ferruginea Geonoma_multisecta 56\n",
            "Geonoma_ferruginea Geonoma_occidentalis 53\n",
            "Geonoma_ferruginea Geonoma_orbignyana 67\n",
            "Geonoma_ferruginea Geonoma_paradoxa 59\n",
            "Geonoma_ferruginea Geonoma_pauciflora 61\n",
            "Geonoma_ferruginea Geonoma_pohliana 65\n",
            "Geonoma_ferruginea Geonoma_procumbens 56\n",
            "Geonoma_ferruginea Geonoma_rubescens 63\n",
            "Geonoma_ferruginea Geonoma_santanderensis 55\n",
            "Geonoma_ferruginea Geonoma_schottiana 61\n",
            "Geonoma_ferruginea Geonoma_stricta 67\n",
            "Geonoma_ferruginea Geonoma_triandra 65\n",
            "Geonoma_ferruginea Geonoma_undata 69\n",
            "Geonoma_ferruginea Geonoma_aspdiifolia 59\n",
            "Geonoma_ferruginea Geonoma_atrovirens_(Chamaedorea) 48\n",
            "Geonoma_ferruginea Geonoma_epetiolata 61\n",
            "Geonoma_ferruginea Geonoma_euspatha 65\n",
            "Geonoma_ferruginea Geonoma_hollinensis 59\n",
            "Geonoma_ferruginea Geonoma_lanata 62\n",
            "Geonoma_ferruginea Geonoma_laxiflora 63\n",
            "Geonoma_ferruginea Geonoma_leptospadix 54\n",
            "Geonoma_ferruginea Geonoma_longepedunculata 57\n",
            "Geonoma_ferruginea Geonoma_oldemanii 57\n",
            "Geonoma_ferruginea Geonoma_oligoclona 61\n",
            "Geonoma_ferruginea Geonoma_pinnatifrons 58\n",
            "Geonoma_ferruginea Geonoma_poeppigiana 70\n",
            "Geonoma_ferruginea Geonoma_triglochin 67\n",
            "Geonoma_ferruginea Geonoma_umbraculiformis 54\n",
            "Geonoma_tenuissima Geonoma_baculifera 56\n",
            "Geonoma_tenuissima Geonoma_brongniartii 58\n",
            "Geonoma_tenuissima Geonoma_calyptrogynoidea 52\n",
            "Geonoma_tenuissima Geonoma_camana 62\n",
            "Geonoma_tenuissima Geonoma_chlamydostachys 49\n",
            "Geonoma_tenuissima Geonoma_concinna 65\n",
            "Geonoma_tenuissima Geonoma_congesta 59\n",
            "Geonoma_tenuissima Geonoma_cuneata 61\n",
            "Geonoma_tenuissima Geonoma_deversa 67\n",
            "Geonoma_tenuissima Geonoma_elegans 67\n",
            "Geonoma_tenuissima Geonoma_fosteri 67\n",
            "Geonoma_tenuissima Geonoma_frontinensis 68\n",
            "Geonoma_tenuissima Geonoma_galeanoae 63\n",
            "Geonoma_tenuissima Geonoma_hugonis 67\n",
            "Geonoma_tenuissima Geonoma_interrupta 67\n",
            "Geonoma_tenuissima Geonoma_lehmannii 63\n",
            "Geonoma_tenuissima Geonoma_longivaginata 51\n",
            "Geonoma_tenuissima Geonoma_macrostachys 53\n",
            "Geonoma_tenuissima Geonoma_maxima 69\n",
            "Geonoma_tenuissima Geonoma_monospatha 56\n",
            "Geonoma_tenuissima Geonoma_mooreana 65\n",
            "Geonoma_tenuissima Geonoma_multisecta 67\n",
            "Geonoma_tenuissima Geonoma_occidentalis 63\n",
            "Geonoma_tenuissima Geonoma_orbignyana 56\n",
            "Geonoma_tenuissima Geonoma_paradoxa 53\n",
            "Geonoma_tenuissima Geonoma_pauciflora 61\n",
            "Geonoma_tenuissima Geonoma_pohliana 59\n",
            "Geonoma_tenuissima Geonoma_procumbens 61\n",
            "Geonoma_tenuissima Geonoma_rubescens 63\n",
            "Geonoma_tenuissima Geonoma_santanderensis 65\n",
            "Geonoma_tenuissima Geonoma_schottiana 61\n",
            "Geonoma_tenuissima Geonoma_stricta 67\n",
            "Geonoma_tenuissima Geonoma_triandra 65\n",
            "Geonoma_tenuissima Geonoma_undata 62\n",
            "Geonoma_tenuissima Geonoma_aspdiifolia 59\n",
            "Geonoma_tenuissima Geonoma_atrovirens_(Chamaedorea) 56\n",
            "Geonoma_tenuissima Geonoma_epetiolata 61\n",
            "Geonoma_tenuissima Geonoma_euspatha 59\n",
            "Geonoma_tenuissima Geonoma_hollinensis 65\n",
            "Geonoma_tenuissima Geonoma_lanata 62\n",
            "Geonoma_tenuissima Geonoma_laxiflora 57\n",
            "Geonoma_tenuissima Geonoma_leptospadix 54\n",
            "Geonoma_tenuissima Geonoma_longepedunculata 48\n",
            "Geonoma_tenuissima Geonoma_oldemanii 63\n",
            "Geonoma_tenuissima Geonoma_oligoclona 56\n",
            "Geonoma_tenuissima Geonoma_pinnatifrons 58\n",
            "Geonoma_tenuissima Geonoma_poeppigiana 59\n",
            "Geonoma_tenuissima Geonoma_triglochin 56\n",
            "Geonoma_tenuissima Geonoma_umbraculiformis 54\n",
            "Geonoma_bernalii Geonoma_baculifera 71\n",
            "Geonoma_bernalii Geonoma_brongniartii 78\n",
            "Geonoma_bernalii Geonoma_calyptrogynoidea 55\n",
            "Geonoma_bernalii Geonoma_camana 67\n",
            "Geonoma_bernalii Geonoma_chlamydostachys 46\n",
            "Geonoma_bernalii Geonoma_concinna 62\n",
            "Geonoma_bernalii Geonoma_congesta 62\n",
            "Geonoma_bernalii Geonoma_cuneata 65\n",
            "Geonoma_bernalii Geonoma_deversa 71\n",
            "Geonoma_bernalii Geonoma_elegans 65\n",
            "Geonoma_bernalii Geonoma_fosteri 71\n",
            "Geonoma_bernalii Geonoma_frontinensis 61\n",
            "Geonoma_bernalii Geonoma_galeanoae 61\n",
            "Geonoma_bernalii Geonoma_hugonis 65\n",
            "Geonoma_bernalii Geonoma_interrupta 65\n",
            "Geonoma_bernalii Geonoma_lehmannii 73\n",
            "Geonoma_bernalii Geonoma_longivaginata 54\n",
            "Geonoma_bernalii Geonoma_macrostachys 56\n",
            "Geonoma_bernalii Geonoma_maxima 67\n",
            "Geonoma_bernalii Geonoma_monospatha 59\n",
            "Geonoma_bernalii Geonoma_mooreana 69\n",
            "Geonoma_bernalii Geonoma_multisecta 59\n",
            "Geonoma_bernalii Geonoma_occidentalis 72\n",
            "Geonoma_bernalii Geonoma_orbignyana 65\n",
            "Geonoma_bernalii Geonoma_paradoxa 62\n",
            "Geonoma_bernalii Geonoma_pauciflora 59\n",
            "Geonoma_bernalii Geonoma_pohliana 62\n",
            "Geonoma_bernalii Geonoma_procumbens 65\n",
            "Geonoma_bernalii Geonoma_rubescens 67\n",
            "Geonoma_bernalii Geonoma_santanderensis 63\n",
            "Geonoma_bernalii Geonoma_schottiana 59\n",
            "Geonoma_bernalii Geonoma_stricta 65\n",
            "Geonoma_bernalii Geonoma_triandra 69\n",
            "Geonoma_bernalii Geonoma_undata 67\n",
            "Geonoma_bernalii Geonoma_aspdiifolia 63\n",
            "Geonoma_bernalii Geonoma_atrovirens_(Chamaedorea) 46\n",
            "Geonoma_bernalii Geonoma_epetiolata 59\n",
            "Geonoma_bernalii Geonoma_euspatha 62\n",
            "Geonoma_bernalii Geonoma_hollinensis 63\n",
            "Geonoma_bernalii Geonoma_lanata 67\n",
            "Geonoma_bernalii Geonoma_laxiflora 61\n",
            "Geonoma_bernalii Geonoma_leptospadix 63\n",
            "Geonoma_bernalii Geonoma_longepedunculata 55\n",
            "Geonoma_bernalii Geonoma_oldemanii 73\n",
            "Geonoma_bernalii Geonoma_oligoclona 59\n",
            "Geonoma_bernalii Geonoma_pinnatifrons 61\n",
            "Geonoma_bernalii Geonoma_poeppigiana 63\n",
            "Geonoma_bernalii Geonoma_triglochin 59\n",
            "Geonoma_bernalii Geonoma_umbraculiformis 72\n",
            "Geonoma_chlamydostachys Geonoma_baculifera 49\n",
            "Geonoma_chlamydostachys Geonoma_brongniartii 47\n",
            "Geonoma_chlamydostachys Geonoma_calyptrogynoidea 47\n",
            "Geonoma_chlamydostachys Geonoma_camana 65\n",
            "Geonoma_chlamydostachys Geonoma_chlamydostachys 100\n",
            "Geonoma_chlamydostachys Geonoma_concinna 51\n",
            "Geonoma_chlamydostachys Geonoma_congesta 67\n",
            "Geonoma_chlamydostachys Geonoma_cuneata 63\n",
            "Geonoma_chlamydostachys Geonoma_deversa 47\n",
            "Geonoma_chlamydostachys Geonoma_elegans 58\n",
            "Geonoma_chlamydostachys Geonoma_fosteri 58\n",
            "Geonoma_chlamydostachys Geonoma_frontinensis 51\n",
            "Geonoma_chlamydostachys Geonoma_galeanoae 60\n",
            "Geonoma_chlamydostachys Geonoma_hugonis 58\n",
            "Geonoma_chlamydostachys Geonoma_interrupta 49\n",
            "Geonoma_chlamydostachys Geonoma_lehmannii 50\n",
            "Geonoma_chlamydostachys Geonoma_longivaginata 55\n",
            "Geonoma_chlamydostachys Geonoma_macrostachys 79\n",
            "Geonoma_chlamydostachys Geonoma_maxima 59\n",
            "Geonoma_chlamydostachys Geonoma_monospatha 63\n",
            "Geonoma_chlamydostachys Geonoma_mooreana 51\n",
            "Geonoma_chlamydostachys Geonoma_multisecta 54\n",
            "Geonoma_chlamydostachys Geonoma_occidentalis 60\n",
            "Geonoma_chlamydostachys Geonoma_orbignyana 49\n",
            "Geonoma_chlamydostachys Geonoma_paradoxa 62\n",
            "Geonoma_chlamydostachys Geonoma_pauciflora 54\n",
            "Geonoma_chlamydostachys Geonoma_pohliana 62\n",
            "Geonoma_chlamydostachys Geonoma_procumbens 54\n",
            "Geonoma_chlamydostachys Geonoma_rubescens 50\n",
            "Geonoma_chlamydostachys Geonoma_santanderensis 53\n",
            "Geonoma_chlamydostachys Geonoma_schottiana 59\n",
            "Geonoma_chlamydostachys Geonoma_stricta 58\n",
            "Geonoma_chlamydostachys Geonoma_triandra 56\n",
            "Geonoma_chlamydostachys Geonoma_undata 59\n",
            "Geonoma_chlamydostachys Geonoma_aspdiifolia 48\n",
            "Geonoma_chlamydostachys Geonoma_atrovirens_(Chamaedorea) 55\n",
            "Geonoma_chlamydostachys Geonoma_epetiolata 59\n",
            "Geonoma_chlamydostachys Geonoma_euspatha 51\n",
            "Geonoma_chlamydostachys Geonoma_hollinensis 57\n",
            "Geonoma_chlamydostachys Geonoma_lanata 65\n",
            "Geonoma_chlamydostachys Geonoma_laxiflora 60\n",
            "Geonoma_chlamydostachys Geonoma_leptospadix 57\n",
            "Geonoma_chlamydostachys Geonoma_longepedunculata 55\n",
            "Geonoma_chlamydostachys Geonoma_oldemanii 50\n",
            "Geonoma_chlamydostachys Geonoma_oligoclona 54\n",
            "Geonoma_chlamydostachys Geonoma_pinnatifrons 51\n",
            "Geonoma_chlamydostachys Geonoma_poeppigiana 48\n",
            "Geonoma_chlamydostachys Geonoma_triglochin 49\n",
            "Geonoma_chlamydostachys Geonoma_umbraculiformis 52\n",
            "Geonoma_elegans Geonoma_baculifera 61\n",
            "Geonoma_elegans Geonoma_brongniartii 57\n",
            "Geonoma_elegans Geonoma_calyptrogynoidea 51\n",
            "Geonoma_elegans Geonoma_camana 69\n",
            "Geonoma_elegans Geonoma_chlamydostachys 58\n",
            "Geonoma_elegans Geonoma_concinna 58\n",
            "Geonoma_elegans Geonoma_congesta 65\n",
            "Geonoma_elegans Geonoma_cuneata 67\n",
            "Geonoma_elegans Geonoma_deversa 73\n",
            "Geonoma_elegans Geonoma_elegans 100\n",
            "Geonoma_elegans Geonoma_fosteri 60\n",
            "Geonoma_elegans Geonoma_frontinensis 63\n",
            "Geonoma_elegans Geonoma_galeanoae 75\n",
            "Geonoma_elegans Geonoma_hugonis 73\n",
            "Geonoma_elegans Geonoma_interrupta 61\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KBLsXlJ-rVnw"
      },
      "source": [
        "And the very last thing, export the dataframe."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WJA6N3t9rYaI"
      },
      "source": [
        "df.to_csv('./palm_metadata_%s_v2.txt'%(str(datetime.datetime.now()).split(' ')[0]), sep = '\\t')"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4_kCnsPUqS6o"
      },
      "source": [
        "You can import your own data into Colab notebooks from your Google Drive account, including from spreadsheets, as well as from Github and many other sources. To learn more about importing data, and how Colab can be used for data science, see the links below under [Working with Data](#working-with-data)."
      ]
    }
  ]
}