{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "American_palm_data.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/mftorres/APP/blob/main/American_palm_data.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5fCEDCU_qrC0"
      },
      "source": [
        "# Leaving this here in case you need info about Colaboratory in Google\n",
        "\n",
        "[link text](https://)<p><img alt=\"Colaboratory logo\" height=\"45px\" src=\"/img/colab_favicon.ico\" align=\"left\" hspace=\"10px\" vspace=\"0px\"></p>\n",
        "\n",
        "<h1>What is Colaboratory?</h1>\n",
        "\n",
        "Colaboratory, or \"Colab\" for short, allows you to write and execute Python in your browser, with \n",
        "- Zero configuration required\n",
        "- Free access to GPUs\n",
        "- Easy sharing\n",
        "\n",
        "Whether you're a **student**, a **data scientist** or an **AI researcher**, Colab can make your work easier. Watch [Introduction to Colab](https://www.youtube.com/watch?v=inN8seMm7UI) to learn more, or just get started below!"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GJBs_flRovLc"
      },
      "source": [
        "----\n",
        "\n",
        "----\n",
        "\n",
        "# **American Palm Phylogeny data exploration/management**\n",
        "\n",
        "Document authored by **Maria Fernanda Torres Jimenez**  \n",
        "Date: 2021 May 05\n",
        "\n",
        "____\n",
        "\n",
        "\n",
        "The aim of this notebook is to show you how to retrieve data from the American Palm Phylogeny metadata, how to query, and how to update records.\n",
        "\n",
        "Maintaining metadata associated to files requires you to do version control, properly dated, and to keep backups at all times. You should be prepared to accidentally change/delete metadata and you should always be able to go back to a stable version (but that depends on you and your data practices).\n",
        "\n",
        "\n",
        "## 0. Importing packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eZXbXzVW8Bj6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "13a753ba-9b97-4774-f19d-6ce57e91b121"
      },
      "source": [
        "import pandas as pd\n",
        "import re\n",
        "import matplotlib as mpl\n",
        "from matplotlib import pyplot as plt\n",
        "import requests as rs\n",
        "import datetime\n",
        "import numpy as np\n",
        "\n",
        "!pip install fuzzywuzzy # install package to do partial text match\n",
        "from fuzzywuzzy import fuzz\n",
        "from fuzzywuzzy import process"
      ],
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: fuzzywuzzy in /usr/local/lib/python3.7/dist-packages (0.18.0)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rKut_-9OVXnU"
      },
      "source": [
        "The database is stored in a google spreadsheet. It is good to have it in the cloud, google has version control and everyone can have access to the document without it staying with a single person.\n",
        "\n",
        "Creating a function to download the data:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 609
        },
        "id": "gJr_9dXGpJ05",
        "outputId": "7596df3e-a976-4d87-8d0e-8b4bb6272f68"
      },
      "source": [
        "def get_data():\n",
        "  return pd.read_csv('./palm_metadata_%s.txt'%(str(datetime.datetime.now()).split(' ')[0]), sep = '\\t')\n",
        "\n",
        "# try reading the file if exists, keep in mind the date\n",
        "try:\n",
        "  df = get_data()\n",
        "except:\n",
        "  url = 'https://raw.githubusercontent.com/mftorres/APP/main/data/Metadata_palm_gap_2021-05-05.txt' # shared link careful who you share this with\n",
        "  print('Downloading')\n",
        "  res = rs.get(url = url)\n",
        "  with open('./palm_metadata_%s.txt'%(str(datetime.datetime.now()).split(' ')[0]), 'wb') as file: # Writing document in binary mode so python doesn't make changes\n",
        "    file.write(res.content)\n",
        "  df = get_data()\n",
        "\n",
        "df # check the database"
      ],
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>filename</th>\n",
              "      <th>sense</th>\n",
              "      <th>filecode</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "      <th>botanic_garden</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>ifmorphotype</th>\n",
              "      <th>ifpopulation</th>\n",
              "      <th>sent_to_Cano</th>\n",
              "      <th>newfilename</th>\n",
              "      <th>raw_reads</th>\n",
              "      <th>reads_trimed_paired_flag</th>\n",
              "      <th>reads_trimed_paired</th>\n",
              "      <th>reads_trimed_single</th>\n",
              "      <th>percentage_lost_tosingles</th>\n",
              "      <th>pcr_filtered</th>\n",
              "      <th>collection_year</th>\n",
              "      <th>Continent</th>\n",
              "      <th>country</th>\n",
              "      <th>long</th>\n",
              "      <th>lat</th>\n",
              "      <th>flag</th>\n",
              "      <th>notes</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AKG91_P3252_1182_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AKG91_P3252_1182</td>\n",
              "      <td>AKG91_1182</td>\n",
              "      <td>CTAGCT</td>\n",
              "      <td>JBP</td>\n",
              "      <td>Lorenzi_&amp;_Soares_6762</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_emensis</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_eme_AKG911182_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3044</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_vulgata_Standley67344_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>Standley67344</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_vulgata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3045</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>Azofeifa484</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3046</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>Azofeifa484</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3047</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...</td>\n",
              "      <td>R1</td>\n",
              "      <td>Azofeifa484_S36_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3048</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...</td>\n",
              "      <td>R2</td>\n",
              "      <td>Azofeifa484_S36_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3049 rows × 30 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "            dataset        provider pi_code  ... lat flag notes\n",
              "0              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "1              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "2              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "3              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "4              Cano     Angela Cano      AC  ... NaN  NaN   NaN\n",
              "...             ...             ...     ...  ...  ..  ...   ...\n",
              "3044  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3045  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3046  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3047  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3048  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "\n",
              "[3049 rows x 30 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MjHzGbeechIv"
      },
      "source": [
        "The first thing to check is the columns on the data and what they mean"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HBaHmjmGckXD",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8c59bd83-c238-4577-f71b-c17e426a2512"
      },
      "source": [
        "list(df.columns)"
      ],
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['dataset',\n",
              " 'provider',\n",
              " 'pi_code',\n",
              " 'originalmetadata',\n",
              " 'filename',\n",
              " 'sense',\n",
              " 'filecode',\n",
              " 'samplecode',\n",
              " 'library_index',\n",
              " 'botanic_garden',\n",
              " 'voucher',\n",
              " 'taxgenus',\n",
              " 'taxspecies',\n",
              " 'ifmorphotype',\n",
              " 'ifpopulation',\n",
              " 'sent_to_Cano',\n",
              " 'newfilename',\n",
              " 'raw_reads',\n",
              " 'reads_trimed_paired_flag',\n",
              " 'reads_trimed_paired',\n",
              " 'reads_trimed_single',\n",
              " 'percentage_lost_tosingles',\n",
              " 'pcr_filtered',\n",
              " 'collection_year',\n",
              " 'Continent',\n",
              " 'country',\n",
              " 'long',\n",
              " 'lat',\n",
              " 'flag',\n",
              " 'notes']"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 18
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tD7Ifx2zx0Jv"
      },
      "source": [
        "**dataset** bundle in which files are provided. e.g. CANO: sequences extracted by Angela Cano  \n",
        "**provider** who provided the files or is responsible for sending the files  \n",
        "**pi_code** two-letter code to shorten the *provider*  \n",
        "**originalmetadata** name of original metadata provided by provider. That file should never be modified and only annotated  \n",
        "**filename** original file name, usually as delivered by the sequencing facility  \n",
        "**sense** whether file contains forward, reverse, interleaved, single reads    \n",
        "**filecode** code provided by sequencing facility, often links to sequencing experiment, lane and plate information  \n",
        "**samplecode** sample ID in the plate or as delivered to the sequencing facility\n",
        "**library_index** sequencing index if provided or if in header of fastq. When double, each index is separated with a '+'  \n",
        "**botanic_garden** if sample was obtained from a botanic garden rather than collected  \n",
        "**voucher** voucher of the tissue or collecting code  \n",
        "**taxgenus** taxonomic genus of the sample  \n",
        "**taxspecies** taxonomic species with genus, separated by an underscore '_'  \n",
        "**ifmorphotype** if the sample has a morphotype, variety or subspecies assigned  \n",
        "**ifpopulation** if the sample belongs to a population-wide sampling effort, which population it was collected from  \n",
        "**sent_to_Cano** files sent to Angela Cano who's analysing the data and leading the project  \n",
        "**newfilename** new file name assigned to the sample. This is to annotate provider, dataset, species and voucher in the name. See Appendix for the code used to name files  \n",
        "**raw_reads** number of raw reads before processing the data  -- filled as analysis progresses  \n",
        "**reads_trimed_paired_flag** if trimmed -- filled as analysis progresses  \n",
        "**reads_trimed_paired** number of reads after trimming -- filled as analysis progresses  \n",
        "**reads_trimed_single** number of unpaired reads after trimming -- filled as analysis progresses  \n",
        "**percentage_lost_tosingles** percentage of unpaired reads from the total of reads -- filled as analysis progresses  \n",
        "**pcr_filtered** if PCR duplicate reads were filtered out using samtools or picard-- filled as analysis progresses  \n",
        "**collection_year** year that sample was collected  \n",
        "**Continent** continent. The project involves species from America but the metadata keeps track of other data from other continents  \n",
        "**country** which country was the sample collected at if known from collection or voucher  \n",
        "**long** longitude coordinates if available, should be standardised to decimals  \n",
        "**lat**  latitude coordinates if available, should be standardised to decimals  \n",
        "**flag** single word descriptors separated by spaces (never colons, semicolons or tabs) that describe the flaws of the file. e.g. misingtaxspecies  \n",
        "**notes** elaborated notes if needed but avoid using commas, semicolons or tabs.  "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2fhs6GZ4qFMx"
      },
      "source": [
        "The second interesting thing to do is filtering unique species for which we have data. The files here represent species that might or might not be good, but at this point of the analysis (which will be on a sequence by sequence basis) we can only know what we have sequenced.\n",
        "\n",
        "Another thing is to filter out species that are not in America."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-gE-Ez1qtyIA",
        "outputId": "085ab642-5323-4951-dc74-03a4eee2dea1"
      },
      "source": [
        "df['Continent'].unique()"
      ],
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['americas', 'out_americas', nan], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 19
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1GQlEP_TiZ4u"
      },
      "source": [
        "Some species don't have a continent specified. Let's see which ones:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Y9MDwf5NioT4",
        "outputId": "a681f951-ced6-472a-cc90-d3aee0d186fb"
      },
      "source": [
        "# filtering pandas dataframe\n",
        "df[df['Continent'].isna()]['taxspecies'].unique()"
      ],
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Ceroxylon_alpinum', 'Ceroxylon_ceriferum',\n",
              "       'Ceroxylon_echinulatum', 'Ceroxylon_parvifrons',\n",
              "       'Ceroxylon_parvum', 'Ceroxylon_quindiuense',\n",
              "       'Ceroxylon_ventricosum', 'Ceroxylon_vogelianum', 'Chamaedorea',\n",
              "       'Cocos', nan, 'Phytelephas_aecuatorialis',\n",
              "       'Phytelephas_macrocarpa', 'Phytelephas', 'Phytelephas_tumacana',\n",
              "       'Ravenea_sambiranensis', 'Chamaedorea_nationsiana',\n",
              "       'Chamaedorea_arenbergiana', 'Chamaedorea_binderi',\n",
              "       'Chamaedorea_brachyclada', 'Chamaedorea_seifrizii',\n",
              "       'Chamaedorea_fractiflexa', 'Chamaedorea_graminifolia',\n",
              "       'Chamaedorea_hodelii', 'Chamaedorea_ibarrae',\n",
              "       'Chamaedorea_incrustata', 'Chamaedorea_keelerorum',\n",
              "       'Chamaedorea_lehmannii', 'Chamaedorea_liebmannii',\n",
              "       'Chamaedorea_pachenoana', 'Chamaedorea_parvisecta',\n",
              "       'Chamaedorea_pauciflora', 'Chamaedorea_piscifolia',\n",
              "       'Chamaedorea_pumila', 'Chamaedorea_costaricana',\n",
              "       'Chamaedorea_rigida', 'Chamaedorea_schideana',\n",
              "       'Chamaedorea_simplex', 'Chamaedorea_skutchii',\n",
              "       'Chamaedorea_stenocarpa', 'Chamaedorea_subjectifolia',\n",
              "       'Chamaedorea_tenerrima', 'Chamaedorea_verapazensis',\n",
              "       'Chamaedorea_volcanensis', 'Chamaedorea_vulgata',\n",
              "       'Chamaedorea_zamorae'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 20
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3d2HoUqji2WB",
        "outputId": "3426a80f-bce5-49a7-b8c0-34ce1af7a173"
      },
      "source": [
        "# using a diferent syntax that can help with readability\n",
        "df[df.Continent.isna()]['taxspecies'].unique()"
      ],
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Ceroxylon_alpinum', 'Ceroxylon_ceriferum',\n",
              "       'Ceroxylon_echinulatum', 'Ceroxylon_parvifrons',\n",
              "       'Ceroxylon_parvum', 'Ceroxylon_quindiuense',\n",
              "       'Ceroxylon_ventricosum', 'Ceroxylon_vogelianum', 'Chamaedorea',\n",
              "       'Cocos', nan, 'Phytelephas_aecuatorialis',\n",
              "       'Phytelephas_macrocarpa', 'Phytelephas', 'Phytelephas_tumacana',\n",
              "       'Ravenea_sambiranensis', 'Chamaedorea_nationsiana',\n",
              "       'Chamaedorea_arenbergiana', 'Chamaedorea_binderi',\n",
              "       'Chamaedorea_brachyclada', 'Chamaedorea_seifrizii',\n",
              "       'Chamaedorea_fractiflexa', 'Chamaedorea_graminifolia',\n",
              "       'Chamaedorea_hodelii', 'Chamaedorea_ibarrae',\n",
              "       'Chamaedorea_incrustata', 'Chamaedorea_keelerorum',\n",
              "       'Chamaedorea_lehmannii', 'Chamaedorea_liebmannii',\n",
              "       'Chamaedorea_pachenoana', 'Chamaedorea_parvisecta',\n",
              "       'Chamaedorea_pauciflora', 'Chamaedorea_piscifolia',\n",
              "       'Chamaedorea_pumila', 'Chamaedorea_costaricana',\n",
              "       'Chamaedorea_rigida', 'Chamaedorea_schideana',\n",
              "       'Chamaedorea_simplex', 'Chamaedorea_skutchii',\n",
              "       'Chamaedorea_stenocarpa', 'Chamaedorea_subjectifolia',\n",
              "       'Chamaedorea_tenerrima', 'Chamaedorea_verapazensis',\n",
              "       'Chamaedorea_volcanensis', 'Chamaedorea_vulgata',\n",
              "       'Chamaedorea_zamorae'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 21
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ag5uTCqAjIIi",
        "outputId": "0f6e3f53-24d8-4197-f5a5-7694c74e0ab6"
      },
      "source": [
        "df[df.Continent.isna()]['taxgenus'].unique()"
      ],
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Ceroxylon', 'Chamaedorea', 'Cocos', nan, 'Phytelephas', 'Ravenea'],\n",
              "      dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 22
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IQmksWQZjDGx"
      },
      "source": [
        "Most species are from the genera *Ceroxylon*, *Chamaedorea*, *Cocos*, *Phytelephas*, and *Ravenea*. I can already remember the dataset they come from and I can tell they don't have a continent assigned because I haven't updated the dataframe (as is not my main priority)\n",
        "\n",
        "we can totally fix that today though. Let's iterate through rows in the dataset to assign the continent information. We also know all but *Cocos* (which is cosmopolita) are from America."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WIpoqq1ojBWq",
        "outputId": "dc84d5fc-3306-43d3-f991-9b40254f27a0"
      },
      "source": [
        "# itertuples iterates through tuples (pairs) of index,row and different columns can be acceced using the column label. Similar to the iterrows() method but faster.\n",
        "# itertuples() only allows for the \"dot\" syntax for filtering pandas columns.\n",
        "\n",
        "for row in df[df['Continent'].isna()].itertuples():\n",
        "  # print(row.Continent,'empty') # test\n",
        "  df.loc[row.Index,'Continent'] = 'americas' # loc assigns a value in a cell with coordinates [row, column] labels\n",
        "df['Continent'].unique() # no nan should appear"
      ],
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['americas', 'out_americas'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 23
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZDgD093GuFbP"
      },
      "source": [
        "Nice. Now we can filter all species in America and count/list them"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sljrTqz4uLuO",
        "outputId": "6d114a68-74ae-4032-e157-c5f41614c96c"
      },
      "source": [
        "df[df['Continent'] == 'americas']['taxspecies'].unique()"
      ],
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Acrocomia_aculeata', 'Acrocomia_crispa', 'Acrocomia_emensis',\n",
              "       'Acrocomia_glaucescens', 'Acrocomia_intumescens',\n",
              "       'Acrocomia_totai', 'Aiphanes_acaulis', 'Aiphanes_buenaventurae',\n",
              "       'Aiphanes_concinna', 'Aiphanes_erinacea', 'Aiphanes_gelatinosa',\n",
              "       'Aiphanes_hirsuta', 'Aiphanes_horrida', 'Aiphanes_killipii',\n",
              "       'Aiphanes_leiostachys', 'Aiphanes_lindeniana', 'Aiphanes_linearis',\n",
              "       'Aiphanes_macroloba', 'Aiphanes_minima', 'Aiphanes_parvifolia',\n",
              "       'Aiphanes_pilaris', 'Aiphanes_simplex', 'Aiphanes_tricuspidata',\n",
              "       'Aiphanes_ulei', 'Allagoptera_arenaria', 'Allagoptera_brevicalyx',\n",
              "       'Allagoptera_caudescens', 'Allagoptera_leucocalyx',\n",
              "       'Allagoptera_pectinata', 'Ammandra_decasperma', 'Aphandra_natalia',\n",
              "       'Astrocaryum_acaule', 'Astrocaryum_aculeatum',\n",
              "       'Astrocaryum_alatum', 'Astrocaryum_campestre',\n",
              "       'Astrocaryum_carnosum', 'Astrocaryum_chambira',\n",
              "       'Astrocaryum_chonta', 'Astrocaryum_ciliatum',\n",
              "       'Astrocaryum_echinatum', 'Astrocaryum_faranae',\n",
              "       'Astrocaryum_ferrugineum', 'Astrocaryum_gratum',\n",
              "       'Astrocaryum_gynacanthum', 'Astrocaryum_huicungo',\n",
              "       'Astrocaryum_javarense', 'Astrocaryum_macrocalyx',\n",
              "       'Astrocaryum_malybo', 'Asterogyne_martiana',\n",
              "       'Astrocaryum_mexicanum', 'Astrocaryum_murumuru',\n",
              "       'Astrocaryum_perangustatum', 'Astrocaryum_sciophilum',\n",
              "       'Astrocaryum_scopatum', 'Astrocaryum_standleyanum',\n",
              "       'Astrocaryum_triandrum', 'Astrocaryum_ulei',\n",
              "       'Astrocaryum_urostachys', 'Attalea_allenii', 'Attalea_amygdalina',\n",
              "       'Attalea_anisitsiana', 'Attalea_apoda', 'Attalea_attaleoides',\n",
              "       'Attalea_barreirensis', 'Attalea_brasiliensis',\n",
              "       'Attalea_brejinhoensis', 'Attalea_burretiana', 'Attalea_butyracea',\n",
              "       'Attalea_cohune', 'Attalea_colenda', 'Attalea_crassispatha',\n",
              "       'Attalea_cuatrecasana', 'Attalea_cuatrecasa', 'Attalea_dubia',\n",
              "       'Attalea_eichleri', 'Attalea_funifera', 'Attalea_gearensis',\n",
              "       'Attalea_geraensis', 'Attalea_guacuyule', 'Attalea_humilis',\n",
              "       'Attalea_iguadummat', 'Attalea_insignis', 'Attalea_leandroana',\n",
              "       'Attalea_luetzelburgii', 'Attalea_maripa', 'Attalea_microcarpa',\n",
              "       'Attalea_moorei', 'Attalea_nucifera', 'Attalea_oleifera',\n",
              "       'Attalea_phalerata', 'Attalea_pindobassu', 'Attalea_plowmanii',\n",
              "       'Attalea_princeps', 'Attalea_sagotii', 'Attalea_seabrensis',\n",
              "       'Attalea_spectabilis', 'Attalea_speciosa', 'Attalea_tessmannii',\n",
              "       'Attalea_vitrivir', 'Bactris_barronis', 'Bactris_bidentula',\n",
              "       'Bactris_brongniartii', 'Bactris_caryotifolia',\n",
              "       'Bactris_charnleyae', 'Bactris_coloniata', 'Bactris_coloradonis',\n",
              "       'Bactris_concinna', 'Bactris_elegans', 'Bactris_fissifrons',\n",
              "       'Bactris_gasipaes', 'Bactris_glandulosa', 'Bactris_glaucescens',\n",
              "       'Bactris_gracilior', 'Bactris_hirta', 'Bactris_killipii',\n",
              "       'Bactris_macroacantha', 'Bactris_major', 'Bactris_manriquei',\n",
              "       'Bactris_maraja', 'Bactris_mexicana', 'Bactris_militaris',\n",
              "       'Bactris_neomilitaris', 'Bactris_obovata', 'Bactris_panamensis',\n",
              "       'Bactris_pickellii', 'Bactris_pilosa', 'Bactris_plumeriana',\n",
              "       'Bactris_setulosa', 'Bactris_setosa', 'Bactris_simplicifrons',\n",
              "       'Barcella_odora', 'Brahea_aculeata', 'Brahea_armata',\n",
              "       'Brahea_bella', 'Brahea_dulcis', 'Brahea_nitida',\n",
              "       'Brahea_sarukhanii', 'Butia_archeri', 'Butia_campicola',\n",
              "       'Butia_capitata', 'Butia_catharinensis', 'Butia_eriospatha',\n",
              "       'Butia_exospadix', 'Butia_lallemantii', 'Butia_leiospatha',\n",
              "       'Butia_lepidotispatha', 'Butia_leptospatha', 'Butia_marmori',\n",
              "       'Butia_matogrossensis', 'Butia_microspadix', 'Butia_noblickii',\n",
              "       'Butia_odorata', 'Butia_paraguayensis', 'Butia_pubispatha',\n",
              "       'Butia_purpurascens', 'Butia_witecki', 'Butia_yatay',\n",
              "       'Calyptrogyne_allenii', 'Calyptrogyne_anomala',\n",
              "       'Calyptrogyne_costatifrons', 'Calyptrogyne_ghiesbreghtiana',\n",
              "       'Calyptronoma_occidentalis', 'Calyptrogyne_panamensis',\n",
              "       'Calyptronoma_rivalis', 'Calyptrogyne_tutensis',\n",
              "       'Ceroxylon_alpinum', 'Ceroxylon_amazonicum', 'Ceroxylon_ceriferum',\n",
              "       'Ceroxylon_echinulatum', 'Ceroxylon_parvifrons',\n",
              "       'Ceroxylon_parvum', 'Ceroxylon_peruvianum',\n",
              "       'Ceroxylon_pityrophyllum', 'Ceroxylon_quindiuense',\n",
              "       'Ceroxylon_sasaimae', 'Ceroxylon_ventricosum',\n",
              "       'Ceroxylon_vogelianum', 'Chamaedorea_adscendens',\n",
              "       'Chamaedorea_allenii', 'Chamaedorea_alternans',\n",
              "       'Chamaedorea_anemophila', 'Chamaedorea_arenbergiana',\n",
              "       'Chamaedorea_benziei', 'Chamaedorea_brachypoda',\n",
              "       'Chamaedorea_carchensis', 'Chamaedorea_cataractarum',\n",
              "       'Chamaedorea_correae', 'Chamaedorea_costaricana',\n",
              "       'Chamaedorea_crucensis', 'Chamaedorea_dammeriana',\n",
              "       'Chamaedorea_deckeriana', 'Chamaedorea_deneversiana',\n",
              "       'Chamaedorea_elatior', 'Chamaedorea_elegans',\n",
              "       'Chamaedorea_ernesti-augustii', 'Chamaedorea_falcifera',\n",
              "       'Chamaedorea_fragrans', 'Chamaedorea_frondosa',\n",
              "       'Chamaedorea_geonomiformis', 'Chamaedorea_glaucifolia',\n",
              "       'Chamaedorea_graminifolia', 'Chamaedorea_guntheriana',\n",
              "       'Chamaedorea_hooperiana', 'Chamaedorea_klotzschiana',\n",
              "       'Chamaedorea_linearis', 'Chamaedorea_lucidifrons',\n",
              "       'Chamaedorea_macrospadix', 'Chamaedorea_matae',\n",
              "       'Chamaedorea_metallica', 'Chamaedorea_microspadix',\n",
              "       'Chamaedorea_microphylla', 'Chamaedorea_murriensis',\n",
              "       'Chamaedorea_nationsiana', 'Chamaedorea_neurochlamys',\n",
              "       'Chamaedorea_nubium', 'Chamaedorea_oblongata',\n",
              "       'Chamaedorea_oreophila', 'Chamaedorea_palmeriana',\n",
              "       'Chamaedorea_parvifolia', 'Chamaedorea_pedunculata',\n",
              "       'Chamaedorea_pinnatifrons', 'Chamaedorea_pittieri',\n",
              "       'Chamaedorea_plumosa', 'Chamaedorea_pochutlensis',\n",
              "       'Chamaedorea_ponderosa', 'Chamaedorea_pygmaea',\n",
              "       'Chamaedorea_quezalteca', 'Chamaedorea_radicalis',\n",
              "       'Chamaedorea_recurvata', 'Chamaedorea_rhizomatosa',\n",
              "       'Chamaedorea_ricardoi', 'Chamaedorea_robertii',\n",
              "       'Chamaedorea_rojasiana', 'Chamaedorea_rossteniorum',\n",
              "       'Chamaedorea_sartorii', 'Chamaedorea_scheryi',\n",
              "       'Chamaedorea_schippii', 'Chamaedorea_schiediana',\n",
              "       'Chamaedorea_seifrizii', 'Chamaedorea_spp', 'Chamaedorea',\n",
              "       'Chamaedorea_stolonifera', 'Chamaedorea_stricta',\n",
              "       'Chamaedorea_sullivaniorum', 'Chamaedorea_tenella',\n",
              "       'Chamaedorea_tepejilote', 'Chamaedorea_tuerckheimii',\n",
              "       'Chamaedorea_warscewiczii', 'Chamaedorea_verecunda',\n",
              "       'Chamaedorea_woodsoniana', 'Chamaedorea_zamorae',\n",
              "       'Chelyocarpus_chuco', 'Chelyocarpus_dianeurus',\n",
              "       'Chelyocarpus_repens', 'Chelyocarpus_ulei',\n",
              "       'Coccothrinax_alexandria', 'Coccothrinax_alta',\n",
              "       'Coccothrinax_argentata', 'Coccothrinax_argentea',\n",
              "       'Coccothrinax_barbadensis', 'Coccothrinax_borhidiana',\n",
              "       'Coccothrinax_clarensis', 'Coccothrinax_crinita',\n",
              "       'Coccothrinax_cupularis', 'Coccothrinax_ekmanii',\n",
              "       'Coccothrinax_elegans', 'Coccothrinax_fragrans',\n",
              "       'Coccothrinax_gracilis', 'Coccothrinax_guantanamensis',\n",
              "       'Coccothrinax_gundlachii', 'Coccothrinax_hioramii',\n",
              "       'Coccothrinax_inaguensis', 'Coccothrinax_jamaicensis',\n",
              "       'Coccothrinax_litoralis', 'Coccothrinax_macroglossa',\n",
              "       'Coccothrinax_miraguama', 'Coccothrinax_montana',\n",
              "       'Coccothrinax_munizii', 'Coccothrinax_orientalis',\n",
              "       'Coccothrinax_proctorii', 'Coccothrinax_readii',\n",
              "       'Coccothrinax_salvatoris', 'Coccothrinax_saxicola',\n",
              "       'Coccothrinax_spissa', 'Cocos', 'Coccothrinax_torrida',\n",
              "       'Colpothrinax_aphanopetala', 'Colpothrinax_wrightii',\n",
              "       'Copernicia_alba', 'Copernicia_baileyana', 'Copernicia_berteroana',\n",
              "       'Copernicia_ekmanii', 'Copernicia_fallaensis', 'Copernicia_gigas',\n",
              "       'Copernicia_glabrescens', 'Copernicia_hospita',\n",
              "       'Copernicia_macroglossa', 'Copernicia_prunifera',\n",
              "       'Copernicia_rigida', 'Copernicia_tectorum', 'Copernicia_yarey',\n",
              "       'Cryosophila_bartlettii', 'Cryosophila_guagara',\n",
              "       'Cryosophila_kalbreyeri', 'Cryosophila_nana',\n",
              "       'Cryosophila_stauracantha', 'Cryosophila_warscewiczii',\n",
              "       'Cryosophila_williamsii', 'Desmoncus_mitis',\n",
              "       'Desmoncus_orthacanthos', 'Desmoncus_polyacanthos',\n",
              "       'Dictyocaryum_lamarckianum', 'Elaeis_oleifera',\n",
              "       'Euterpe_broadwayi', 'Euterpe_catinga', 'Euterpe_edulis',\n",
              "       'Euterpe_espiritosantensis', 'Euterpe_longivaginata',\n",
              "       'Euterpe_oleracea', 'Euterpe_precatoria', 'Gaussia_attenuata',\n",
              "       'Gaussia_gomez-pompae', 'Gaussia_maya', 'Gaussia_princeps',\n",
              "       'Gaussia_spirituana', 'Geonoma_baculifera', 'Geonoma_brongniartii',\n",
              "       'Geonoma_calyptrogynoidea', 'Geonoma_camana',\n",
              "       'Geonoma_chlamydostachys', 'Geonoma_concinna', 'Geonoma_congesta',\n",
              "       'Geonoma_cuneata', 'Geonoma_deversa', 'Geonoma_elegans',\n",
              "       'Geonoma_fosteri', 'Geonoma_frontinensis', 'Geonoma_galeanoae',\n",
              "       'Geonoma_hugonis', 'Geonoma_interrupta', 'Geonoma_lehmannii',\n",
              "       'Geonoma_longivaginata', 'Geonoma_macrostachys', 'Geonoma_maxima',\n",
              "       'Geonoma_monospatha', 'Geonoma_mooreana', 'Geonoma_multisecta',\n",
              "       'Geonoma_occidentalis', 'Geonoma_orbignyana', 'Geonoma_paradoxa',\n",
              "       'Geonoma_pauciflora', 'Geonoma_pohliana', 'Geonoma_procumbens',\n",
              "       'Geonoma_rubescens', 'Geonoma_santanderensis',\n",
              "       'Geonoma_schottiana', 'Geonoma_stricta', 'Geonoma_triandra',\n",
              "       'Geonoma_undata', 'Hemithrinax_ekmaniana', 'Hyospathe_elegans',\n",
              "       'Hyospathe_pittieri', 'Iriartea_deltoidea', 'Itaya_amicorum',\n",
              "       'Jubaea_chilensis', 'Leopoldinia_pulchra', 'Lepidocaryum_tenue',\n",
              "       'Leucothrinax_morrisii', 'Manicaria_saccifera',\n",
              "       'Mauritiella_aculeata', 'Mauritiella_armata', 'Mauritia_Carana',\n",
              "       'Mauritiella_disticha', 'Mauritia_flexuosa',\n",
              "       'Mauritiella_macroclada', 'Mauritiella_pumila', nan,\n",
              "       'Neonicholsonia_watsonii', 'Oenocarpus_balickii',\n",
              "       'Oenocarpus_bataua', 'Oenocarpus_distichus', 'Oenocarpus_mapora',\n",
              "       'Oenocarpus_minor', 'Pholidostachys_dactyloides',\n",
              "       'Pholidostachys_kalbreyeri', 'Pholidostachys_panamensis',\n",
              "       'Pholidostachys_pulchra', 'Pholidostachys_sanluisensis',\n",
              "       'Phytelephas_aecuatorialis', 'Phytelephas_aequatorialis',\n",
              "       'Phytelephas_macrocarpa', 'Phytelephas', 'Phytelephas_tenuicaulis',\n",
              "       'Phytelephas_tumacana', 'Prestoea_acuminata', 'Prestoea_decurrens',\n",
              "       'Prestoea_ensiformis', 'Prestoea_longepetiolata',\n",
              "       'Prestoea_pubens', 'Prestoea_schultzeana', 'Pseudophoenix_ekmanii',\n",
              "       'Pseudophoenix_lediniana', 'Pseudophoenix_sargentii',\n",
              "       'Pseudophoenix_vinifera', 'Raphia_taedigera',\n",
              "       'Ravenea_sambiranensis', 'Reinhardtia_gracilis',\n",
              "       'Reinhardtia_koschnyana', 'Reinhardtia_simplex',\n",
              "       'Rhapidophyllum_hystrix', 'Roystonea_altissima',\n",
              "       'Roystonea_borinquena', 'Roystonea_lenis', 'Roystonea_maisiana',\n",
              "       'Roystonea_oleracea', 'Roystonea_regia', 'Sabal_bermudana',\n",
              "       'Sabal_causiarum', 'Sabal_dominguensis', 'Sabal_etonia',\n",
              "       'Sabal_guatemalensis', 'Sabinaria_magnifica', 'Sabal_maritima',\n",
              "       'Sabal_mauritiiformis', 'Sabal_mexicana', 'Sabal_minor',\n",
              "       'Sabal_palmetto', 'Sabal_pumos', 'Sabal_rosei', 'Sabal_uresana',\n",
              "       'Sabal_yapa', 'Schippia_concolor', 'Serenoa_repens',\n",
              "       'Socratea_exhorriza', 'Socratea_exorrhiza',\n",
              "       'Syagrus_allagopteroides', 'Syagrus_amara', 'Syagrus_angustifolia',\n",
              "       'Syagrus_botryophora', 'Syagrus_caerulescens',\n",
              "       'Syagrus_campylospatha', 'Syagrus_campestris',\n",
              "       'Syagrus_cardenasii', 'Syagrus_cerqueirana', 'Syagrus_cocoides',\n",
              "       'Syagrus_comosa', 'Syagrus_coronata', 'Syagrus_deflexa',\n",
              "       'Syagrus_duartei', 'Syagrus_evansiana', 'Syagrus_flexuosa',\n",
              "       'Syagrus_glaucescens', 'Syagrus_glazioviana', 'Syagrus_gouveiana',\n",
              "       'Syagrus_graminifolia', 'Syagrus_guimaraensis', 'Syagrus_harleyi',\n",
              "       'Syagrus_hoehnei', 'Syagrus_inajai', 'Syagrus_insignis',\n",
              "       'Syagrus_itapebiensis', 'Syagrus_itacambirana', 'Syagrus_kellyana',\n",
              "       'Syagrus_liliputiana', 'Syagrus_loefgrenii',\n",
              "       'Syagrus_longipedunculata', 'Syagrus_lorenzoniorum',\n",
              "       'Syagrus_macrocarpa', 'Syagrus_menzeliana', 'Syagrus_mendanhensis',\n",
              "       'Syagrus_microphylla', 'Syagrus_minor', 'Syagrus_oleracea',\n",
              "       'Syagrus_orinocensis', 'Syagrus_petraea', 'Syagrus_picrophylla',\n",
              "       'Syagrus_pleiocladoides', 'Syagrus_procumbens',\n",
              "       'Syagrus_pseudococos', 'Syagrus_romanzoffiana', 'Syagrus_rupicola',\n",
              "       'Syagrus_ruschiana', 'Syagrus_santosii', 'Syagrus_sancona',\n",
              "       'Syagrus_schizophylla', 'Syagrus_smithii', 'Syagrus_smith',\n",
              "       'Syagrus_stenopetala', 'Syagrus_stratincola', 'Syagrus_vagans',\n",
              "       'Syagrus_weddelianum', 'Syagrus_werdermanii',\n",
              "       'Syagrus_vermicularis', 'Syagrus_yungascensis',\n",
              "       'Synechanthus_fibrosus', 'Synechanthus_warscewiczianus',\n",
              "       'Thrinax_parviflora', 'Thrinax_radiata', 'Trithrinax_brasiliensis',\n",
              "       'Trithrinax_campestris', 'Trithrinax_schizophylla',\n",
              "       'Washingtonia_filifera', 'Welfia_regia', 'Wendlandiella_gracilis',\n",
              "       'Wettinia_augusta', 'Wettinia_aequalis', 'Zombia_antillarum',\n",
              "       'Chamaedorea_binderi', 'Chamaedorea_brachyclada',\n",
              "       'Chamaedorea_fractiflexa', 'Chamaedorea_hodelii',\n",
              "       'Chamaedorea_ibarrae', 'Chamaedorea_incrustata',\n",
              "       'Chamaedorea_keelerorum', 'Chamaedorea_lehmannii',\n",
              "       'Chamaedorea_liebmannii', 'Chamaedorea_pachenoana',\n",
              "       'Chamaedorea_parvisecta', 'Chamaedorea_pauciflora',\n",
              "       'Chamaedorea_piscifolia', 'Chamaedorea_pumila',\n",
              "       'Chamaedorea_rigida', 'Chamaedorea_schideana',\n",
              "       'Chamaedorea_simplex', 'Chamaedorea_skutchii',\n",
              "       'Chamaedorea_stenocarpa', 'Chamaedorea_subjectifolia',\n",
              "       'Chamaedorea_tenerrima', 'Chamaedorea_verapazensis',\n",
              "       'Chamaedorea_volcanensis', 'Chamaedorea_vulgata'], dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 24
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rG4SjP7_-dzt",
        "outputId": "f880a3fc-68bb-4abf-ab57-d9e40b316e51"
      },
      "source": [
        "print('Number of american palm species sequenced:')\n",
        "len(df[df['Continent'] == 'americas']['taxspecies'].unique())"
      ],
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Number of american palm species sequenced:\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "531"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 25
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nJYca_hk_WmU"
      },
      "source": [
        "We need to know how many species are American species. I initially used this list passed on to me (but it is something to check out more formaly). The Kew checklist, for example, doesn't include some Sabal species for which we have sequences. There has to be a consensus about which list to compare against."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KxW-Tm5Q_kEW",
        "outputId": "96b66dde-89ca-4ab7-918b-f15f14d0306a"
      },
      "source": [
        "palms_americas=['Schippia_concolor','Pseudophoenix_ekmanii','Pseudophoenix_lediniana','Pseudophoenix_sargentii','Pseudophoenix_vinifera','Acoelorraphe_wrightii','Pholidostachys_amazonensis','Pholidostachys_dactyloides','Pholidostachys_kalbreyeri','Pholidostachys_occidentalis','Pholidostachys_panamensis','Pholidostachys_pulchra','Pholidostachys_sanluisensis','Pholidostachys_synanthera','Lepidocaryum_tenue','Hyospathe_elegans','Hyospathe_frontinensis','Hyospathe_macrorhachis','Hyospathe_peruviana','Hyospathe_pittieri','Hyospathe_wendlandiana','Chelyocarpus_chuco','Chelyocarpus_dianeurus','Chelyocarpus_repens','Chelyocarpus_ulei','Acrocomia_aculeata','Acrocomia_crispa','Acrocomia_emensis','Acrocomia_glaucescens','Acrocomia_hassleri','Acrocomia_intumescens','Acrocomia_media','Acrocomia_totai','Aiphanes_acanthophylla','Aiphanes_acaulis','Aiphanes_bicornis','Aiphanes_buenaventurae','Aiphanes_chiribogensis','Aiphanes_concinna','Aiphanes_deltoidea','Aiphanes_duquei','Aiphanes_eggersii','Aiphanes_erinacea','Aiphanes_gelatinosa','Aiphanes_graminifolia','Aiphanes_grandis','Aiphanes_hirsuta','Aiphanes_horrida','Aiphanes_killipii','Aiphanes_leiostachys','Aiphanes_lindeniana','Aiphanes_linearis','Aiphanes_macroloba','Aiphanes_minima','Aiphanes_multiplex','Aiphanes_parvifolia','Aiphanes_pilaris','Aiphanes_simplex','Aiphanes_spicata','Aiphanes_stergiosii','Aiphanes_tricuspidata','Aiphanes_ulei','Aiphanes_verrucosa','Aiphanes_weberbaueri','Allagoptera_arenaria','Allagoptera_brevicalyx','Allagoptera_campestris','Allagoptera_caudescens','Allagoptera_leucocalyx','Ammandra_decasperma','Aphandra_natalia','Asterogyne_guianensis','Asterogyne_martiana','Asterogyne_ramosa','Asterogyne_spicata','Asterogyne_yaracuyense','Astrocaryum_acaule','Astrocaryum_aculeatissimum','Astrocaryum_aculeatum','Astrocaryum_alatum','Astrocaryum_campestre','Astrocaryum_carnosum','Astrocaryum_chambira','Astrocaryum_chonta','Astrocaryum_ciliatum','Astrocaryum_confertum','Astrocaryum_cuatrecasanum','Astrocaryum_echinatum','Astrocaryum_faranae','Astrocaryum_farinosum','Astrocaryum_ferrugineum','Astrocaryum_giganteum','Astrocaryum_gratum','Astrocaryum_gynacanthum','Astrocaryum_huaimi','Astrocaryum_huicungo','Astrocaryum_jauari','Astrocaryum_javarense','Astrocaryum_macrocalyx','Astrocaryum_malybo','Astrocaryum_mexicanum','Astrocaryum_minus','Astrocaryum_murumuru','Astrocaryum_paramaca','Astrocaryum_perangustatum','Astrocaryum_rodriguesii','Astrocaryum_sciophilum','Astrocaryum_scopatum','Astrocaryum_sociale','Astrocaryum_standleyanum','Astrocaryum_triandrum','Astrocaryum_tucuma','Astrocaryum_ulei','Astrocaryum_urostachys','Astrocaryum_vulgare','Attalea_allenii','Attalea_amygdalina','Attalea_amylacea','Attalea_anisitsiana','Attalea_apoda','Attalea_attaleoides','Attalea_barreirensis','Attalea_bassleriana','Attalea_blepharopus','Attalea_brasiliensis','Attalea_brejinhoensis','Attalea_burretiana','Attalea_butyracea','Attalea_camopiensis','Attalea_cephalotus','Attalea_cohune','Attalea_colenda','Attalea_compta','Attalea_crassispatha','Attalea_cuatrecasana','Attalea_dahlgreniana','Attalea_degranvillei','Attalea_dubia','Attalea_eichleri','Attalea_exigua','Attalea_fairchildensis','Attalea_funifera','Attalea_geraensis','Attalea_guacuyule','Attalea_guianensis','Attalea_hoehnei','Attalea_huebneri','Attalea_humilis','Attalea_iguadummat','Attalea_insignis','Attalea_kewensis','Attalea_lauromuelleriana','Attalea_leandroana','Attalea_luetzelburgii','Attalea_macrolepis','Attalea_magdalenica','Attalea_maracaibensis','Attalea_maripa','Attalea_maripensis','Attalea_microcarpa','Attalea_moorei','Attalea_nucifera','Attalea_oleifera','Attalea_osmantha','Attalea_peruviana','Attalea_phalerata','Attalea_pindobassu','Attalea_plowmanii','Attalea_princeps','Attalea_racemosa','Attalea_rhynchocarpa','Attalea_rostrata','Attalea_salazarii','Attalea_salvadorensis','Attalea_seabrensis','Attalea_septuagenata','Attalea_speciosa','Attalea_spectabilis','Attalea_tessmannii','Attalea_vitrivir','Attalea_weberbaueri','Attalea_wesselsboeri','Bactris_acanthocarpa','Bactris_acanthocarpoides','Bactris_ana-juliae','Bactris_aubletiana','Bactris_bahiensis','Bactris_balanophora','Bactris_barronis','Bactris_bidentula','Bactris_bifida','Bactris_brongniartii','Bactris_campestris','Bactris_caryotifolia','Bactris_caudata','Bactris_charnleyae','Bactris_chaveziae','Bactris_coloniata','Bactris_coloradonis','Bactris_concinna','Bactris_constanciae','Bactris_corossilla','Bactris_cubensis','Bactris_cuspidata','Bactris_dianeura','Bactris_elegans','Bactris_faucium','Bactris_ferruginea','Bactris_fissifrons','Bactris_gasipaes','Bactris_gastoniana','Bactris_glandulosa','Bactris_glassmanii','Bactris_glaucescens','Bactris_gracilior','Bactris_grayumii','Bactris_guineensis','Bactris_halmoorei','Bactris_hatschbachii','Bactris_herrerana','Bactris_hirta','Bactris_hondurensis','Bactris_horridispatha','Bactris_jamaicana','Bactris_killipii','Bactris_kunorum','Bactris_longiseta','Bactris_macroacantha','Bactris_major','Bactris_maraja','Bactris_martiana','Bactris_mexicana','Bactris_militaris','Bactris_nancibaensis','Bactris_obovata','Bactris_oligocarpa','Bactris_oligoclada','Bactris_panamensis','Bactris_pickelii','Bactris_pilosa','Bactris_pliniana','Bactris_plumeriana','Bactris_polystachya','Bactris_ptariana','Bactris_rhaphidacantha','Bactris_riparia','Bactris_rostrata','Bactris_schultesii','Bactris_setiflora','Bactris_setosa','Bactris_setulosa','Bactris_simplicifrons','Bactris_soeiroana','Bactris_sphaerocarpa','Bactris_syagroides','Bactris_tefensis','Bactris_timbuiensis','Bactris_tomentosa','Bactris_turbinocarpa','Bactris_vulgaris','Brahea_aculeata','Brahea_armata','Brahea_brandegeei','Brahea_calcarea','Brahea_decumbens','Brahea_dulcis','Brahea_edulis','Brahea_moorei','Brahea_pimo','Brahea_salvadorensis','Brahea_sarukhanii','Butia_archeri','Butia_campicola','Butia_capitata','Butia_catarinensis','Butia_eriospatha','Butia_exilata','Butia_exospadix','Butia_lallemantii','Butia_leptospatha','Butia_marmorii','Butia_matogrossensis','Butia_microspadix','Butia_missionera','Butia_noblickii','Butia_odorata','Butia_paraguayensis','Butia_purpurascens','Butia_quaraimana','Butia_stolonifera','Butia_yatay','Calyptrogyne_allenii','Calyptrogyne_anomala','Calyptrogyne_baudensis','Calyptrogyne_coloradensis','Calyptrogyne_condensata','Calyptrogyne_costatifrons','Calyptrogyne_deneversii','Calyptrogyne_fortunensis','Calyptrogyne_ghiesbreghtiana','Calyptrogyne_herrerae','Calyptrogyne_kunorum','Calyptrogyne_osensis','Calyptrogyne_panamensis','Calyptrogyne_pubescens','Calyptrogyne_sanblasensis','Calyptrogyne_trichostachys','Calyptrogyne_tutensis','Calyptronoma_occidentalis','Calyptronoma_plumeriana','Calyptronoma_rivalis','Ceroxylon_alpinum','Ceroxylon_amazonicum','Ceroxylon_ceriferum','Ceroxylon_echinulatum','Ceroxylon_parvifrons','Ceroxylon_parvum','Ceroxylon_peruvianum','Ceroxylon_pityrophyllum','Ceroxylon_quindiuense','Ceroxylon_sasaimae','Ceroxylon_ventricosum','Ceroxylon_vogelianum','Chamaedorea_adscendens','Chamaedorea_allenii','Chamaedorea_alternans','Chamaedorea_amabilis','Chamaedorea_anemophila','Chamaedorea_angustisecta','Chamaedorea_arenbergiana','Chamaedorea_atrovirens','Chamaedorea_benziei','Chamaedorea_binderi','Chamaedorea_brachyclada','Chamaedorea_brachypoda','Chamaedorea_carchensis','Chamaedorea_castillo-montii','Chamaedorea_cataractarum','Chamaedorea_christinae','Chamaedorea_correae','Chamaedorea_costaricana','Chamaedorea_crucensis','Chamaedorea_dammeriana','Chamaedorea_deckeriana','Chamaedorea_deneversiana','Chamaedorea_elatior','Chamaedorea_elegans','Chamaedorea_ernesti-augusti','Chamaedorea_falcifera','Chamaedorea_foveata','Chamaedorea_fractiflexa','Chamaedorea_fragrans','Chamaedorea_frondosa','Chamaedorea_geonomiformis','Chamaedorea_glaucifolia','Chamaedorea_graminifolia','Chamaedorea_guntheriana','Chamaedorea_hodelii','Chamaedorea_hooperiana','Chamaedorea_ibarrae','Chamaedorea_incrustata','Chamaedorea_keelerorum','Chamaedorea_klotzschiana','Chamaedorea_latisecta','Chamaedorea_lehmannii','Chamaedorea_liebmannii','Chamaedorea_linearis','Chamaedorea_lucidifrons','Chamaedorea_macrospadix','Chamaedorea_matae','Chamaedorea_metallica','Chamaedorea_microphylla','Chamaedorea_microspadix','Chamaedorea_moliniana','Chamaedorea_murriensis','Chamaedorea_nationsiana','Chamaedorea_neurochlamys','Chamaedorea_nubium','Chamaedorea_oblongata','Chamaedorea_oreophila','Chamaedorea_pachecoana','Chamaedorea_palmeriana','Chamaedorea_parvifolia','Chamaedorea_parvisecta','Chamaedorea_pauciflora','Chamaedorea_pedunculata','Chamaedorea_pinnatifrons','Chamaedorea_piscifolia','Chamaedorea_pittieri','Chamaedorea_plumosa','Chamaedorea_pochutlensis','Chamaedorea_ponderosa','Chamaedorea_pumila','Chamaedorea_pygmaea','Chamaedorea_queroana','Chamaedorea_radicalis','Chamaedorea_recurvata','Chamaedorea_rhizomatosa','Chamaedorea_ricardoi','Chamaedorea_rigida','Chamaedorea_robertii','Chamaedorea_rojasiana','Chamaedorea_rosibeliae','Chamaedorea_rossteniorum','Chamaedorea_sartorii','Chamaedorea_scheryi','Chamaedorea_schiedeana','Chamaedorea_schippii','Chamaedorea_seifrizii','Chamaedorea_serpens','Chamaedorea_simplex','Chamaedorea_skutchii','Chamaedorea_smithii','Chamaedorea_stenocarpa','Chamaedorea_stolonifera','Chamaedorea_stricta','Chamaedorea_subjectifolia','Chamaedorea_tenerrima','Chamaedorea_tepejilote','Chamaedorea_tuerckheimii','Chamaedorea_undulatifolia','Chamaedorea_verapazensis','Chamaedorea_verecunda','Chamaedorea_volcanensis','Chamaedorea_vulgata','Chamaedorea_warscewiczii','Chamaedorea_whitelockiana','Chamaedorea_woodsoniana','Chamaedorea_zamorae','Coccothrinax_acunana','Coccothrinax_alexandri','Coccothrinax_argentata','Coccothrinax_argentea','Coccothrinax_baracoensis','Coccothrinax_barbadensis','Coccothrinax_bermudezii','Coccothrinax_borhidiana','Coccothrinax_boschiana','Coccothrinax_camagueyana','Coccothrinax_clarensis','Coccothrinax_concolor','Coccothrinax_crinita','Coccothrinax_cupularis','Coccothrinax_ekmanii','Coccothrinax_elegans','Coccothrinax_fagildei','Coccothrinax_fragrans','Coccothrinax_garciana','Coccothrinax_gracilis','Coccothrinax_guantanamensis','Coccothrinax_gundlachii','Coccothrinax_hioramii','Coccothrinax_inaguensis','Coccothrinax_jamaicensis','Coccothrinax_leonis','Coccothrinax_litoralis','Coccothrinax_macroglossa','Coccothrinax_microphylla','Coccothrinax_miraguama','Coccothrinax_moaensis','Coccothrinax_montana','Coccothrinax_munizii','Coccothrinax_muricata','Coccothrinax_nipensis','Coccothrinax_orientalis','Coccothrinax_pauciramosa','Coccothrinax_proctorii','Coccothrinax_pseudorigida','Coccothrinax_pumila','Coccothrinax_readii','Coccothrinax_rigida','Coccothrinax_salvatoris','Coccothrinax_saxicola','Coccothrinax_scoparia','Coccothrinax_spissa','Coccothrinax_torrida','Coccothrinax_trinitensis','Coccothrinax_victorini','Coccothrinax_yunquensis','Coccothrinax_yuraguana','Colpothrinax_aphanopetala','Colpothrinax_cookii','Colpothrinax_wrightii','Copernicia_alba','Copernicia_baileyana','Copernicia_berteroana','Copernicia_brittonorum','Copernicia_cowellii','Copernicia_curbeloi','Copernicia_curtissii','Copernicia_ekmanii','Copernicia_fallaensis','Copernicia_gigas','Copernicia_glabrescens','Copernicia_hospita','Copernicia_humicola','Copernicia_longiglossa','Copernicia_macroglossa','Copernicia_molineti','Copernicia_oxycalyx','Copernicia_prunifera','Copernicia_rigida','Copernicia_roigii','Copernicia_tectorum','Copernicia_yarey','Cryosophila_bartlettii','Cryosophila_cookii','Cryosophila_grayumii','Cryosophila_guagara','Cryosophila_kalbreyeri','Cryosophila_macrocarpa','Cryosophila_nana','Cryosophila_stauracantha','Cryosophila_warscewiczii','Cryosophila_williamsii','Desmoncus_chinantlensis','Desmoncus_cirrhifer','Desmoncus_costaricensis','Desmoncus_giganteus','Desmoncus_horridus','Desmoncus_interjectus','Desmoncus_kunarius','Desmoncus_latisectus','Desmoncus_leptoclonos','Desmoncus_loretanus','Desmoncus_madrensis','Desmoncus_mitis','Desmoncus_moorei','Desmoncus_myriacanthos','Desmoncus_obovoideus','Desmoncus_orthacanthos','Desmoncus_osensis','Desmoncus_parvulus','Desmoncus_polyacanthos','Desmoncus_prunifer','Desmoncus_pumilus','Desmoncus_setosus','Desmoncus_stans','Desmoncus_vacivus','Dictyocaryum_fuscum','Dictyocaryum_lamarckianum','Dictyocaryum_ptarianum','Elaeis_oleifera','Euterpe_broadwayi','Euterpe_catinga','Euterpe_edulis','Euterpe_longibracteata','Euterpe_luminosa','Euterpe_oleracea','Euterpe_precatoria','Gaussia_attenuata','Gaussia_gomez-pompae','Gaussia_maya','Gaussia_princeps','Gaussia_spirituana','Geonoma_aspidiifolia','Geonoma_baculifera','Geonoma_bernalii','Geonoma_braunii','Geonoma_brenesii','Geonoma_brongniartii','Geonoma_calyptrogynoidea','Geonoma_camana','Geonoma_chlamydostachys','Geonoma_chococola','Geonoma_concinna','Geonoma_concinnoidea','Geonoma_congesta','Geonoma_cuneata','Geonoma_deneversii','Geonoma_deversa','Geonoma_dindoensis','Geonoma_divisa','Geonoma_elegans','Geonoma_epetiolata','Geonoma_euspatha','Geonoma_ferruginea','Geonoma_fosteri','Geonoma_frontinensis','Geonoma_galeanoae','Geonoma_gentryi','Geonoma_hollinensis','Geonoma_hugonis','Geonoma_interrupta','Geonoma_lanata','Geonoma_laxiflora','Geonoma_lehmannii','Geonoma_leptospadix','Geonoma_longipedunculata','Geonoma_longivaginata','Geonoma_macrostachys','Geonoma_maxima','Geonoma_monospatha','Geonoma_mooreana','Geonoma_multisecta','Geonoma_occidentalis','Geonoma_oldemanii','Geonoma_oligoclona','Geonoma_operculata','Geonoma_orbignyana','Geonoma_paradoxa','Geonoma_pauciflora','Geonoma_peruviana','Geonoma_pinnatifrons','Geonoma_poeppigiana','Geonoma_pohliana','Geonoma_poiteauana','Geonoma_sanmartinensis','Geonoma_santanderensis','Geonoma_schizocarpa','Geonoma_schottiana','Geonoma_scoparia','Geonoma_simplicifrons','Geonoma_spinescens','Geonoma_stricta','Geonoma_talamancana','Geonoma_tenuissima','Geonoma_triandra','Geonoma_triglochin','Geonoma_trigona','Geonoma_umbraculiformis','Geonoma_undata','Geonoma_venosa','Guihaia_argyrata','Guihaia_grossifibrosa','Guihaia_lancifolia','Hemithrinax_compacta','Hemithrinax_ekmaniana','Hemithrinax_rivularis','Iriartea_deltoidea','Iriartella_setigera','Iriartella_stenocarpa','Itaya_amicorum','Juania_australis','Jubaea_chilensis','Leopoldinia_piassaba','Leopoldinia_pulchra','Leucothrinax_morrisii','Lytocaryum_hoehnei','Lytocaryum_insigne','Lytocaryum_itapebiense','Lytocaryum_weddellianum','Manicaria_martiana','Manicaria_saccifera','Mauritia_carana','Mauritia_flexuosa','Mauritiella_aculeata','Mauritiella_armata','Mauritiella_macroclada','Mauritiella_pumila','Neonicholsonia_watsonii','Oenocarpus_bacaba','Oenocarpus_balickii','Oenocarpus_bataua','Oenocarpus_circumtextus','Oenocarpus_distichus','Oenocarpus_makeru','Oenocarpus_mapora','Oenocarpus_minor','Oenocarpus_simplex','Phytelephas_macrocarpa','Phytelephas_schottii','Phytelephas_seemannii','Phytelephas_tenuicaulis','Phytelephas_tumacana','Prestoea_acuminata','Prestoea_carderi','Prestoea_decurrens','Prestoea_ensiformis','Prestoea_longepetiolata','Prestoea_pubens','Prestoea_pubigera','Prestoea_schultzeana','Prestoea_simplicifolia','Prestoea_tenuiramosa','Raphia_taedigera','Reinhardtia_elegans','Reinhardtia_gracilis','Reinhardtia_koschnyana','Reinhardtia_latisecta','Reinhardtia_paiewonskiana','Reinhardtia_simplex','Rhapidophyllum_hystrix','Roystonea_altissima','Roystonea_borinquena','Roystonea_dunlapiana','Roystonea_lenis','Roystonea_maisiana','Roystonea_oleracea','Roystonea_princeps','Roystonea_regia','Roystonea_stellata','Roystonea_violacea','Sabal_bermudana','Sabal_causiarum','Sabal_domingensis','Sabal_etonia','Sabal_gretherae','Sabal_maritima','Sabal_mauritiiformis','Sabal_mexicana','Sabal_minor','Sabal_palmetto','Sabal_pumos','Sabal_rosei','Sabal_uresana','Sabal_yapa','Serenoa_repens','Socratea_exorrhiza','Socratea_hecatonandra','Socratea_montana','Socratea_rostrata','Socratea_salazarii','Syagrus_allagopteroides','Syagrus_amara','Syagrus_angustifolia','Syagrus_botryophora','Syagrus_caerulescens','Syagrus_campestris','Syagrus_campylospatha','Syagrus_cardenasii','Syagrus_cataphracta','Syagrus_cearensis','Syagrus_cerqueirana','Syagrus_cocoides','Syagrus_comosa','Syagrus_coronata','Syagrus_deflexa','Syagrus_duartei','Syagrus_emasensis','Syagrus_evansiana','Syagrus_flexuosa','Syagrus_glaucescens','Syagrus_glazioviana','Syagrus_gouveiana','Syagrus_graminifolia','Syagrus_harleyi','Syagrus_hoehnei','Syagrus_inajai','Syagrus_insignis','Syagrus_itacambirana','Syagrus_itapebiensis','Syagrus_kellyana','Syagrus_lilliputiana','Syagrus_loefgrenii','Syagrus_longipedunculata','Syagrus_lorenzoniorum','Syagrus_macrocarpa','Syagrus_mendanhensis','Syagrus_menzeliana','Syagrus_microphylla','Syagrus_minor','Syagrus_oleracea','Syagrus_orinocensis','Syagrus_petraea','Syagrus_picrophylla','Syagrus_pimentae','Syagrus_pleioclada','Syagrus_pleiocladoides','Syagrus_pompeoi','Syagrus_procumbens','Syagrus_pseudococos','Syagrus_romanzoffiana','Syagrus_rupicola','Syagrus_ruschiana','Syagrus_sancona','Syagrus_santosii','Syagrus_schizophylla','Syagrus_smithii','Syagrus_stenopetala','Syagrus_stratincola','Syagrus_vagans','Syagrus_vermicularis','Syagrus_werdermannii','Syagrus_yungasensis','Synechanthus_fibrosus','Synechanthus_warscewiczianus','Thrinax_excelsa','Thrinax_parviflora','Thrinax_radiata','Trithrinax_acanthocoma','Trithrinax_brasiliensis','Trithrinax_campestris','Trithrinax_schizophylla','Washingtonia_filifera','Washingtonia_robusta','Welfia_regia','Wendlandiella_gracilis','Wettinia_aequalis','Wettinia_aequatorialis','Wettinia_anomala','Wettinia_augusta','Wettinia_castanea','Wettinia_disticha','Wettinia_drudei','Wettinia_fascicularis','Wettinia_hirsuta','Wettinia_kalbreyeri','Wettinia_lanata','Wettinia_longipetala','Wettinia_maynensis','Wettinia_microcarpa','Wettinia_minima','Wettinia_oxycarpa','Wettinia_panamensis','Wettinia_praemorsa','Wettinia_quinaria','Wettinia_radiata','Wettinia_verruculosa','Zombia_antillarum']\n",
        "\n",
        "print('Number of american palms:')\n",
        "len(palms_americas)"
      ],
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Number of american palms:\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "810"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 26
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yTdA2VFAA6PK"
      },
      "source": [
        "So, what have we sequenced?"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "o-3Ff_prCUOo",
        "outputId": "ff4ae5d4-c518-4ffc-ebaa-1e82ceb10746"
      },
      "source": [
        "# first, we create sets of items to compare, using list comprehention. Sets should be defined within curly brackets\n",
        "sppseqed = {spp for spp in df[df['Continent'] == 'americas']['taxspecies'].unique()} #list comprehension are loops of a single line\n",
        "apspp = {spp for spp in palms_americas} # american palm species\n",
        "# we can apply set theory to find the differences (not sequenced or sequenced but not in the list of american palms)\n",
        "# items in setA that are not shared with setB and viceversa\n",
        "# symmetric difference \n",
        "print(apspp - sppseqed) # niceee, nothing is missing\n",
        "sppmissing = len(apspp - sppseqed)\n",
        "print('\\nAlleged missing species: %s'%(sppmissing))\n",
        "print('Percentage of species missing: %s'%((sppmissing * 100)/len(palms_americas)))\n",
        "print('Percentage of species sequenced: %s'%(100-(sppmissing * 100)/len(palms_americas)))"
      ],
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "{'Chamaedorea_undulatifolia', 'Desmoncus_osensis', 'Attalea_macrolepis', 'Bactris_balanophora', 'Brahea_edulis', 'Copernicia_humicola', 'Chamaedorea_smithii', 'Bactris_grayumii', 'Calyptrogyne_kunorum', 'Oenocarpus_simplex', 'Colpothrinax_cookii', 'Chamaedorea_latisecta', 'Desmoncus_myriacanthos', 'Reinhardtia_paiewonskiana', 'Bactris_nancibaensis', 'Brahea_calcarea', 'Oenocarpus_circumtextus', 'Attalea_wesselsboeri', 'Copernicia_molineti', 'Coccothrinax_scoparia', 'Geonoma_chococola', 'Roystonea_violacea', 'Geonoma_trigona', 'Oenocarpus_makeru', 'Coccothrinax_bermudezii', 'Socratea_rostrata', 'Calyptrogyne_trichostachys', 'Aiphanes_bicornis', 'Dictyocaryum_fuscum', 'Acoelorraphe_wrightii', 'Geonoma_longipedunculata', 'Wettinia_longipetala', 'Attalea_fairchildensis', 'Dictyocaryum_ptarianum', 'Attalea_lauromuelleriana', 'Wettinia_oxycarpa', 'Bactris_cubensis', 'Bactris_sphaerocarpa', 'Wettinia_anomala', 'Geonoma_scoparia', 'Chamaedorea_queroana', 'Socratea_hecatonandra', 'Astrocaryum_paramaca', 'Bactris_bifida', 'Aiphanes_deltoidea', 'Desmoncus_obovoideus', 'Attalea_septuagenata', 'Wettinia_minima', 'Astrocaryum_farinosum', 'Hemithrinax_compacta', 'Geonoma_leptospadix', 'Coccothrinax_boschiana', 'Desmoncus_pumilus', 'Astrocaryum_cuatrecasanum', 'Geonoma_dindoensis', 'Geonoma_operculata', 'Bactris_ferruginea', 'Geonoma_venosa', 'Bactris_hondurensis', 'Coccothrinax_concolor', 'Coccothrinax_pseudorigida', 'Manicaria_martiana', 'Bactris_glassmanii', 'Astrocaryum_vulgare', 'Chamaedorea_whitelockiana', 'Astrocaryum_huaimi', 'Bactris_setiflora', 'Brahea_brandegeei', 'Attalea_cephalotus', 'Wettinia_lanata', 'Desmoncus_loretanus', 'Calyptrogyne_herrerae', 'Desmoncus_setosus', 'Bactris_cuspidata', 'Coccothrinax_yunquensis', 'Geonoma_umbraculiformis', 'Euterpe_longibracteata', 'Geonoma_brenesii', 'Attalea_salazarii', 'Euterpe_luminosa', 'Coccothrinax_microphylla', 'Iriartella_setigera', 'Desmoncus_kunarius', 'Iriartella_stenocarpa', 'Lytocaryum_hoehnei', 'Coccothrinax_muricata', 'Prestoea_pubigera', 'Aiphanes_acanthophylla', 'Prestoea_carderi', 'Attalea_dahlgreniana', 'Butia_missionera', 'Calyptrogyne_deneversii', 'Attalea_huebneri', 'Bactris_constanciae', 'Syagrus_pimentae', 'Coccothrinax_garciana', 'Attalea_racemosa', 'Roystonea_stellata', 'Chamaedorea_rosibeliae', 'Calyptrogyne_condensata', 'Asterogyne_yaracuyense', 'Geonoma_deneversii', 'Attalea_camopiensis', 'Attalea_blepharopus', 'Butia_catarinensis', 'Geonoma_triglochin', 'Wettinia_verruculosa', 'Aiphanes_grandis', 'Chamaedorea_atrovirens', 'Astrocaryum_sociale', 'Hyospathe_frontinensis', 'Syagrus_cataphracta', 'Wettinia_fascicularis', 'Aiphanes_eggersii', 'Copernicia_longiglossa', 'Bactris_jamaicana', 'Reinhardtia_elegans', 'Bactris_martiana', 'Attalea_kewensis', 'Bactris_dianeura', 'Coccothrinax_pumila', 'Syagrus_pompeoi', 'Syagrus_emasensis', 'Bactris_schultesii', 'Bactris_soeiroana', 'Geonoma_bernalii', 'Geonoma_lanata', 'Bactris_polystachya', 'Socratea_montana', 'Guihaia_lancifolia', 'Asterogyne_spicata', 'Desmoncus_madrensis', 'Attalea_rostrata', 'Brahea_decumbens', 'Coccothrinax_nipensis', 'Coccothrinax_alexandri', 'Bactris_syagroides', 'Astrocaryum_rodriguesii', 'Acrocomia_hassleri', 'Coccothrinax_fagildei', 'Attalea_amylacea', 'Geonoma_braunii', 'Mauritia_carana', 'Attalea_rhynchocarpa', 'Bactris_acanthocarpa', 'Calyptrogyne_coloradensis', 'Copernicia_brittonorum', 'Bactris_longiseta', 'Calyptrogyne_baudensis', 'Wettinia_castanea', 'Wettinia_aequatorialis', 'Pholidostachys_occidentalis', 'Copernicia_roigii', 'Prestoea_tenuiramosa', 'Bactris_horridispatha', 'Geonoma_concinnoidea', 'Juania_australis', 'Chamaedorea_serpens', 'Bactris_oligoclada', 'Desmoncus_parvulus', 'Wettinia_panamensis', 'Copernicia_cowellii', 'Attalea_exigua', 'Asterogyne_ramosa', 'Geonoma_euspatha', 'Geonoma_pinnatifrons', 'Desmoncus_cirrhifer', 'Pholidostachys_amazonensis', 'Guihaia_grossifibrosa', 'Calyptronoma_plumeriana', 'Chamaedorea_christinae', 'Bactris_herrerana', 'Geonoma_simplicifrons', 'Hyospathe_wendlandiana', 'Prestoea_simplicifolia', 'Desmoncus_prunifer', 'Astrocaryum_minus', 'Wettinia_hirsuta', 'Wettinia_maynensis', 'Bactris_tefensis', 'Wettinia_drudei', 'Geonoma_schizocarpa', 'Lytocaryum_weddellianum', 'Bactris_vulgaris', 'Attalea_maripensis', 'Geonoma_ferruginea', 'Astrocaryum_tucuma', 'Bactris_gastoniana', 'Coccothrinax_rigida', 'Coccothrinax_moaensis', 'Desmoncus_stans', 'Attalea_degranvillei', 'Desmoncus_giganteus', 'Aiphanes_verrucosa', 'Phytelephas_seemannii', 'Attalea_salvadorensis', 'Chamaedorea_ernesti-augusti', 'Aiphanes_graminifolia', 'Syagrus_werdermannii', 'Bactris_hatschbachii', 'Coccothrinax_baracoensis', 'Geonoma_poeppigiana', 'Copernicia_oxycalyx', 'Desmoncus_chinantlensis', 'Coccothrinax_camagueyana', 'Desmoncus_latisectus', 'Brahea_moorei', 'Guihaia_argyrata', 'Bactris_tomentosa', 'Bactris_oligocarpa', 'Butia_marmorii', 'Attalea_hoehnei', 'Aiphanes_duquei', 'Chamaedorea_castillo-montii', 'Attalea_guianensis', 'Acrocomia_media', 'Cryosophila_grayumii', 'Coccothrinax_victorini', 'Desmoncus_leptoclonos', 'Bactris_campestris', 'Bactris_riparia', 'Trithrinax_acanthocoma', 'Hyospathe_peruviana', 'Coccothrinax_trinitensis', 'Roystonea_princeps', 'Geonoma_laxiflora', 'Geonoma_aspidiifolia', 'Bactris_halmoorei', 'Butia_stolonifera', 'Aiphanes_spicata', 'Desmoncus_interjectus', 'Sabal_domingensis', 'Wettinia_radiata', 'Geonoma_spinescens', 'Syagrus_pleioclada', 'Butia_quaraimana', 'Wettinia_microcarpa', 'Aiphanes_chiribogensis', 'Butia_exilata', 'Desmoncus_moorei', 'Desmoncus_costaricensis', 'Copernicia_curtissii', 'Wettinia_kalbreyeri', 'Chamaedorea_amabilis', 'Lytocaryum_insigne', 'Aiphanes_multiplex', 'Bactris_ptariana', 'Asterogyne_guianensis', 'Coccothrinax_acunana', 'Desmoncus_horridus', 'Astrocaryum_aculeatissimum', 'Lytocaryum_itapebiense', 'Hemithrinax_rivularis', 'Geonoma_hollinensis', 'Geonoma_poiteauana', 'Geonoma_oligoclona', 'Bactris_kunorum', 'Leopoldinia_piassaba', 'Geonoma_oldemanii', 'Bactris_timbuiensis', 'Brahea_pimo', 'Aiphanes_stergiosii', 'Bactris_guineensis', 'Calyptrogyne_pubescens', 'Oenocarpus_bacaba', 'Attalea_maracaibensis', 'Bactris_acanthocarpoides', 'Phytelephas_schottii', 'Copernicia_curbeloi', 'Wettinia_praemorsa', 'Coccothrinax_yuraguana', 'Bactris_pliniana', 'Attalea_compta', 'Desmoncus_vacivus', 'Bactris_turbinocarpa', 'Attalea_magdalenica', 'Cryosophila_macrocarpa', 'Aiphanes_weberbaueri', 'Chamaedorea_moliniana', 'Brahea_salvadorensis', 'Calyptrogyne_osensis', 'Roystonea_dunlapiana', 'Calyptrogyne_fortunensis', 'Geonoma_tenuissima', 'Attalea_bassleriana', 'Washingtonia_robusta', 'Bactris_rostrata', 'Thrinax_excelsa', 'Bactris_bahiensis', 'Chamaedorea_foveata', 'Geonoma_peruviana', 'Wettinia_quinaria', 'Wettinia_disticha', 'Calyptrogyne_sanblasensis', 'Cryosophila_cookii', 'Bactris_corossilla', 'Geonoma_epetiolata', 'Chamaedorea_pachecoana', 'Hyospathe_macrorhachis', 'Syagrus_cearensis', 'Attalea_osmantha', 'Sabal_gretherae', 'Chamaedorea_schiedeana', 'Syagrus_yungasensis', 'Geonoma_gentryi', 'Allagoptera_campestris', 'Syagrus_lilliputiana', 'Astrocaryum_jauari', 'Geonoma_divisa', 'Geonoma_sanmartinensis', 'Coccothrinax_pauciramosa', 'Attalea_peruviana', 'Bactris_pickelii', 'Geonoma_talamancana', 'Chamaedorea_angustisecta', 'Bactris_caudata', 'Bactris_faucium', 'Bactris_rhaphidacantha', 'Attalea_weberbaueri', 'Bactris_ana-juliae', 'Coccothrinax_leonis', 'Socratea_salazarii', 'Bactris_aubletiana', 'Bactris_chaveziae', 'Reinhardtia_latisecta', 'Astrocaryum_confertum', 'Pholidostachys_synanthera', 'Astrocaryum_giganteum'}\n",
            "\n",
            "Alleged missing species: 328\n",
            "Percentage of species missing: 40.49382716049383\n",
            "Percentage of species sequenced: 59.50617283950617\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "c5I5tHfWlTyh"
      },
      "source": [
        "That looks like a very low percentage. But there are a bunch of sequences we have not logged in the metadata. We can have a look at lists of species other people sequenced and count them in.\n",
        "\n",
        "The datasets that are missing:  \n",
        "- Chamaedoreas from Elliot Gardner\n",
        "- Species extracted by Paola Lima\n",
        "- Acrocomia from Brazil researchers\n",
        "- Geonoma from Loiseau et al. 2019\n",
        "\n",
        "We can check lists of files and metadata for those datasets\n",
        "\n",
        "First, Chamaedoreas. They are conveniently named already. All I'll do is to create a list of files in a text file and work from there.\n",
        "\n",
        "If running this on a local disk, you can use jupyter notebook's magic commands for bash\n",
        "\n",
        "\n",
        "```\n",
        "%%bash\n",
        "\n",
        "ls *gz > files_chamaedorea.txt\n",
        "```\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 436
        },
        "id": "6IeqWbaqCU7o",
        "outputId": "996f937b-2aee-4411-8ac9-d8dd11f129ec"
      },
      "source": [
        "def get_data():\n",
        "  return pd.read_csv('./files_chamaedorea.txt', sep = '\\t', names = ['file']) # the file doesn't have a header\n",
        "\n",
        "# try reading the file if exists, keep in mind the date\n",
        "try:\n",
        "  eg = get_data()\n",
        "except:\n",
        "  url = 'https://raw.githubusercontent.com/mftorres/APP/main/data/files_chamaedorea.txt' # shared link careful who you share this with\n",
        "  print('Downloading')\n",
        "  res = rs.get(url = url)\n",
        "  with open('./files_chamaedorea.txt', 'wb') as file: # Writing document in binary mode so python doesn't make changes\n",
        "    file.write(res.content)\n",
        "  eg = get_data()\n",
        "\n",
        "eg # check the database"
      ],
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Downloading\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>file</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Chamaedorea_aff_nationsiana_Steyermark41863_S2...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Chamaedorea_aff_nationsiana_Steyermark41863_S2...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Chamaedorea_arenbergiana_Hawkins228_R1_001.fas...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Chamaedorea_arenbergiana_Hawkins228_R2_001.fas...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Chamaedorea_binderi_Grayum3351_R1_001.fastq.gz*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>75</th>\n",
              "      <td>Chamaedorea_vulgata_Standley67344_R2_001.fastq...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>76</th>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>77</th>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz*</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>79</th>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>80 rows × 1 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                 file\n",
              "0   Chamaedorea_aff_nationsiana_Steyermark41863_S2...\n",
              "1   Chamaedorea_aff_nationsiana_Steyermark41863_S2...\n",
              "2   Chamaedorea_arenbergiana_Hawkins228_R1_001.fas...\n",
              "3   Chamaedorea_arenbergiana_Hawkins228_R2_001.fas...\n",
              "4     Chamaedorea_binderi_Grayum3351_R1_001.fastq.gz*\n",
              "..                                                ...\n",
              "75  Chamaedorea_vulgata_Standley67344_R2_001.fastq...\n",
              "76   Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz*\n",
              "77   Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz*\n",
              "78  Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...\n",
              "79  Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...\n",
              "\n",
              "[80 rows x 1 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 28
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TIpiqQXW0Rqk"
      },
      "source": [
        "That file has 80 files of compressed paired-end reads, forward and reverse.\n",
        "We need to add the file names into the database. We keep every file in the database to be able to mix single and paired-end reads.\n",
        "\n",
        "The column for files looks like"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yXVqI43XCU-q",
        "outputId": "ef20ec4d-9248-4a11-b5a5-d7be3b14aa93"
      },
      "source": [
        "df.columns\n",
        "df['filename']"
      ],
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0          1_151124_000000000-AJE6N_P3252_1001_1.fastq.gz\n",
              "1          1_151124_000000000-AJE6N_P3252_1001_2.fastq.gz\n",
              "2          1_151130_000000000-AK5EU_P3252_1126_1.fastq.gz\n",
              "3          1_151130_000000000-AK5EU_P3252_1126_2.fastq.gz\n",
              "4          1_151130_000000000-AKG91_P3252_1182_1.fastq.gz\n",
              "                              ...                        \n",
              "3044    Chamaedorea_vulgata_Standley67344_R2_001.fastq.gz\n",
              "3045      Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz\n",
              "3046      Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz\n",
              "3047    Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...\n",
              "3048    Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...\n",
              "Name: filename, Length: 3049, dtype: object"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 29
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "S-Hd1aem0-cc"
      },
      "source": [
        "well, looks like I did update that data on the metadata. Let's see if the other fields are filled. Looks pretty much complete. We can go ahead with other datasets then."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 609
        },
        "id": "q20OOxN8CVDx",
        "outputId": "d29cd1a7-6036-41b9-eb39-88af7996d456"
      },
      "source": [
        "df[df['filename'].str.contains('Chamaedorea')] # filters the database by looking a string match"
      ],
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>filename</th>\n",
              "      <th>sense</th>\n",
              "      <th>filecode</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "      <th>botanic_garden</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>ifmorphotype</th>\n",
              "      <th>ifpopulation</th>\n",
              "      <th>sent_to_Cano</th>\n",
              "      <th>newfilename</th>\n",
              "      <th>raw_reads</th>\n",
              "      <th>reads_trimed_paired_flag</th>\n",
              "      <th>reads_trimed_paired</th>\n",
              "      <th>reads_trimed_single</th>\n",
              "      <th>percentage_lost_tosingles</th>\n",
              "      <th>pcr_filtered</th>\n",
              "      <th>collection_year</th>\n",
              "      <th>Continent</th>\n",
              "      <th>country</th>\n",
              "      <th>long</th>\n",
              "      <th>lat</th>\n",
              "      <th>flag</th>\n",
              "      <th>notes</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>2969</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_aff_nationsiana_Steyermark41863_S2...</td>\n",
              "      <td>R1</td>\n",
              "      <td>Steyermark41863_S26_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_nationsiana</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2970</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_aff_nationsiana_Steyermark41863_S2...</td>\n",
              "      <td>R2</td>\n",
              "      <td>Steyermark41863_S26_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_nationsiana</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2971</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_arenbergiana_Hawkins228_R1_001.fas...</td>\n",
              "      <td>R1</td>\n",
              "      <td>Hawkins228</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_arenbergiana</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2972</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_arenbergiana_Hawkins228_R2_001.fas...</td>\n",
              "      <td>R2</td>\n",
              "      <td>Hawkins228</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_arenbergiana</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2973</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_binderi_Grayum3351_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>Grayum3351</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_binderi</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3044</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_vulgata_Standley67344_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>Standley67344</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_vulgata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3045</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R1_001.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>Azofeifa484</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3046</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_R2_001.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>Azofeifa484</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3047</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R1_00...</td>\n",
              "      <td>R1</td>\n",
              "      <td>Azofeifa484_S36_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3048</th>\n",
              "      <td>EliottGardner</td>\n",
              "      <td>Eliott Gardner</td>\n",
              "      <td>EGR</td>\n",
              "      <td>none</td>\n",
              "      <td>Chamaedorea_zamorae_Azofeifa484_S36_L001_R2_00...</td>\n",
              "      <td>R2</td>\n",
              "      <td>Azofeifa484_S36_L001</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Chamaedorea</td>\n",
              "      <td>Chamaedorea_zamorae</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>80 rows × 30 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "            dataset        provider pi_code  ... lat flag notes\n",
              "2969  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "2970  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "2971  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "2972  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "2973  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "...             ...             ...     ...  ...  ..  ...   ...\n",
              "3044  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3045  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3046  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3047  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "3048  EliottGardner  Eliott Gardner     EGR  ... NaN  NaN   NaN\n",
              "\n",
              "[80 rows x 30 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 30
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qx1KxDSY23k9"
      },
      "source": [
        "Let's check the data sequenced by Paola. There's a list of species "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 436
        },
        "id": "ri4iuevFCVHF",
        "outputId": "d1a9050d-9b05-45de-ae18-d061195fa81e"
      },
      "source": [
        "def get_data():\n",
        "  return pd.read_csv('./metadata_machinereadable_Paola_2020.txt', sep = '\\t') # the file doesn't have a header\n",
        "\n",
        "# try reading the file if exists, keep in mind the date\n",
        "try:\n",
        "  seqp = get_data()\n",
        "except:\n",
        "  url = 'https://raw.githubusercontent.com/mftorres/APP/main/data/metadata_machinereadable_Paola_2020.txt' # shared link careful who you share this with\n",
        "  print('Downloading')\n",
        "  res = rs.get(url = url)\n",
        "  with open('./metadata_machinereadable_Paola_2020.txt', 'wb') as file: # Writing document in binary mode so python doesn't make changes\n",
        "    file.write(res.content)\n",
        "  seqp = get_data()\n",
        "\n",
        "seqp # check the database"
      ],
      "execution_count": 31,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Downloading\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>plate</th>\n",
              "      <th>Rapid_Genomics_Identification</th>\n",
              "      <th>Plate</th>\n",
              "      <th>Column Sort</th>\n",
              "      <th>Silica number</th>\n",
              "      <th>Species</th>\n",
              "      <th>Collector and Number</th>\n",
              "      <th>Final Concentration (ng/ul)</th>\n",
              "      <th>Nanodrop (260/280)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA01</td>\n",
              "      <td>A1</td>\n",
              "      <td>1</td>\n",
              "      <td>3174</td>\n",
              "      <td>Acoelorrhaphe wrightii</td>\n",
              "      <td>H. Balslev 8156</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.76</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA02</td>\n",
              "      <td>A2</td>\n",
              "      <td>9</td>\n",
              "      <td>1156</td>\n",
              "      <td>Aiphanes bicornis</td>\n",
              "      <td>J. West s.n.</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.82</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA03</td>\n",
              "      <td>A3</td>\n",
              "      <td>17</td>\n",
              "      <td>261</td>\n",
              "      <td>Aiphanes chiribogensis</td>\n",
              "      <td>F. Borchsenius 652</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.80</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA04</td>\n",
              "      <td>A4</td>\n",
              "      <td>25</td>\n",
              "      <td>3563</td>\n",
              "      <td>Aiphanes deltoidea</td>\n",
              "      <td>L.A. Nuñez s.n.</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.84</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA05</td>\n",
              "      <td>A5</td>\n",
              "      <td>33</td>\n",
              "      <td>397</td>\n",
              "      <td>Aiphanes duquei</td>\n",
              "      <td>R. Bernal 2852</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.78</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>118</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P002_WB11</td>\n",
              "      <td>B11</td>\n",
              "      <td>82</td>\n",
              "      <td>431</td>\n",
              "      <td>Wettinia praemorsa</td>\n",
              "      <td>R. Bernal 2883</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.82</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>119</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P002_WB12</td>\n",
              "      <td>B12</td>\n",
              "      <td>90</td>\n",
              "      <td>434</td>\n",
              "      <td>Wettinia quinaria</td>\n",
              "      <td>R. Bernal 2491</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.73</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>120</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P002_WC01</td>\n",
              "      <td>C1</td>\n",
              "      <td>3</td>\n",
              "      <td>436</td>\n",
              "      <td>Wettinia radiata</td>\n",
              "      <td>R. Bernal 2190</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.75</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P002_WC02</td>\n",
              "      <td>C2</td>\n",
              "      <td>11</td>\n",
              "      <td>3514</td>\n",
              "      <td>Wettinia verruculosa</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1.90</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>122</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P002_WC03</td>\n",
              "      <td>C3</td>\n",
              "      <td>19</td>\n",
              "      <td>2299</td>\n",
              "      <td>Mauritiella armata</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>50.0</td>\n",
              "      <td>2.11</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>123 rows × 9 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     plate  ... Nanodrop (260/280)\n",
              "0        1  ...               1.76\n",
              "1        1  ...               1.82\n",
              "2        1  ...               1.80\n",
              "3        1  ...               1.84\n",
              "4        1  ...               1.78\n",
              "..     ...  ...                ...\n",
              "118      2  ...               1.82\n",
              "119      2  ...               1.73\n",
              "120      2  ...               1.75\n",
              "121      2  ...               1.90\n",
              "122      2  ...               2.11\n",
              "\n",
              "[123 rows x 9 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 31
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 436
        },
        "id": "lRmy13YlB4PQ",
        "outputId": "f64c4561-a709-4077-ba8d-725f3b1f301f"
      },
      "source": [
        "# now the rapid genomics data\n",
        "def get_data():\n",
        "  return pd.read_csv('./GOT_130404_SampleSheet.csv', sep = '\\t') # the file doesn't have a header\n",
        "\n",
        "# try reading the file if exists, keep in mind the date\n",
        "try:\n",
        "  seqprapid = get_data()\n",
        "except:\n",
        "  url = 'https://raw.githubusercontent.com/mftorres/APP/main/data/GOT_130404_SampleSheet.csv' # shared link careful who you share this with\n",
        "  print('Downloading')\n",
        "  res = rs.get(url = url)\n",
        "  with open('./GOT_130404_SampleSheet.csv', 'wb') as file: # Writing document in binary mode so python doesn't make changes\n",
        "    file.write(res.content)\n",
        "  seqprapid = get_data()\n",
        "\n",
        "seqprapid # check the database"
      ],
      "execution_count": 32,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Downloading\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>RG_Sample_Code</th>\n",
              "      <th>Customer_Code</th>\n",
              "      <th>i5_Barcode_Seq</th>\n",
              "      <th>i7_Barcode_Seq</th>\n",
              "      <th>Sequence_Name</th>\n",
              "      <th>Sequencing_Cycle</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>GOT_130404_P001_WA01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>GTGTTCTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>GOT_130404_P001_WA02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>AGTCACTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>GOT_130404_P001_WA03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>ATTGGCTC</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>GOT_130404_P001_WA04</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>CAGATCTG</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA04_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>GOT_130404_P001_WA05</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>CCGTGAGA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA05_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>118</th>\n",
              "      <td>GOT_130404_P002_WB11</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>ACCACTGT</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WB11_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>119</th>\n",
              "      <td>GOT_130404_P002_WB12</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>ACAGATTC</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WB12_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>120</th>\n",
              "      <td>GOT_130404_P002_WC01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGAACAA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121</th>\n",
              "      <td>GOT_130404_P002_WC02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGTGGTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>122</th>\n",
              "      <td>GOT_130404_P002_WC03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>CGAACTTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>123 rows × 6 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "           RG_Sample_Code  ...  Sequencing_Cycle\n",
              "0    GOT_130404_P001_WA01  ...             2x150\n",
              "1    GOT_130404_P001_WA02  ...             2x150\n",
              "2    GOT_130404_P001_WA03  ...             2x150\n",
              "3    GOT_130404_P001_WA04  ...             2x150\n",
              "4    GOT_130404_P001_WA05  ...             2x150\n",
              "..                    ...  ...               ...\n",
              "118  GOT_130404_P002_WB11  ...             2x150\n",
              "119  GOT_130404_P002_WB12  ...             2x150\n",
              "120  GOT_130404_P002_WC01  ...             2x150\n",
              "121  GOT_130404_P002_WC02  ...             2x150\n",
              "122  GOT_130404_P002_WC03  ...             2x150\n",
              "\n",
              "[123 rows x 6 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 32
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wfH1Jc63CVJ6",
        "outputId": "c2d81586-3a10-40a1-b188-b37f8b7bbe19"
      },
      "source": [
        "df.columns\n",
        "df['dataset'].unique() # looks like these sequences aren't there. We can fill up the dataset then."
      ],
      "execution_count": 33,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Cano', 'Lausanne_Cerox', 'American Palms', 'Astrocaryum',\n",
              "       'Mauritia', 'Zizka', 'Raphia', 'Geonoma', 'EliottGardner'],\n",
              "      dtype=object)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 33
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_ux58BM0-qya"
      },
      "source": [
        "We can create new columns in seqp to contain the information needed in df and then concatenate both.\n",
        "\n",
        "File names can be extractede from the 'Sequence_name' column in seqpraid. We can add the species and voucher from seqp to seqpraid, then duplicate rows to create the R2 read entries, rename columns and concatenate with df"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8f8kXdXD-oQl",
        "outputId": "e563d74b-b1a0-47c1-f54b-63379d0f767b"
      },
      "source": [
        "# create a column with file information, forward and reverse\n",
        "# will create empty lists that I will join in a dataframe later\n",
        "\n",
        "# first, a dictionary onject linking sample codes with species and vouchers\n",
        "# a dictionary has keys (unique) and values, the latest can be strings, lists, other dictionaries, etc.\n",
        "\n",
        "# in this case keys are the rapid genomics identifications\n",
        "# values are tuples of species (replacing space with underscore) and voucher\n",
        "id_spp_dict = dict(zip(seqp['Rapid_Genomics_Identification'],zip(seqp['Species'].str.replace(' ','_'),seqp['Collector and Number']))) # think of zip as a jacket zip: makes tuples with values in the same indexing\n",
        "\n",
        "# looks like this\n",
        "id_spp_dict"
      ],
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'GOT_130404_P001_WA01': ('Acoelorrhaphe_wrightii', 'H. Balslev 8156'),\n",
              " 'GOT_130404_P001_WA02': ('Aiphanes_bicornis', 'J. West s.n.'),\n",
              " 'GOT_130404_P001_WA03': ('Aiphanes_chiribogensis', 'F. Borchsenius 652'),\n",
              " 'GOT_130404_P001_WA04': ('Aiphanes_deltoidea', 'L.A. Nuñez s.n.'),\n",
              " 'GOT_130404_P001_WA05': ('Aiphanes_duquei', 'R. Bernal 2852'),\n",
              " 'GOT_130404_P001_WA06': ('Aiphanes_eggersii', 'J.C.Pintaud 1476'),\n",
              " 'GOT_130404_P001_WA07': ('Aiphanes_graminifolia', 'R.Bernal 4825'),\n",
              " 'GOT_130404_P001_WA08': ('Aiphanes_grandis', 'F. Borchsenius 648'),\n",
              " 'GOT_130404_P001_WA09': ('Aiphanes_verrucosa', 'F. Borchsenius 641'),\n",
              " 'GOT_130404_P001_WA10': ('Aiphanes_weberbaueri', 'H.Balslev 7769'),\n",
              " 'GOT_130404_P001_WA11': ('Asterogyne_guianensis', 'J. Roncal 406'),\n",
              " 'GOT_130404_P001_WA12': ('Asterogyne_ramosa', 'Private collection'),\n",
              " 'GOT_130404_P001_WB01': ('Astrocaryum_chambira', 'T. Emilio 864'),\n",
              " 'GOT_130404_P001_WB02': ('Astrocaryum_faranae', 'H. Balslev 7897'),\n",
              " 'GOT_130404_P001_WB03': ('Astrocaryum_huaimi', 'T.P.L. Couvreur 289'),\n",
              " 'GOT_130404_P001_WB04': ('Astrocaryum_jauari', 'H. Balslev 7347'),\n",
              " 'GOT_130404_P001_WB05': ('Mauritiella_pumila', 'R. Bernal 4373'),\n",
              " 'GOT_130404_P001_WB06': ('Astrocaryum_minus', 'De Granville, J.-J. 14511'),\n",
              " 'GOT_130404_P001_WB07': ('Astrocaryum_paramaca', 'J.C.Pintaud 901'),\n",
              " 'GOT_130404_P001_WB08': ('Astrocaryum_sciophilum', 'B. Maguire 54126'),\n",
              " 'GOT_130404_P001_WB09': ('Astrocaryum_urostachys',\n",
              "  'F. Borchsenius, no voucher'),\n",
              " 'GOT_130404_P001_WB10': ('Astrocaryum_vulgare', 'M.J. Balic 1626'),\n",
              " 'GOT_130404_P001_WB11': ('Attalea_funifera', 'L.A. Matos Silva 747'),\n",
              " 'GOT_130404_P001_WB12': ('Attalea_racemosa', 'H. Balslev 7272'),\n",
              " 'GOT_130404_P001_WC01': ('Bactris_acanthocarpoides', 'T. Emilio 831'),\n",
              " 'GOT_130404_P001_WC02': ('Bactris_aubletiana', 'ENV 1593'),\n",
              " 'GOT_130404_P001_WC03': ('Bactris_balanophora', 'R. Bernal 4417'),\n",
              " 'GOT_130404_P001_WC04': ('Bactris_bifida', 'H. Balslev 6594'),\n",
              " 'GOT_130404_P001_WC05': ('Bactris_campestris', 'J.C. Pintaud 1448'),\n",
              " 'GOT_130404_P001_WC06': ('Bactris_chaveziae', 'T.L.P. Couvreur 279'),\n",
              " 'GOT_130404_P001_WC07': ('Bactris_constanciae', 'De Granville 4850'),\n",
              " 'GOT_130404_P001_WC08': ('Bactris_corossilla', 'H. Balslev 7294'),\n",
              " 'GOT_130404_P001_WC09': ('Bactris_gastoniana', 'J.C.Pintaud 1446'),\n",
              " 'GOT_130404_P001_WC10': ('Bactris_guineensis', 'JBP s.n.'),\n",
              " 'GOT_130404_P001_WC11': ('Bactris_halmoorei', 'H. Balslev 7267'),\n",
              " 'GOT_130404_P001_WC12': ('Bactris_killipii', 'J.C.Pintaud 1428'),\n",
              " 'GOT_130404_P001_WD01': ('Bactris_longiseta', 'J.C.Pintaud 330'),\n",
              " 'GOT_130404_P001_WD02': ('Bactris_maraja', 'H. Balslev 6565'),\n",
              " 'GOT_130404_P001_WD03': ('Bactris_martiana', 'H. Balslev 6549'),\n",
              " 'GOT_130404_P001_WD04': ('Bactris_oligocarpa', 'T. Emilio 528'),\n",
              " 'GOT_130404_P001_WD05': ('Bactris_pliniana', 'J.C.Pintaud 1452'),\n",
              " 'GOT_130404_P001_WD06': ('Bactris_riparia', 'H. Balslev 7953'),\n",
              " 'GOT_130404_P001_WD07': ('Bactris_schultesii', 'H. Balslev 7284'),\n",
              " 'GOT_130404_P001_WD08': ('Bactris_setiflora', 'F. Borchsenius 629'),\n",
              " 'GOT_130404_P001_WD09': ('Bactris_simplicifrons', 'H. Balslev 6780'),\n",
              " 'GOT_130404_P001_WD10': ('Bactris_syagroides', 'T. Emilio 801'),\n",
              " 'GOT_130404_P001_WD11': ('Calyptrogyne_baudensis', 'Samín et al. 05'),\n",
              " 'GOT_130404_P001_WD12': ('Calyptrogyne_fortunensis', 'J. Roncal 370'),\n",
              " 'GOT_130404_P001_WE01': ('Calyptrogyne_trichostachys', 'J. Roncal 414'),\n",
              " 'GOT_130404_P001_WE02': ('Calyptronoma_occidentalis', 'FTG 71375D'),\n",
              " 'GOT_130404_P001_WE03': ('Calyptronoma_plumeriana', 'FTG 82492D'),\n",
              " 'GOT_130404_P001_WE04': ('Ceroxylon_peruvianum', 'Samín et al. 05'),\n",
              " 'GOT_130404_P001_WE05': ('Chamaedorea_crucensis', 'Hod No ID'),\n",
              " 'GOT_130404_P001_WE06': ('Chamaedorea_fragrans', 'J.C. Pintaud 1490'),\n",
              " 'GOT_130404_P001_WE07': ('Desmoncus_cirrhifer', 'R. Bernal 4989'),\n",
              " 'GOT_130404_P001_WE08': ('Desmoncus_giganteus', 'L.A. Nuñez s.n.'),\n",
              " 'GOT_130404_P001_WE09': ('Desmoncus_myriacanthos', 'R. Bernal 4945'),\n",
              " 'GOT_130404_P001_WE10': ('Desmoncus_vacivus', 'MFT 703'),\n",
              " 'GOT_130404_P001_WE11': ('Dictyocaryon_ptarianum', 'Pintaud ENV 1185'),\n",
              " 'GOT_130404_P001_WE12': ('Geonoma_aspdiifolia', 'C. Farney 1752'),\n",
              " 'GOT_130404_P001_WF01': ('Geonoma_atrovirens_(Chamaedorea)', 'T. Emilio 761'),\n",
              " 'GOT_130404_P001_WF02': ('Geonoma_baculifera', 'M.J. Jansen-Jacobs 2386'),\n",
              " 'GOT_130404_P001_WF03': ('Geonoma_epetiolata', 'J. Roncal 316'),\n",
              " 'GOT_130404_P001_WF04': ('Geonoma_euspatha', 'J.J.Granville 11989'),\n",
              " 'GOT_130404_P001_WF05': ('Geonoma_hollinensis', 'H. Balslev 6418'),\n",
              " 'GOT_130404_P001_WF06': ('Geonoma_hugonis', 'J. Roncal 358'),\n",
              " 'GOT_130404_P001_WF07': ('Geonoma_lanata', 'F. Borchsenius 334'),\n",
              " 'GOT_130404_P001_WF08': ('Geonoma_laxiflora', 'H. Balslev 7273'),\n",
              " 'GOT_130404_P001_WF09': ('Geonoma_leptospadix', 'T. Emilio 714'),\n",
              " 'GOT_130404_P001_WF10': ('Geonoma_longepedunculata', 'Gloria Galeano 9186'),\n",
              " 'GOT_130404_P001_WF11': ('Geonoma_oldemanii', 'J. Roncal 403'),\n",
              " 'GOT_130404_P001_WF12': ('Geonoma_oligoclona', 'T. Emilio 731'),\n",
              " 'GOT_130404_P001_WG01': ('Geonoma_pinnatifrons', 'BG 534'),\n",
              " 'GOT_130404_P001_WG02': ('Geonoma_poeppigiana', 'R. Bernal 4621'),\n",
              " 'GOT_130404_P001_WG03': ('Mauritiella_armata', 'R. Bernal 4469'),\n",
              " 'GOT_130404_P001_WG04': ('Geonoma_triglochin', 'T. Emilio 820'),\n",
              " 'GOT_130404_P001_WG05': ('Geonoma_umbraculiformis', 'T. Emilio 780'),\n",
              " 'GOT_130404_P001_WG06': ('Guiahia_argyrata', '3757'),\n",
              " 'GOT_130404_P001_WG07': ('Hyospathe_frontinensis', 'R. Bernal 4955'),\n",
              " 'GOT_130404_P001_WG08': ('Iriartella_setigera', 'R. Bernal 4462'),\n",
              " 'GOT_130404_P001_WG09': ('Iriartella_stenocarpa', 'MFT 704'),\n",
              " 'GOT_130404_P001_WG10': ('Juania_australis', 'JCP 1454'),\n",
              " 'GOT_130404_P001_WG11': ('Manicaria_saccifera', 'H. Balslev 8114'),\n",
              " 'GOT_130404_P001_WG12': ('Oenocarpus_bacaba', 'M.J. Jansen-Jacobs 2387'),\n",
              " 'GOT_130404_P001_WH01': ('Oenocarpus_simplex', 'Rodrigo Cámara-Leret 1711'),\n",
              " 'GOT_130404_P001_WH02': ('Parajubaea_cocoides', 'Austria 80 **'),\n",
              " 'GOT_130404_P001_WH03': ('Pholidostachys_synanthera', 'T. Couvreur 119'),\n",
              " 'GOT_130404_P001_WH04': ('Phytelephas_tumacana', 'F.Borchsenius 659'),\n",
              " 'GOT_130404_P001_WH05': ('Prestoea_pubigera', 'A. Henderson 1623'),\n",
              " 'GOT_130404_P001_WH06': ('Pritchardia_minor', 'Kwood 7367'),\n",
              " 'GOT_130404_P001_WH07': ('Pritchardia_pericularum', 'JVM 1 (Niau?)'),\n",
              " 'GOT_130404_P001_WH08': ('Pritchardia_viscosa', 'SP Perlman 18712'),\n",
              " 'GOT_130404_P001_WH09': ('Pritchardia_vuylstekeana', 'JVM 2 (Makatea?)'),\n",
              " 'GOT_130404_P001_WH10': ('Sabal_antillensis', 'M.P. Griffith 388'),\n",
              " 'GOT_130404_P001_WH11': ('Sabal_gretherae', 'H. Balslev 8170'),\n",
              " 'GOT_130404_P001_WH12': ('Sabal_lougheediana', 'M.P. Griffith 392'),\n",
              " 'GOT_130404_P002_WA01': ('Sabal_minor', 'M. Eason 1142'),\n",
              " 'GOT_130404_P002_WA02': ('Sabal_yapa', 'H. Balslev 8166'),\n",
              " 'GOT_130404_P002_WA03': ('Schippia_concolor', '59-1200 D'),\n",
              " 'GOT_130404_P002_WA04': ('Socratea_hecatonandra', 'GG 8119'),\n",
              " 'GOT_130404_P002_WA05': ('Socratea_rostrata', 'R. Bernal 2493'),\n",
              " 'GOT_130404_P002_WA06': ('Socratea_salazarii', 'H. Balslev 7298'),\n",
              " 'GOT_130404_P002_WA07': ('Syagrus_amara', 'Noblick 5560'),\n",
              " 'GOT_130404_P002_WA08': ('Washingtonia_robusta', 'Austria 81 **'),\n",
              " 'GOT_130404_P002_WA09': ('Wettinia_aequatorialis', 'H. Balslev 6462'),\n",
              " 'GOT_130404_P002_WA10': ('Wettinia_anomala', 'R. Bernal 2487'),\n",
              " 'GOT_130404_P002_WA11': ('Wettinia_disticha', 'R. Bernal 2838'),\n",
              " 'GOT_130404_P002_WA12': ('Wettinia_castanea', 'R. Bernal 2825'),\n",
              " 'GOT_130404_P002_WB01': ('Wettinia_drudei', 'T. Emilio 810'),\n",
              " 'GOT_130404_P002_WB02': ('Wettinia_fascicularis', 'R. Bernal 2830'),\n",
              " 'GOT_130404_P002_WB03': ('Wettinia_hirsuta', 'R. Bernal 4737'),\n",
              " 'GOT_130404_P002_WB04': ('Wettinia_kalbreyeri', 'R. Bernal 4726'),\n",
              " 'GOT_130404_P002_WB05': ('Wettinia_lanata', 'R. Bernal 4775'),\n",
              " 'GOT_130404_P002_WB06': ('Wettinia_longipetala', 'H. Balslev 6545'),\n",
              " 'GOT_130404_P002_WB07': ('Wettinia_maynensis', 'A. Barford 60538'),\n",
              " 'GOT_130404_P002_WB08': ('Wettinia_microcarpa', 'R. Bernal 3481'),\n",
              " 'GOT_130404_P002_WB09': ('Wettinia_oxycarpa', 'R. Bernal 2498'),\n",
              " 'GOT_130404_P002_WB10': ('Wettinia_panamensis', 'A.J. Henderson 2045'),\n",
              " 'GOT_130404_P002_WB11': ('Wettinia_praemorsa', 'R. Bernal 2883'),\n",
              " 'GOT_130404_P002_WB12': ('Wettinia_quinaria', 'R. Bernal 2491'),\n",
              " 'GOT_130404_P002_WC01': ('Wettinia_radiata', 'R. Bernal 2190'),\n",
              " 'GOT_130404_P002_WC02': ('Wettinia_verruculosa', 'R. Bernal 2500'),\n",
              " 'GOT_130404_P002_WC03': ('Mauritiella_armata', 'J.C.Pintaud 1488')}"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 34
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "cRo7MFt9l5EH",
        "outputId": "ea2abc8b-a0ee-4401-ebc2-54316db3eb40"
      },
      "source": [
        "# now, a dictionary for sample code\n",
        "id_code_dict = dict(zip(seqp['Rapid_Genomics_Identification'],seqp['Plate']+'_'+seqp['Collector and Number'].str.replace('[^a-zA-Z \\d]','').str.replace(' ','')))\n",
        "id_code_dict"
      ],
      "execution_count": 35,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'GOT_130404_P001_WA01': 'A1_HBalslev8156',\n",
              " 'GOT_130404_P001_WA02': 'A2_JWestsn',\n",
              " 'GOT_130404_P001_WA03': 'A3_FBorchsenius652',\n",
              " 'GOT_130404_P001_WA04': 'A4_LANuezsn',\n",
              " 'GOT_130404_P001_WA05': 'A5_RBernal2852',\n",
              " 'GOT_130404_P001_WA06': 'A6_JCPintaud1476',\n",
              " 'GOT_130404_P001_WA07': 'A7_RBernal4825',\n",
              " 'GOT_130404_P001_WA08': 'A8_FBorchsenius648',\n",
              " 'GOT_130404_P001_WA09': 'A9_FBorchsenius641',\n",
              " 'GOT_130404_P001_WA10': 'A10_HBalslev7769',\n",
              " 'GOT_130404_P001_WA11': 'A11_JRoncal406',\n",
              " 'GOT_130404_P001_WA12': 'A12_Privatecollection',\n",
              " 'GOT_130404_P001_WB01': 'B1_TEmilio864',\n",
              " 'GOT_130404_P001_WB02': 'B2_HBalslev7897',\n",
              " 'GOT_130404_P001_WB03': 'B3_TPLCouvreur289',\n",
              " 'GOT_130404_P001_WB04': 'B4_HBalslev7347',\n",
              " 'GOT_130404_P001_WB05': 'B5_RBernal4373',\n",
              " 'GOT_130404_P001_WB06': 'B6_DeGranvilleJJ14511',\n",
              " 'GOT_130404_P001_WB07': 'B7_JCPintaud901',\n",
              " 'GOT_130404_P001_WB08': 'B8_BMaguire54126',\n",
              " 'GOT_130404_P001_WB09': 'B9_FBorchseniusnovoucher',\n",
              " 'GOT_130404_P001_WB10': 'B10_MJBalic1626',\n",
              " 'GOT_130404_P001_WB11': 'B11_LAMatosSilva747',\n",
              " 'GOT_130404_P001_WB12': 'B12_HBalslev7272',\n",
              " 'GOT_130404_P001_WC01': 'C1_TEmilio831',\n",
              " 'GOT_130404_P001_WC02': 'C2_ENV1593',\n",
              " 'GOT_130404_P001_WC03': 'C3_RBernal4417',\n",
              " 'GOT_130404_P001_WC04': 'C4_HBalslev6594',\n",
              " 'GOT_130404_P001_WC05': 'C5_JCPintaud1448',\n",
              " 'GOT_130404_P001_WC06': 'C6_TLPCouvreur279',\n",
              " 'GOT_130404_P001_WC07': 'C7_DeGranville4850',\n",
              " 'GOT_130404_P001_WC08': 'C8_HBalslev7294',\n",
              " 'GOT_130404_P001_WC09': 'C9_JCPintaud1446',\n",
              " 'GOT_130404_P001_WC10': 'C10_JBPsn',\n",
              " 'GOT_130404_P001_WC11': 'C11_HBalslev7267',\n",
              " 'GOT_130404_P001_WC12': 'C12_JCPintaud1428',\n",
              " 'GOT_130404_P001_WD01': 'D1_JCPintaud330',\n",
              " 'GOT_130404_P001_WD02': 'D2_HBalslev6565',\n",
              " 'GOT_130404_P001_WD03': 'D3_HBalslev6549',\n",
              " 'GOT_130404_P001_WD04': 'D4_TEmilio528',\n",
              " 'GOT_130404_P001_WD05': 'D5_JCPintaud1452',\n",
              " 'GOT_130404_P001_WD06': 'D6_HBalslev7953',\n",
              " 'GOT_130404_P001_WD07': 'D7_HBalslev7284',\n",
              " 'GOT_130404_P001_WD08': 'D8_FBorchsenius629',\n",
              " 'GOT_130404_P001_WD09': 'D9_HBalslev6780',\n",
              " 'GOT_130404_P001_WD10': 'D10_TEmilio801',\n",
              " 'GOT_130404_P001_WD11': 'D11_Samnetal05',\n",
              " 'GOT_130404_P001_WD12': 'D12_JRoncal370',\n",
              " 'GOT_130404_P001_WE01': 'E1_JRoncal414',\n",
              " 'GOT_130404_P001_WE02': 'E2_FTG71375D',\n",
              " 'GOT_130404_P001_WE03': 'E3_FTG82492D',\n",
              " 'GOT_130404_P001_WE04': 'E4_Samnetal05',\n",
              " 'GOT_130404_P001_WE05': 'E5_HodNoID',\n",
              " 'GOT_130404_P001_WE06': 'E6_JCPintaud1490',\n",
              " 'GOT_130404_P001_WE07': 'E7_RBernal4989',\n",
              " 'GOT_130404_P001_WE08': 'E8_LANuezsn',\n",
              " 'GOT_130404_P001_WE09': 'E9_RBernal4945',\n",
              " 'GOT_130404_P001_WE10': 'E10_MFT703',\n",
              " 'GOT_130404_P001_WE11': 'E11_PintaudENV1185',\n",
              " 'GOT_130404_P001_WE12': 'E12_CFarney1752',\n",
              " 'GOT_130404_P001_WF01': 'F1_TEmilio761',\n",
              " 'GOT_130404_P001_WF02': 'F2_MJJansenJacobs2386',\n",
              " 'GOT_130404_P001_WF03': 'F3_JRoncal316',\n",
              " 'GOT_130404_P001_WF04': 'F4_JJGranville11989',\n",
              " 'GOT_130404_P001_WF05': 'F5_HBalslev6418',\n",
              " 'GOT_130404_P001_WF06': 'F6_JRoncal358',\n",
              " 'GOT_130404_P001_WF07': 'F7_FBorchsenius334',\n",
              " 'GOT_130404_P001_WF08': 'F8_HBalslev7273',\n",
              " 'GOT_130404_P001_WF09': 'F9_TEmilio714',\n",
              " 'GOT_130404_P001_WF10': 'F10_GloriaGaleano9186',\n",
              " 'GOT_130404_P001_WF11': 'F11_JRoncal403',\n",
              " 'GOT_130404_P001_WF12': 'F12_TEmilio731',\n",
              " 'GOT_130404_P001_WG01': 'G1_BG534',\n",
              " 'GOT_130404_P001_WG02': 'G2_RBernal4621',\n",
              " 'GOT_130404_P001_WG03': 'G3_RBernal4469',\n",
              " 'GOT_130404_P001_WG04': 'G4_TEmilio820',\n",
              " 'GOT_130404_P001_WG05': 'G5_TEmilio780',\n",
              " 'GOT_130404_P001_WG06': 'G6_3757',\n",
              " 'GOT_130404_P001_WG07': 'G7_RBernal4955',\n",
              " 'GOT_130404_P001_WG08': 'G8_RBernal4462',\n",
              " 'GOT_130404_P001_WG09': 'G9_MFT704',\n",
              " 'GOT_130404_P001_WG10': 'G10_JCP1454',\n",
              " 'GOT_130404_P001_WG11': 'G11_HBalslev8114',\n",
              " 'GOT_130404_P001_WG12': 'G12_MJJansenJacobs2387',\n",
              " 'GOT_130404_P001_WH01': 'H1_RodrigoCmaraLeret1711',\n",
              " 'GOT_130404_P001_WH02': 'H2_Austria80',\n",
              " 'GOT_130404_P001_WH03': 'H3_TCouvreur119',\n",
              " 'GOT_130404_P001_WH04': 'H4_FBorchsenius659',\n",
              " 'GOT_130404_P001_WH05': 'H5_AHenderson1623',\n",
              " 'GOT_130404_P001_WH06': 'H6_Kwood7367',\n",
              " 'GOT_130404_P001_WH07': 'H7_JVM1Niau',\n",
              " 'GOT_130404_P001_WH08': 'H8_SPPerlman18712',\n",
              " 'GOT_130404_P001_WH09': 'H9_JVM2Makatea',\n",
              " 'GOT_130404_P001_WH10': 'H10_MPGriffith388',\n",
              " 'GOT_130404_P001_WH11': 'H11_HBalslev8170',\n",
              " 'GOT_130404_P001_WH12': 'H12_MPGriffith392',\n",
              " 'GOT_130404_P002_WA01': 'A1_MEason1142',\n",
              " 'GOT_130404_P002_WA02': 'A2_HBalslev8166',\n",
              " 'GOT_130404_P002_WA03': 'A3_591200D',\n",
              " 'GOT_130404_P002_WA04': 'A4_GG8119',\n",
              " 'GOT_130404_P002_WA05': 'A5_RBernal2493',\n",
              " 'GOT_130404_P002_WA06': 'A6_HBalslev7298',\n",
              " 'GOT_130404_P002_WA07': 'A7_Noblick5560',\n",
              " 'GOT_130404_P002_WA08': 'A8_Austria81',\n",
              " 'GOT_130404_P002_WA09': 'A9_HBalslev6462',\n",
              " 'GOT_130404_P002_WA10': 'A10_RBernal2487',\n",
              " 'GOT_130404_P002_WA11': 'A11_RBernal2838',\n",
              " 'GOT_130404_P002_WA12': 'A12_RBernal2825',\n",
              " 'GOT_130404_P002_WB01': 'B1_TEmilio810',\n",
              " 'GOT_130404_P002_WB02': 'B2_RBernal2830',\n",
              " 'GOT_130404_P002_WB03': 'B3_RBernal4737',\n",
              " 'GOT_130404_P002_WB04': 'B4_RBernal4726',\n",
              " 'GOT_130404_P002_WB05': 'B5_RBernal4775',\n",
              " 'GOT_130404_P002_WB06': 'B6_HBalslev6545',\n",
              " 'GOT_130404_P002_WB07': 'B7_ABarford60538',\n",
              " 'GOT_130404_P002_WB08': 'B8_RBernal3481',\n",
              " 'GOT_130404_P002_WB09': 'B9_RBernal2498',\n",
              " 'GOT_130404_P002_WB10': 'B10_AJHenderson2045',\n",
              " 'GOT_130404_P002_WB11': 'B11_RBernal2883',\n",
              " 'GOT_130404_P002_WB12': 'B12_RBernal2491',\n",
              " 'GOT_130404_P002_WC01': 'C1_RBernal2190',\n",
              " 'GOT_130404_P002_WC02': 'C2_RBernal2500',\n",
              " 'GOT_130404_P002_WC03': 'C3_JCPintaud1488'}"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 35
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rH8gRCEHCVMk",
        "outputId": "0b8e7c14-a580-4db5-8b41-5e0011032337"
      },
      "source": [
        "# you can access the values for a particular key like this\n",
        "# dictionaries are a cool object that R doesn't have..jejeje\n",
        "# GOT_13etc is the key\n",
        "print(id_spp_dict['GOT_130404_P001_WA01'])\n",
        "\n",
        "# to access only the species:\n",
        "print('\\n',id_spp_dict['GOT_130404_P001_WA01'][0]) # python, unlike R, is zero-index. The first item on anything has index 0\n",
        "\n",
        "# to access only the voucher:\n",
        "print('\\n',id_spp_dict['GOT_130404_P001_WA01'][1])"
      ],
      "execution_count": 36,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "('Acoelorrhaphe_wrightii', 'H. Balslev 8156')\n",
            "\n",
            " Acoelorrhaphe_wrightii\n",
            "\n",
            " H. Balslev 8156\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 779
        },
        "id": "ekf8vKl_CVPY",
        "outputId": "11b7f97e-4368-494e-ec17-dcf9b3e881b0"
      },
      "source": [
        "# creating a column with species and voucher in the rapidgenomics dataframe\n",
        "seqprapid.columns # print them to make sure you aren't replacing any column\n",
        "\n",
        "seqprapid['taxspecies'] = ''\n",
        "seqprapid['voucher'] = ''\n",
        "seqprapid['taxgenus'] = ''\n",
        "for row in seqprapid.itertuples():\n",
        "  seqprapid.loc[row.Index,'taxspecies'] = id_spp_dict[row.RG_Sample_Code][0]\n",
        "  seqprapid.loc[row.Index,'taxgenus'] = id_spp_dict[row.RG_Sample_Code][0].split('_')[0]\n",
        "  seqprapid.loc[row.Index,'voucher'] = id_spp_dict[row.RG_Sample_Code][1]\n",
        "seqprapid['dataset'] = 'APP'\n",
        "seqprapid['provider'] = 'Paola Lima'\n",
        "seqprapid['pi_code'] = 'PL'\n",
        "seqprapid['Continent'] = 'americas'\n",
        "seqprapid['sense'] = 'R1'\n",
        "\n",
        "seqprapid['samplecode'] = seqprapid['RG_Sample_Code'].map(id_code_dict)\n",
        "\n",
        "seqprapid['originalmetadata'] = 'GOT_130404_SampleSheet metadata_machinereadable_Paola_2020'\n",
        "seqprapid['library_index'] = seqprapid['i5_Barcode_Seq']+'+'+seqprapid['i7_Barcode_Seq']\n",
        "\n",
        "seqprapid.rename(columns = {'Sequence_Name':'filename'}, inplace = True) # uses a dictionary with keys (old name) and value (new name)\n",
        "seqprapid"
      ],
      "execution_count": 43,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>RG_Sample_Code</th>\n",
              "      <th>Customer_Code</th>\n",
              "      <th>i5_Barcode_Seq</th>\n",
              "      <th>i7_Barcode_Seq</th>\n",
              "      <th>filename</th>\n",
              "      <th>Sequencing_Cycle</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>Continent</th>\n",
              "      <th>sense</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>GOT_130404_P001_WA01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>GTGTTCTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Acoelorrhaphe_wrightii</td>\n",
              "      <td>H. Balslev 8156</td>\n",
              "      <td>Acoelorrhaphe</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>A1_HBalslev8156</td>\n",
              "      <td>TATGCAGT+GTGTTCTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>GOT_130404_P001_WA02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>AGTCACTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_bicornis</td>\n",
              "      <td>J. West s.n.</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>A2_JWestsn</td>\n",
              "      <td>TATGCAGT+AGTCACTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>GOT_130404_P001_WA03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>ATTGGCTC</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_chiribogensis</td>\n",
              "      <td>F. Borchsenius 652</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>A3_FBorchsenius652</td>\n",
              "      <td>TATGCAGT+ATTGGCTC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>GOT_130404_P001_WA04</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>CAGATCTG</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA04_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_deltoidea</td>\n",
              "      <td>L.A. Nuñez s.n.</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>A4_LANuezsn</td>\n",
              "      <td>TATGCAGT+CAGATCTG</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>GOT_130404_P001_WA05</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>CCGTGAGA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA05_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_duquei</td>\n",
              "      <td>R. Bernal 2852</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>A5_RBernal2852</td>\n",
              "      <td>TATGCAGT+CCGTGAGA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>118</th>\n",
              "      <td>GOT_130404_P002_WB11</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>ACCACTGT</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WB11_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_praemorsa</td>\n",
              "      <td>R. Bernal 2883</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>B11_RBernal2883</td>\n",
              "      <td>TACTCCTT+ACCACTGT</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>119</th>\n",
              "      <td>GOT_130404_P002_WB12</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>ACAGATTC</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WB12_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_quinaria</td>\n",
              "      <td>R. Bernal 2491</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>B12_RBernal2491</td>\n",
              "      <td>TACTCCTT+ACAGATTC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>120</th>\n",
              "      <td>GOT_130404_P002_WC01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGAACAA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_radiata</td>\n",
              "      <td>R. Bernal 2190</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>C1_RBernal2190</td>\n",
              "      <td>TACTCCTT+TGGAACAA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121</th>\n",
              "      <td>GOT_130404_P002_WC02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGTGGTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_verruculosa</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>C2_RBernal2500</td>\n",
              "      <td>TACTCCTT+TGGTGGTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>122</th>\n",
              "      <td>GOT_130404_P002_WC03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>CGAACTTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Mauritiella_armata</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>Mauritiella</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>C3_JCPintaud1488</td>\n",
              "      <td>TACTCCTT+CGAACTTA</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>123 rows × 17 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "           RG_Sample_Code  Customer_Code  ...          samplecode      library_index\n",
              "0    GOT_130404_P001_WA01            NaN  ...     A1_HBalslev8156  TATGCAGT+GTGTTCTA\n",
              "1    GOT_130404_P001_WA02            NaN  ...          A2_JWestsn  TATGCAGT+AGTCACTA\n",
              "2    GOT_130404_P001_WA03            NaN  ...  A3_FBorchsenius652  TATGCAGT+ATTGGCTC\n",
              "3    GOT_130404_P001_WA04            NaN  ...         A4_LANuezsn  TATGCAGT+CAGATCTG\n",
              "4    GOT_130404_P001_WA05            NaN  ...      A5_RBernal2852  TATGCAGT+CCGTGAGA\n",
              "..                    ...            ...  ...                 ...                ...\n",
              "118  GOT_130404_P002_WB11            NaN  ...     B11_RBernal2883  TACTCCTT+ACCACTGT\n",
              "119  GOT_130404_P002_WB12            NaN  ...     B12_RBernal2491  TACTCCTT+ACAGATTC\n",
              "120  GOT_130404_P002_WC01            NaN  ...      C1_RBernal2190  TACTCCTT+TGGAACAA\n",
              "121  GOT_130404_P002_WC02            NaN  ...      C2_RBernal2500  TACTCCTT+TGGTGGTA\n",
              "122  GOT_130404_P002_WC03            NaN  ...    C3_JCPintaud1488  TACTCCTT+CGAACTTA\n",
              "\n",
              "[123 rows x 17 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 43
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MlCxAxl3IzcQ",
        "outputId": "43189920-9f5e-4f28-b525-f502cb2babca"
      },
      "source": [
        "# duplicate the dataframe to make up for the file with reverse files\n",
        "seqprapidR2 = seqprapid.copy(deep = True)\n",
        "seqprapidR2['filename'] = seqprapidR2['filename'].str.replace('_R1_','_R2_')\n",
        "seqprapidR2['sense'] = seqprapidR2['sense'].str.replace('R1','R2')\n",
        "\n",
        "list(seqprapidR2['filename'])"
      ],
      "execution_count": 51,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['RAPiD-Genomics_F147_GOT_130404_P001_WA01_i5-514_i7-59_S42_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA02_i5-514_i7-27_S43_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA03_i5-514_i7-82_S44_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA04_i5-514_i7-7_S45_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA05_i5-514_i7-38_S46_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA06_i5-514_i7-74_S47_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA07_i5-514_i7-77_S48_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA08_i5-514_i7-16_S49_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA09_i5-514_i7-36_S50_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA10_i5-514_i7-54_S51_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA11_i5-514_i7-25_S52_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WA12_i5-514_i7-23_S53_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB01_i5-514_i7-31_S54_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB02_i5-514_i7-39_S55_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB03_i5-514_i7-2_S56_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB04_i5-514_i7-20_S57_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB05_i5-514_i7-90_S58_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB06_i5-514_i7-6_S59_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB07_i5-514_i7-29_S60_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB08_i5-514_i7-57_S61_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB09_i5-514_i7-63_S62_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB10_i5-514_i7-55_S63_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB11_i5-514_i7-5_S64_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WB12_i5-514_i7-75_S65_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC01_i5-514_i7-65_S66_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC02_i5-514_i7-67_S67_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC03_i5-514_i7-40_S68_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC04_i5-514_i7-19_S69_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC05_i5-514_i7-81_S70_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC06_i5-514_i7-3_S71_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC07_i5-514_i7-62_S72_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC08_i5-514_i7-17_S73_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC09_i5-514_i7-93_S74_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC10_i5-514_i7-47_S75_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC11_i5-514_i7-70_S76_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WC12_i5-514_i7-43_S77_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD01_i5-514_i7-52_S78_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD02_i5-514_i7-72_S79_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD03_i5-514_i7-30_S80_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD04_i5-514_i7-91_S81_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD05_i5-514_i7-71_S82_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD06_i5-514_i7-50_S83_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD07_i5-514_i7-86_S84_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD08_i5-514_i7-80_S85_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD09_i5-514_i7-64_S86_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD10_i5-514_i7-8_S87_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD11_i5-514_i7-58_S88_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WD12_i5-514_i7-9_S89_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE01_i5-514_i7-48_S90_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE02_i5-514_i7-89_S91_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE03_i5-514_i7-87_S92_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE04_i5-514_i7-21_S93_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE05_i5-514_i7-94_S94_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE06_i5-514_i7-35_S95_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE07_i5-514_i7-44_S96_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE08_i5-514_i7-4_S97_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE09_i5-514_i7-24_S98_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE10_i5-514_i7-61_S99_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE11_i5-514_i7-13_S100_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WE12_i5-514_i7-76_S101_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF01_i5-514_i7-28_S102_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF02_i5-514_i7-78_S103_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF03_i5-514_i7-79_S104_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF04_i5-514_i7-45_S150_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF05_i5-514_i7-92_S105_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF06_i5-514_i7-60_S106_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF07_i5-514_i7-33_S107_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF08_i5-514_i7-53_S108_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF09_i5-514_i7-46_S109_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF10_i5-514_i7-26_S110_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF11_i5-514_i7-51_S111_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WF12_i5-514_i7-66_S112_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG01_i5-514_i7-84_S113_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG02_i5-514_i7-12_S114_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG03_i5-514_i7-22_S115_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG04_i5-514_i7-10_S116_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG05_i5-514_i7-69_S117_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG06_i5-514_i7-14_S118_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG07_i5-514_i7-18_S119_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG08_i5-514_i7-68_S151_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG09_i5-514_i7-41_S120_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG10_i5-514_i7-37_S121_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG11_i5-514_i7-96_S122_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WG12_i5-514_i7-42_S123_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH01_i5-514_i7-11_S124_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH02_i5-514_i7-32_S125_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH03_i5-514_i7-88_S126_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH04_i5-514_i7-73_S127_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH05_i5-514_i7-15_S128_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH06_i5-514_i7-49_S129_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH07_i5-514_i7-1_S130_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH08_i5-514_i7-85_S131_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH09_i5-514_i7-83_S132_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH10_i5-514_i7-34_S133_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH11_i5-514_i7-95_S134_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P001_WH12_i5-514_i7-56_S152_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA01_i5-515_i7-59_S135_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA02_i5-515_i7-27_S136_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA03_i5-515_i7-82_S137_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA04_i5-515_i7-7_S138_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA05_i5-515_i7-38_S139_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA06_i5-515_i7-74_S140_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA07_i5-515_i7-77_S141_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA08_i5-515_i7-16_S142_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA09_i5-515_i7-36_S143_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA10_i5-515_i7-54_S144_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA11_i5-515_i7-25_S145_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WA12_i5-515_i7-23_S146_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB01_i5-515_i7-31_S147_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB02_i5-515_i7-39_S148_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB03_i5-515_i7-2_S149_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB04_i5-515_i7-20_S153_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB05_i5-515_i7-90_S154_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB06_i5-515_i7-6_S155_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB07_i5-515_i7-29_S156_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB08_i5-515_i7-57_S157_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB09_i5-515_i7-63_S158_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB10_i5-515_i7-55_S159_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB11_i5-515_i7-5_S160_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WB12_i5-515_i7-75_S161_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WC01_i5-515_i7-65_S162_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-515_i7-67_S163_L001_R2_001.fastq.gz',\n",
              " 'RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-515_i7-40_S164_L001_R2_001.fastq.gz']"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 51
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 779
        },
        "id": "DFo3wwcVKZ5H",
        "outputId": "9f528896-1410-4f5f-f822-54afe07ad149"
      },
      "source": [
        "# concatenate dataframes for forward and reverse files and sort by filename\n",
        "seqprapidall = pd.concat([seqprapid, seqprapidR2]).sort_values(by = 'filename').reset_index()\n",
        "seqprapidall"
      ],
      "execution_count": 52,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>index</th>\n",
              "      <th>RG_Sample_Code</th>\n",
              "      <th>Customer_Code</th>\n",
              "      <th>i5_Barcode_Seq</th>\n",
              "      <th>i7_Barcode_Seq</th>\n",
              "      <th>filename</th>\n",
              "      <th>Sequencing_Cycle</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>Continent</th>\n",
              "      <th>sense</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>GOT_130404_P001_WA01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>GTGTTCTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Acoelorrhaphe_wrightii</td>\n",
              "      <td>H. Balslev 8156</td>\n",
              "      <td>Acoelorrhaphe</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>A1_HBalslev8156</td>\n",
              "      <td>TATGCAGT+GTGTTCTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>0</td>\n",
              "      <td>GOT_130404_P001_WA01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>GTGTTCTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Acoelorrhaphe_wrightii</td>\n",
              "      <td>H. Balslev 8156</td>\n",
              "      <td>Acoelorrhaphe</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R2</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>A1_HBalslev8156</td>\n",
              "      <td>TATGCAGT+GTGTTCTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>AGTCACTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_bicornis</td>\n",
              "      <td>J. West s.n.</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>A2_JWestsn</td>\n",
              "      <td>TATGCAGT+AGTCACTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1</td>\n",
              "      <td>GOT_130404_P001_WA02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>AGTCACTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_bicornis</td>\n",
              "      <td>J. West s.n.</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R2</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>A2_JWestsn</td>\n",
              "      <td>TATGCAGT+AGTCACTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>2</td>\n",
              "      <td>GOT_130404_P001_WA03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TATGCAGT</td>\n",
              "      <td>ATTGGCTC</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P001_WA03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Aiphanes_chiribogensis</td>\n",
              "      <td>F. Borchsenius 652</td>\n",
              "      <td>Aiphanes</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>A3_FBorchsenius652</td>\n",
              "      <td>TATGCAGT+ATTGGCTC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>241</th>\n",
              "      <td>120</td>\n",
              "      <td>GOT_130404_P002_WC01</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGAACAA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC01_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_radiata</td>\n",
              "      <td>R. Bernal 2190</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R2</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>C1_RBernal2190</td>\n",
              "      <td>TACTCCTT+TGGAACAA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>242</th>\n",
              "      <td>121</td>\n",
              "      <td>GOT_130404_P002_WC02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGTGGTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_verruculosa</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>C2_RBernal2500</td>\n",
              "      <td>TACTCCTT+TGGTGGTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>243</th>\n",
              "      <td>121</td>\n",
              "      <td>GOT_130404_P002_WC02</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>TGGTGGTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Wettinia_verruculosa</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R2</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>C2_RBernal2500</td>\n",
              "      <td>TACTCCTT+TGGTGGTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>244</th>\n",
              "      <td>122</td>\n",
              "      <td>GOT_130404_P002_WC03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>CGAACTTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Mauritiella_armata</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>Mauritiella</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R1</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>C3_JCPintaud1488</td>\n",
              "      <td>TACTCCTT+CGAACTTA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>245</th>\n",
              "      <td>122</td>\n",
              "      <td>GOT_130404_P002_WC03</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TACTCCTT</td>\n",
              "      <td>CGAACTTA</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>2x150</td>\n",
              "      <td>Mauritiella_armata</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>Mauritiella</td>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>americas</td>\n",
              "      <td>R2</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>C3_JCPintaud1488</td>\n",
              "      <td>TACTCCTT+CGAACTTA</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>246 rows × 18 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     index        RG_Sample_Code  ...          samplecode      library_index\n",
              "0        0  GOT_130404_P001_WA01  ...     A1_HBalslev8156  TATGCAGT+GTGTTCTA\n",
              "1        0  GOT_130404_P001_WA01  ...     A1_HBalslev8156  TATGCAGT+GTGTTCTA\n",
              "2        1  GOT_130404_P001_WA02  ...          A2_JWestsn  TATGCAGT+AGTCACTA\n",
              "3        1  GOT_130404_P001_WA02  ...          A2_JWestsn  TATGCAGT+AGTCACTA\n",
              "4        2  GOT_130404_P001_WA03  ...  A3_FBorchsenius652  TATGCAGT+ATTGGCTC\n",
              "..     ...                   ...  ...                 ...                ...\n",
              "241    120  GOT_130404_P002_WC01  ...      C1_RBernal2190  TACTCCTT+TGGAACAA\n",
              "242    121  GOT_130404_P002_WC02  ...      C2_RBernal2500  TACTCCTT+TGGTGGTA\n",
              "243    121  GOT_130404_P002_WC02  ...      C2_RBernal2500  TACTCCTT+TGGTGGTA\n",
              "244    122  GOT_130404_P002_WC03  ...    C3_JCPintaud1488  TACTCCTT+CGAACTTA\n",
              "245    122  GOT_130404_P002_WC03  ...    C3_JCPintaud1488  TACTCCTT+CGAACTTA\n",
              "\n",
              "[246 rows x 18 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 52
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hIz2L-PyIRl0"
      },
      "source": [
        "Now, we can concatenate the dataframes"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 694
        },
        "id": "lq-9XIitH-fB",
        "outputId": "11bd7657-c678-464e-edb4-65c4b3cbd303"
      },
      "source": [
        "df = pd.concat([df, seqprapidall[['filename','taxspecies', 'voucher','taxgenus',\n",
        "       'dataset', 'provider', 'pi_code', 'Continent', 'sense',\n",
        "       'originalmetadata', 'samplecode', 'library_index' ]]])\n",
        "df"
      ],
      "execution_count": 53,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset</th>\n",
              "      <th>provider</th>\n",
              "      <th>pi_code</th>\n",
              "      <th>originalmetadata</th>\n",
              "      <th>filename</th>\n",
              "      <th>sense</th>\n",
              "      <th>filecode</th>\n",
              "      <th>samplecode</th>\n",
              "      <th>library_index</th>\n",
              "      <th>botanic_garden</th>\n",
              "      <th>voucher</th>\n",
              "      <th>taxgenus</th>\n",
              "      <th>taxspecies</th>\n",
              "      <th>ifmorphotype</th>\n",
              "      <th>ifpopulation</th>\n",
              "      <th>sent_to_Cano</th>\n",
              "      <th>newfilename</th>\n",
              "      <th>raw_reads</th>\n",
              "      <th>reads_trimed_paired_flag</th>\n",
              "      <th>reads_trimed_paired</th>\n",
              "      <th>reads_trimed_single</th>\n",
              "      <th>percentage_lost_tosingles</th>\n",
              "      <th>pcr_filtered</th>\n",
              "      <th>collection_year</th>\n",
              "      <th>Continent</th>\n",
              "      <th>country</th>\n",
              "      <th>long</th>\n",
              "      <th>lat</th>\n",
              "      <th>flag</th>\n",
              "      <th>notes</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151124_000000000-AJE6N_P3252_1001_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AJE6N_P3252_1001</td>\n",
              "      <td>AJE6N_1001</td>\n",
              "      <td>CGATGT</td>\n",
              "      <td>G</td>\n",
              "      <td>Cano_A._etal__ACS338</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_aculeata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_acu_AJE6N1001_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>Panama</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AK5EU_P3252_1126_2.fastq.gz</td>\n",
              "      <td>R2</td>\n",
              "      <td>AK5EU_P3252_1126</td>\n",
              "      <td>AK5EU_1126</td>\n",
              "      <td>CACCGG</td>\n",
              "      <td>FTBG</td>\n",
              "      <td>_FTBG_20040120A</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_crispa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_cri_AK5EU1126_AC_R2.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>failed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Cano</td>\n",
              "      <td>Angela Cano</td>\n",
              "      <td>AC</td>\n",
              "      <td>Sent Appendix_SamplingPhylogeny_CentralAmerica...</td>\n",
              "      <td>1_151130_000000000-AKG91_P3252_1182_1.fastq.gz</td>\n",
              "      <td>R1</td>\n",
              "      <td>AKG91_P3252_1182</td>\n",
              "      <td>AKG91_1182</td>\n",
              "      <td>CTAGCT</td>\n",
              "      <td>JBP</td>\n",
              "      <td>Lorenzi_&amp;_Soares_6762</td>\n",
              "      <td>Acrocomia</td>\n",
              "      <td>Acrocomia_emensis</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>sent</td>\n",
              "      <td>Acr_eme_AKG911182_AC_R1.fastq.gz</td>\n",
              "      <td>NaN</td>\n",
              "      <td>passed</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>241</th>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC01_i5-51...</td>\n",
              "      <td>R2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>C1_RBernal2190</td>\n",
              "      <td>TACTCCTT+TGGAACAA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>R. Bernal 2190</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>Wettinia_radiata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>242</th>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>R1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>C2_RBernal2500</td>\n",
              "      <td>TACTCCTT+TGGTGGTA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>Wettinia_verruculosa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>243</th>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC02_i5-51...</td>\n",
              "      <td>R2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>C2_RBernal2500</td>\n",
              "      <td>TACTCCTT+TGGTGGTA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>R. Bernal 2500</td>\n",
              "      <td>Wettinia</td>\n",
              "      <td>Wettinia_verruculosa</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>244</th>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>R1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>C3_JCPintaud1488</td>\n",
              "      <td>TACTCCTT+CGAACTTA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>Mauritiella</td>\n",
              "      <td>Mauritiella_armata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>245</th>\n",
              "      <td>APP</td>\n",
              "      <td>Paola Lima</td>\n",
              "      <td>PL</td>\n",
              "      <td>GOT_130404_SampleSheet metadata_machinereadabl...</td>\n",
              "      <td>RAPiD-Genomics_F147_GOT_130404_P002_WC03_i5-51...</td>\n",
              "      <td>R2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>C3_JCPintaud1488</td>\n",
              "      <td>TACTCCTT+CGAACTTA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>J.C.Pintaud 1488</td>\n",
              "      <td>Mauritiella</td>\n",
              "      <td>Mauritiella_armata</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>americas</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3418 rows × 30 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "    dataset     provider pi_code  ... lat flag notes\n",
              "0      Cano  Angela Cano      AC  ... NaN  NaN   NaN\n",
              "1      Cano  Angela Cano      AC  ... NaN  NaN   NaN\n",
              "2      Cano  Angela Cano      AC  ... NaN  NaN   NaN\n",
              "3      Cano  Angela Cano      AC  ... NaN  NaN   NaN\n",
              "4      Cano  Angela Cano      AC  ... NaN  NaN   NaN\n",
              "..      ...          ...     ...  ...  ..  ...   ...\n",
              "241     APP   Paola Lima      PL  ... NaN  NaN   NaN\n",
              "242     APP   Paola Lima      PL  ... NaN  NaN   NaN\n",
              "243     APP   Paola Lima      PL  ... NaN  NaN   NaN\n",
              "244     APP   Paola Lima      PL  ... NaN  NaN   NaN\n",
              "245     APP   Paola Lima      PL  ... NaN  NaN   NaN\n",
              "\n",
              "[3418 rows x 30 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 53
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8tmr99mGLlGV"
      },
      "source": [
        "Let's count species again"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GQZ5zBZYCVRt",
        "outputId": "95aa8cca-1745-4f7a-bef1-6ab8719107fa"
      },
      "source": [
        "# first, we create sets of items to compare, using list comprehention. Sets should be defined within curly brackets\n",
        "sppseqed = {spp for spp in df[df['Continent'] == 'americas']['taxspecies'].unique()} #list comprehension are loops of a single line\n",
        "apspp = {spp for spp in palms_americas} # american palm species\n",
        "# we can apply set theory to find the differences (not sequenced or sequenced but not in the list of american palms)\n",
        "# items in setA that are not shared with setB and viceversa\n",
        "# symmetric difference \n",
        "print(apspp - sppseqed) # niceee, nothing is missing\n",
        "sppmissing = len(apspp - sppseqed)\n",
        "print('\\nAlleged missing species: %s'%(sppmissing))\n",
        "print('Percentage of species missing: %s'%((sppmissing * 100)/len(palms_americas)))\n",
        "print('Percentage of species sequenced: %s'%(100-(sppmissing * 100)/len(palms_americas)))"
      ],
      "execution_count": 54,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "{'Attalea_exigua', 'Chamaedorea_undulatifolia', 'Euterpe_longibracteata', 'Lytocaryum_insigne', 'Desmoncus_osensis', 'Attalea_macrolepis', 'Geonoma_brenesii', 'Brahea_edulis', 'Copernicia_humicola', 'Chamaedorea_smithii', 'Bactris_grayumii', 'Attalea_salazarii', 'Aiphanes_multiplex', 'Euterpe_luminosa', 'Bactris_ptariana', 'Calyptrogyne_kunorum', 'Pholidostachys_amazonensis', 'Guihaia_grossifibrosa', 'Colpothrinax_cookii', 'Chamaedorea_latisecta', 'Chamaedorea_christinae', 'Coccothrinax_microphylla', 'Bactris_herrerana', 'Coccothrinax_acunana', 'Desmoncus_kunarius', 'Desmoncus_horridus', 'Reinhardtia_paiewonskiana', 'Bactris_nancibaensis', 'Lytocaryum_hoehnei', 'Brahea_calcarea', 'Geonoma_simplicifrons', 'Coccothrinax_muricata', 'Oenocarpus_circumtextus', 'Hyospathe_wendlandiana', 'Astrocaryum_aculeatissimum', 'Aiphanes_acanthophylla', 'Prestoea_carderi', 'Hemithrinax_rivularis', 'Attalea_wesselsboeri', 'Copernicia_molineti', 'Geonoma_poiteauana', 'Attalea_dahlgreniana', 'Coccothrinax_scoparia', 'Geonoma_chococola', 'Bactris_kunorum', 'Butia_missionera', 'Prestoea_simplicifolia', 'Leopoldinia_piassaba', 'Desmoncus_prunifer', 'Calyptrogyne_deneversii', 'Roystonea_violacea', 'Geonoma_trigona', 'Oenocarpus_makeru', 'Bactris_timbuiensis', 'Attalea_huebneri', 'Brahea_pimo', 'Aiphanes_stergiosii', 'Syagrus_pimentae', 'Coccothrinax_bermudezii', 'Bactris_tefensis', 'Coccothrinax_garciana', 'Calyptrogyne_pubescens', 'Attalea_maracaibensis', 'Roystonea_stellata', 'Dictyocaryum_fuscum', 'Geonoma_schizocarpa', 'Lytocaryum_weddellianum', 'Bactris_vulgaris', 'Attalea_maripensis', 'Acoelorraphe_wrightii', 'Geonoma_ferruginea', 'Chamaedorea_rosibeliae', 'Geonoma_longipedunculata', 'Phytelephas_schottii', 'Astrocaryum_tucuma', 'Copernicia_curbeloi', 'Calyptrogyne_condensata', 'Asterogyne_yaracuyense', 'Geonoma_deneversii', 'Coccothrinax_rigida', 'Coccothrinax_yuraguana', 'Attalea_fairchildensis', 'Coccothrinax_moaensis', 'Dictyocaryum_ptarianum', 'Desmoncus_stans', 'Attalea_lauromuelleriana', 'Attalea_degranvillei', 'Attalea_camopiensis', 'Attalea_blepharopus', 'Attalea_compta', 'Bactris_turbinocarpa', 'Butia_catarinensis', 'Bactris_cubensis', 'Attalea_magdalenica', 'Cryosophila_macrocarpa', 'Phytelephas_seemannii', 'Attalea_salvadorensis', 'Chamaedorea_ernesti-augusti', 'Chamaedorea_moliniana', 'Syagrus_werdermannii', 'Bactris_sphaerocarpa', 'Bactris_hatschbachii', 'Chamaedorea_atrovirens', 'Brahea_salvadorensis', 'Astrocaryum_sociale', 'Coccothrinax_baracoensis', 'Calyptrogyne_osensis', 'Roystonea_dunlapiana', 'Syagrus_cataphracta', 'Geonoma_scoparia', 'Chamaedorea_queroana', 'Copernicia_longiglossa', 'Copernicia_oxycalyx', 'Desmoncus_chinantlensis', 'Geonoma_tenuissima', 'Attalea_bassleriana', 'Bactris_jamaicana', 'Coccothrinax_camagueyana', 'Reinhardtia_elegans', 'Desmoncus_latisectus', 'Attalea_kewensis', 'Bactris_dianeura', 'Coccothrinax_pumila', 'Bactris_rostrata', 'Syagrus_pompeoi', 'Thrinax_excelsa', 'Bactris_bahiensis', 'Brahea_moorei', 'Chamaedorea_foveata', 'Syagrus_emasensis', 'Geonoma_peruviana', 'Guihaia_argyrata', 'Desmoncus_obovoideus', 'Attalea_septuagenata', 'Wettinia_minima', 'Astrocaryum_farinosum', 'Hemithrinax_compacta', 'Bactris_soeiroana', 'Calyptrogyne_sanblasensis', 'Bactris_tomentosa', 'Butia_marmorii', 'Cryosophila_cookii', 'Attalea_hoehnei', 'Chamaedorea_castillo-montii', 'Attalea_guianensis', 'Chamaedorea_pachecoana', 'Hyospathe_macrorhachis', 'Syagrus_cearensis', 'Geonoma_bernalii', 'Bactris_polystachya', 'Socratea_montana', 'Acrocomia_media', 'Cryosophila_grayumii', 'Coccothrinax_boschiana', 'Guihaia_lancifolia', 'Coccothrinax_victorini', 'Desmoncus_pumilus', 'Asterogyne_spicata', 'Desmoncus_madrensis', 'Desmoncus_leptoclonos', 'Attalea_osmantha', 'Attalea_rostrata', 'Lytocaryum_itapebiense', 'Chamaedorea_schiedeana', 'Syagrus_yungasensis', 'Brahea_decumbens', 'Astrocaryum_cuatrecasanum', 'Coccothrinax_nipensis', 'Geonoma_gentryi', 'Allagoptera_campestris', 'Coccothrinax_alexandri', 'Geonoma_dindoensis', 'Geonoma_operculata', 'Trithrinax_acanthocoma', 'Bactris_ferruginea', 'Geonoma_venosa', 'Astrocaryum_rodriguesii', 'Hyospathe_peruviana', 'Bactris_hondurensis', 'Coccothrinax_trinitensis', 'Coccothrinax_concolor', 'Syagrus_lilliputiana', 'Geonoma_divisa', 'Coccothrinax_pseudorigida', 'Manicaria_martiana', 'Acrocomia_hassleri', 'Coccothrinax_fagildei', 'Roystonea_princeps', 'Attalea_amylacea', 'Geonoma_braunii', 'Geonoma_sanmartinensis', 'Bactris_glassmanii', 'Mauritia_carana', 'Geonoma_aspidiifolia', 'Coccothrinax_pauciramosa', 'Attalea_rhynchocarpa', 'Bactris_acanthocarpa', 'Attalea_peruviana', 'Bactris_pickelii', 'Geonoma_talamancana', 'Chamaedorea_whitelockiana', 'Calyptrogyne_coloradensis', 'Copernicia_brittonorum', 'Butia_stolonifera', 'Pholidostachys_occidentalis', 'Brahea_brandegeei', 'Aiphanes_spicata', 'Chamaedorea_angustisecta', 'Attalea_cephalotus', 'Bactris_caudata', 'Desmoncus_interjectus', 'Copernicia_roigii', 'Prestoea_tenuiramosa', 'Desmoncus_loretanus', 'Bactris_faucium', 'Sabal_domingensis', 'Bactris_rhaphidacantha', 'Calyptrogyne_herrerae', 'Attalea_weberbaueri', 'Bactris_ana-juliae', 'Bactris_horridispatha', 'Geonoma_spinescens', 'Coccothrinax_leonis', 'Geonoma_concinnoidea', 'Syagrus_pleioclada', 'Butia_quaraimana', 'Desmoncus_setosus', 'Bactris_cuspidata', 'Butia_exilata', 'Reinhardtia_latisecta', 'Desmoncus_moorei', 'Chamaedorea_serpens', 'Bactris_oligoclada', 'Coccothrinax_yunquensis', 'Astrocaryum_confertum', 'Desmoncus_parvulus', 'Desmoncus_costaricensis', 'Copernicia_cowellii', 'Astrocaryum_giganteum', 'Copernicia_curtissii', 'Chamaedorea_amabilis'}\n",
            "\n",
            "Alleged missing species: 241\n",
            "Percentage of species missing: 29.753086419753085\n",
            "Percentage of species sequenced: 70.24691358024691\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "G4IH81SXL42z"
      },
      "source": [
        "Looks better, huh?\n",
        "We can also double check that we are not missing species because of mislabeled continents. We already have the list of american species and that column is redundant... but better to have it"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pDK-iVJZi-KA",
        "outputId": "23f65d0e-385d-4b89-9ef1-6f96ddd45161"
      },
      "source": [
        "# first, we create sets of items to compare, using list comprehention. Sets should be defined within curly brackets\n",
        "sppseqed = {spp for spp in df['taxspecies'].unique()} #list comprehension are loops of a single line\n",
        "apspp = {spp for spp in palms_americas} # american palm species\n",
        "# we can apply set theory to find the differences (not sequenced or sequenced but not in the list of american palms)\n",
        "# items in setA that are not shared with setB and viceversa\n",
        "# symmetric difference \n",
        "print(apspp - sppseqed) # niceee, nothing is missing\n",
        "sppmissing = len(apspp - sppseqed)\n",
        "print('\\nAlleged missing species: %s'%(sppmissing))\n",
        "print('Percentage of species missing: %s'%((sppmissing * 100)/len(palms_americas)))\n",
        "print('Percentage of species sequenced: %s'%(100-(sppmissing * 100)/len(palms_americas)))"
      ],
      "execution_count": 55,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "{'Attalea_exigua', 'Chamaedorea_undulatifolia', 'Euterpe_longibracteata', 'Lytocaryum_insigne', 'Desmoncus_osensis', 'Attalea_macrolepis', 'Geonoma_brenesii', 'Brahea_edulis', 'Copernicia_humicola', 'Chamaedorea_smithii', 'Bactris_grayumii', 'Attalea_salazarii', 'Aiphanes_multiplex', 'Euterpe_luminosa', 'Bactris_ptariana', 'Calyptrogyne_kunorum', 'Pholidostachys_amazonensis', 'Guihaia_grossifibrosa', 'Colpothrinax_cookii', 'Chamaedorea_latisecta', 'Chamaedorea_christinae', 'Coccothrinax_microphylla', 'Bactris_herrerana', 'Coccothrinax_acunana', 'Desmoncus_kunarius', 'Desmoncus_horridus', 'Reinhardtia_paiewonskiana', 'Bactris_nancibaensis', 'Lytocaryum_hoehnei', 'Brahea_calcarea', 'Geonoma_simplicifrons', 'Coccothrinax_muricata', 'Oenocarpus_circumtextus', 'Hyospathe_wendlandiana', 'Astrocaryum_aculeatissimum', 'Aiphanes_acanthophylla', 'Prestoea_carderi', 'Hemithrinax_rivularis', 'Attalea_wesselsboeri', 'Copernicia_molineti', 'Geonoma_poiteauana', 'Attalea_dahlgreniana', 'Coccothrinax_scoparia', 'Geonoma_chococola', 'Bactris_kunorum', 'Butia_missionera', 'Prestoea_simplicifolia', 'Leopoldinia_piassaba', 'Desmoncus_prunifer', 'Calyptrogyne_deneversii', 'Roystonea_violacea', 'Geonoma_trigona', 'Oenocarpus_makeru', 'Bactris_timbuiensis', 'Attalea_huebneri', 'Brahea_pimo', 'Aiphanes_stergiosii', 'Syagrus_pimentae', 'Coccothrinax_bermudezii', 'Bactris_tefensis', 'Coccothrinax_garciana', 'Calyptrogyne_pubescens', 'Attalea_maracaibensis', 'Roystonea_stellata', 'Dictyocaryum_fuscum', 'Geonoma_schizocarpa', 'Lytocaryum_weddellianum', 'Bactris_vulgaris', 'Attalea_maripensis', 'Acoelorraphe_wrightii', 'Geonoma_ferruginea', 'Chamaedorea_rosibeliae', 'Geonoma_longipedunculata', 'Phytelephas_schottii', 'Astrocaryum_tucuma', 'Copernicia_curbeloi', 'Calyptrogyne_condensata', 'Asterogyne_yaracuyense', 'Geonoma_deneversii', 'Coccothrinax_rigida', 'Coccothrinax_yuraguana', 'Attalea_fairchildensis', 'Coccothrinax_moaensis', 'Dictyocaryum_ptarianum', 'Desmoncus_stans', 'Attalea_lauromuelleriana', 'Attalea_degranvillei', 'Attalea_camopiensis', 'Attalea_blepharopus', 'Attalea_compta', 'Bactris_turbinocarpa', 'Butia_catarinensis', 'Bactris_cubensis', 'Attalea_magdalenica', 'Cryosophila_macrocarpa', 'Phytelephas_seemannii', 'Attalea_salvadorensis', 'Chamaedorea_ernesti-augusti', 'Chamaedorea_moliniana', 'Syagrus_werdermannii', 'Bactris_sphaerocarpa', 'Bactris_hatschbachii', 'Chamaedorea_atrovirens', 'Brahea_salvadorensis', 'Astrocaryum_sociale', 'Coccothrinax_baracoensis', 'Calyptrogyne_osensis', 'Roystonea_dunlapiana', 'Syagrus_cataphracta', 'Geonoma_scoparia', 'Chamaedorea_queroana', 'Copernicia_longiglossa', 'Copernicia_oxycalyx', 'Desmoncus_chinantlensis', 'Geonoma_tenuissima', 'Attalea_bassleriana', 'Bactris_jamaicana', 'Coccothrinax_camagueyana', 'Reinhardtia_elegans', 'Desmoncus_latisectus', 'Attalea_kewensis', 'Bactris_dianeura', 'Coccothrinax_pumila', 'Bactris_rostrata', 'Syagrus_pompeoi', 'Thrinax_excelsa', 'Bactris_bahiensis', 'Brahea_moorei', 'Chamaedorea_foveata', 'Syagrus_emasensis', 'Geonoma_peruviana', 'Guihaia_argyrata', 'Desmoncus_obovoideus', 'Attalea_septuagenata', 'Wettinia_minima', 'Astrocaryum_farinosum', 'Hemithrinax_compacta', 'Bactris_soeiroana', 'Calyptrogyne_sanblasensis', 'Bactris_tomentosa', 'Butia_marmorii', 'Cryosophila_cookii', 'Attalea_hoehnei', 'Chamaedorea_castillo-montii', 'Attalea_guianensis', 'Chamaedorea_pachecoana', 'Hyospathe_macrorhachis', 'Syagrus_cearensis', 'Geonoma_bernalii', 'Bactris_polystachya', 'Socratea_montana', 'Acrocomia_media', 'Cryosophila_grayumii', 'Coccothrinax_boschiana', 'Guihaia_lancifolia', 'Coccothrinax_victorini', 'Desmoncus_pumilus', 'Asterogyne_spicata', 'Desmoncus_madrensis', 'Desmoncus_leptoclonos', 'Attalea_osmantha', 'Attalea_rostrata', 'Lytocaryum_itapebiense', 'Chamaedorea_schiedeana', 'Syagrus_yungasensis', 'Brahea_decumbens', 'Astrocaryum_cuatrecasanum', 'Coccothrinax_nipensis', 'Geonoma_gentryi', 'Allagoptera_campestris', 'Coccothrinax_alexandri', 'Geonoma_dindoensis', 'Geonoma_operculata', 'Trithrinax_acanthocoma', 'Bactris_ferruginea', 'Geonoma_venosa', 'Astrocaryum_rodriguesii', 'Hyospathe_peruviana', 'Bactris_hondurensis', 'Coccothrinax_trinitensis', 'Coccothrinax_concolor', 'Syagrus_lilliputiana', 'Geonoma_divisa', 'Coccothrinax_pseudorigida', 'Manicaria_martiana', 'Acrocomia_hassleri', 'Coccothrinax_fagildei', 'Roystonea_princeps', 'Attalea_amylacea', 'Geonoma_braunii', 'Geonoma_sanmartinensis', 'Bactris_glassmanii', 'Mauritia_carana', 'Geonoma_aspidiifolia', 'Coccothrinax_pauciramosa', 'Attalea_rhynchocarpa', 'Bactris_acanthocarpa', 'Attalea_peruviana', 'Bactris_pickelii', 'Geonoma_talamancana', 'Chamaedorea_whitelockiana', 'Calyptrogyne_coloradensis', 'Copernicia_brittonorum', 'Butia_stolonifera', 'Pholidostachys_occidentalis', 'Brahea_brandegeei', 'Aiphanes_spicata', 'Chamaedorea_angustisecta', 'Attalea_cephalotus', 'Bactris_caudata', 'Desmoncus_interjectus', 'Copernicia_roigii', 'Prestoea_tenuiramosa', 'Desmoncus_loretanus', 'Bactris_faucium', 'Sabal_domingensis', 'Bactris_rhaphidacantha', 'Calyptrogyne_herrerae', 'Attalea_weberbaueri', 'Bactris_ana-juliae', 'Bactris_horridispatha', 'Geonoma_spinescens', 'Coccothrinax_leonis', 'Geonoma_concinnoidea', 'Syagrus_pleioclada', 'Butia_quaraimana', 'Desmoncus_setosus', 'Bactris_cuspidata', 'Butia_exilata', 'Reinhardtia_latisecta', 'Desmoncus_moorei', 'Chamaedorea_serpens', 'Bactris_oligoclada', 'Coccothrinax_yunquensis', 'Astrocaryum_confertum', 'Desmoncus_parvulus', 'Desmoncus_costaricensis', 'Copernicia_cowellii', 'Astrocaryum_giganteum', 'Copernicia_curtissii', 'Chamaedorea_amabilis'}\n",
            "\n",
            "Alleged missing species: 241\n",
            "Percentage of species missing: 29.753086419753085\n",
            "Percentage of species sequenced: 70.24691358024691\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7i6Y9yx4MOL5"
      },
      "source": [
        "well.. same thing. But it is always GOOD to double check things.\n",
        "\n",
        "Next thing to do are the Acrocromias. First... how many species do we have until now?"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "A77ZxZhnMXfq"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4_kCnsPUqS6o"
      },
      "source": [
        "You can import your own data into Colab notebooks from your Google Drive account, including from spreadsheets, as well as from Github and many other sources. To learn more about importing data, and how Colab can be used for data science, see the links below under [Working with Data](#working-with-data)."
      ]
    }
  ]
}